{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Large Test Set #\n",
    "## Goal ##\n",
    "The goal of this test set is to show that what was found in the smaller data set is generalizable to a larger number of proteins with a wide range of sizes and available homologs, orthologs, and paralogs. This set of proteins was used previously (Marcos et al., 2011) to evaluate covariance predictions. By using this dataset it should be possible to show the generalizability of the method as well as make a better comparison to previous methods.\n",
    "## Warning ##\n",
    "Before attempting to use this notebook make sure that your .env file has been properly setup to reflect the correct locations of command line tools and the location of files and directories needed for execution.\n",
    "### Initial Import###\n",
    "This first cell performs the necessary imports required to begin this notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from dotenv import find_dotenv, load_dotenv\n",
    "try:\n",
    "    dotenv_path = find_dotenv(raise_error_if_not_found=True)\n",
    "except IOError:\n",
    "    dotenv_path = find_dotenv(raise_error_if_not_found=True, usecwd=True)\n",
    "load_dotenv(dotenv_path)\n",
    "import os\n",
    "import sys\n",
    "sys.path.append(os.path.join(os.environ.get('PROJECT_PATH'), 'src'))\n",
    "sys.path.append(os.path.join(os.environ.get('PROJECT_PATH'), 'src', 'SupportingClasses'))\n",
    "input_dir = os.path.join(os.environ.get('INPUT_PATH'), 'Large_Test_Set')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Set Construction ##\n",
    "The first task required to test the data set is to download the required data and construct any necessary input files for all down stream analyses.\n",
    "In this case that means:\n",
    "* Downloading PDB files for the proteins in our small test set.\n",
    "* Extracting a query sequence from each PDB file.\n",
    "* Searching for paralogs, homologs, and orthologs in a custom BLAST database built by filtering the Uniref90 database.\n",
    "* Filtering the hits from the BLAST search to meet minimum and maximum length requirements, as well as minimum and maximum identity requirements.\n",
    "* Building alignments using CULSTALW in both the fasta and msf formats since some of the tools which will be used for comparison need different formats.\n",
    "* Filtering the alignment for maximum identity similarity between seqeunces.\n",
    "* Re-aligning the filtered sequences using CLUSTALW.\n",
    "This is all handeled by the DataSetGenerator class found in the src/SupportingClasses folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "STARTING identify_protein_sequences\n",
      "IMPORTED PROTEIN LIST\n",
      "PROTEIN ID: 153lA\n",
      "PROTEIN ID: 154lA\n",
      "PROTEIN ID: 1a04A\n",
      "PROTEIN ID: 1a0bA\n",
      "PROTEIN ID: 1a0pA\n",
      "PROTEIN ID: 1ae9A\n",
      "PROTEIN ID: 1al3A\n",
      "PROTEIN ID: 1atgA\n",
      "PROTEIN ID: 1b7eA\n",
      "PROTEIN ID: 1b9mA\n",
      "PROTEIN ID: 1b9nA\n",
      "PROTEIN ID: 1biaA\n",
      "PROTEIN ID: 1bibA\n",
      "PROTEIN ID: 1bl0A\n",
      "PROTEIN ID: 1booA\n",
      "PROTEIN ID: 1bslA\n",
      "PROTEIN ID: 1byiA\n",
      "PROTEIN ID: 1byqA\n",
      "PROTEIN ID: 1c02A\n",
      "PROTEIN ID: 1c52A\n",
      "PROTEIN ID: 1c5kA\n",
      "PROTEIN ID: 1c75A\n",
      "PROTEIN ID: 1cb7A\n",
      "PROTEIN ID: 1ccwA\n",
      "PROTEIN ID: 1cp2A\n",
      "PROTEIN ID: 1crxA\n",
      "PROTEIN ID: 1crzA\n",
      "PROTEIN ID: 1ctjA\n",
      "PROTEIN ID: 1d4aA\n",
      "PROTEIN ID: 1d5yA\n",
      "PROTEIN ID: 1dadA\n",
      "PROTEIN ID: 1daeA\n",
      "PROTEIN ID: 1dagA\n",
      "PROTEIN ID: 1dahA\n",
      "PROTEIN ID: 1daiA\n",
      "PROTEIN ID: 1dakA\n",
      "PROTEIN ID: 1dd9A\n",
      "PROTEIN ID: 1ddeA\n",
      "PROTEIN ID: 1di6A\n",
      "PROTEIN ID: 1di7A\n",
      "PROTEIN ID: 1dljA\n",
      "PROTEIN ID: 1dtsA\n",
      "PROTEIN ID: 1durA\n",
      "PROTEIN ID: 1e2xA\n",
      "PROTEIN ID: 1e3uA\n",
      "PROTEIN ID: 1e4dA\n",
      "PROTEIN ID: 1e4fT\n",
      "PROTEIN ID: 1e4gT\n",
      "PROTEIN ID: 1e8cA\n",
      "PROTEIN ID: 1eclA\n",
      "PROTEIN ID: 1efaA\n",
      "PROTEIN ID: 1efdN\n",
      "PROTEIN ID: 1eg2A\n",
      "PROTEIN ID: 1ek9A\n",
      "PROTEIN ID: 1eszA\n",
      "PROTEIN ID: 1etkA\n",
      "PROTEIN ID: 1etoA\n",
      "PROTEIN ID: 1etvA\n",
      "PROTEIN ID: 1etwA\n",
      "PROTEIN ID: 1etxA\n",
      "PROTEIN ID: 1etyA\n",
      "PROTEIN ID: 1ezwA\n",
      "PROTEIN ID: 1f07A\n",
      "PROTEIN ID: 1f1uA\n",
      "PROTEIN ID: 1f44A\n",
      "PROTEIN ID: 1f48A\n",
      "PROTEIN ID: 1f5vA\n",
      "PROTEIN ID: 1f9iA\n",
      "PROTEIN ID: 1fcaA\n",
      "PROTEIN ID: 1fdnA\n",
      "PROTEIN ID: 1fepA\n",
      "PROTEIN ID: 1fiaA\n",
      "PROTEIN ID: 1fipA\n",
      "PROTEIN ID: 1fp6A\n",
      "PROTEIN ID: 1fr3A\n",
      "PROTEIN ID: 1fseA\n",
      "PROTEIN ID: 1fxoA\n",
      "PROTEIN ID: 1g1lA\n",
      "PROTEIN ID: 1g1mA\n",
      "PROTEIN ID: 1g20A\n",
      "PROTEIN ID: 1g28A\n",
      "PROTEIN ID: 1g5pA\n",
      "PROTEIN ID: 1g60A\n",
      "PROTEIN ID: 1g6oA\n",
      "PROTEIN ID: 1g72A\n",
      "PROTEIN ID: 1g8kA\n",
      "PROTEIN ID: 1gbsA\n",
      "PROTEIN ID: 1gdtA\n",
      "PROTEIN ID: 1gg4A\n",
      "PROTEIN ID: 1gqyA\n",
      "PROTEIN ID: 1gu9A\n",
      "PROTEIN ID: 1gugA\n",
      "PROTEIN ID: 1gunA\n",
      "PROTEIN ID: 1gusA\n",
      "PROTEIN ID: 1gutA\n",
      "PROTEIN ID: 1h3lA\n",
      "PROTEIN ID: 1h4iA\n",
      "PROTEIN ID: 1h7lA\n",
      "PROTEIN ID: 1h7qA\n",
      "PROTEIN ID: 1h8zA\n",
      "PROTEIN ID: 1h98A\n",
      "PROTEIN ID: 1h9gA\n",
      "PROTEIN ID: 1h9jA\n",
      "PROTEIN ID: 1h9kA\n",
      "PROTEIN ID: 1h9mA\n",
      "PROTEIN ID: 1h9sA\n",
      "PROTEIN ID: 1hfeS\n",
      "PROTEIN ID: 1hfeL\n",
      "PROTEIN ID: 1hm9A\n",
      "PROTEIN ID: 1hw1A\n",
      "PROTEIN ID: 1hxdA\n",
      "PROTEIN ID: 1i0rA\n",
      "PROTEIN ID: 1i1gA\n",
      "PROTEIN ID: 1i52A\n",
      "PROTEIN ID: 1i58A\n",
      "PROTEIN ID: 1i5nA\n",
      "PROTEIN ID: 1i74A\n",
      "PROTEIN ID: 1i8oA\n",
      "PROTEIN ID: 1i9cA\n",
      "PROTEIN ID: 1icrA\n",
      "PROTEIN ID: 1id0A\n",
      "PROTEIN ID: 1id1A\n",
      "PROTEIN ID: 1ihcA\n",
      "PROTEIN ID: 1ihrA\n",
      "PROTEIN ID: 1ihuA\n",
      "PROTEIN ID: 1ii0A\n",
      "PROTEIN ID: 1ii9A\n",
      "PROTEIN ID: 1iniA\n",
      "PROTEIN ID: 1injA\n",
      "PROTEIN ID: 1ir6A\n",
      "PROTEIN ID: 1iujA\n",
      "PROTEIN ID: 1ixcA\n",
      "PROTEIN ID: 1ixgA\n",
      "PROTEIN ID: 1ixhA\n",
      "PROTEIN ID: 1iz1A\n",
      "PROTEIN ID: 1j5yA\n",
      "PROTEIN ID: 1j6uA\n",
      "PROTEIN ID: 1jbgA\n",
      "PROTEIN ID: 1jbwA\n",
      "PROTEIN ID: 1je8A\n",
      "PROTEIN ID: 1jetA\n",
      "PROTEIN ID: 1jeuA\n",
      "PROTEIN ID: 1jevA\n",
      "PROTEIN ID: 1jftA\n",
      "PROTEIN ID: 1jh9A\n",
      "PROTEIN ID: 1jiwP\n",
      "PROTEIN ID: 1jiwI\n",
      "PROTEIN ID: 1jljA\n",
      "PROTEIN ID: 1jnuA\n",
      "PROTEIN ID: 1jpuA\n",
      "PROTEIN ID: 1jq5A\n",
      "PROTEIN ID: 1jykA\n",
      "PROTEIN ID: 1k20A\n",
      "PROTEIN ID: 1k2vN\n",
      "PROTEIN ID: 1k38A\n",
      "PROTEIN ID: 1k4fA\n",
      "PROTEIN ID: 1k54A\n",
      "PROTEIN ID: 1k56A\n",
      "PROTEIN ID: 1kapP\n",
      "PROTEIN ID: 1kb0A\n",
      "PROTEIN ID: 1kbuA\n",
      "PROTEIN ID: 1kgsA\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/Seq.py:175: BiopythonWarning: Biopython Seq objects now use string comparison. Older versions of Biopython used object comparison. During this transition, please use hash(id(my_seq)) or my_dict[id(my_seq)] if you want the old behaviour, or use hash(str(my_seq)) or my_dict[str(my_seq)] for the new string hashing behaviour.\n",
      "  \"the new string hashing behaviour.\", BiopythonWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PROTEIN ID: 1kmoA\n",
      "PROTEIN ID: 1kmpA\n",
      "PROTEIN ID: 1kq3A\n",
      "PROTEIN ID: 1ku3A\n",
      "PROTEIN ID: 1ku7A\n",
      "PROTEIN ID: 1kv9A\n",
      "PROTEIN ID: 1kw3B\n",
      "PROTEIN ID: 1kw6B\n",
      "PROTEIN ID: 1l3lA\n",
      "PROTEIN ID: 1lj9A\n",
      "PROTEIN ID: 1lq9A\n",
      "PROTEIN ID: 1lqkA\n",
      "PROTEIN ID: 1lqpA\n",
      "PROTEIN ID: 1lr0A\n",
      "PROTEIN ID: 1ls9A\n",
      "PROTEIN ID: 1lspA\n",
      "PROTEIN ID: 1lssA\n",
      "PROTEIN ID: 1lucA\n",
      "PROTEIN ID: 1lvwA\n",
      "PROTEIN ID: 1m65A\n",
      "PROTEIN ID: 1m68A\n",
      "PROTEIN ID: 1m6kA\n",
      "PROTEIN ID: 1m70A\n",
      "PROTEIN ID: 1m7jA\n",
      "PROTEIN ID: 1ma7A\n",
      "PROTEIN ID: 1mb3A\n",
      "PROTEIN ID: 1mdoA\n",
      "PROTEIN ID: 1mkmA\n",
      "PROTEIN ID: 1mkzA\n",
      "PROTEIN ID: 1mm8A\n",
      "PROTEIN ID: 1mnzA\n",
      "PROTEIN ID: 1moqA\n",
      "PROTEIN ID: 1muhA\n",
      "Desired structure doesn't exists\n",
      "PROTEIN ID: 1murA\n",
      "PROTEIN ID: 1musA\n",
      "PROTEIN ID: 1muwA\n",
      "PROTEIN ID: 1mv8A\n",
      "PROTEIN ID: 1mw8X\n",
      "PROTEIN ID: 1mw9X\n",
      "PROTEIN ID: 1n2zA\n",
      "PROTEIN ID: 1n9lA\n",
      "PROTEIN ID: 1n9nA\n",
      "PROTEIN ID: 1nfpA\n",
      "PROTEIN ID: 1nkiA\n",
      "PROTEIN ID: 1nlyA\n",
      "PROTEIN ID: 1nnfA\n",
      "PROTEIN ID: 1noxA\n",
      "PROTEIN ID: 1nqeA\n",
      "PROTEIN ID: 1nw5A\n",
      "PROTEIN ID: 1nw6A\n",
      "PROTEIN ID: 1nw7A\n",
      "PROTEIN ID: 1nw8A\n",
      "PROTEIN ID: 1nwzA\n",
      "PROTEIN ID: 1ny5A\n",
      "PROTEIN ID: 1ny6A\n",
      "PROTEIN ID: 1o1hA\n",
      "PROTEIN ID: 1o2dA\n",
      "PROTEIN ID: 1o61A\n",
      "PROTEIN ID: 1o69A\n",
      "PROTEIN ID: 1o7lA\n",
      "PROTEIN ID: 1oadA\n",
      "PROTEIN ID: 1oapA\n",
      "PROTEIN ID: 1oddA\n",
      "PROTEIN ID: 1odvA\n",
      "PROTEIN ID: 1oj7A\n",
      "PROTEIN ID: 1oltA\n",
      "PROTEIN ID: 1opcA\n",
      "PROTEIN ID: 1opxA\n",
      "PROTEIN ID: 1or7A\n",
      "PROTEIN ID: 1ot6A\n",
      "PROTEIN ID: 1ot9A\n",
      "PROTEIN ID: 1otaA\n",
      "PROTEIN ID: 1otbA\n",
      "PROTEIN ID: 1oxkA\n",
      "PROTEIN ID: 1p2fA\n",
      "PROTEIN ID: 1p31A\n",
      "PROTEIN ID: 1p3dA\n",
      "PROTEIN ID: 1p7dA\n",
      "PROTEIN ID: 1p9rA\n",
      "PROTEIN ID: 1p9wA\n",
      "PROTEIN ID: 1pb0A\n",
      "PROTEIN ID: 1pb7A\n",
      "PROTEIN ID: 1pb8A\n",
      "PROTEIN ID: 1pjrA\n",
      "PROTEIN ID: 1pnzA\n",
      "PROTEIN ID: 1po0A\n",
      "PROTEIN ID: 1pt7A\n",
      "PROTEIN ID: 1pvpA\n",
      "PROTEIN ID: 1q05A\n",
      "PROTEIN ID: 1q06A\n",
      "PROTEIN ID: 1q07A\n",
      "PROTEIN ID: 1q08A\n",
      "PROTEIN ID: 1q09A\n",
      "PROTEIN ID: 1q0aA\n",
      "PROTEIN ID: 1q35A\n",
      "PROTEIN ID: 1q7eA\n",
      "PROTEIN ID: 1qg8A\n",
      "PROTEIN ID: 1qgqA\n",
      "PROTEIN ID: 1qgsA\n",
      "PROTEIN ID: 1qhgA\n",
      "PROTEIN ID: 1qhhA\n",
      "PROTEIN ID: 1qksA\n",
      "PROTEIN ID: 1qpzA\n",
      "PROTEIN ID: 1qsaA\n",
      "PROTEIN ID: 1qteA\n",
      "PROTEIN ID: 1qtwA\n",
      "PROTEIN ID: 1qu7A\n",
      "PROTEIN ID: 1qwyA\n",
      "PROTEIN ID: 1qxxA\n",
      "PROTEIN ID: 1r1mA\n",
      "PROTEIN ID: 1r1tA\n",
      "PROTEIN ID: 1r1uA\n",
      "PROTEIN ID: 1r23A\n",
      "PROTEIN ID: 1r62A\n",
      "PROTEIN ID: 1r8dA\n",
      "PROTEIN ID: 1r8eA\n",
      "PROTEIN ID: 1r9xA\n",
      "PROTEIN ID: 1r9yA\n",
      "PROTEIN ID: 1r9zA\n",
      "PROTEIN ID: 1ra0A\n",
      "PROTEIN ID: 1ra5A\n",
      "PROTEIN ID: 1rakA\n",
      "PROTEIN ID: 1reqA\n",
      "PROTEIN ID: 1rhcA\n",
      "PROTEIN ID: 1rioA\n",
      "PROTEIN ID: 1rk6A\n",
      "PROTEIN ID: 1rp3A\n",
      "PROTEIN ID: 1rrmA\n",
      "PROTEIN ID: 1rttA\n",
      "PROTEIN ID: 1rzuA\n",
      "PROTEIN ID: 1rzvA\n",
      "PROTEIN ID: 1s5mA\n",
      "PROTEIN ID: 1s5nA\n",
      "PROTEIN ID: 1s8nA\n",
      "PROTEIN ID: 1sfxA\n",
      "PROTEIN ID: 1sg0A\n",
      "PROTEIN ID: 1si0A\n",
      "PROTEIN ID: 1sigA\n",
      "PROTEIN ID: 1slyA\n",
      "PROTEIN ID: 1sqeA\n",
      "PROTEIN ID: 1sqsA\n",
      "PROTEIN ID: 1sumB\n",
      "PROTEIN ID: 1suuA\n",
      "PROTEIN ID: 1t3tA\n",
      "PROTEIN ID: 1t5bA\n",
      "PROTEIN ID: 1t72A\n",
      "PROTEIN ID: 1ta9A\n",
      "PROTEIN ID: 1td5A\n",
      "PROTEIN ID: 1tf1A\n",
      "PROTEIN ID: 1tqgA\n",
      "PROTEIN ID: 1tqqA\n",
      "PROTEIN ID: 1tv8A\n",
      "PROTEIN ID: 1tvlA\n",
      "PROTEIN ID: 1tzbA\n",
      "PROTEIN ID: 1tzcA\n",
      "PROTEIN ID: 1u07A\n",
      "PROTEIN ID: 1u2wA\n",
      "PROTEIN ID: 1u8bA\n",
      "PROTEIN ID: 1u8tA\n",
      "PROTEIN ID: 1uaaA\n",
      "PROTEIN ID: 1uc8A\n",
      "PROTEIN ID: 1uc9A\n",
      "PROTEIN ID: 1us4A\n",
      "PROTEIN ID: 1us5A\n",
      "PROTEIN ID: 1uscA\n",
      "PROTEIN ID: 1usfA\n",
      "PROTEIN ID: 1uuxA\n",
      "PROTEIN ID: 1uuyA\n",
      "PROTEIN ID: 1uylA\n",
      "PROTEIN ID: 1v4yA\n",
      "PROTEIN ID: 1v51A\n",
      "PROTEIN ID: 1v8pA\n",
      "PROTEIN ID: 1v96A\n",
      "PROTEIN ID: 1vctA\n",
      "PROTEIN ID: 1ve2A\n",
      "PROTEIN ID: 1vf7A\n",
      "PROTEIN ID: 1vgtA\n",
      "PROTEIN ID: 1vgwA\n",
      "PROTEIN ID: 1vhdA\n",
      "PROTEIN ID: 1vhvA\n",
      "PROTEIN ID: 1vimA\n",
      "PROTEIN ID: 1vj7A\n",
      "PROTEIN ID: 1vkeA\n",
      "PROTEIN ID: 1vljA\n",
      "PROTEIN ID: 1vz0A\n",
      "PROTEIN ID: 1w55A\n",
      "PROTEIN ID: 1w6sA\n",
      "PROTEIN ID: 1w77A\n",
      "PROTEIN ID: 1w78A\n",
      "PROTEIN ID: 1w8iA\n",
      "PROTEIN ID: 1wetA\n",
      "PROTEIN ID: 1wmiA\n",
      "PROTEIN ID: 1woqA\n",
      "PROTEIN ID: 1wp1A\n",
      "PROTEIN ID: 1wpmA\n",
      "PROTEIN ID: 1wpnA\n",
      "PROTEIN ID: 1wppA\n",
      "PROTEIN ID: 1ws6A\n",
      "PROTEIN ID: 1x74A\n",
      "PROTEIN ID: 1x9hA\n",
      "PROTEIN ID: 1x9iA\n",
      "PROTEIN ID: 1xa3A\n",
      "PROTEIN ID: 1xc3A\n",
      "PROTEIN ID: 1xd7A\n",
      "PROTEIN ID: 1xi2A\n",
      "PROTEIN ID: 1xjaA\n",
      "PROTEIN ID: 1xk6A\n",
      "PROTEIN ID: 1xk7A\n",
      "PROTEIN ID: 1xkwA\n",
      "PROTEIN ID: 1xkzA\n",
      "PROTEIN ID: 1xmaA\n",
      "PROTEIN ID: 1xo0A\n",
      "PROTEIN ID: 1xocA\n",
      "PROTEIN ID: 1xw3A\n",
      "PROTEIN ID: 1y0hA\n",
      "PROTEIN ID: 1y1zA\n",
      "Importing the PDB file took 0.000726171334584554 min\n",
      "{'UNP': {'A': [('P35439', 'NMDZ1_RAT', 394, 544), ('P35439', 'NMZ1_RAT', 663, 800)]}}\n",
      "PROTEIN ID: 1y20A\n",
      "Importing the PDB file took 0.0014162302017211915 min\n",
      "{'UNP': {'A': [('P35439', 'NMDZ1_RAT', 394, 544), ('P35439', 'NMZ1_RAT', 663, 800)]}}\n",
      "PROTEIN ID: 1y7mA\n",
      "PROTEIN ID: 1y7yA\n",
      "PROTEIN ID: 1y80A\n",
      "PROTEIN ID: 1y82A\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/Seq.py:175: BiopythonWarning: Biopython Seq objects now use string comparison. Older versions of Biopython used object comparison. During this transition, please use hash(id(my_seq)) or my_dict[id(my_seq)] if you want the old behaviour, or use hash(str(my_seq)) or my_dict[str(my_seq)] for the new string hashing behaviour.\n",
      "  \"the new string hashing behaviour.\", BiopythonWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PROTEIN ID: 1y9uA\n",
      "PROTEIN ID: 1yc9A\n",
      "PROTEIN ID: 1ydxA\n",
      "PROTEIN ID: 1ye5A\n",
      "PROTEIN ID: 1yf2A\n",
      "PROTEIN ID: 1yg2A\n",
      "PROTEIN ID: 1yioA\n",
      "PROTEIN ID: 1yiqA\n",
      "PROTEIN ID: 1ylfA\n",
      "PROTEIN ID: 1yoyA\n",
      "PROTEIN ID: 1yspA\n",
      "PROTEIN ID: 1ysqA\n",
      "PROTEIN ID: 1yviA\n",
      "PROTEIN ID: 1z05A\n",
      "PROTEIN ID: 1z19A\n",
      "PROTEIN ID: 1z7uA\n",
      "PROTEIN ID: 1zatA\n",
      "PROTEIN ID: 1zi0A\n",
      "PROTEIN ID: 1zljA\n",
      "PROTEIN ID: 1zvtA\n",
      "PROTEIN ID: 1zvuA\n",
      "PROTEIN ID: 1zzcA\n",
      "PROTEIN ID: 2a0bA\n",
      "PROTEIN ID: 2a3nA\n",
      "PROTEIN ID: 2a5hA\n",
      "PROTEIN ID: 2a5lA\n",
      "PROTEIN ID: 2a61A\n",
      "PROTEIN ID: 2aa4A\n",
      "PROTEIN ID: 2aacA\n",
      "PROTEIN ID: 2ad6A\n",
      "PROTEIN ID: 2ad7A\n",
      "PROTEIN ID: 2ad8A\n",
      "PROTEIN ID: 2aefA\n",
      "PROTEIN ID: 2aejA\n",
      "PROTEIN ID: 2afhA\n",
      "PROTEIN ID: 2am1A\n",
      "PROTEIN ID: 2anuA\n",
      "PROTEIN ID: 2ap1A\n",
      "PROTEIN ID: 2ar0A\n",
      "PROTEIN ID: 2araA\n",
      "PROTEIN ID: 2arcA\n",
      "PROTEIN ID: 2aznA\n",
      "PROTEIN ID: 2b02A\n",
      "PROTEIN ID: 2b0pA\n",
      "PROTEIN ID: 2b13A\n",
      "PROTEIN ID: 2b3zA\n",
      "PROTEIN ID: 2b44A\n",
      "PROTEIN ID: 2basA\n",
      "PROTEIN ID: 2bfwA\n",
      "PROTEIN ID: 2bknA\n",
      "PROTEIN ID: 2bkoA\n",
      "PROTEIN ID: 2bkpA\n",
      "PROTEIN ID: 2bm4A\n",
      "PROTEIN ID: 2bm5A\n",
      "PROTEIN ID: 2bm6A\n",
      "PROTEIN ID: 2bm7A\n",
      "PROTEIN ID: 2bnmA\n",
      "PROTEIN ID: 2brcA\n",
      "PROTEIN ID: 2byiA\n",
      "PROTEIN ID: 2c2aA\n",
      "PROTEIN ID: 2c81A\n",
      "PROTEIN ID: 2ce0A\n",
      "PROTEIN ID: 2cg4A\n",
      "PROTEIN ID: 2ch7A\n",
      "PROTEIN ID: 2cviA\n",
      "PROTEIN ID: 2cwqA\n",
      "PROTEIN ID: 2cyyA\n",
      "PROTEIN ID: 2d1hA\n",
      "PROTEIN ID: 2d1vA\n",
      "PROTEIN ID: 2d5mA\n",
      "PROTEIN ID: 2d5nA\n",
      "PROTEIN ID: 2d5wA\n",
      "PROTEIN ID: 2dbbA\n",
      "PROTEIN ID: 2dekA\n",
      "PROTEIN ID: 2df8A\n",
      "PROTEIN ID: 2dg6A\n",
      "PROTEIN ID: 2di3A\n",
      "PROTEIN ID: 2dqlA\n",
      "PROTEIN ID: 2dvzA\n",
      "PROTEIN ID: 2dxwA\n",
      "PROTEIN ID: 2dxxA\n",
      "PROTEIN ID: 2e15A\n",
      "PROTEIN ID: 2e1nA\n",
      "PROTEIN ID: 2e4nA\n",
      "PROTEIN ID: 2e5fA\n",
      "PROTEIN ID: 2e7wA\n",
      "PROTEIN ID: 2e7xA\n",
      "PROTEIN ID: 2e7zA\n",
      "PROTEIN ID: 2eb7A\n",
      "PROTEIN ID: 2ecuA\n",
      "PROTEIN ID: 2efnA\n",
      "PROTEIN ID: 2eh3A\n",
      "PROTEIN ID: 2ehlA\n",
      "PROTEIN ID: 2ehzA\n",
      "PROTEIN ID: 2ek5A\n",
      "PROTEIN ID: 2eshA\n",
      "PROTEIN ID: 2esnA\n",
      "PROTEIN ID: 2esrA\n",
      "PROTEIN ID: 2ewnA\n",
      "PROTEIN ID: 2ewvA\n",
      "PROTEIN ID: 2eyuA\n",
      "PROTEIN ID: 2f00A\n",
      "PROTEIN ID: 2f2eA\n",
      "PROTEIN ID: 2f5xA\n",
      "PROTEIN ID: 2f6gA\n",
      "PROTEIN ID: 2f6pA\n",
      "PROTEIN ID: 2f7aA\n",
      "PROTEIN ID: 2f7bA\n",
      "PROTEIN ID: 2f8lA\n",
      "PROTEIN ID: 2f9fA\n",
      "PROTEIN ID: 2fa1A\n",
      "PROTEIN ID: 2fa5A\n",
      "PROTEIN ID: 2fb2A\n",
      "PROTEIN ID: 2fbhA\n",
      "PROTEIN ID: 2fcjA\n",
      "PROTEIN ID: 2fdnA\n",
      "PROTEIN ID: 2fe1A\n",
      "PROTEIN ID: 2fezA\n",
      "PROTEIN ID: 2ff4A\n",
      "PROTEIN ID: 2ffuA\n",
      "PROTEIN ID: 2fhpA\n",
      "PROTEIN ID: 2fn9A\n",
      "PROTEIN ID: 2fnuA\n",
      "PROTEIN ID: 2fpoA\n",
      "PROTEIN ID: 2fswA\n",
      "PROTEIN ID: 2fvyA\n",
      "PROTEIN ID: 2fw0A\n",
      "PROTEIN ID: 2g2cA\n",
      "PROTEIN ID: 2g6vA\n",
      "PROTEIN ID: 2g7uA\n",
      "PROTEIN ID: 2gaiA\n",
      "PROTEIN ID: 2gajA\n",
      "PROTEIN ID: 2gciA\n",
      "PROTEIN ID: 2gd0A\n",
      "PROTEIN ID: 2gd2A\n",
      "PROTEIN ID: 2gd9A\n",
      "PROTEIN ID: 2gj3A\n",
      "PROTEIN ID: 2gjgA\n",
      "PROTEIN ID: 2gkgA\n",
      "PROTEIN ID: 2glkA\n",
      "PROTEIN ID: 2gm5A\n",
      "PROTEIN ID: 2gmsA\n",
      "PROTEIN ID: 2gmyA\n",
      "PROTEIN ID: 2gqpA\n",
      "PROTEIN ID: 2gskA\n",
      "PROTEIN ID: 2gu1A\n",
      "PROTEIN ID: 2gufA\n",
      "PROTEIN ID: 2guhA\n",
      "PROTEIN ID: 2gupA\n",
      "PROTEIN ID: 2gxgA\n",
      "PROTEIN ID: 2gzaA\n",
      "PROTEIN ID: 2h1cA\n",
      "PROTEIN ID: 2h98A\n",
      "PROTEIN ID: 2h99A\n",
      "PROTEIN ID: 2h9bA\n",
      "PROTEIN ID: 2hawA\n",
      "PROTEIN ID: 2hekA\n",
      "PROTEIN ID: 2heuA\n",
      "PROTEIN ID: 2hklA\n",
      "PROTEIN ID: 2hmtA\n",
      "PROTEIN ID: 2hmuA\n",
      "PROTEIN ID: 2hmvA\n",
      "PROTEIN ID: 2hnhA\n",
      "PROTEIN ID: 2hoeA\n",
      "PROTEIN ID: 2hofA\n",
      "PROTEIN ID: 2hphA\n",
      "PROTEIN ID: 2hq0A\n",
      "PROTEIN ID: 2hqsA\n",
      "PROTEIN ID: 2hs5A\n",
      "PROTEIN ID: 2hsgA\n",
      "PROTEIN ID: 2hsiA\n",
      "PROTEIN ID: 2hwvA\n",
      "PROTEIN ID: 2hxvA\n",
      "PROTEIN ID: 2i0mA\n",
      "PROTEIN ID: 2i5rA\n",
      "PROTEIN ID: 2ia2A\n",
      "PROTEIN ID: 2ia4A\n",
      "PROTEIN ID: 2ibdA\n",
      "PROTEIN ID: 2ictA\n",
      "PROTEIN ID: 2iftA\n",
      "PROTEIN ID: 2ikkA\n",
      "PROTEIN ID: 2iplA\n",
      "PROTEIN ID: 2ipmA\n",
      "PROTEIN ID: 2ipnA\n",
      "PROTEIN ID: 2is1A\n",
      "PROTEIN ID: 2is2A\n",
      "PROTEIN ID: 2is4A\n",
      "PROTEIN ID: 2is6A\n",
      "PROTEIN ID: 2is8A\n",
      "PROTEIN ID: 2iu5A\n",
      "PROTEIN ID: 2iuyA\n",
      "PROTEIN ID: 2iv7A\n",
      "PROTEIN ID: 2iw1A\n",
      "PROTEIN ID: 2iw4A\n",
      "PROTEIN ID: 2iwxA\n",
      "PROTEIN ID: 2jbaA\n",
      "PROTEIN ID: 2jcgA\n",
      "PROTEIN ID: 2jfgA\n",
      "PROTEIN ID: 2nipA\n",
      "PROTEIN ID: 2npnA\n",
      "PROTEIN ID: 2nq2A\n",
      "PROTEIN ID: 2nq9A\n",
      "PROTEIN ID: 2nqhA\n",
      "PROTEIN ID: 2nt3A\n",
      "PROTEIN ID: 2nt4A\n",
      "PROTEIN ID: 2o08A\n",
      "PROTEIN ID: 2o0yA\n",
      "PROTEIN ID: 2o3jA\n",
      "PROTEIN ID: 2o4dA\n",
      "PROTEIN ID: 2o7iA\n",
      "PROTEIN ID: 2o7pA\n",
      "PROTEIN ID: 2o8xA\n",
      "PROTEIN ID: 2o99A\n",
      "PROTEIN ID: 2o9aA\n",
      "PROTEIN ID: 2obcA\n",
      "PROTEIN ID: 2ofyA\n",
      "PROTEIN ID: 2ogiA\n",
      "PROTEIN ID: 2ojhA\n",
      "PROTEIN ID: 2okcA\n",
      "PROTEIN ID: 2olbA\n",
      "PROTEIN ID: 2oocA\n",
      "PROTEIN ID: 2oqgA\n",
      "PROTEIN ID: 2oqrA\n",
      "PROTEIN ID: 2oxoA\n",
      "PROTEIN ID: 2oyoA\n",
      "PROTEIN ID: 2p19A\n",
      "PROTEIN ID: 2p4gA\n",
      "PROTEIN ID: 2p5vA\n",
      "PROTEIN ID: 2p7oA\n",
      "PROTEIN ID: 2paqA\n",
      "PROTEIN ID: 2pbqA\n",
      "PROTEIN ID: 2pfxA\n",
      "PROTEIN ID: 2ph1A\n",
      "PROTEIN ID: 2pjrA\n",
      "PROTEIN ID: 2pkhA\n",
      "PROTEIN ID: 2pmhA\n",
      "PROTEIN ID: 2pn6A\n",
      "PROTEIN ID: 2pq7A\n",
      "PROTEIN ID: 2pt7A\n",
      "PROTEIN ID: 2pucA\n",
      "PROTEIN ID: 2pudA\n",
      "PROTEIN ID: 2px7A\n",
      "PROTEIN ID: 2q0oA\n",
      "PROTEIN ID: 2q0tA\n",
      "PROTEIN ID: 2q1zA\n",
      "PROTEIN ID: 2q4fA\n",
      "PROTEIN ID: 2q8pA\n",
      "PROTEIN ID: 2qb6A\n",
      "PROTEIN ID: 2qb7A\n",
      "PROTEIN ID: 2qb8A\n",
      "PROTEIN ID: 2qczA\n",
      "PROTEIN ID: 2qdfA\n",
      "PROTEIN ID: 2qdlA\n",
      "PROTEIN ID: 2qeuA\n",
      "PROTEIN ID: 2qgqA\n",
      "PROTEIN ID: 2qgzA\n",
      "PROTEIN ID: 2qi9A\n",
      "PROTEIN ID: 2qj7A\n",
      "PROTEIN ID: 2qm1A\n",
      "PROTEIN ID: 2qmoA\n",
      "PROTEIN ID: 2qpqA\n",
      "PROTEIN ID: 2qsxA\n",
      "PROTEIN ID: 2qwxA\n",
      "PROTEIN ID: 2qx4A\n",
      "PROTEIN ID: 2qx6A\n",
      "PROTEIN ID: 2qx8A\n",
      "PROTEIN ID: 2r01A\n",
      "PROTEIN ID: 2r0xA\n",
      "PROTEIN ID: 2r1jA\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 2187.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 2272.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain L is discontinuous at line 2346.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain R is discontinuous at line 2424.\n",
      "  PDBConstructionWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Importing the PDB file took 0.0008391976356506347 min\n",
      "{'UNP': {'L': [('P69202', 'RPC2_BPP22', 1, 68)], 'R': [('P69202', 'RPC2_BPP22', 1, 68)]}}\n",
      "PROTEIN ID: 2r25A\n",
      "PROTEIN ID: 2r4tA\n",
      "PROTEIN ID: 2r6gA\n",
      "PROTEIN ID: 2r6oA\n",
      "PROTEIN ID: 2r6vA\n",
      "PROTEIN ID: 2ra5A\n",
      "PROTEIN ID: 2rb9A\n",
      "PROTEIN ID: 2rc7A\n",
      "PROTEIN ID: 2rc8A\n",
      "PROTEIN ID: 2rcaA\n",
      "PROTEIN ID: 2rdeA\n",
      "PROTEIN ID: 2riiA\n",
      "PROTEIN ID: 2rilA\n",
      "PROTEIN ID: 2rslA\n",
      "PROTEIN ID: 2uagA\n",
      "PROTEIN ID: 2v25A\n",
      "PROTEIN ID: 2v2kA\n",
      "PROTEIN ID: 2v9yA\n",
      "PROTEIN ID: 2vhaA\n",
      "PROTEIN ID: 2vjqA\n",
      "PROTEIN ID: 2vk2A\n",
      "PROTEIN ID: 2vkeA\n",
      "PROTEIN ID: 2vkrA\n",
      "PROTEIN ID: 2vlgA\n",
      "PROTEIN ID: 2vmaA\n",
      "PROTEIN ID: 2vmbA\n",
      "PROTEIN ID: 2vpzA\n",
      "PROTEIN ID: 2vshA\n",
      "PROTEIN ID: 2w27A\n",
      "PROTEIN ID: 2w8bA\n",
      "PROTEIN ID: 2w8iA\n",
      "PROTEIN ID: 2yveA\n",
      "PROTEIN ID: 2yx0A\n",
      "PROTEIN ID: 2yxbA\n",
      "PROTEIN ID: 2yxoA\n",
      "PROTEIN ID: 2yxzA\n",
      "PROTEIN ID: 2yyeA\n",
      "PROTEIN ID: 2yz5A\n",
      "PROTEIN ID: 2z1eA\n",
      "PROTEIN ID: 2z1fA\n",
      "PROTEIN ID: 2z1uA\n",
      "PROTEIN ID: 2z2lA\n",
      "PROTEIN ID: 2z2mA\n",
      "PROTEIN ID: 2z4gA\n",
      "PROTEIN ID: 2z4pA\n",
      "PROTEIN ID: 2z6rA\n",
      "PROTEIN ID: 2z8xA\n",
      "PROTEIN ID: 2z98A\n",
      "PROTEIN ID: 2z9bA\n",
      "PROTEIN ID: 2zauA\n",
      "PROTEIN ID: 2zbcA\n",
      "PROTEIN ID: 2zc3A\n",
      "PROTEIN ID: 2zc4A\n",
      "PROTEIN ID: 2zcmA\n",
      "PROTEIN ID: 2zdpA\n",
      "PROTEIN ID: 2zf8A\n",
      "PROTEIN ID: 2zieA\n",
      "PROTEIN ID: 2zifA\n",
      "PROTEIN ID: 2zigA\n",
      "PROTEIN ID: 2zkiA\n",
      "PROTEIN ID: 2zkzA\n",
      "PROTEIN ID: 2zodA\n",
      "PROTEIN ID: 2zovA\n",
      "PROTEIN ID: 2zxjA\n",
      "PROTEIN ID: 3b4yA\n",
      "PROTEIN ID: 3b6iA\n",
      "PROTEIN ID: 3b8xA\n",
      "PROTEIN ID: 3b9oA\n",
      "PROTEIN ID: 3bcvA\n",
      "PROTEIN ID: 3be6A\n",
      "PROTEIN ID: 3bemA\n",
      "PROTEIN ID: 3bg2A\n",
      "PROTEIN ID: 3bhqA\n",
      "PROTEIN ID: 3bkhA\n",
      "PROTEIN ID: 3bkvA\n",
      "PROTEIN ID: 3bm7A\n",
      "PROTEIN ID: 3bpkA\n",
      "PROTEIN ID: 3bpqA\n",
      "PROTEIN ID: 3bpvA\n",
      "PROTEIN ID: 3bqxA\n",
      "PROTEIN ID: 3breA\n",
      "PROTEIN ID: 3bs3A\n",
      "PROTEIN ID: 3bvpA\n",
      "PROTEIN ID: 3bwgA\n",
      "PROTEIN ID: 3c1qA\n",
      "PROTEIN ID: 3c29A\n",
      "PROTEIN ID: 3c3wA\n",
      "PROTEIN ID: 3c48A\n",
      "PROTEIN ID: 3c57A\n",
      "PROTEIN ID: 3c7jA\n",
      "PROTEIN ID: 3c85A\n",
      "PROTEIN ID: 3c8fA\n",
      "PROTEIN ID: 3c8nA\n",
      "PROTEIN ID: 3c9uA\n",
      "PROTEIN ID: 3canA\n",
      "PROTEIN ID: 3ccgA\n",
      "PROTEIN ID: 3cijA\n",
      "PROTEIN ID: 3cixA\n",
      "PROTEIN ID: 3ckjA\n",
      "PROTEIN ID: 3cknA\n",
      "PROTEIN ID: 3ckvA\n",
      "PROTEIN ID: 3cloA\n",
      "PROTEIN ID: 3cnrA\n",
      "PROTEIN ID: 3cnvA\n",
      "PROTEIN ID: 3cp5A\n",
      "PROTEIN ID: 3ctpA\n",
      "PROTEIN ID: 3cuoA\n",
      "PROTEIN ID: 3cwrA\n",
      "PROTEIN ID: 3cx4A\n",
      "PROTEIN ID: 3cyiA\n",
      "PROTEIN ID: 3cypB\n",
      "PROTEIN ID: 3cypC\n",
      "PROTEIN ID: 3cypD\n",
      "PROTEIN ID: 3cyqA\n",
      "PROTEIN ID: 3d5kA\n",
      "PROTEIN ID: 3d6zA\n",
      "PROTEIN ID: 3d7iA\n",
      "PROTEIN ID: 3dboA\n",
      "PROTEIN ID: 3df7A\n",
      "PROTEIN ID: 3df8A\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/Seq.py:175: BiopythonWarning: Biopython Seq objects now use string comparison. Older versions of Biopython used object comparison. During this transition, please use hash(id(my_seq)) or my_dict[id(my_seq)] if you want the old behaviour, or use hash(str(my_seq)) or my_dict[str(my_seq)] for the new string hashing behaviour.\n",
      "  \"the new string hashing behaviour.\", BiopythonWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Desired structure doesn't exists\n",
      "PROTEIN ID: 3dmaA\n",
      "PROTEIN ID: 3dr4A\n",
      "PROTEIN ID: 3drfA\n",
      "PROTEIN ID: 3drjA\n",
      "PROTEIN ID: 3dsgA\n",
      "PROTEIN ID: 3du1X\n",
      "Importing the PDB file took 0.0007538636525472005 min\n",
      "{'UNP': {'X': [('Q8YQS9', 'Q8YQS9_ANASP', 1, 237)]}}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/Seq.py:175: BiopythonWarning: Biopython Seq objects now use string comparison. Older versions of Biopython used object comparison. During this transition, please use hash(id(my_seq)) or my_dict[id(my_seq)] if you want the old behaviour, or use hash(str(my_seq)) or my_dict[str(my_seq)] for the new string hashing behaviour.\n",
      "  \"the new string hashing behaviour.\", BiopythonWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 6270.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 6326.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 6381.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 6471.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/Seq.py:175: BiopythonWarning: Biopython Seq objects now use string comparison. Older versions of Biopython used object comparison. During this transition, please use hash(id(my_seq)) or my_dict[id(my_seq)] if you want the old behaviour, or use hash(str(my_seq)) or my_dict[str(my_seq)] for the new string hashing behaviour.\n",
      "  \"the new string hashing behaviour.\", BiopythonWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PROTEIN ID: 3e10A\n",
      "PROTEIN ID: 3e38A\n",
      "PROTEIN ID: 3e4rA\n",
      "PROTEIN ID: 3e4vA\n",
      "PROTEIN ID: 3e7lA\n",
      "PROTEIN ID: 3e8oA\n",
      "PROTEIN ID: 3eagA\n",
      "PROTEIN ID: 3ec2A\n",
      "PROTEIN ID: 3eccA\n",
      "PROTEIN ID: 3echA\n",
      "PROTEIN ID: 3ecpA\n",
      "PROTEIN ID: 3edpA\n",
      "PROTEIN ID: 3eetA\n",
      "PROTEIN ID: 3efmA\n",
      "PROTEIN ID: 3eiwA\n",
      "PROTEIN ID: 3eixA\n",
      "PROTEIN ID: 3ekoA\n",
      "PROTEIN ID: 3elkA\n",
      "PROTEIN ID: 3eusA\n",
      "PROTEIN ID: 3ex8A\n",
      "PROTEIN ID: 3eywA\n",
      "Importing the PDB file took 0.002374744415283203 min\n",
      "{'UNP': {'A': [('P03819', 'KEFC_ECOLI', 401, 620), ('P0A754', 'KEFF_ECOLI', 1, 176)], 'B': [('P03819', 'KEFC_ECOLI', 401, 620), ('P0A754', 'KEFF_ECOLI', 1, 176)]}}\n",
      "PROTEIN ID: 3ezuA\n",
      "PROTEIN ID: 3f1cA\n",
      "PROTEIN ID: 3f1nA\n",
      "PROTEIN ID: 3f1oA\n",
      "PROTEIN ID: 3f1pA\n",
      "PROTEIN ID: 3f2bA\n",
      "PROTEIN ID: 3f44A\n",
      "PROTEIN ID: 3f52A\n",
      "PROTEIN ID: 3f6cA\n",
      "PROTEIN ID: 3f6oA\n",
      "PROTEIN ID: 3f6vA\n",
      "PROTEIN ID: 3f8bA\n",
      "PROTEIN ID: 3f8cA\n",
      "PROTEIN ID: 3f8fA\n",
      "PROTEIN ID: 3fd3A\n",
      "PROTEIN ID: 3fgvA\n",
      "PROTEIN ID: 3fisA\n",
      "PROTEIN ID: 3fmsA\n",
      "PROTEIN ID: 3fwyA\n",
      "PROTEIN ID: 3fwzA\n",
      "PROTEIN ID: 3fxaA\n",
      "PROTEIN ID: 3fzvA\n",
      "PROTEIN ID: 3g13A\n",
      "PROTEIN ID: 3g5oA\n",
      "PROTEIN ID: 3g7rA\n",
      "PROTEIN ID: 3gdiA\n",
      "PROTEIN ID: 3gfaA\n",
      "PROTEIN ID: 3gfvA\n",
      "PROTEIN ID: 3gfxA\n",
      "PROTEIN ID: 3gfyA\n",
      "PROTEIN ID: 3gfzA\n",
      "PROTEIN ID: 3gg0A\n",
      "PROTEIN ID: 3gg1A\n",
      "PROTEIN ID: 3gg2A\n",
      "PROTEIN ID: 3ghjA\n",
      "PROTEIN ID: 3gp4A\n",
      "PROTEIN ID: 3gpvA\n",
      "PROTEIN ID: 3gr3A\n",
      "PROTEIN ID: 3guvA\n",
      "PROTEIN ID: 3h4oA\n",
      "PROTEIN ID: 3h5tA\n",
      "PROTEIN ID: 3h87A\n",
      "PROTEIN ID: 3hfiA\n",
      "PROTEIN ID: 3hh0A\n",
      "PROTEIN ID: 3hhhA\n",
      "PROTEIN ID: 3hl0A\n",
      "PROTEIN ID: 3hmzA\n",
      "PROTEIN ID: 3hn7A\n",
      "PROTEIN ID: 3hoiA\n",
      "PROTEIN ID: 3htvA\n",
      "PROTEIN ID: 3hvwA\n",
      "PROTEIN ID: 3pypA\n",
      "PROTEIN ID: 3uagA\n",
      "PROTEIN ID: 4aahA\n",
      "PROTEIN ID: 4crxA\n",
      "PROTEIN ID: 4reqA\n",
      "PROTEIN ID: 4uagA\n",
      "PROTEIN ID: 5reqA\n",
      "PROTEIN ID: 6reqA\n",
      "PROTEIN ID: 7reqA\n",
      "PROTEIN ID: 8abpA\n",
      "CLOSED PROCESSING POOL\n",
      "WROTE OUT ALL SEQS\n",
      "BLASTing query sequences\n",
      "/home/daniel/lib/ncbi-blast-2.9.0+-src/c++/ReleaseMT/bin/blastp -out /media/daniel/ExtraDrive1/Results/Newest_Covariation/Input/Large_Test_Set/BLAST/LargeDataSet_All_Seqs.xml -outfmt 5 -query /media/daniel/ExtraDrive1/Results/Newest_Covariation/Input/Large_Test_Set/Sequences/LargeDataSet.fasta -db /media/daniel/ExtraDrive1/blast_databases/uniref90_05122020/custom_uniref90_05122020.fasta -evalue 0.05 -max_target_seqs 20000 -num_threads 10\n"
     ]
    },
    {
     "ename": "ApplicationError",
     "evalue": "Non-zero return code 137 from '/home/daniel/lib/ncbi-blast-2.9.0+-src/c++/ReleaseMT/bin/blastp -out /media/daniel/ExtraDrive1/Results/Newest_Covariation/Input/Large_Test_Set/BLAST/LargeDataSet_All_Seqs.xml -outfmt 5 -query /media/daniel/ExtraDrive1/Results/Newest_Covariation/Input/Large_Test_Set/Sequences/LargeDataSet.fasta -db /media/daniel/ExtraDrive1/blast_databases/uniref90_05122020/custom_uniref90_05122020.fasta -evalue 0.05 -max_target_seqs 20000 -num_threads 10', message 'Killed'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mApplicationError\u001b[0m                          Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-2-f48dfbfc2ab3>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     95\u001b[0m                                                 \u001b[0mmax_target_seqs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m20000\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0me_value_threshold\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0.05\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mremote\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     96\u001b[0m                                                 \u001b[0mmin_fraction\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0.7\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmin_identity\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0.40\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmax_identity\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0.98\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 97\u001b[0;31m                                                 msf=True, fasta=True, sources=['UNP', 'GB', 'PDB'], verbose=False)\n\u001b[0m\u001b[1;32m     98\u001b[0m \u001b[0msummary\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Chain'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msummary\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Protein_ID'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mgenerator\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mprotein_data\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Chain'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     99\u001b[0m \u001b[0msummary\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Accession'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msummary\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Protein_ID'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mgenerator\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mprotein_data\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Accession'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Documents/git/ETMIP/src/SupportingClasses/DataSetGenerator.py\u001b[0m in \u001b[0;36mbuild_pdb_alignment_dataset\u001b[0;34m(self, protein_list_fn, processes, max_target_seqs, e_value_threshold, database, remote, min_fraction, min_identity, max_identity, msf, fasta, sources, verbose)\u001b[0m\n\u001b[1;32m    251\u001b[0m         \u001b[0;31m# # pool2 = Pool(1, initializer=init_filtering_and_alignment_pool,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    252\u001b[0m         \u001b[0;31m#              initargs=(max_target_seqs, e_value_threshold, database, remote, min_fraction, min_identity,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 253\u001b[0;31m         \u001b[0;31m#                        max_identity, msf, fasta, blast_fn, self.filtered_blast_path, self.alignment_path,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    254\u001b[0m         \u001b[0;31m#                        self.filtered_alignment_path, self.final_alignment_path, verbose))\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    255\u001b[0m         \u001b[0;31m# res = pool2.map_async(filtering_and_alignment, [(p_id, self.protein_data[p_id]['Sequence'])\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Documents/git/ETMIP/src/SupportingClasses/DataSetGenerator.py\u001b[0m in \u001b[0;36mblast_query_sequence\u001b[0;34m(protein_id, blast_path, sequence_fn, evalue, processes, max_target_seqs, database, remote, verbose)\u001b[0m\n\u001b[1;32m    569\u001b[0m     \u001b[0;32mwith\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mblast_fn\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'r'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mblast_handle\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    570\u001b[0m         \u001b[0mpdb_id_pattern\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcompile\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mr'^([0-9][a-z0-9]{3}).*$'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 571\u001b[0;31m         \u001b[0mblast_iter\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mNCBIXML\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mblast_handle\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    572\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mblast_record\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mblast_iter\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    573\u001b[0m             \u001b[0mpdb_id_match\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpdb_id_pattern\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmatch\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mblast_record\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mquery\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/Application/__init__.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, stdin, stdout, stderr, cwd, env)\u001b[0m\n\u001b[1;32m    525\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mreturn_code\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    526\u001b[0m             raise ApplicationError(return_code, str(self),\n\u001b[0;32m--> 527\u001b[0;31m                                    stdout_str, stderr_str)\n\u001b[0m\u001b[1;32m    528\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mstdout_str\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstderr_str\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    529\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mApplicationError\u001b[0m: Non-zero return code 137 from '/home/daniel/lib/ncbi-blast-2.9.0+-src/c++/ReleaseMT/bin/blastp -out /media/daniel/ExtraDrive1/Results/Newest_Covariation/Input/Large_Test_Set/BLAST/LargeDataSet_All_Seqs.xml -outfmt 5 -query /media/daniel/ExtraDrive1/Results/Newest_Covariation/Input/Large_Test_Set/Sequences/LargeDataSet.fasta -db /media/daniel/ExtraDrive1/blast_databases/uniref90_05122020/custom_uniref90_05122020.fasta -evalue 0.05 -max_target_seqs 20000 -num_threads 10', message 'Killed'"
     ]
    }
   ],
   "source": [
    "from time import time\n",
    "from DataSetGenerator import DataSetGenerator\n",
    "protein_list_dir = os.path.join(input_dir, 'ProteinLists')\n",
    "if not os.path.isdir(protein_list_dir):\n",
    "    os.makedirs(protein_list_dir)\n",
    "large_list_fn = os.path.join(protein_list_dir, 'LargeDataSet.txt')\n",
    "if not os.path.isfile(large_list_fn):\n",
    "    proteins_of_interest = ['153lA', '154lA', '1a04A', '1a0bA', '1a0pA', '1ae9A', '1al3A', '1atgA', '1b7eA', '1b9mA', '1b9nA',\n",
    "                            '1biaA', '1bibA', '1bl0A', '1booA', '1bslA', '1byiA', '1byqA', '1c02A', '1c52A', '1c5kA', '1c75A',\n",
    "                            '1cb7A', '1ccwA', '1cp2A', '1crxA', '1crzA', '1ctjA', '1d4aA', '1d5yA', '1dadA', '1daeA', '1dagA',\n",
    "                            '1dahA', '1daiA', '1dakA', '1dd9A', '1ddeA', '1di6A', '1di7A', '1dljA', '1dtsA', '1durA', '1e2xA',\n",
    "                            '1e3uA', '1e4dA', '1e4fT', '1e4gT', '1e8cA', '1eclA', '1efaA', '1efdN', '1eg2A', '1ek9A', '1eszA',\n",
    "                            '1etkA', '1etoA', '1etvA', '1etwA', '1etxA', '1etyA', '1ezwA', '1f07A', '1f1uA', '1f44A', '1f48A',\n",
    "                            '1f5vA', '1f9iA', '1fcaA', '1fdnA', '1fepA', '1fiaA', '1fipA', '1fp6A', '1fr3A', '1fseA', '1fxoA',\n",
    "                            '1g1lA', '1g1mA', '1g20A', '1g28A', '1g5pA', '1g60A', '1g6oA', '1g72A', '1g8kA', '1gbsA', '1gdtA',\n",
    "                            '1gg4A', '1gqyA', '1gu9A', '1gugA', '1gunA', '1gusA', '1gutA', '1h3lA', '1h4iA', '1h7lA', '1h7qA',\n",
    "                            '1h8zA', '1h98A', '1h9gA', '1h9jA', '1h9kA', '1h9mA', '1h9sA', '1hfeS', '1hfeL', '1hm9A', '1hw1A',\n",
    "                            '1hxdA',\n",
    "                            '1i0rA', '1i1gA', '1i52A', '1i58A', '1i5nA', '1i74A', '1i8oA', '1i9cA', '1icrA', '1id0A', '1id1A',\n",
    "                            '1ihcA', '1ihrA', '1ihuA', '1ii0A', '1ii9A', '1iniA', '1injA', '1ir6A', '1iujA', '1ixcA', '1ixgA',\n",
    "                            '1ixhA', '1iz1A', '1j5yA', '1j6uA', '1jbgA', '1jbwA', '1je8A', '1jetA', '1jeuA', '1jevA', '1jftA',\n",
    "                            '1jh9A', '1jiwP', '1jiwI', '1jljA', '1jnuA', '1jpuA', '1jq5A', '1jykA', '1k20A', '1k2vN', '1k38A',\n",
    "                            '1k4fA',\n",
    "                            '1k54A', '1k56A', '1kapP', '1kb0A', '1kbuA', '1kgsA', '1kmoA', '1kmpA', '1kq3A', '1ku3A', '1ku7A',\n",
    "                            '1kv9A', '1kw3B', '1kw6B', '1l3lA', '1lj9A', '1lq9A', '1lqkA', '1lqpA', '1lr0A', '1ls9A', '1lspA',\n",
    "                            '1lssA', '1lucA', '1lvwA', '1m65A', '1m68A', '1m6kA', '1m70A', '1m7jA', '1ma7A', '1mb3A', '1mdoA',\n",
    "                            '1mkmA', '1mkzA', '1mm8A', '1mnzA', '1moqA', '1muhA', '1murA', '1musA', '1muwA', '1mv8A', '1mw8X',\n",
    "                            '1mw9X', '1n2zA', '1n9lA', '1n9nA', '1nfpA', '1nkiA', '1nlyA', '1nnfA', '1noxA', '1nqeA', '1nw5A',\n",
    "                            '1nw6A', '1nw7A', '1nw8A', '1nwzA', '1ny5A', '1ny6A', '1o1hA', '1o2dA', '1o61A', '1o69A', '1o7lA',\n",
    "                            '1oadA', '1oapA', '1oddA', '1odvA', '1oj7A', '1oltA', '1opcA', '1opxA', '1or7A', '1ot6A', '1ot9A',\n",
    "                            '1otaA', '1otbA', '1oxkA', '1p2fA', '1p31A', '1p3dA', '1p7dA', '1p9rA', '1p9wA', '1pb0A', '1pb7A',\n",
    "                            '1pb8A', '1pjrA', '1pnzA', '1po0A', '1pt7A', '1pvpA', '1q05A', '1q06A', '1q07A', '1q08A', '1q09A',\n",
    "                            '1q0aA', '1q35A', '1q7eA', '1qg8A', '1qgqA', '1qgsA', '1qhgA', '1qhhA', '1qksA', '1qpzA', '1qsaA',\n",
    "                            '1qteA', '1qtwA', '1qu7A', '1qwyA', '1qxxA', '1r1mA', '1r1tA', '1r1uA', '1r23A', '1r62A', '1r8dA',\n",
    "                            '1r8eA', '1r9xA', '1r9yA', '1r9zA', '1ra0A', '1ra5A', '1rakA', '1reqA', '1rhcA', '1rioA', '1rk6A',\n",
    "                            '1rp3A', '1rrmA', '1rttA', '1rzuA', '1rzvA', '1s5mA', '1s5nA', '1s8nA', '1sfxA', '1sg0A', '1si0A',\n",
    "                            '1sigA', '1slyA', '1sqeA', '1sqsA', '1sumB', '1suuA', '1t3tA', '1t5bA', '1t72A', '1ta9A', '1td5A',\n",
    "                            '1tf1A', '1tqgA', '1tqqA', '1tv8A', '1tvlA', '1tzbA', '1tzcA', '1u07A', '1u2wA', '1u8bA', '1u8tA',\n",
    "                            '1uaaA', '1uc8A', '1uc9A', '1us4A', '1us5A', '1uscA', '1usfA', '1uuxA', '1uuyA', '1uylA', '1v4yA',\n",
    "                            '1v51A', '1v8pA', '1v96A', '1vctA', '1ve2A', '1vf7A', '1vgtA', '1vgwA', '1vhdA', '1vhvA', '1vimA',\n",
    "                            '1vj7A', '1vkeA', '1vljA', '1vz0A', '1w55A', '1w6sA', '1w77A', '1w78A', '1w8iA', '1wetA', '1wmiA',\n",
    "                            '1woqA', '1wp1A', '1wpmA', '1wpnA', '1wppA', '1ws6A', '1x74A', '1x9hA', '1x9iA', '1xa3A', '1xc3A',\n",
    "                            '1xd7A', '1xi2A', '1xjaA', '1xk6A', '1xk7A', '1xkwA', '1xkzA', '1xmaA', '1xo0A', '1xocA', '1xw3A',\n",
    "                            '1y0hA', '1y1zA', '1y20A', '1y7mA', '1y7yA', '1y80A', '1y82A', '1y9uA', '1yc9A', '1ydxA', '1ye5A',\n",
    "                            '1yf2A', '1yg2A', '1yioA', '1yiqA', '1ylfA', '1yoyA', '1yspA', '1ysqA', '1yviA', '1z05A', '1z19A',\n",
    "                            '1z7uA', '1zatA', '1zi0A', '1zljA', '1zvtA', '1zvuA', '1zzcA', '2a0bA', '2a3nA', '2a5hA', '2a5lA',\n",
    "                            '2a61A', '2aa4A', '2aacA', '2ad6A', '2ad7A', '2ad8A', '2aefA', '2aejA', '2afhA', '2am1A', '2anuA',\n",
    "                            '2ap1A', '2ar0A', '2araA', '2arcA', '2aznA', '2b02A', '2b0pA', '2b13A', '2b3zA', '2b44A', '2basA',\n",
    "                            '2bfwA', '2bknA', '2bkoA', '2bkpA', '2bm4A', '2bm5A', '2bm6A', '2bm7A', '2bnmA', '2brcA', '2byiA',\n",
    "                            '2c2aA', '2c81A', '2ce0A', '2cg4A', '2ch7A', '2cviA', '2cwqA', '2cyyA', '2d1hA', '2d1vA', '2d5mA',\n",
    "                            '2d5nA', '2d5wA', '2dbbA', '2dekA', '2df8A', '2dg6A', '2di3A', '2dqlA', '2dvzA', '2dxwA', '2dxxA',\n",
    "                            '2e15A', '2e1nA', '2e4nA', '2e5fA', '2e7wA', '2e7xA', '2e7zA', '2eb7A', '2ecuA', '2efnA', '2eh3A',\n",
    "                            '2ehlA', '2ehzA', '2ek5A', '2eshA', '2esnA', '2esrA', '2ewnA', '2ewvA', '2eyuA', '2f00A', '2f2eA',\n",
    "                            '2f5xA', '2f6gA', '2f6pA', '2f7aA', '2f7bA', '2f8lA', '2f9fA', '2fa1A', '2fa5A', '2fb2A', '2fbhA',\n",
    "                            '2fcjA', '2fdnA', '2fe1A', '2fezA', '2ff4A', '2ffuA', '2fhpA', '2fn9A', '2fnuA', '2fpoA', '2fswA',\n",
    "                            '2fvyA', '2fw0A', '2g2cA', '2g6vA', '2g7uA', '2gaiA', '2gajA', '2gciA', '2gd0A', '2gd2A', '2gd9A',\n",
    "                            '2gj3A', '2gjgA', '2gkgA', '2glkA', '2gm5A', '2gmsA', '2gmyA', '2gqpA', '2gskA', '2gu1A', '2gufA',\n",
    "                            '2guhA', '2gupA', '2gxgA', '2gzaA', '2h1cA', '2h98A', '2h99A', '2h9bA', '2hawA', '2hekA', '2heuA',\n",
    "                            '2hklA', '2hmtA', '2hmuA', '2hmvA', '2hnhA', '2hoeA', '2hofA', '2hphA', '2hq0A', '2hqsA', '2hs5A',\n",
    "                            '2hsgA', '2hsiA', '2hwvA', '2hxvA', '2i0mA', '2i5rA', '2ia2A', '2ia4A', '2ibdA', '2ictA', '2iftA',\n",
    "                            '2ikkA', '2iplA', '2ipmA', '2ipnA', '2is1A', '2is2A', '2is4A', '2is6A', '2is8A', '2iu5A', '2iuyA',\n",
    "                            '2iv7A', '2iw1A', '2iw4A', '2iwxA', '2jbaA', '2jcgA', '2jfgA', '2nipA', '2npnA', '2nq2A', '2nq9A',\n",
    "                            '2nqhA', '2nt3A', '2nt4A', '2o08A', '2o0yA', '2o3jA', '2o4dA', '2o7iA', '2o7pA', '2o8xA', '2o99A',\n",
    "                            '2o9aA', '2obcA', '2ofyA', '2ogiA', '2ojhA', '2okcA', '2olbA', '2oocA', '2oqgA', '2oqrA', '2oxoA',\n",
    "                            '2oyoA', '2p19A', '2p4gA', '2p5vA', '2p7oA', '2paqA', '2pbqA', '2pfxA', '2ph1A', '2pjrA', '2pkhA',\n",
    "                            '2pmhA', '2pn6A', '2pq7A', '2pt7A', '2pucA', '2pudA', '2px7A', '2q0oA', '2q0tA', '2q1zA', '2q4fA',\n",
    "                            '2q8pA', '2qb6A', '2qb7A', '2qb8A', '2qczA', '2qdfA', '2qdlA', '2qeuA', '2qgqA', '2qgzA', '2qi9A',\n",
    "                            '2qj7A', '2qm1A', '2qmoA', '2qpqA', '2qsxA', '2qwxA', '2qx4A', '2qx6A', '2qx8A', '2r01A', '2r0xA',\n",
    "                            '2r1jA', '2r25A', '2r4tA', '2r6gA', '2r6oA', '2r6vA', '2ra5A', '2rb9A', '2rc7A', '2rc8A', '2rcaA',\n",
    "                            '2rdeA', '2riiA', '2rilA', '2rslA', '2uagA', '2v25A', '2v2kA', '2v9yA', '2vhaA', '2vjqA', '2vk2A',\n",
    "                            '2vkeA', '2vkrA', '2vlgA', '2vmaA', '2vmbA', '2vpzA', '2vshA', '2w27A', '2w8bA', '2w8iA', '2yveA',\n",
    "                            '2yx0A', '2yxbA', '2yxoA', '2yxzA', '2yyeA', '2yz5A', '2z1eA', '2z1fA', '2z1uA', '2z2lA', '2z2mA',\n",
    "                            '2z4gA', '2z4pA', '2z6rA', '2z8xA', '2z98A', '2z9bA', '2zauA', '2zbcA', '2zc3A', '2zc4A', '2zcmA',\n",
    "                            '2zdpA', '2zf8A', '2zieA', '2zifA', '2zigA', '2zkiA', '2zkzA', '2zodA', '2zovA', '2zxjA', '3b4yA',\n",
    "                            '3b6iA', '3b8xA', '3b9oA', '3bcvA', '3be6A', '3bemA', '3bg2A', '3bhqA', '3bkhA', '3bkvA', '3bm7A',\n",
    "                            '3bpkA', '3bpqA', '3bpvA', '3bqxA', '3breA', '3bs3A', '3bvpA', '3bwgA', '3c1qA', '3c29A', '3c3wA',\n",
    "                            '3c48A', '3c57A', '3c7jA', '3c85A', '3c8fA', '3c8nA', '3c9uA', '3canA', '3ccgA', '3cijA', '3cixA',\n",
    "                            '3ckjA', '3cknA', '3ckvA', '3cloA', '3cnrA', '3cnvA', '3cp5A', '3ctpA', '3cuoA', '3cwrA', '3cx4A',\n",
    "                            '3cyiA', '3cypB', '3cypC', '3cypD', '3cyqA', '3d5kA', '3d6zA', '3d7iA', '3dboA', '3df7A', '3df8A', '3dmaA', '3dr4A',\n",
    "                            '3drfA', '3drjA', '3dsgA', '3du1X', '3e10A', '3e38A', '3e4rA', '3e4vA', '3e7lA', '3e8oA', '3eagA',\n",
    "                            '3ec2A', '3eccA', '3echA', '3ecpA', '3edpA', '3eetA', '3efmA', '3eiwA', '3eixA', '3ekoA', '3elkA',\n",
    "                            '3eusA', '3ex8A', '3eywA', '3ezuA', '3f1cA', '3f1nA', '3f1oA', '3f1pA', '3f2bA', '3f44A', '3f52A',\n",
    "                            '3f6cA', '3f6oA', '3f6vA', '3f8bA', '3f8cA', '3f8fA', '3fd3A', '3fgvA', '3fisA', '3fmsA', '3fwyA',\n",
    "                            '3fwzA', '3fxaA', '3fzvA', '3g13A', '3g5oA', '3g7rA', '3gdiA', '3gfaA', '3gfvA', '3gfxA', '3gfyA',\n",
    "                            '3gfzA', '3gg0A', '3gg1A', '3gg2A', '3ghjA', '3gp4A', '3gpvA', '3gr3A', '3guvA', '3h4oA', '3h5tA',\n",
    "                            '3h87A', '3hfiA', '3hh0A', '3hhhA', '3hl0A', '3hmzA', '3hn7A', '3hoiA', '3htvA', '3hvwA', '3pypA',\n",
    "                            '3uagA', '4aahA', '4crxA', '4reqA', '4uagA', '5reqA', '6reqA', '7reqA', '8abpA']\n",
    "    with open(large_list_fn, 'w') as large_list_handle:\n",
    "        for p_id in proteins_of_interest:\n",
    "            large_list_handle.write('{}\\n'.format(p_id))\n",
    "generator = DataSetGenerator(input_dir)\n",
    "start = time()\n",
    "summary = generator.build_pdb_alignment_dataset(protein_list_fn=os.path.basename(large_list_fn), processes=10,\n",
    "                                                database='uniref90_05122020/custom_uniref90_05122020.fasta',\n",
    "                                                max_target_seqs=20000, e_value_threshold=0.05, remote=False,\n",
    "                                                min_fraction=0.7, min_identity=0.40, max_identity=0.98,\n",
    "                                                msf=True, fasta=True, sources=['UNP', 'GB', 'PDB'], verbose=False)\n",
    "summary['Chain'] = summary['Protein_ID'].apply(lambda x: generator.protein_data[x]['Chain'])\n",
    "summary['Accession'] = summary['Protein_ID'].apply(lambda x: generator.protein_data[x]['Accession'])\n",
    "summary['Length'] = summary['Protein_ID'].apply(lambda x: generator.protein_data[x]['Length'])\n",
    "summary['Total_Size'] = summary.apply(lambda x: float(x['Length']) * float(x['Filtered_Alignment']), axis=1)\n",
    "summary.sort_values(by=['Filtered_Alignment', 'Length'], axis=0, inplace=True)\n",
    "summary_columns = ['Protein_ID', 'Chain', 'Accession', 'BLAST_Hits', 'Filtered_BLAST',\n",
    "                   'Filtered_Alignment', 'Length',\n",
    "                   'Total_Size']\n",
    "print(summary[summary_columns])\n",
    "end = time()\n",
    "print('It took {} min to generate the data set.'.format((end - start) / 60.0))\n",
    "summary.to_csv(os.path.join(input_dir, 'large_data_set_summary.tsv'), sep='\\t', index=False, header=True,\n",
    "               columns=summary_columns)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a location to store the output of this validation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_dir = os.environ.get('OUTPUT_PATH')\n",
    "large_set_out_dir = os.path.join(output_dir, 'LargeTestSet')\n",
    "if not os.path.isdir(large_set_out_dir):\n",
    "    os.makedirs(large_set_out_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setting Up Scoring For Each Method\n",
    "To reduce memory load during prediction and evaluation, the scoring objects needed to compute the metrics used to compare methods will be created ahead of time so they are available to each method when it computes its predictions for a given protein. This will ensure that results do not need to be kept in memory while waiting for all other results to be computed, only the metrics measured for each method will be recorded."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from SeqAlignment import SeqAlignment\n",
    "from PDBReference import PDBReference\n",
    "from ContactScorer import ContactScorer, plot_z_scores\n",
    "protein_order = list(summary['Protein_ID'])\n",
    "method_order = ['DCA', 'EV Couplings', 'EV Couplings MF', 'ET-MIp', 'cET-MIp']\n",
    "sequence_separation_order = ['Any', 'Neighbors', 'Short', 'Medium', 'Long']\n",
    "protein_scorers = {}\n",
    "counts = {'success': 0, 'attribute': 0, 'value': 0}\n",
    "for p_id in summary['Protein_ID']:\n",
    "    try:\n",
    "        protein_scorers[p_id] = {}\n",
    "        # Import alignment and remove gaps\n",
    "        full_aln = SeqAlignment(file_name=generator.protein_data[p_id]['Final_FA_Aln'], query_id=p_id)\n",
    "        full_aln.import_alignment()\n",
    "        non_gap_aln = full_aln.remove_gaps()\n",
    "        # Import structure\n",
    "        pdb_structure = PDBReference(pdb_file=generator.protein_data[p_id]['PDB'])\n",
    "        pdb_structure.import_pdb(structure_id=p_id)\n",
    "        protein_scorers[p_id]['Structure'] = pdb_structure\n",
    "        # Initialize Beta Carbon distance scorer\n",
    "        contact_scorer_cb = ContactScorer(query=p_id, seq_alignment=non_gap_aln,\n",
    "                                          pdb_reference=pdb_structure, cutoff=8.0)\n",
    "        contact_scorer_cb.best_chain = generator.protein_data[p_id]['Chain']\n",
    "        contact_scorer_cb.fit()\n",
    "        contact_scorer_cb.measure_distance(method='CB')\n",
    "        protein_scorers[p_id]['Scorer_CB'] = contact_scorer_cb\n",
    "        # Initialize distance scorer minimizing distance between any atoms\n",
    "        contact_scorer_any = ContactScorer(query=p_id, seq_alignment=non_gap_aln,\n",
    "                                           pdb_reference=pdb_structure, cutoff=8.0)\n",
    "        contact_scorer_any.best_chain = generator.protein_data[p_id]['Chain']\n",
    "        contact_scorer_any.fit()\n",
    "        contact_scorer_any.measure_distance(method='Any')\n",
    "        protein_scorers[p_id]['Scorer_Any'] = contact_scorer_any\n",
    "        # Initialize z-scoring subproblems\n",
    "        protein_scorers[p_id]['biased_w2_ave'] = None\n",
    "        protein_scorers[p_id]['unbiased_w2_ave'] = None\n",
    "        print('Successfully initialized scoring for: {}'.format(p_id))\n",
    "        counts['success'] += 1\n",
    "    except AttributeError:\n",
    "        print('Could not initialize scoring for: {}\\tATTRIBUTE'.format(p_id))\n",
    "        counts['attribute'] += 1\n",
    "    except ValueError:\n",
    "        print('Could not initialize scoring for: {}\\tVALUE'.format(p_id))\n",
    "        counts['value'] += 1\n",
    "print('Completed initializing scorers:\\nSuccesses:\\t{}\\nAttribute:\\t{}\\nValue:\\t{}'.format(counts['success'], counts['attribute'],\n",
    "                                                                                           counts['value']))\n",
    "output_columns = ['Protein', 'Protein Length', 'Alignment Size', 'Method', 'Distance', 'Init Time', 'Import Time', 'Dist Tree Time', 'Trace Time', 'Total Time', \n",
    "                  'Sequence_Separation', 'AUROC', 'AUPRC',\n",
    "                  'Top K Predictions', 'Precision', 'Recall', 'F1 Score',\n",
    "                  'Biased Z-Score at 10%', 'Biased Z-Score at 30%', 'Max Biased Z-Score', 'AUC Biased Z-Score',\n",
    "                  'Unbiased Z-Score at 10%', 'Unbiased Z-Score at 30%', 'Max Unbiased Z-Score', 'AUC Unbiased Z-Score']\n",
    "large_comparison_df = None\n",
    "large_comparison_fn = os.path.join(large_set_out_dir, 'Large_Comparision_Data.csv')\n",
    "if os.path.isfile(large_comparison_fn):\n",
    "    large_comparison_df = pd.read_csv(large_comparison_fn, sep='\\t', header=0, index_col=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generating Values For Comparision#\n",
    "To determine the effectiveness of the new method and implementation the covariation of the same proteins will be computed using the previous Evolutionary Trace covariation method (ET-MIp) and other methods in the field.\n",
    "\n",
    "## ET-MIp##\n",
    "Scoring the the covariation of the proteins using the previous Evolutionary Trace covariation method (ET-MIp)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# from ETMIPWrapper import ETMIPWrapper\n",
    "# etmip_out_dir = os.path.join(large_set_out_dir, 'ET-MIp')\n",
    "# if not os.path.isdir(etmip_out_dir):\n",
    "#     os.makedirs(etmip_out_dir)\n",
    "# etmip_scores = {}\n",
    "# counts = {'success':0, 'value': 0, 'attribute':0}\n",
    "# for p_id in generator.protein_data:\n",
    "#     print('Attempting to calculate ET-MIp covariance for: {}'.format(p_id))\n",
    "#     try:\n",
    "#         protein_out_dir = os.path.join(etmip_out_dir, p_id)\n",
    "#         if not os.path.isdir(protein_out_dir):\n",
    "#             os.makedirs(protein_out_dir)\n",
    "##         curr_aln = SeqAlignment(file_name=generator.protein_data[p_id]['Final_FA_Aln'], query_id=p_id, polymer_type='Protein')\n",
    "##         curr_aln.import_alignment()\n",
    "#         curr_etmip = ETMIPWrapper(query=p_id, aln_file=generator.protein_data[p_id]['Final_FA_Aln'],\n",
    "#                                   out_dir=protein_out_dir)\n",
    "#         curr_etmip.calculate_scores(delete_files=False)\n",
    "#         etmip_scores[p_id] = curr_etmip\n",
    "#         print('Successfully computed ET-MIp covariance for: {}'.format(p_id))\n",
    "#         counts['success'] += 1\n",
    "#     except ValueError:\n",
    "#         print('Could not compute ET-MIp covariance for: {} with seq_length: {} and size: {}'.format(\n",
    "#             p_id, curr_aln.seq_length, curr_aln.size))\n",
    "#         counts['value'] += 1\n",
    "#     except AttributeError:\n",
    "#         print('Could not compute ET-MIp covariance for: {} no alignment'.format(p_id))\n",
    "#         counts['attribute'] += 1\n",
    "# print('{}\\tSuccesses\\n{}\\tValue Errors\\n{}\\tAttribute Errors'.format(counts['success'], counts['value'],\n",
    "#                                                                      counts['attribute']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ET-MIp (Continued)\n",
    "The previous implementation is not able to run for alignments of the size used here. Instead we use the new implementation with the same parameterization used by the previous implementation (Distance Model - blosum62 similarity, Tree - ET UPGMA variant, Scoring Metric - filtered average product corrected mutual information, Ranks - all)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from EvolutionaryTrace import EvolutionaryTrace\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "if not os.path.isfile(large_comparison_fn):\n",
    "    etmip_out_dir = os.path.join(large_set_out_dir, 'ET-MIp')\n",
    "    if not os.path.isdir(etmip_out_dir):\n",
    "        os.makedirs(etmip_out_dir)\n",
    "    etmip_method_fn = os.path.join(etmip_out_dir, 'ET-MIp_Method_Data.csv')\n",
    "    if os.path.isfile(etmip_method_fn):\n",
    "        etmip_method_df = pd.read_csv(etmip_method_fn, sep='\\t', header=0, index_col=False)\n",
    "    else:    \n",
    "        etmip_method_df = None\n",
    "        counts = {'success':0, 'value': 0, 'attribute':0, 'type': 0}\n",
    "        for p_id in summary['Protein_ID']: # generator.protein_data:\n",
    "            print('Attempting to calculate ET-MIp covariance for: {}'.format(p_id))\n",
    "            protein_dir = os.path.join(etmip_out_dir, p_id)\n",
    "            if not os.path.isdir(protein_dir):\n",
    "                os.makedirs(protein_dir)\n",
    "            protein_fn = os.path.join(protein_dir, '{}_Protein_Data.csv'.format(p_id))\n",
    "            if os.path.isfile(protein_fn):\n",
    "                protein_df = pd.read_csv(protein_fn, sep='\\t', header=0, index_col=False)\n",
    "            else:\n",
    "                try:\n",
    "\n",
    "#                     start_time = time()\n",
    "                    curr_etmip = EvolutionaryTrace(query=p_id, polymer_type='Protein',\n",
    "                                                   aln_file=generator.protein_data[p_id]['Final_FA_Aln'], et_distance=True,\n",
    "                                                   distance_model='blosum62', tree_building_method='et', tree_building_options={},\n",
    "                                                   ranks=None, position_type='pair',\n",
    "                                                   scoring_metric='filtered_average_product_corrected_mutual_information',\n",
    "                                                   gap_correction=None, out_dir=protein_dir,\n",
    "                                                   output_files={'original_aln', 'non_gap_aln', 'tree', 'scores'},\n",
    "                                                   processors=10, low_memory=True)\n",
    "#                     init_time = time()\n",
    "#                     curr_etmip.import_and_process_aln()\n",
    "#                     import_time = time()\n",
    "#                     curr_etmip.compute_distance_matrix_tree_and_assignments()\n",
    "#                     dist_tree_time = time()\n",
    "#                     curr_etmip.perform_trace()\n",
    "#                     end_time = time()\n",
    "                    curr_etmip.calculate_scores()\n",
    "                    print('Successfully computed ET-MIp covariance for: {}'.format(p_id))\n",
    "                    # Compute statistics for the final scores of the ET-MIp model\n",
    "                    protein_df, _, _ = protein_scorers[p_id]['Scorer_CB'].evaluate_predictor(\n",
    "                        predictor=curr_etmip, verbosity=2, out_dir=protein_dir, dist='CB', biased_w2_ave=None,\n",
    "                        unbiased_w2_ave=None, processes=10, threshold=0.5, pos_size=curr_etmip.scorer.position_size,\n",
    "                        rank_type=curr_etmip.scorer.rank_type, file_prefix='ET-MIp_Scores_', plots=True)\n",
    "                    # Score Prediction Clustering\n",
    "                    z_score_fn = os.path.join(protein_dir, 'ET-MIp_Scores_Dist-Any_{}_ZScores.tsv')\n",
    "                    z_score_plot_fn = os.path.join(protein_dir, 'ET-MIp_Scores_Dist-Any_{}_ZScores.png')\n",
    "                    z_score_biased, biased_w2_ave, biased_scw_z_auc = protein_scorers[p_id]['Scorer_Any'].score_clustering_of_contact_predictions(\n",
    "                        1.0 - curr_etmip.coverage, bias=True, file_path=z_score_fn.format('Biased'),\n",
    "                        w2_ave_sub=protein_scorers[p_id]['biased_w2_ave'], processes=10)\n",
    "                    if protein_scorers[p_id]['biased_w2_ave'] is None:\n",
    "                        protein_scorers[p_id]['biased_w2_ave'] = biased_w2_ave\n",
    "                    biased_z_score_array = np.array(pd.to_numeric(z_score_biased['Z-Score'], errors='coerce'))\n",
    "                    protein_df['Max Biased Z-Score'] = np.nanmax(biased_z_score_array)\n",
    "                    protein_df['Biased Z-Score at 10%'] = biased_z_score_array[z_score_biased['Num_Residues'] >= float(len(protein_scorers[p_id]['Scorer_Any'].query_structure.seq[protein_scorers[p_id]['Scorer_Any'].best_chain])) * 0.10][0]\n",
    "                    protein_df['Biased Z-Score at 30%'] = biased_z_score_array[z_score_biased['Num_Residues'] >= float(len(protein_scorers[p_id]['Scorer_Any'].query_structure.seq[protein_scorers[p_id]['Scorer_Any'].best_chain])) * 0.30][0]\n",
    "                    protein_df['AUC Biased Z-Score'] = biased_scw_z_auc\n",
    "                    plot_z_scores(z_score_biased, z_score_plot_fn.format('Biased'))\n",
    "                    z_score_unbiased, unbiased_w2_ave, unbiased_scw_z_auc = protein_scorers[p_id]['Scorer_Any'].score_clustering_of_contact_predictions(\n",
    "                        1.0 - curr_etmip.coverage, bias=False, file_path=z_score_fn.format('Unbiased'),\n",
    "                        w2_ave_sub=protein_scorers[p_id]['unbiased_w2_ave'], processes=10)\n",
    "                    if protein_scorers[p_id]['unbiased_w2_ave'] is None:\n",
    "                        protein_scorers[p_id]['unbiased_w2_ave'] = unbiased_w2_ave\n",
    "                    unbiased_z_score_array = np.array(pd.to_numeric(z_score_unbiased['Z-Score'], errors='coerce'))\n",
    "                    protein_df['Max Unbiased Z-Score'] = np.nanmax(unbiased_z_score_array)\n",
    "                    protein_df['Unbiased Z-Score at 10%'] = unbiased_z_score_array[z_score_unbiased['Num_Residues'] >= float(len(protein_scorers[p_id]['Scorer_Any'].query_structure.seq[protein_scorers[p_id]['Scorer_Any'].best_chain])) * 0.10][0]\n",
    "                    protein_df['Unbiased Z-Score at 30%'] = unbiased_z_score_array[z_score_unbiased['Num_Residues'] >= float(len(protein_scorers[p_id]['Scorer_Any'].query_structure.seq[protein_scorers[p_id]['Scorer_Any'].best_chain])) * 0.30][0]\n",
    "                    protein_df['AUC Unbiased Z-Score'] = unbiased_scw_z_auc\n",
    "                    plot_z_scores(z_score_unbiased, z_score_plot_fn.format('Unbiased'))\n",
    "                    # Record execution times\n",
    "                    protein_df['Init Time'] = None # init_time - start_time\n",
    "                    protein_df['Import Time'] = None # import_time - init_time\n",
    "                    protein_df['Dist Tree Time'] = None # dist_tree_time - import_time\n",
    "                    protein_df['Trace Time'] = None # end_time - dist_tree_time\n",
    "                    protein_df['Total Time'] = curr_etmip.time # end_time - start_time\n",
    "                    # Record static data for this protein\n",
    "                    protein_df['Protein'] = p_id\n",
    "                    protein_df['Method'] = 'ET-MIp'\n",
    "                    protein_df['Protein Length'] = generator.protein_data[p_id]['Length']\n",
    "                    protein_df['Alignment Size'] = summary['Filtered_Alignment'].values[summary['Protein_ID'] == p_id][0]\n",
    "                    protein_df.to_csv(protein_fn, sep='\\t', header=True, index=False, columns=output_columns)\n",
    "                    temp_data = os.path.join(protein_dir, 'unique_node_data')\n",
    "                    for temp_fn in os.listdir(temp_data):\n",
    "                        if not temp_fn.endswith(\"_pair_rank_filtered_average_product_corrected_mutual_information_score.npz\"):\n",
    "                            os.remove(os.path.join(temp_data, temp_fn))\n",
    "                    print('Metrics meastured for ET-MIp covariance for: {}'.format(p_id))\n",
    "                    counts['success'] += 1\n",
    "                except ValueError:\n",
    "                    print('Could not compute ET-MIp covariance for: {} with seq_length: {} and size: {}'.format(\n",
    "                        p_id, curr_etmip.original_aln.seq_length, curr_etmip.original_aln.size))\n",
    "                    counts['value'] += 1\n",
    "                    continue\n",
    "                except AttributeError:\n",
    "                    print('Could not compute ET-MIp covariance for: {} no alignment'.format(p_id))\n",
    "                    counts['attribute'] += 1\n",
    "                    continue\n",
    "                except TypeError:\n",
    "                    print('Could not compute ET-MIp coooooovariance for: {} type error'.format(p_id))\n",
    "                    counts['type'] += 1\n",
    "                    continue\n",
    "            if etmip_method_df is None:\n",
    "                etmip_method_df = protein_df\n",
    "            else:\n",
    "                etmip_method_df = etmip_method_df.append(protein_df)\n",
    "        print('{}\\tSuccesses\\n{}\\tValue Errors\\n{}\\tAttribute Errors'.format(counts['success'], counts['value'],\n",
    "                                                                             counts['attribute']))\n",
    "        etmip_method_df.to_csv(etmip_method_fn, sep='\\t', header=True, index=False, columns=output_columns)\n",
    "    if large_comparison_df is None:\n",
    "        large_comparison_df = etmip_method_df\n",
    "    else:\n",
    "        large_comparison_df = large_comparison_df.append(etmip_method_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## cET-MIp\n",
    "This segment the ET-MIp method, when constrained to an arbitrary set of nodes (1, 2, 3, 5, 7, 10, 25) at the top of the phylogenetic tree."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.isfile(large_comparison_fn):\n",
    "    cetmip_out_dir = os.path.join(large_set_out_dir, 'cET-MIp')\n",
    "    if not os.path.isdir(cetmip_out_dir):\n",
    "        os.makedirs(cetmip_out_dir)\n",
    "    cetmip_method_fn = os.path.join(cetmip_out_dir, 'cET-MIp_Method_Data.csv')\n",
    "    if os.path.isfile(cetmip_method_fn):\n",
    "        cetmip_method_df = pd.read_csv(cetmip_method_fn, sep='\\t', header=0, index_col=False)\n",
    "    else:\n",
    "        cetmip_method_df = None\n",
    "        counts = {'success':0, 'value': 0, 'attribute':0, 'key': 0}\n",
    "        for p_id in generator.protein_data:\n",
    "            print('Attempting to calculate cET-MIp covariance for: {}'.format(p_id))\n",
    "            protein_dir = os.path.join(cetmip_out_dir, p_id)\n",
    "            if not os.path.isdir(protein_dir):\n",
    "                os.makedirs(protein_dir)\n",
    "            protein_fn = os.path.join(protein_dir, '{}_Protein_Data.csv'.format(p_id))\n",
    "            if os.path.isfile(protein_fn):\n",
    "                protein_df = pd.read_csv(protein_fn, sep='\\t', header=0, index_col=False)\n",
    "            else:\n",
    "                try:\n",
    "#                     start_time = time()\n",
    "                    curr_cetmip = EvolutionaryTrace(query=p_id, polymer_type='Protein',\n",
    "                                                   aln_file=generator.protein_data[p_id]['Final_FA_Aln'], et_distance=True,\n",
    "                                                   distance_model='blosum62', tree_building_method='et', tree_building_options={},\n",
    "                                                   ranks=[1, 2, 3, 5, 7, 10, 25], position_type='pair',\n",
    "                                                   scoring_metric='filtered_average_product_corrected_mutual_information',\n",
    "                                                   gap_correction=None, out_dir=protein_dir,\n",
    "                                                   output_files={'original_aln', 'non_gap_aln', 'tree', 'scores'},\n",
    "                                                   processors=10, low_memory=True)\n",
    "#                     init_time = time()\n",
    "#                     curr_cetmip.import_and_process_aln()\n",
    "#                     import_time = time()\n",
    "#                     curr_cetmip.compute_distance_matrix_tree_and_assignments()\n",
    "#                     dist_tree_time = time()\n",
    "#                     curr_cetmip.perform_trace()\n",
    "#                     end_time = time()\n",
    "                    curr_cetmip.calculate_scores()\n",
    "                    # Compute statistics for the final scores of the ET-MIp model\n",
    "                    protein_df, _, _ = protein_scorers[p_id]['Scorer_CB'].evaluate_predictor(\n",
    "                        predictor=curr_cetmip, verbosity=2, out_dir=protein_dir, dist='CB', biased_w2_ave=None,\n",
    "                        unbiased_w2_ave=None, processes=10, threshold=0.5, pos_size=curr_cetmip.scorer.position_size,\n",
    "                        rank_type=curr_cetmip.scorer.rank_type, file_prefix='cET-MIp_Scores_', plots=True)\n",
    "                    # Score Prediction Clustering\n",
    "                    z_score_fn = os.path.join(protein_dir, 'cET-MIp_Scores_Dist-Any_{}_ZScores.tsv')\n",
    "                    z_score_plot_fn = os.path.join(protein_dir, 'cET-MIp_Scores_Dist-Any_{}_ZScores.png')\n",
    "                    z_score_biased, biased_w2_ave, biased_scw_z_auc = protein_scorers[p_id]['Scorer_Any'].score_clustering_of_contact_predictions(\n",
    "                        1.0 - curr_cetmip.coverage, bias=True, file_path=z_score_fn.format('Biased'),\n",
    "                        w2_ave_sub=protein_scorers[p_id]['biased_w2_ave'], processes=10)\n",
    "                    if protein_scorers[p_id]['biased_w2_ave'] is None:\n",
    "                        protein_scorers[p_id]['biased_w2_ave'] = biased_w2_ave\n",
    "                    biased_z_score_array = np.array(pd.to_numeric(z_score_biased['Z-Score'], errors='coerce'))\n",
    "                    protein_df['Max Biased Z-Score'] = np.nanmax(biased_z_score_array)\n",
    "                    protein_df['Biased Z-Score at 10%'] = biased_z_score_array[z_score_biased['Num_Residues'] >= float(len(protein_scorers[p_id]['Scorer_Any'].query_structure.seq[protein_scorers[p_id]['Scorer_Any'].best_chain])) * 0.10][0]\n",
    "                    protein_df['Biased Z-Score at 30%'] = biased_z_score_array[z_score_biased['Num_Residues'] >= float(len(protein_scorers[p_id]['Scorer_Any'].query_structure.seq[protein_scorers[p_id]['Scorer_Any'].best_chain])) * 0.30][0]\n",
    "                    protein_df['AUC Biased Z-Score'] = biased_scw_z_auc\n",
    "                    plot_z_scores(z_score_biased, z_score_plot_fn.format('Biased'))\n",
    "                    z_score_unbiased, unbiased_w2_ave, unbiased_scw_z_auc = protein_scorers[p_id]['Scorer_Any'].score_clustering_of_contact_predictions(\n",
    "                        1.0 - curr_cetmip.coverage, bias=False, file_path=z_score_fn.format('Unbiased'),\n",
    "                        w2_ave_sub=protein_scorers[p_id]['unbiased_w2_ave'], processes=10)\n",
    "                    if protein_scorers[p_id]['unbiased_w2_ave'] is None:\n",
    "                        protein_scorers[p_id]['unbiased_w2_ave'] = unbiased_w2_ave\n",
    "                    unbiased_z_score_array = np.array(pd.to_numeric(z_score_unbiased['Z-Score'], errors='coerce'))\n",
    "                    protein_df['Max Unbiased Z-Score'] = np.nanmax(unbiased_z_score_array)\n",
    "                    protein_df['Unbiased Z-Score at 10%'] = unbiased_z_score_array[z_score_unbiased['Num_Residues'] >= float(len(protein_scorers[p_id]['Scorer_Any'].query_structure.seq[protein_scorers[p_id]['Scorer_Any'].best_chain])) * 0.10][0]\n",
    "                    protein_df['Unbiased Z-Score at 30%'] = unbiased_z_score_array[z_score_unbiased['Num_Residues'] >= float(len(protein_scorers[p_id]['Scorer_Any'].query_structure.seq[protein_scorers[p_id]['Scorer_Any'].best_chain])) * 0.30][0]\n",
    "                    protein_df['AUC Unbiased Z-Score'] = unbiased_scw_z_auc\n",
    "                    plot_z_scores(z_score_unbiased, z_score_plot_fn.format('Unbiased'))\n",
    "                    # Record execution times\n",
    "                    protein_df['Init Time'] = None # init_time - start_time\n",
    "                    protein_df['Import Time'] = None # import_time - init_time\n",
    "                    protein_df['Dist Tree Time'] = None # dist_tree_time - import_time\n",
    "                    protein_df['Trace Time'] = None # end_time - dist_tree_time\n",
    "                    protein_df['Total Time'] = curr_cetmip.time # end_time - start_time\n",
    "                    # Record static data for this protein\n",
    "                    protein_df['Protein'] = p_id\n",
    "                    protein_df['Method'] = 'cET-MIp'\n",
    "                    protein_df['Protein Length'] = generator.protein_data[p_id]['Length']\n",
    "                    protein_df['Alignment Size'] = summary['Filtered_Alignment'].values[summary['Protein_ID'] == p_id][0]\n",
    "                    protein_df.to_csv(protein_fn, sep='\\t', header=True, index=False, columns=output_columns)\n",
    "                    temp_data = os.path.join(protein_dir, 'unique_node_data')\n",
    "                    for temp_fn in os.listdir(temp_data):\n",
    "                        if not temp_fn.endswith(\"_pair_rank_filtered_average_product_corrected_mutual_information_score.npz\"):\n",
    "                            os.remove(os.path.join(temp_data, temp_fn))\n",
    "                    print('Successfully computed cET-MIp covariance for: {}'.format(p_id))\n",
    "                    counts['success'] += 1\n",
    "                except ValueError:\n",
    "                    print('Could not compute cET-MIp covariance for: {} with seq_length: {} and size: {}'.format(\n",
    "                        p_id, curr_cetmip.original_aln.seq_length, curr_etmip.original_aln.size))\n",
    "                    counts['value'] += 1\n",
    "                    continue\n",
    "                except AttributeError:\n",
    "                    print('Could not compute cET-MIp covariance for: {} no alignment'.format(p_id))\n",
    "                    counts['attribute'] += 1\n",
    "                    continue\n",
    "                except KeyError:\n",
    "                    print('Could not compute cET-MIp covariance for: {} not enough sequences'.format('p_ied'))\n",
    "                    counts['key'] += 1\n",
    "                    continue\n",
    "            if cetmip_method_df is None:\n",
    "                cetmip_method_df = protein_df\n",
    "            else:\n",
    "                cetmip_method_df = cetmip_method_df.append(protein_df)\n",
    "        print('{}\\tSuccesses\\n{}\\tValue Errors\\n{}\\tAttribute Errors'.format(counts['success'], counts['value'],\n",
    "                                                                             counts['attribute']))\n",
    "        cetmip_method_df.to_csv(cetmip_method_fn, sep='\\t', header=True, index=False, columns=output_columns)\n",
    "    if large_comparison_df is None:\n",
    "        large_comparison_df = cetmip_method_df\n",
    "    else:\n",
    "        large_comparison_df = large_comparison_df.append(cetmip_method_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DCA##\n",
    "Scoring the the covariation of the proteins using a DCA julia implementation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from DCAWrapper import DCAWrapper\n",
    "from utils import compute_rank_and_coverage\n",
    "if not os.path.isfile(large_comparison_fn):\n",
    "    dca_out_dir = os.path.join(large_set_out_dir, 'DCA')\n",
    "    if not os.path.isdir(dca_out_dir):\n",
    "        os.makedirs(dca_out_dir)\n",
    "    dca_method_fn = os.path.join(dca_out_dir, 'DCA_Method_Data.csv')\n",
    "    if olarges.path.isfile(dca_method_fn):\n",
    "        dca_method_df = pd.read_csv(dca_method_fn, sep='\\t', header=0, index_col=False)\n",
    "    else:\n",
    "        dca_method_df = None\n",
    "        counts = {'success':0, 'value': 0, 'attribute':0}\n",
    "        for p_id in generator.protein_data:\n",
    "            print('Attempting to calculate DCA covariance for: {}'.format(p_id))\n",
    "            protein_dir = os.path.join(dca_out_dir, p_id)\n",
    "            if not os.path.isdir(protein_dir):\n",
    "                os.makedirs(protein_dir)\n",
    "            protein_fn = os.path.join(protein_dir, '{}_Protein_Data.csv'.format(p_id))\n",
    "            if os.path.isfile(protein_fn):\n",
    "                protein_df = pd.read_csv(protein_fn, sep='\\t', header=0, index_col=False)\n",
    "            else:\n",
    "                try:\n",
    "#                     curr_aln = SeqAlignment(file_name=generator.protein_data[p_id]['Final_FA_Aln'], query_id=p_id,\n",
    "#                                             polymer_type='Protein')\n",
    "#                     curr_aln.import_alignment()\n",
    "                    # Since the DCA implementation used here does not provide a way to specify the query sequence we remove the gaps\n",
    "                    # from the query sequences so positions will be referenced correctly for that sequence (and unnecessary\n",
    "                    # computations can be avoided).\n",
    "#                     curr_aln = curr_aln.remove_gaps()\n",
    "#                     new_aln_fn = os.path.join(protein_dir, '{}_no_gap.fasta'.format(p_id))\n",
    "#                     curr_aln.write_out_alignment(new_aln_fn)\n",
    "#                     curr_aln.file_name = new_aln_fn\n",
    "#                     curr_dca = DCAWrapper(alignment=curr_aln)\n",
    "                    curr_dca = DCAWrapper(query=p_id, aln_file=generator.protein_data[p_id]['Final_FA_Aln'], out_dir=protein_dir)\n",
    "#                     curr_dca.calculate_scores(out_dir=protein_dir, delete_file=False)\n",
    "                    curr_dca.calculate_scores(delete_file=False)\n",
    "                    # Compute statistics for the final scores of the ET-MIp model\n",
    "                    protein_df, _, _ = protein_scorers[p_id]['Scorer_CB'].evaluate_predictor(\n",
    "                        predictor=curr_dca, verbosity=2, out_dir=protein_dir, dist='CB', biased_w2_ave=None,\n",
    "                        unbiased_w2_ave=None, processes=10, threshold=0.5, pos_size=2, rank_type='max', file_prefix='DCA_Scores_', plots=True)\n",
    "                    # Score Prediction Clustering\n",
    "                    _, dca_coverage  = compute_rank_and_coverage(seq_length=curr_dca.alignment.seq_length, scores=curr_dca.scores, pos_size=2,\n",
    "                        rank_type='max')\n",
    "                    z_score_fn = os.path.join(protein_dir, 'DCA_Scores_Dist-Any_{}_ZScores.tsv')\n",
    "                    z_score_plot_fn = os.path.join(protein_dir, 'DCA_Scores_Dist-Any_{}_ZScores.png')\n",
    "                    z_score_biased, biased_w2_ave, biased_scw_z_auc = protein_scorers[p_id]['Scorer_Any'].score_clustering_of_contact_predictions(\n",
    "                        1.0 - dca_coverage, bias=True, file_path=z_score_fn.format('Biased'),\n",
    "                        w2_ave_sub=protein_scorers[p_id]['biased_w2_ave'], processes=10)\n",
    "                    if protein_scorers[p_id]['biased_w2_ave'] is None:\n",
    "                        protein_scorers[p_id]['biased_w2_ave'] = biased_w2_ave\n",
    "                    biased_z_score_array = np.array(pd.to_numeric(z_score_biased['Z-Score'], errors='coerce'))\n",
    "                    protein_df['Max Biased Z-Score'] = np.nanmax(biased_z_score_array)\n",
    "                    protein_df['Biased Z-Score at 10%'] = biased_z_score_array[z_score_biased['Num_Residues'] >= float(len(protein_scorers[p_id]['Scorer_Any'].query_structure.seq[protein_scorers[p_id]['Scorer_Any'].best_chain])) * 0.10][0]\n",
    "                    protein_df['Biased Z-Score at 30%'] = biased_z_score_array[z_score_biased['Num_Residues'] >= float(len(protein_scorers[p_id]['Scorer_Any'].query_structure.seq[protein_scorers[p_id]['Scorer_Any'].best_chain])) * 0.30][0]\n",
    "                    protein_df['AUC Biased Z-Score'] = biased_scw_z_auc\n",
    "                    plot_z_scores(z_score_biased, z_score_plot_fn.format('Biased'))\n",
    "                    z_score_unbiased, unbiased_w2_ave, unbiased_scw_z_auc = protein_scorers[p_id]['Scorer_Any'].score_clustering_of_contact_predictions(\n",
    "                        1.0 - dca_coverage, bias=False, file_path=z_score_fn.format('Unbiased'),\n",
    "                        w2_ave_sub=protein_scorers[p_id]['unbiased_w2_ave'], processes=10)\n",
    "                    if protein_scorers[p_id]['unbiased_w2_ave'] is None:\n",
    "                        protein_scorers[p_id]['unbiased_w2_ave'] = unbiased_w2_ave\n",
    "                    unbiased_z_score_array = np.array(pd.to_numeric(z_score_unbiased['Z-Score'], errors='coerce'))\n",
    "                    protein_df['Max Unbiased Z-Score'] = np.nanmax(unbiased_z_score_array)\n",
    "                    protein_df['Unbiased Z-Score at 10%'] = unbiased_z_score_array[z_score_unbiased['Num_Residues'] >= float(len(protein_scorers[p_id]['Scorer_Any'].query_structure.seq[protein_scorers[p_id]['Scorer_Any'].best_chain])) * 0.10][0]\n",
    "                    protein_df['Unbiased Z-Score at 30%'] = unbiased_z_score_array[z_score_unbiased['Num_Residues'] >= float(len(protein_scorers[p_id]['Scorer_Any'].query_structure.seq[protein_scorers[p_id]['Scorer_Any'].best_chain])) * 0.30][0]\n",
    "                    protein_df['AUC Unbiased Z-Score'] = unbiased_scw_z_auc\n",
    "                    plot_z_scores(z_score_unbiased, z_score_plot_fn.format('Unbiased'))\n",
    "                    # Record execution times\n",
    "                    protein_df['Init Time'] = None\n",
    "                    protein_df['Import Time'] = None\n",
    "                    protein_df['Dist Tree Time'] = None\n",
    "                    protein_df['Trace Time'] = None\n",
    "                    protein_df['Total Time'] = None\n",
    "                    # Record static data for this protein\n",
    "                    protein_df['Protein'] = p_id\n",
    "                    protein_df['Method'] = 'DCA'\n",
    "                    protein_df['Protein Length'] = generator.protein_data[p_id]['Length']\n",
    "                    protein_df['Alignment Size'] = summary['Filtered_Alignment'].values[summary['Protein_ID'] == p_id][0]\n",
    "                    protein_df.to_csv(protein_fn, sep='\\t', header=True, index=False, columns=output_columns)\n",
    "                    print('Successfully computed DCA covariance for: {}'.format(p_id))\n",
    "                    counts['success'] += 1\n",
    "                except ValueError:\n",
    "                    print('Could not compute DCA covariance for: {} with seq_length: {} and size: {}'.format(\n",
    "                        p_id, curr_aln.seq_length, curr_aln.size))\n",
    "                    counts['value'] += 1\n",
    "                    continue\n",
    "                except AttributeError:\n",
    "                    print('Could not compute DCA covariance for: {} no alignment'.format(p_id))\n",
    "                    counts['attribute'] += 1\n",
    "                    continue\n",
    "            if dca_method_df is None:\n",
    "                dca_method_df = protein_df\n",
    "            else:\n",
    "                dca_method_df = dca_method_df.append(protein_df)\n",
    "        print('{}\\tSuccesses\\n{}\\tValue Errors\\n{}\\tAttribute Errors'.format(counts['success'], counts['value'],\n",
    "                                                                             counts['attribute']))\n",
    "        dca_method_df.to_csv(dca_method_fn, sep='\\t', header=True, index=False, columns=output_columns)\n",
    "    if large_comparison_df is None:\n",
    "        large_comparison_df = dca_method_df\n",
    "    else:\n",
    "        large_comparison_df = large_comparison_df.append(dca_method_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## EVCouplings##\n",
    "Scoring the the covariation of the proteins using the EVCouplings method standard protocol."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from EVCouplingsWrapper import EVCouplingsWrapper\n",
    "if not os.path.isfile(large_comparison_fn):\n",
    "    evc_standard_out_dir = os.path.join(large_set_out_dir, 'EVCouplings_Standard')\n",
    "    if not os.path.isdir(evc_standard_out_dir):\n",
    "        os.makedirs(evc_standard_out_dir)\n",
    "    evc_standard_method_fn = os.path.join(evc_standard_out_dir, 'EVCouplings_Standard_Method_Data.csv')\n",
    "    if os.path.isfile(evc_standard_method_fn):\n",
    "        evc_standard_method_df = pd.read_csv(evc_standard_method_fn, sep='\\t', header=0, index_col=False)\n",
    "    else:\n",
    "        evc_standard_method_df = None\n",
    "        counts = {'success':0, 'value': 0, 'attribute':0}\n",
    "        for p_id in generator.protein_data:\n",
    "            print('Attempting to calculate EV couplings standard protocol covariance for: {}'.format(p_id))\n",
    "            protein_dir = os.path.join(evc_standard_out_dir, p_id)\n",
    "            if not os.path.isdir(protein_dir):\n",
    "                os.makedirs(protein_dir)\n",
    "            protein_fn = os.path.join(protein_dir, '{}_Protein_Data.csv'.format(p_id))\n",
    "            if os.path.isfile(protein_fn):\n",
    "                protein_df = pd.read_csv(protein_fn, sep='\\t', header=0, index_col=False)\n",
    "            else:\n",
    "                try:\n",
    "#                     curr_aln = SeqAlignment(file_name=generator.protein_data[p_id]['Final_FA_Aln'], query_id=p_id,\n",
    "#                                             polymer_type='Protein')\n",
    "#                     curr_aln.import_alignment()\n",
    "#                     curr_evc = EVCouplingsWrapper(alignment=curr_aln, protocol='standard')\n",
    "#                     curr_evc.calculate_scores(out_dir=protein_dir, cores=10, delete_files=True)\n",
    "                    curr_evc = EVCouplingsWrapper(query=p_id, aln_file=generator.protein_data[p_id]['Final_FA_Aln'], protocol='standard', out_dir=protein_dir)\n",
    "                    curr_evc.calculate_scores(cores=10, delete_files=True)\n",
    "                    # Compute statistics for the final scores of the ET-MIp model\n",
    "                    protein_df, _, _ = protein_scorers[p_id]['Scorer_CB'].evaluate_predictor(\n",
    "                        predictor=curr_evc, verbosity=2, out_dir=protein_dir, dist='CB', biased_w2_ave=None,\n",
    "                        unbiased_w2_ave=None, processes=10, threshold=0.5, pos_size=2,\n",
    "                        rank_type='max', file_prefix='EVC_Standard_Scores_', plots=True)\n",
    "                    # Score Prediction Clustering\n",
    "                    _, evc_standard_coverage  = compute_rank_and_coverage(seq_length=curr_evc.alignment.seq_length, scores=curr_evc.scores, pos_size=2,\n",
    "                        rank_type='max')\n",
    "                    z_score_fn = os.path.join(protein_dir, 'EVC_Standard_Scores_Dist-Any_{}_ZScores.tsv')\n",
    "                    z_score_plot_fn = os.path.join(protein_dir, 'EVC_Standard_Scores_Dist-Any_{}_ZScores.png')\n",
    "                    z_score_biased, biased_w2_ave, biased_scw_z_auc = protein_scorers[p_id]['Scorer_Any'].score_clustering_of_contact_predictions(\n",
    "                        1.0 - evc_standard_coverage, bias=True, file_path=z_score_fn.format('Biased'),\n",
    "                        w2_ave_sub=protein_scorers[p_id]['biased_w2_ave'], processes=10)\n",
    "                    if protein_scorers[p_id]['biased_w2_ave'] is None:\n",
    "                        protein_scorers[p_id]['biased_w2_ave'] = biased_w2_ave\n",
    "                    biased_z_score_array = np.array(pd.to_numeric(z_score_biased['Z-Score'], errors='coerce'))\n",
    "                    protein_df['Max Biased Z-Score'] = np.nanmax(biased_z_score_array)\n",
    "                    protein_df['Biased Z-Score at 10%'] = biased_z_score_array[z_score_biased['Num_Residues'] >= float(len(protein_scorers[p_id]['Scorer_Any'].query_structure.seq[protein_scorers[p_id]['Scorer_Any'].best_chain])) * 0.10][0]\n",
    "                    protein_df['Biased Z-Score at 30%'] = biased_z_score_array[z_score_biased['Num_Residues'] >= float(len(protein_scorers[p_id]['Scorer_Any'].query_structure.seq[protein_scorers[p_id]['Scorer_Any'].best_chain])) * 0.30][0]\n",
    "                    protein_df['AUC Biased Z-Score'] = biased_scw_z_auc\n",
    "                    plot_z_scores(z_score_biased, z_score_plot_fn.format('Biased'))\n",
    "                    z_score_unbiased, unbiased_w2_ave, unbiased_scw_z_auc = protein_scorers[p_id]['Scorer_Any'].score_clustering_of_contact_predictions(\n",
    "                        1.0 - evc_standard_coverage, bias=False, file_path=z_score_fn.format('Unbiased'),\n",
    "                        w2_ave_sub=protein_scorers[p_id]['unbiased_w2_ave'], processes=10)\n",
    "                    if protein_scorers[p_id]['unbiased_w2_ave'] is None:\n",
    "                        protein_scorers[p_id]['unbiased_w2_ave'] = unbiased_w2_ave\n",
    "                    unbiased_z_score_array = np.array(pd.to_numeric(z_score_unbiased['Z-Score'], errors='coerce'))\n",
    "                    protein_df['Max Unbiased Z-Score'] = np.nanmax(unbiased_z_score_array)\n",
    "                    protein_df['Unbiased Z-Score at 10%'] = unbiased_z_score_array[z_score_unbiased['Num_Residues'] >= float(len(protein_scorers[p_id]['Scorer_Any'].query_structure.seq[protein_scorers[p_id]['Scorer_Any'].best_chain])) * 0.10][0]\n",
    "                    protein_df['Unbiased Z-Score at 30%'] = unbiased_z_score_array[z_score_unbiased['Num_Residues'] >= float(len(protein_scorers[p_id]['Scorer_Any'].query_structure.seq[protein_scorers[p_id]['Scorer_Any'].best_chain])) * 0.30][0]\n",
    "                    protein_df['AUC Unbiased Z-Score'] = unbiased_scw_z_auc\n",
    "                    plot_z_scores(z_score_unbiased, z_score_plot_fn.format('Unbiased'))\n",
    "                    # Record execution times\n",
    "                    protein_df['Init Time'] = None\n",
    "                    protein_df['Import Time'] = None\n",
    "                    protein_df['Dist Tree Time'] = None\n",
    "                    protein_df['Trace Time'] = None\n",
    "                    protein_df['Total Time'] = None\n",
    "                    # Record static data for this protein\n",
    "                    protein_df['Protein'] = p_id\n",
    "                    protein_df['Method'] = 'EVC Standard'\n",
    "                    protein_df['Protein Length'] = generator.protein_data[p_id]['Length']\n",
    "                    protein_df['Alignment Size'] = summary['Filtered_Alignment'].values[summary['Protein_ID'] == p_id][0]\n",
    "                    protein_df.to_csv(protein_fn, sep='\\t', header=True, index=False, columns=output_columns)\n",
    "                    print('Successfully computed EV couplings standard protocol covariance for: {}'.format(p_id))\n",
    "                    counts['success'] += 1\n",
    "                except ValueError:\n",
    "                    print('Could not compute EV couplings standard protocol covariance for: {} with seq_length: {} and size: {}'.format(\n",
    "                        p_id, curr_aln.seq_length, curr_aln.size))\n",
    "                    counts['value'] += 1\n",
    "                    continue\n",
    "                except AttributeError:\n",
    "                    print('Could not compute EV couplings standard protocol covariance for: {} no alignment'.format(p_id))\n",
    "                    counts['attribute'] += 1\n",
    "                    continue\n",
    "            if evc_standard_method_df is None:\n",
    "                evc_standard_method_df = protein_df\n",
    "            else:\n",
    "                evc_standard_method_df = evc_standard_method_df.append(protein_df)\n",
    "        print('{}\\tSuccesses\\n{}\\tValue Errors\\n{}\\tAttribute Errors'.format(counts['success'], counts['value'],\n",
    "                                                                             counts['attribute']))\n",
    "        evc_standard_method_df.to_csv(evc_standard_method_fn, sep='\\t', header=True, index=False, columns=output_columns)\n",
    "    if large_comparison_df is None:\n",
    "        large_comparison_df = evc_standard_method_df\n",
    "    else:\n",
    "        large_comparison_df = large_comparison_df.append(evc_standard_method_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Scoring the covariation of the proteins using the EVCouplings method mean field protocol."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.isfile(large_comparison_fn):\n",
    "    evc_mf_out_dir = os.path.join(large_set_out_dir, 'EVCouplings_Mean_Field')\n",
    "    if not os.path.isdir(evc_mf_out_dir):\n",
    "        os.makedirs(evc_mf_out_dir)\n",
    "    evc_mf_method_fn = os.path.join(evc_mf_out_dir, 'EVCouplings_Mean_Field_Method_Data.csv')\n",
    "    if os.path.isfile(evc_mf_method_fn):\n",
    "        evc_mf_method_df = pd.read_csv(evc_mf_method_fn, sep='\\t', header=0, index_col=False)\n",
    "    else:\n",
    "        evc_mf_method_df = None\n",
    "        counts = {'success':0, 'value': 0, 'attribute':0}\n",
    "        for p_id in generator.protein_data:\n",
    "            print('Attempting to calculate EV couplings covariance for: {}'.format(p_id))\n",
    "            try:\n",
    "                protein_dir = os.path.join(evc_mf_out_dir, p_id)\n",
    "                if not os.path.isdir(protein_dir):\n",
    "                    os.makedirs(protein_dir)\n",
    "#                 curr_aln = SeqAlignment(file_name=generator.protein_data[p_id]['Final_FA_Aln'], query_id=p_id,\n",
    "#                                         polymer_type='Protein')\n",
    "#                 curr_aln.import_alignment()\n",
    "#                 curr_evc = EVCouplingsWrapper(alignment=curr_aln, protocol='mean_field')\n",
    "#                 curr_evc.calculate_scores(out_dir=protein_dir, cores=10, delete_files=True)\n",
    "                curr_evc = EVCouplingsWrapper(query=p_id, aln_file=generator.protein_data[p_id]['Final_FA_Aln'], protocol='mean_field', out_dir=protein_dir)\n",
    "                curr_evc.calculate_scores(cores=10, delete_files=True)\n",
    "                # Compute statistics for the final scores of the ET-MIp model\n",
    "                protein_df, _, _ = protein_scorers[p_id]['Scorer_CB'].evaluate_predictor(\n",
    "                    predictor=curr_evc, verbosity=2, out_dir=protein_dir, dist='CB', biased_w2_ave=None,\n",
    "                    unbiased_w2_ave=None, processes=10, threshold=0.5, pos_size=2, rank_type='max',\n",
    "                    file_prefix='EVC_Standard_Scores_', plots=True)\n",
    "                # Score Prediction Clustering\n",
    "                _, evc_mf_coverage  = compute_rank_and_coverage(seq_length=curr_evc.alignment.seq_length, scores=curr_evc.scores, pos_size=2,\n",
    "                    rank_type='max')\n",
    "                z_score_fn = os.path.join(protein_dir, 'EVC_Mean_Field_Scores_Dist-Any_{}_ZScores.tsv')\n",
    "                z_score_plot_fn = os.path.join(protein_dir, 'EVC_Mean_Field_Scores_Dist-Any_{}_ZScores.png')\n",
    "                z_score_biased, biased_w2_ave, biased_scw_z_auc = protein_scorers[p_id]['Scorer_Any'].score_clustering_of_contact_predictions(\n",
    "                    1.0 - evc_mf_coverage, bias=True, file_path=z_score_fn.format('Biased'),\n",
    "                    w2_ave_sub=protein_scorers[p_id]['biased_w2_ave'], processes=10)\n",
    "                if protein_scorers[p_id]['biased_w2_ave'] is None:\n",
    "                        protein_scorers[p_id]['biased_w2_ave'] = biased_w2_ave\n",
    "                biased_z_score_array = np.array(pd.to_numeric(z_score_biased['Z-Score'], errors='coerce'))\n",
    "                protein_df['Max Biased Z-Score'] = np.nanmax(biased_z_score_array)\n",
    "                protein_df['Biased Z-Score at 10%'] = biased_z_score_array[z_score_biased['Num_Residues'] >= float(len(protein_scorers[p_id]['Scorer_Any'].query_structure.seq[protein_scorers[p_id]['Scorer_Any'].best_chain])) * 0.10][0]\n",
    "                protein_df['Biased Z-Score at 30%'] = biased_z_score_array[z_score_biased['Num_Residues'] >= float(len(protein_scorers[p_id]['Scorer_Any'].query_structure.seq[protein_scorers[p_id]['Scorer_Any'].best_chain])) * 0.30][0]\n",
    "                protein_df['AUC Biased Z-Score'] = biased_scw_z_auc\n",
    "                plot_z_scores(z_score_biased, z_score_plot_fn.format('Biased'))\n",
    "                z_score_unbiased, unbiased_w2_ave, unbiased_scw_z_auc = protein_scorers[p_id]['Scorer_Any'].score_clustering_of_contact_predictions(\n",
    "                    1.0 - evc_mf_coverage, bias=False, file_path=z_score_fn.format('Unbiased'),\n",
    "                    w2_ave_sub=protein_scorers[p_id]['unbiased_w2_ave'], processes=10)\n",
    "                if protein_scorers[p_id]['unbiased_w2_ave'] is None:\n",
    "                        protein_scorers[p_id]['unbiased_w2_ave'] = unbiased_w2_ave\n",
    "                unbiased_z_score_array = np.array(pd.to_numeric(z_score_unbiased['Z-Score'], errors='coerce'))\n",
    "                protein_df['Max Unbiased Z-Score'] = np.nanmax(unbiased_z_score_array)\n",
    "                protein_df['Unbiased Z-Score at 10%'] = unbiased_z_score_array[z_score_unbiased['Num_Residues'] >= float(len(protein_scorers[p_id]['Scorer_Any'].query_structure.seq[protein_scorers[p_id]['Scorer_Any'].best_chain])) * 0.10][0]\n",
    "                protein_df['Unbiased Z-Score at 30%'] = unbiased_z_score_array[z_score_unbiased['Num_Residues'] >= float(len(protein_scorers[p_id]['Scorer_Any'].query_structure.seq[protein_scorers[p_id]['Scorer_Any'].best_chain])) * 0.30][0]\n",
    "                protein_df['AUC Unbiased Z-Score'] = unbiased_scw_z_auc\n",
    "                plot_z_scores(z_score_unbiased, z_score_plot_fn.format('Unbiased'))\n",
    "                # Record execution times\n",
    "                protein_df['Init Time'] = None\n",
    "                protein_df['Import Time'] = None\n",
    "                protein_df['Dist Tree Time'] = None\n",
    "                protein_df['Trace Time'] = None\n",
    "                protein_df['Total Time'] = None\n",
    "                # Record static data for this protein\n",
    "                protein_df['Protein'] = p_id\n",
    "                protein_df['Method'] = 'EVC Mean Field'\n",
    "                protein_df['Protein Length'] = generator.protein_data[p_id]['Length']\n",
    "                protein_df['Alignment Size'] = summary['Filtered_Alignment'].values[summary['Protein_ID'] == p_id][0]\n",
    "                protein_df.to_csv(protein_fn, sep='\\t', header=True, index=False, columns=output_columns)\n",
    "                print('Successfully computed EV couplings covariance for: {}'.format(p_id))\n",
    "                counts['success'] += 1\n",
    "            except ValueError:\n",
    "                print('Could not compute EV couplings covariance for: {} with seq_length: {} and size: {}'.format(\n",
    "                    p_id, curr_aln.seq_length, curr_aln.size))\n",
    "                counts['value'] += 1\n",
    "                continue\n",
    "            except AttributeError:\n",
    "                print('Could not compute EV couplings covariance for: {} no alignment'.format(p_id))\n",
    "                counts['attribute'] += 1\n",
    "                continue\n",
    "            if evc_mf_method_df is None:\n",
    "                evc_mf_method_df = protein_df\n",
    "            else:\n",
    "                evc_mf_method_df = evc_mf_method_df.append(protein_df)\n",
    "        print('{}\\tSuccesses\\n{}\\tValue Errors\\n{}\\tAttribute Errors'.format(counts['success'], counts['value'],\n",
    "                                                                             counts['attribute']))\n",
    "        evc_mf_method_df.to_csv(evc_mf_method_fn, sep='\\t', header=True, index=False, columns=output_columns)\n",
    "    if large_comparison_df is None:\n",
    "        large_comparison_df = evc_mf_method_df\n",
    "    else:\n",
    "        large_comparison_df = large_comparison_df.append(evc_mf_method_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write out final comparison data so it can be loaded later for generating figures.\n",
    "large_comparison_df.to_csv(large_comparison_fn, sep='\\t', header=True, index=False, columns=output_columns)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python3 (PyET3)",
   "language": "python",
   "name": "pyet3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
