{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Method Characteriation #\n",
    "## Goal ##\n",
    "The goal of this test set is to perform proof of concept testing on a small number of proteins with a wide range of sizes and available homologs, orthologs, and paralogs. By doing so it should be possible to test the best parameterization for this tool as well as identifying the strengths and weaknesses of the tool using various measurments as end points.\n",
    "## Warning ##\n",
    "Before attempting to use this notebook make sure that your .env file has been properly setup to reflect the correct locations of command line tools and the location of files and directories needed for execution.\n",
    "### Initial Import###\n",
    "This first cell performs the necessary imports required to begin this notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from dotenv import find_dotenv, load_dotenv\n",
    "try:\n",
    "    dotenv_path = find_dotenv(raise_error_if_not_found=True)\n",
    "except IOError:\n",
    "    dotenv_path = find_dotenv(raise_error_if_not_found=True, usecwd=True)\n",
    "load_dotenv(dotenv_path)\n",
    "import os\n",
    "import sys\n",
    "sys.path.append(os.path.join(os.environ.get('PROJECT_PATH'), 'src'))\n",
    "sys.path.append(os.path.join(os.environ.get('PROJECT_PATH'), 'src', 'SupportingClasses'))\n",
    "input_dir = os.environ.get('INPUT_PATH')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Set Construction ##\n",
    "The first task required to test the data set is to download the required data and construct any necessary input files for all down stream analyses.\n",
    "In this case that means:\n",
    "* Downloading PDB files for the proteins in our small test set.\n",
    "* Extracting a query sequence from each PDB file.\n",
    "* Searching for paralogs, homologs, and orthologs in a custom BLAST database built by filtering the Uniref90 database.\n",
    "* Filtering the hits from the BLAST search to meet minimum and maximum length requirements, as well as minimum and maximum identity requirements.\n",
    "* Building alignments using CLUSTALW in both the fasta and msf formats since some of the tools which will be used for comparison need different formats.\n",
    "* Filtering the alignment for maximum identity similarity between seqeunces.\n",
    "* Re-aligning the filtered sequences using CLUSTALW.\n",
    "This is all handeled by the DataSetGenerator class found in the src/SupportingClasses folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Importing protein list\n",
      "Downloading structures and parsing in query sequences\n",
      "BLASTing query sequences\n",
      "Filtering BLAST hits, aligning, filtering by identity, and re-aligning\n",
      "   Protein_ID       UniProt  BLAST_Hits  Filtered_BLAST  Filtered_Alignment  \\\n",
      "21       2b59    CIPA_CLOTM        1606               4                   4   \n",
      "5        206l      LYS_BPT4        1039             131                  92   \n",
      "16       1c0k    OXDA_RHOTO        2500              51                  51   \n",
      "7        1bol    RNRH_RHINI        2500             131                 127   \n",
      "3        7hvp     POL_HV1A2        2500              33                  33   \n",
      "1        1c17    ATPL_ECOLI        1671             850                 813   \n",
      "14       1jwl    LACI_ECOLI        2500             228                 226   \n",
      "8        3q05     P53_HUMAN         932             306                 210   \n",
      "4        135l    LYSC_MELGA        1913             853                 818   \n",
      "13       2z0e   ATG4B_HUMAN        2111             360                 344   \n",
      "10       2rh1   ADRB2_HUMAN        2500             416                 399   \n",
      "15       1a26   PARP1_CHICK        2500             267                 261   \n",
      "12       3b6v   KIF3C_HUMAN        2500             481                 441   \n",
      "22       1h1v    GELS_HUMAN        2500             654                 599   \n",
      "0        2ysd   MAGI1_HUMAN        2500             412                 318   \n",
      "11       1hck    CDK2_HUMAN        2500            2468                1681   \n",
      "9        1axb    BLAT_ECOLI        2500            2102                2087   \n",
      "17       4lli   MYO5A_HUMAN        2500             441                 352   \n",
      "2        3tnu   K1C14_HUMAN        2500            1681                1518   \n",
      "20       2zxe  Q4H132_SQUAC        2500             880                 795   \n",
      "6        2wer   HSP82_YEAST        2500            1327                1207   \n",
      "18       4ycu     SYK_HUMAN        2500            2421                2347   \n",
      "19       2iop    HTPG_ECOLI        2500            2478                2472   \n",
      "\n",
      "    Length  Total_Size  \n",
      "21    1853      7412.0  \n",
      "5      164     15088.0  \n",
      "16     368     18768.0  \n",
      "7      238     30226.0  \n",
      "3     1437     47421.0  \n",
      "1       79     64227.0  \n",
      "14     360     81360.0  \n",
      "8      393     82530.0  \n",
      "4      147    120246.0  \n",
      "13     393    135192.0  \n",
      "10     413    164787.0  \n",
      "15    1011    263871.0  \n",
      "12     793    349713.0  \n",
      "22     782    468418.0  \n",
      "0     1491    474138.0  \n",
      "11     298    500938.0  \n",
      "9      286    596882.0  \n",
      "17    1855    652960.0  \n",
      "2      472    716496.0  \n",
      "20    1028    817260.0  \n",
      "6      709    855763.0  \n",
      "18     597   1401159.0  \n",
      "19     624   1542528.0  \n",
      "It took 0.15634393692016602 min to generate the data set.\n"
     ]
    }
   ],
   "source": [
    "from time import time\n",
    "from DataSetGenerator import DataSetGenerator\n",
    "protein_list_dir = os.path.join(input_dir, 'ProteinLists')\n",
    "if not os.path.isdir(protein_list_dir):\n",
    "    os.makedirs(protein_list_dir)\n",
    "small_list_fn = os.path.join(protein_list_dir, 'SmallDataSet.txt')\n",
    "if not os.path.isfile(small_list_fn):\n",
    "    proteins_of_interest = ['2ysdA', '1c17A', '3tnuA', '7hvpA', '135lA', '206lA', '2werA', '1bolA', '3q05A', '1axbA',\n",
    "                            '2rh1A', '1hckA', '3b6vA', '2z0eA', '1jwlA', '1a26A', '1c0kA', '4lliA', '4ycuA', '2iopA',\n",
    "                            '2zxeA', '2b59B', '1h1vG']\n",
    "    with open(small_list_fn, 'w') as small_list_handle:\n",
    "        for p_id in proteins_of_interest:\n",
    "            small_list_handle.write('{}\\n'.format(p_id))\n",
    "generator = DataSetGenerator(input_dir)\n",
    "start = time()\n",
    "summary = generator.build_pdb_alignment_dataset(protein_list_fn=os.path.basename(small_list_fn), num_threads=10,\n",
    "                                                database='customuniref90.fasta', max_target_seqs=2500, remote=False, verbose=False)\n",
    "summary['Accession'] = summary['Protein_ID'].apply(lambda x: generator.protein_data[x]['Accession'])\n",
    "summary['Length'] = summary['Protein_ID'].apply(lambda x: generator.protein_data[x]['Length'])\n",
    "summary['Total_Size'] = summary.apply(lambda x: float(x['Length']) * float(x['Filtered_Alignment']), axis=1)\n",
    "summary.sort_values(by=['Total_Size', 'Length', 'Filtered_Alignment'], axis=0, inplace=True)\n",
    "print(summary[['Protein_ID', 'Accession', 'BLAST_Hits', 'Filtered_BLAST', 'Filtered_Alignment', 'Length', 'Total_Size']])\n",
    "end = time()\n",
    "print('It took {} min to generate the data set.'.format((end - start) / 60.0))\n",
    "summary.to_csv(os.path.join(input_dir, 'small_data_set_summary.tsv'), sep='\\t', index=False, header=True,\n",
    "               columns=['Protein_ID', 'Accession', 'BLAST_Hits', 'Filtered_BLAST', 'Filtered_Alignment', 'Length', 'Total_Size'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a location to store the output of this parameter tuning."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_dir = os.environ.get('OUTPUT_PATH')\n",
    "characterization_out_dir = os.path.join(output_dir, 'Characterization')\n",
    "if not os.path.isdir(characterization_out_dir):\n",
    "    os.makedirs(characterization_out_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Method Characterization##\n",
    "This section performs the Evoluationary Trace method for covariation ('pair's of residues) with different parameters to help determine which provide the optimal behavior (in this case measured by ability to predict structural contacts: AUROC, Precision, and clustering: Biased SCW Z-Score, Unbiased SCW Z-Score).\n",
    "### Distance Metric Parameters###\n",
    "* identity', False - Uses the identity metric to compute the distance between the sequences in the provided alignment.\n",
    "* 'blosum62', True - Uses the similarity metric (as defined by the off diagonal values from the 'blosum62' distance matrix) to compute the distance between the sequences in the provided alignment.\n",
    "* 'blosum62', False - Uses the 'blosum62' scoring matrix to compute the edit distance between sequences in the provided alignment.\n",
    "### Tree Construction Parameters###\n",
    "* 'et' - A phylogenetic tree related to the UPGMA tree, but where the distance update does not use the average of columns from the current step, but rather the average distance between all contributing terminal nodes. This has been used in previous methods by our group.\n",
    "* 'upgma' - A phylogenetic tree constructed using the standard UPGMA algorithm.\n",
    "* 'agglomerative' (affinity='euclidean', linkage='ward') - A tree constructed based on agglomerative/hierarchical clustering over the distance matrix using the specified affinity and linkage.\n",
    "### Scoring Parameters###\n",
    "* 'identity' - A binary scoring of invariance at each level and node in the tree, which yields an integer score for each pair of positions (lower score means the pair position was fixed higher in the tree and therefore more important, higher score means the pair of positions became fixed lower down in the score and is therefore less important).\n",
    "* 'plain_entropy' - The joint entropy between a pair of positions. This provides a real valued (floating point score) for each pair of positions, which can be interpreted the same way as the 'identity' scoring metric but with greater resolution/ability to separate positions.\n",
    "* 'mutual_information' - The mutual informaiton score for each pair of positions within a node of the phylogenetic tree. This is built up over levels of the tree using the trace methodology, in this case the higher the score the better.\n",
    "* 'normalized_mutual_information' - The normalized mutual informaiton score for each pair of positions within a node of the phylogenetic tree. This is built up over levels of the tree using the trace methodology, in this case the higher the score the better.\n",
    "* 'average_product_corrected_mutual_information' - The average product corrected mutual informaiton (MIp) score for each pair of positions within a node of the phylogenetic tree. This is built up over levels of the tree using the trace methodology, in this case the higher the score the better (some low scores my be negative).\n",
    "* 'filtered_average_product_corrected_mutual_information' - The average product corrected mutual informaiton (MIp) score for each pair of positions within a node of the phylogenetic tree. This is built up over levels of the tree using the trace methodology, in this case the higher the score the better (some low scores my be negative). This score is additionally filtered at the node level so that positions with a mutual information <= 0.0001 are set to 0 (this was done in a previously released paper by our lab on this topic)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Characterizing protein: 2b59\n",
      "Distance model: identity, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Importing the PDB file took 0.0017962058385213215 min\n",
      "<class 'SupportingClasses.SeqAlignment.SeqAlignment'>\n",
      "Removing gaps took 0.0005907098452250163 min\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 2960.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 3114.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 2960.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 3114.\n",
      "  PDBConstructionWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Importing the PDB file took 0.0013850212097167968 min\n",
      "Mapping query sequence and pdb took 0.0027697404225667317 min\n",
      "Computing the distance matrix based on the PDB file took 0.0039414207140604654 min\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: True\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Characterizing protein: 206l\n",
      "Distance model: identity, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Importing the PDB file took 0.0010275761286417642 min\n",
      "<class 'SupportingClasses.SeqAlignment.SeqAlignment'>\n",
      "Removing gaps took 0.002173348267873128 min\n",
      "Importing the PDB file took 0.0005202015240987142 min\n",
      "Mapping query sequence and pdb took 0.0031191349029541016 min\n",
      "Computing the distance matrix based on the PDB file took 0.004554923375447591 min\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: True\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Characterizing protein: 1c0k\n",
      "Distance model: identity, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Importing the PDB file took 0.0028552730878194173 min\n",
      "<class 'SupportingClasses.SeqAlignment.SeqAlignment'>\n",
      "Removing gaps took 0.0012175440788269043 min\n",
      "Importing the PDB file took 0.0015178720156351726 min\n",
      "Mapping query sequence and pdb took 0.0033608078956604004 min\n",
      "Computing the distance matrix based on the PDB file took 0.016396959622701008 min\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: True\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Characterizing protein: 1bol\n",
      "Distance model: identity, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Importing the PDB file took 0.0008846243222554524 min\n",
      "<class 'SupportingClasses.SeqAlignment.SeqAlignment'>\n",
      "Removing gaps took 0.0010811368624369304 min\n",
      "Importing the PDB file took 0.0004590193430582682 min\n",
      "Mapping query sequence and pdb took 0.00204011599222819 min\n",
      "Computing the distance matrix based on the PDB file took 0.0063018004099528 min\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: True\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Characterizing protein: 7hvp\n",
      "Distance model: identity, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Importing the PDB file took 0.001327045758565267 min\n",
      "<class 'SupportingClasses.SeqAlignment.SeqAlignment'>\n",
      "Removing gaps took 0.0008839805920918783 min\n",
      "Importing the PDB file took 0.0006072044372558594 min\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 2100.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 2138.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain C is discontinuous at line 2191.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 2100.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 2138.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain C is discontinuous at line 2191.\n",
      "  PDBConstructionWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mapping query sequence and pdb took 0.0021351774533589682 min\n",
      "Computing the distance matrix based on the PDB file took 0.0015633900960286458 min\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: True\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Characterizing protein: 1c17\n",
      "Distance model: identity, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Importing the PDB file took 0.006720646222432455 min\n",
      "<class 'SupportingClasses.SeqAlignment.SeqAlignment'>\n",
      "Removing gaps took 0.0021963119506835938 min\n",
      "Importing the PDB file took 0.0031335512797037762 min\n",
      "Mapping query sequence and pdb took 0.005725649992624918 min\n",
      "Computing the distance matrix based on the PDB file took 0.0009480675061543782 min\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: True\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Characterizing protein: 1jwl\n",
      "Distance model: identity, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Importing the PDB file took 0.0031699180603027345 min\n",
      "<class 'SupportingClasses.SeqAlignment.SeqAlignment'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 8081.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain C is discontinuous at line 8101.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 8121.\n",
      "  PDBConstructionWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Removing gaps took 0.002110143502553304 min\n",
      "Importing the PDB file took 0.0015636205673217774 min\n",
      "Mapping query sequence and pdb took 0.00416949192682902 min\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 8081.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain C is discontinuous at line 8101.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 8121.\n",
      "  PDBConstructionWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Computing the distance matrix based on the PDB file took 0.014155419667561848 min\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: True\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Characterizing protein: 3q05\n",
      "Distance model: identity, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 9163.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain C is discontinuous at line 9164.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain D is discontinuous at line 9165.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 9166.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 9167.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain C is discontinuous at line 9201.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain D is discontinuous at line 9267.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 9314.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain K is discontinuous at line 9321.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain L is discontinuous at line 9325.\n",
      "  PDBConstructionWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Importing the PDB file took 0.003706196943918864 min\n",
      "<class 'SupportingClasses.SeqAlignment.SeqAlignment'>\n",
      "Removing gaps took 0.00602957010269165 min\n",
      "Importing the PDB file took 0.0019708553949991862 min\n",
      "Mapping query sequence and pdb took 0.008848496278127034 min\n",
      "Computing the distance matrix based on the PDB file took 0.007801071802775065 min"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 9163.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain C is discontinuous at line 9164.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain D is discontinuous at line 9165.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 9166.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 9167.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain C is discontinuous at line 9201.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain D is discontinuous at line 9267.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 9314.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain K is discontinuous at line 9321.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain L is discontinuous at line 9325.\n",
      "  PDBConstructionWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: True\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Characterizing protein: 135l\n",
      "Distance model: identity, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Importing the PDB file took 0.0006554285685221354 min\n",
      "<class 'SupportingClasses.SeqAlignment.SeqAlignment'>\n",
      "Removing gaps took 0.0108033816019694 min\n",
      "Importing the PDB file took 0.00027779340744018557 min\n",
      "Mapping query sequence and pdb took 0.011637671788533529 min\n",
      "Computing the distance matrix based on the PDB file took 0.001934957504272461 min\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: True\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Characterizing protein: 2z0e\n",
      "Distance model: identity, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Importing the PDB file took 0.0018515427907307943 min\n",
      "<class 'SupportingClasses.SeqAlignment.SeqAlignment'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 3897.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 4034.\n",
      "  PDBConstructionWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Removing gaps took 0.010118826230367025 min\n",
      "Importing the PDB file took 0.0007925192515055338 min\n",
      "Mapping query sequence and pdb took 0.01166535218556722 min\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 3897.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 4034.\n",
      "  PDBConstructionWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Computing the distance matrix based on the PDB file took 0.011488986015319825 min\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: True\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Characterizing protein: 2rh1\n",
      "Distance model: identity, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Importing the PDB file took 0.0029030323028564455 min\n",
      "<class 'SupportingClasses.SeqAlignment.SeqAlignment'>\n",
      "Removing gaps took 0.008225754896799723 min\n",
      "Importing the PDB file took 0.0009191513061523437 min\n",
      "Mapping query sequence and pdb took 0.017950967947642008 min\n",
      "Computing the distance matrix based on the PDB file took 0.02283240556716919 min\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: True\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Characterizing protein: 1a26\n",
      "Distance model: identity, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Importing the PDB file took 0.0016844987869262696 min\n",
      "<class 'SupportingClasses.SeqAlignment.SeqAlignment'>\n",
      "Removing gaps took 0.017734567324320476 min\n",
      "Importing the PDB file took 0.0006439884503682455 min\n",
      "Mapping query sequence and pdb took 0.019372018178304036 min\n",
      "Computing the distance matrix based on the PDB file took 0.014081116517384846 min\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: True\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Characterizing protein: 3b6v\n",
      "Distance model: identity, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Importing the PDB file took 0.0026808579762776694 min\n",
      "<class 'SupportingClasses.SeqAlignment.SeqAlignment'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 5418.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/Atom.py:189: PDBConstructionWarning: Used element 'U' for Atom (name=UNK) with given element 'X'\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 5449.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 5482.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 5486.\n",
      "  PDBConstructionWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Removing gaps took 0.028707818190256754 min\n",
      "Importing the PDB file took 0.0010918021202087403 min\n",
      "Mapping query sequence and pdb took 0.03124358654022217 min\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 5418.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/Atom.py:189: PDBConstructionWarning: Used element 'U' for Atom (name=UNK) with given element 'X'\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 5449.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 5482.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 5486.\n",
      "  PDBConstructionWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Computing the distance matrix based on the PDB file took 0.01274248758951823 min\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: True\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Characterizing protein: 1h1v\n",
      "Distance model: identity, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Importing the PDB file took 0.0026745319366455076 min\n",
      "<class 'SupportingClasses.SeqAlignment.SeqAlignment'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 6161.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain G is discontinuous at line 6193.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 6197.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain G is discontinuous at line 6393.\n",
      "  PDBConstructionWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Removing gaps took 0.03830460707346598 min\n",
      "Importing the PDB file took 0.0012879610061645509 min\n",
      "Mapping query sequence and pdb took 0.04090494712193807 min\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 6161.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain G is discontinuous at line 6193.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 6197.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain G is discontinuous at line 6393.\n",
      "  PDBConstructionWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Computing the distance matrix based on the PDB file took 0.01266096035639445 min\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: True\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Characterizing protein: 2ysd\n",
      "Distance model: identity, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Importing the PDB file took 0.006450204054514567 min\n",
      "<class 'SupportingClasses.SeqAlignment.SeqAlignment'>\n",
      "Removing gaps took 0.028473830223083495 min\n",
      "Importing the PDB file took 0.004678893089294434 min\n",
      "Mapping query sequence and pdb took 0.047457293669382734 min\n",
      "Computing the distance matrix based on the PDB file took 0.0005194346110026041 min\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: True\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Characterizing protein: 1hck\n",
      "Distance model: identity, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Importing the PDB file took 0.0012215058008829753 min\n",
      "<class 'SupportingClasses.SeqAlignment.SeqAlignment'>\n",
      "Removing gaps took 0.05669077237447103 min\n",
      "Importing the PDB file took 0.0005538503328959148 min\n",
      "Mapping query sequence and pdb took 0.05932676394780477 min\n",
      "Computing the distance matrix based on the PDB file took 0.009823469320933025 min\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: True\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Characterizing protein: 1axb\n",
      "Distance model: identity, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Importing the PDB file took 0.0010283986727396648 min\n",
      "<class 'SupportingClasses.SeqAlignment.SeqAlignment'>\n",
      "Removing gaps took 0.04774384498596192 min\n",
      "Importing the PDB file took 0.00047981739044189453 min\n",
      "Mapping query sequence and pdb took 0.04914156198501587 min\n",
      "Computing the distance matrix based on the PDB file took 0.008520174026489257 min\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: True\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Characterizing protein: 4lli\n",
      "Distance model: identity, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Importing the PDB file took 0.0028051455815633138 min\n",
      "<class 'SupportingClasses.SeqAlignment.SeqAlignment'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 6531.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 6714.\n",
      "  PDBConstructionWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Removing gaps took 0.0422515074412028 min\n",
      "Importing the PDB file took 0.001342904567718506 min\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 6531.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 6714.\n",
      "  PDBConstructionWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mapping query sequence and pdb took 0.04684976736704508 min\n",
      "Computing the distance matrix based on the PDB file took 0.021047155062357586 min\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: True\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Characterizing protein: 3tnu\n",
      "Distance model: identity, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Importing the PDB file took 0.001785135269165039 min\n",
      "<class 'SupportingClasses.SeqAlignment.SeqAlignment'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 3478.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 3488.\n",
      "  PDBConstructionWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Removing gaps took 0.06559314330418904 min\n",
      "Importing the PDB file took 0.0005948146184285482 min\n",
      "Mapping query sequence and pdb took 0.06802994807561238 min\n",
      "Computing the distance matrix based on the PDB file took 0.001495679219563802 min\n",
      "Scoring metric: plain_entropy\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 3478.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 3488.\n",
      "  PDBConstructionWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: True\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Characterizing protein: 2zxe\n",
      "Distance model: identity, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 11066.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 11103.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 11145.\n",
      "  PDBConstructionWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Importing the PDB file took 0.0041049599647521974 min\n",
      "<class 'SupportingClasses.SeqAlignment.SeqAlignment'>\n",
      "Removing gaps took 0.07236009438832601 min\n",
      "Importing the PDB file took 0.0021209200223286945 min\n",
      "Mapping query sequence and pdb took 0.07675887743631998 min\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 11066.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 11103.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 11145.\n",
      "  PDBConstructionWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Computing the distance matrix based on the PDB file took 0.11188989082972209 min\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: True\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Characterizing protein: 2wer\n",
      "Distance model: identity, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Importing the PDB file took 0.0025065382321675616 min\n",
      "<class 'SupportingClasses.SeqAlignment.SeqAlignment'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 4201.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 4226.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 4251.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 4395.\n",
      "  PDBConstructionWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Removing gaps took 0.08637921412785848 min\n",
      "Importing the PDB file took 0.0008933305740356446 min\n",
      "Mapping query sequence and pdb took 0.08946357170740764 min\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 4201.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 4226.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 4251.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 4395.\n",
      "  PDBConstructionWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Computing the distance matrix based on the PDB file took 0.005767917633056641 min\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: True\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Characterizing protein: 4ycu\n",
      "Distance model: identity, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 17206.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 17292.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 17378.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 18052.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain C is discontinuous at line 18724.\n",
      "  PDBConstructionWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Importing the PDB file took 0.005696499347686767 min\n",
      "<class 'SupportingClasses.SeqAlignment.SeqAlignment'>\n",
      "Removing gaps took 0.13378031651178995 min\n",
      "Importing the PDB file took 0.003286643822987874 min\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 17206.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 17292.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 17378.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 18052.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain C is discontinuous at line 18724.\n",
      "  PDBConstructionWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mapping query sequence and pdb took 0.13995036284128826 min\n",
      "Computing the distance matrix based on the PDB file took 0.04038008451461792 min\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: True\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Characterizing protein: 2iop\n",
      "Distance model: identity, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 20520.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 20547.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain C is discontinuous at line 20574.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain D is discontinuous at line 20601.\n",
      "  PDBConstructionWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Importing the PDB file took 0.0062994877497355144 min\n",
      "<class 'SupportingClasses.SeqAlignment.SeqAlignment'>\n",
      "Removing gaps took 0.09878825743993123 min\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 20520.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 20547.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain C is discontinuous at line 20574.\n",
      "  PDBConstructionWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:91: PDBConstructionWarning: WARNING: Chain D is discontinuous at line 20601.\n",
      "  PDBConstructionWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Importing the PDB file took 0.004011897246042887 min\n",
      "Mapping query sequence and pdb took 0.10498187144597372 min\n",
      "Computing the distance matrix based on the PDB file took 0.045100049177805586 min\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: True\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Distance model: blosum62, ET dist: False\n",
      "Tree construction: et\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: upgma\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Tree construction: agglomerative\n",
      "Scoring metric: identity\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: plain_entropy\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: normalized_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n",
      "Scoring metric: filtered_average_product_corrected_mutual_information\n",
      "Evolutionary Trace analysis with the same parameters already saved to this location.\n"
     ]
    }
   ],
   "source": [
    "%matplotlib inline\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from EvolutionaryTrace import EvolutionaryTrace\n",
    "from SupportingClasses.PDBReference import PDBReference\n",
    "from SupportingClasses.ContactScorer import ContactScorer\n",
    "characterization_fn = os.path.join(characterization_out_dir, 'Characterization_Data.csv')\n",
    "if os.path.isfile(characterization_fn):\n",
    "    characterization_df = pd.read_csv(characterization_fn, sep='\\t', header=0, index_col=False)\n",
    "else:\n",
    "    characterization_df = None\n",
    "    for p_id in summary['Protein_ID']:\n",
    "        protein_df = None\n",
    "        contact_scorer = None\n",
    "        biased_w2_ave = None\n",
    "        unbiased_w2_ave = None\n",
    "        print('Characterizing protein: {}'.format(p_id))\n",
    "        protein_dir = os.path.join(characterization_out_dir, p_id)\n",
    "        if not os.path.isdir(protein_dir):\n",
    "            os.mkdir(protein_dir)\n",
    "        for dist_model, et_dist in [('identity', False), ('blosum62', True), ('blosum62', False)]:\n",
    "            print('Distance model: {}, ET dist: {}'.format(dist_model, et_dist))\n",
    "            for tree_building, tree_options in [('et', {}), ('upgma', {}),\n",
    "                                                ('agglomerative', {'affinity': 'euclidean', 'linkage': 'ward'})]:\n",
    "                print('Tree construction: {}'.format(tree_building))\n",
    "                dist_tree_dir = os.path.join(\n",
    "                    protein_dir, '{}{}{}{}'.format(dist_model, ('_ET' if et_dist else ''), tree_building,\n",
    "                    ('_'.join(['{}_{}'.format(k, v) for k,v in tree_options.items()]) if tree_options else '')))\n",
    "                if not os.path.isdir(dist_tree_dir):\n",
    "                    os.mkdir(dist_tree_dir)\n",
    "                for scoring_metric in ['identity', 'plain_entropy', 'mutual_information', 'normalized_mutual_information',\n",
    "                                       'average_product_corrected_mutual_information',\n",
    "                                       'filtered_average_product_corrected_mutual_information']:\n",
    "                    print('Scoring metric: {}'.format(scoring_metric))\n",
    "                    curr_et = EvolutionaryTrace(query_id=p_id, polymer_type='Protein',\n",
    "                                                aln_fn=generator.protein_data[p_id]['Final_FA_Aln'], et_distance=et_dist,\n",
    "                                                distance_model=dist_model, tree_building_method=tree_building,\n",
    "                                                tree_building_options=tree_options, ranks=None, position_type='pair',\n",
    "                                                scoring_metric=scoring_metric, gap_correction=None, out_dir=protein_dir,\n",
    "                                                output_files={'original_aln', 'non_gap_aln', 'tree', 'scores'},\n",
    "                                                processors=10, low_memory=True)\n",
    "                    curr_et.import_and_process_aln()\n",
    "                    curr_et.out_dir = dist_tree_dir\n",
    "                    curr_et.compute_distance_matrix_tree_and_assignments()\n",
    "                    curr_et.perform_trace()\n",
    "                    if contact_scorer is None:\n",
    "                        pdb_structure = PDBReference(pdb_file=generator.protein_data[p_id]['PDB'])\n",
    "                        pdb_structure.import_pdb(structure_id=p_id)\n",
    "                        print(type(curr_et.non_gapped_aln))\n",
    "                        contact_scorer = ContactScorer(query=p_id, seq_alignment=curr_et.non_gapped_aln,\n",
    "                                                       pdb_reference=pdb_structure, cutoff=8.0)\n",
    "                        contact_scorer.best_chain = generator.protein_data[p_id]['Chain']\n",
    "                        contact_scorer.fit()\n",
    "                        contact_scorer.measure_distance(method='Any')\n",
    "                    curr_df, biased_dict, unbiased_dict = contact_scorer.evaluate_predictor(predictor=curr_et, verbosity=1, out_dir=dist_tree_dir, dist='Any', biased_w2_ave=biased_w2_ave,\n",
    "                                                                                            unbiased_w2_ave=unbiased_w2_ave, processes=10, threshold=0.5, pos_size=curr_et.scorer.position_size,\n",
    "                                                                                            rank_type=curr_et.scorer.rank_type, file_prefix='{}_Scores_'.format(scoring_metric))\n",
    "                    curr_df['Distance Model'] = '{}{}'.format(dist_model, '_similarity' if et_dist else '')\n",
    "                    curr_df['Tree Type'] = tree_building\n",
    "                    curr_df['Scoring Metric'] = scoring_metric\n",
    "                    curr_df['Method'] = '{}{}_{}_{}'.format(dist_model, '_similarity' if et_dist else '', tree_building, scoring_metric)\n",
    "                    if protein_df is None:\n",
    "                        protein_df = curr_df\n",
    "                    else:\n",
    "                        protein_df = protein_df.append(curr_df)\n",
    "        protein_df['Protein'] = p_id\n",
    "        distance_model_order = ['identity', 'blosum62_similarity', 'blosum62']\n",
    "        tree_type_order = ['et', 'upgma', 'agglomerative']\n",
    "        scoring_metric_order = ['identity', 'plain_entropy', 'mutual_information', 'normalized_mutual_information', 'average_product_corrected_mutual_information', 'filtered_average_product_corrected_mutual_information']\n",
    "        sequence_separation_order = ['Any', 'Neighbors', 'Short', 'Medium', 'Long']\n",
    "        #########################################################################################################################################################################################################################\n",
    "        # Plot AUROC Data by Distance model\n",
    "        fn1a = os.path.join(protein_dir, 'AUROC_Distance_Comparison.png')\n",
    "        if not os.path.isfile(fn1a):\n",
    "            auroc = sns.catplot(data=protein_df, x='Distance Model', y='AUROC', legend=True, legend_out=True, sharex=True, sharey=False)\n",
    "            auroc.set(ylim=(0, 1)).set_xticklabels(rotation=30).set_titles(\"Distance Model Comparison\")\n",
    "            plt.show()\n",
    "            auroc.savefig(fn1a, dpi=600, bbox_inches='tight', transparent=True)\n",
    "            plt.close()\n",
    "        # Plot AUROC Data by Distance model add hue (sequence separation)\n",
    "        fn2a = os.path.join(protein_dir, 'AUROC_Distance_Comparison_Hue.png')\n",
    "        if not os.path.isfile(fn2a):\n",
    "            auroc = sns.catplot(data=protein_df, x='Distance Model', y='AUROC', hue='Sequence_Separation', hue_order=sequence_separation_order, legend=True, legend_out=True, sharex=True, sharey=False)\n",
    "            auroc.set(ylim=(0, 1)).set_xticklabels(rotation=30).set_titles(\"Distance Model Comparison\")\n",
    "            auroc.savefig(fn2a, dpi=600, bbox_inches='tight', transparent=True)\n",
    "            plt.close()\n",
    "        # Plot AUROC Data by Distance model add col (tree type)\n",
    "        fn3a = os.path.join(protein_dir, 'AUROC_Distance_Comparison_Col.png')\n",
    "        if not os.path.isfile(fn3a):\n",
    "            auroc = sns.catplot(data=protein_df, x='Distance Model', y='AUROC', col='Tree Type', col_order=tree_type_order,\n",
    "                                hue='Sequence_Separation', hue_order=sequence_separation_order, legend=True, legend_out=True, sharex=True, sharey=False)\n",
    "            auroc.set(ylim=(0, 1)).set_xticklabels(rotation=30).set_titles(\"{col_var}:{col_name}\")\n",
    "            auroc.savefig(fn3a, dpi=600, bbox_inches='tight', transparent=True)\n",
    "            plt.close()\n",
    "        # Plot AUROC Data by Distance model add row (scoring metric)\n",
    "        fn4a = os.path.join(protein_dir, 'AUROC_Distance_Comparison_Row.png')\n",
    "        if not os.path.isfile(fn4a):\n",
    "            auroc = sns.catplot(data=protein_df, x='Distance Model', y='AUROC', row='Scoring Metric', row_order=scoring_metric_order, col='Tree Type', col_order=tree_type_order,\n",
    "                                hue='Sequence_Separation', hue_order=sequence_separation_order, legend=True, legend_out=True, sharex=True, sharey=False)\n",
    "            auroc.set(ylim=(0, 1)).set_xticklabels(rotation=30).set_titles(\"{col_name}:{row_name}\")\n",
    "            auroc.savefig(fn4a, dpi=600, bbox_inches='tight', transparent=True)\n",
    "            plt.close()\n",
    "        #########################################################################################################################################################################################################################\n",
    "        # Plot AUROC Data by Tree type\n",
    "        fn1b = os.path.join(protein_dir, 'AUROC_Tree_Comparison.png')\n",
    "        if not os.path.isfile(fn1b):\n",
    "            auroc = sns.catplot(data=protein_df, x='Tree Type', y='AUROC', legend=True, legend_out=True, sharex=True, sharey=False)\n",
    "            auroc.set(ylim=(0, 1)).set_xticklabels(rotation=30).set_titles(\"Tree Type Comparison\")\n",
    "            plt.show()\n",
    "            auroc.savefig(fn1b, dpi=600, bbox_inches='tight', transparent=True)\n",
    "            plt.close()\n",
    "        # Plot AUROC Data by Tree type add hue (sequence separation)\n",
    "        fn2b = os.path.join(protein_dir, 'AUROC_Tree_Comparison_Hue.png')\n",
    "        if not os.path.isfile(fn2b):\n",
    "            auroc = sns.catplot(data=protein_df, x='Tree Type', y='AUROC', hue='Sequence_Separation', hue_order=sequence_separation_order, legend=True, legend_out=True, sharex=True, sharey=False)\n",
    "            auroc.set(ylim=(0, 1)).set_xticklabels(rotation=30).set_titles(\"Tree Type Comparison\")\n",
    "            auroc.savefig(fn2b, dpi=600, bbox_inches='tight', transparent=True)\n",
    "            plt.close()\n",
    "        # Plot AUROC Data by Tree type add col (distance model)\n",
    "        fn3b = os.path.join(protein_dir, 'AUROC_Tree_Comparison_Col.png')\n",
    "        if not os.path.isfile(fn3b):\n",
    "            auroc = sns.catplot(data=protein_df, x='Tree Type', y='AUROC', col='Distance Model', col_order=distance_model_order,\n",
    "                                hue='Sequence_Separation', hue_order=sequence_separation_order, legend=True, legend_out=True, sharex=True, sharey=False)\n",
    "            auroc.set(ylim=(0, 1)).set_xticklabels(rotation=30).set_titles(\"{col_var}:{col_name}\")\n",
    "            auroc.savefig(fn3b, dpi=600, bbox_inches='tight', transparent=True)\n",
    "            plt.close()\n",
    "        # Plot AUROC Data by Tree type add row (scoring metric)\n",
    "        fn4b = os.path.join(protein_dir, 'AUROC_Tree_Comparison_Row.png')\n",
    "        if not os.path.isfile(fn4b):\n",
    "            auroc = sns.catplot(data=protein_df, x='Tree Type', y='AUROC', row='Scoring Metric', row_order=scoring_metric_order, col='Distance Model', col_order=distance_model_order,\n",
    "                                hue='Sequence_Separation', hue_order=sequence_separation_order, legend=True, legend_out=True, sharex=True, sharey=False)\n",
    "            auroc.set(ylim=(0, 1)).set_xticklabels(rotation=30).set_titles(\"{col_name}:{row_name}\")\n",
    "            auroc.savefig(fn4b, dpi=600, bbox_inches='tight', transparent=True)\n",
    "            plt.close()\n",
    "        #########################################################################################################################################################################################################################\n",
    "        # Plot AUROC Data by Scoring Metric\n",
    "        fn1c = os.path.join(protein_dir, 'AUROC_Metric_Comparison.png')\n",
    "        if not os.path.isfile(fn1c):\n",
    "            auroc = sns.catplot(data=protein_df, x='Scoring Metric', y='AUROC', legend=True, legend_out=True, sharex=True, sharey=False)\n",
    "            auroc.set(ylim=(0, 1)).set_xticklabels(rotation=30).set_titles(\"Scoring Metric Comparison\")\n",
    "            plt.show()\n",
    "            auroc.savefig(fn1c, dpi=600, bbox_inches='tight', transparent=True)\n",
    "            plt.close()\n",
    "        # Plot AUROC Data by Scoring Metric add hue (sequence separation)\n",
    "        fn2c = os.path.join(protein_dir, 'AUROC_Metric_Comparison_Hue.png')\n",
    "        if not os.path.isfile(fn2c):\n",
    "            auroc = sns.catplot(data=protein_df, x='Scoring Metric', y='AUROC', hue='Sequence_Separation', hue_order=sequence_separation_order, legend=True, legend_out=True, sharex=True, sharey=False)\n",
    "            auroc.set(ylim=(0, 1)).set_xticklabels(rotation=30).set_titles(\"Scoring Metric Comparison\")\n",
    "            auroc.savefig(fn2c, dpi=600, bbox_inches='tight', transparent=True)\n",
    "            plt.close()\n",
    "        # Plot AUROC Data by Scoring Metric add col (distance model)\n",
    "        fn3c = os.path.join(protein_dir, 'AUROC_Metric_Comparison_Col.png')\n",
    "        if not os.path.isfile(fn3c):\n",
    "            auroc = sns.catplot(data=protein_df, x='Scoring Metric', y='AUROC', col='Distance Model', col_order=distance_model_order,\n",
    "                                hue='Sequence_Separation', hue_order=sequence_separation_order, legend=True, legend_out=True, sharex=True, sharey=False)\n",
    "            auroc.set(ylim=(0, 1)).set_xticklabels(rotation=30).set_titles(\"{col_var}:{col_name}\")\n",
    "            auroc.savefig(fn3c, dpi=600, bbox_inches='tight', transparent=True)\n",
    "            plt.close()\n",
    "        # Plot AUROC Data by Scoring Metric add row (tree type)\n",
    "        fn4c = os.path.join(protein_dir, 'AUROC_Metric_Comparison_Row.png')\n",
    "        if not os.path.isfile(fn4c):\n",
    "            auroc = sns.catplot(data=protein_df, x='Scoring Metric', y='AUROC', row='Tree Type', row_order=tree_type_order, col='Distance Model', col_order=distance_model_order,\n",
    "                                hue='Sequence_Separation', hue_order=sequence_separation_order, legend=True, legend_out=True, sharex=True, sharey=False)\n",
    "            auroc.set(ylim=(0, 1)).set_xticklabels(rotation=30).set_titles(\"{col_name}:{row_name}\")\n",
    "            auroc.savefig(fn4c, dpi=600, bbox_inches='tight', transparent=True)\n",
    "            plt.close()\n",
    "        #########################################################################################################################################################################################################################\n",
    "        if characterization_df is None:\n",
    "            characterization_df = protein_df\n",
    "        else:\n",
    "            characterization_df = characterization_df.append(protein_df)\n",
    "    characterization_df.to_csv(os.path.join(characterization_out_dir, 'Characterization_Data.csv'), sep='\\t', header=True, index=False,\n",
    "                               columns=['Protein', 'Distance Model', 'Tree Type', 'Scoring Metric', 'Method', 'Distance', 'Sequence_Separation', 'AUROC'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/ipykernel_launcher.py:6: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.\n",
      "  \n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/ipykernel_launcher.py:31: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/ipykernel_launcher.py:37: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/ipykernel_launcher.py:42: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/ipykernel_launcher.py:31: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/ipykernel_launcher.py:37: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/ipykernel_launcher.py:42: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/ipykernel_launcher.py:31: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/ipykernel_launcher.py:37: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/ipykernel_launcher.py:42: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/ipykernel_launcher.py:31: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/ipykernel_launcher.py:37: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/ipykernel_launcher.py:42: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/ipykernel_launcher.py:31: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/ipykernel_launcher.py:37: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/ipykernel_launcher.py:42: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "protein_order = summary['Protein_ID']\n",
    "# Plot AUROC Data by Scoring Metric\n",
    "auroc = sns.catplot(data=characterization_df, x='Scoring Metric', y='AUROC', legend=True, legend_out=True, sharex=True, sharey=False)\n",
    "auroc.set(ylim=(0, 1)).set_xticklabels(rotation=90).set_titles(\"Scoring Metric Comparison\")\n",
    "plt.show()\n",
    "auroc.savefig(os.path.join(characterization_out_dir, 'AUROC_Metric_Comparison.png'), dpi=600, bbox_inches='tight', transparent=True)\n",
    "plt.close()\n",
    "# Plot AUROC Data by Scoring Metric add hue (sequence separation)\n",
    "auroc = sns.catplot(data=characterization_df, x='Scoring Metric', y='AUROC', hue='Sequence_Separation', hue_order=sequence_separation_order, legend=True, legend_out=True, sharex=True, sharey=False)\n",
    "auroc.set(ylim=(0, 1)).set_xticklabels(rotation=90).set_titles(\"Scoring Metric Comparison\")\n",
    "auroc.savefig(os.path.join(characterization_out_dir, 'AUROC_Metric_Comparison_Hue.png'), dpi=600, bbox_inches='tight', transparent=True)\n",
    "plt.close()\n",
    "# Plot AUROC Data by Scoring Metric add col (distance model)\n",
    "auroc = sns.catplot(data=characterization_df, x='Scoring Metric', y='AUROC', col='Distance Model', col_order=distance_model_order,\n",
    "                    hue='Sequence_Separation', hue_order=sequence_separation_order, legend=True, legend_out=True, sharex=True, sharey=False)\n",
    "auroc.set(ylim=(0, 1)).set_xticklabels(rotation=90).set_titles(\"{col_var}:{col_name}\")\n",
    "auroc.savefig(os.path.join(characterization_out_dir, 'AUROC_Metric_Comparison_Col.png'), dpi=600, bbox_inches='tight', transparent=True)\n",
    "plt.close()\n",
    "# Plot AUROC Data by Scoring Metric add row (tree type)\n",
    "auroc = sns.catplot(data=characterization_df, x='Scoring Metric', y='AUROC', row='Tree Type', row_order=tree_type_order, col='Distance Model', col_order=distance_model_order,\n",
    "                    hue='Sequence_Separation', hue_order=sequence_separation_order, legend=True, legend_out=True, sharex=True, sharey=False)\n",
    "auroc.set(ylim=(0, 1)).set_xticklabels(rotation=90).set_titles(\"{col_name}:{row_name}\")\n",
    "auroc.savefig(os.path.join(characterization_out_dir, 'AUROC_Metric_Comparison_Row.png'), dpi=600, bbox_inches='tight', transparent=True)\n",
    "plt.close()\n",
    "for separation in sequence_separation_order:\n",
    "    sub_df = characterization_df[characterization_df['Sequence_Separation'] == separation]\n",
    "    # Plot AUROC Data by Scoring Metric\n",
    "    auroc = sns.catplot(data=sub_df, x='Scoring Metric', y='AUROC', legend=True, legend_out=True, sharex=True, sharey=False)\n",
    "    auroc.set(ylim=(0, 1)).set_xticklabels(rotation=90).set_titles(\"Scoring Metric Comparison\")\n",
    "    plt.show()\n",
    "    auroc.savefig(os.path.join(characterization_out_dir, '{}_AUROC_Metric_Comparison.png'.format(separation)), dpi=600, bbox_inches='tight', transparent=True)\n",
    "    plt.close()\n",
    "    # Plot AUROC Data by Scoring metric summarizing by mean/median\n",
    "    auroc = sns.catplot(data=sub_df[['Scoring Metric', 'AUROC']], x='Scoring Metric', y='AUROC', legend=True, legend_out=True, sharex=True, sharey=False, estimator=np.mean)\n",
    "    auroc.set(ylim=(0, 1)).set_xticklabels(rotation=90).set_titles(\"Scoring Metric Comparison\")\n",
    "    plt.show()\n",
    "    auroc.savefig(os.path.join(characterization_out_dir, '{}_AUROC_Metric_Comparison_Mean.png'.format(separation)), dpi=600, bbox_inches='tight', transparent=True)\n",
    "    plt.close()\n",
    "    auroc = sns.catplot(data=sub_df[['Scoring Metric', 'AUROC']], x='Scoring Metric', y='AUROC', legend=True, legend_out=True, sharex=True, sharey=False, estimator=np.median)\n",
    "    auroc.set(ylim=(0, 1)).set_xticklabels(rotation=90).set_titles(\"Scoring Metric Comparison\")\n",
    "    plt.show()\n",
    "    auroc.savefig(os.path.join(characterization_out_dir, '{}_AUROC_Metric_Comparison_Median.png'.format(separation)), dpi=600, bbox_inches='tight', transparent=True)\n",
    "    plt.close()\n",
    "    # Plot AUROC Data by Scoring Metric add hue (sequence separation)\n",
    "    auroc = sns.catplot(data=sub_df, x='Scoring Metric', y='AUROC', hue='Protein', hue_order=protein_order, legend=True, legend_out=True, sharex=True, sharey=False)\n",
    "    auroc.set(ylim=(0, 1)).set_xticklabels(rotation=90).set_titles(\"Scoring Metric Comparison\")\n",
    "    auroc.savefig(os.path.join(characterization_out_dir, '{}_AUROC_Metric_Comparison_Hue.png'.format(separation)), dpi=600, bbox_inches='tight', transparent=True)\n",
    "    plt.close()\n",
    "    # Plot AUROC Data by Scoring Metric add hue (sequence separation) summarizing by mean/median\n",
    "    auroc = sns.catplot(data=sub_df[['Scoring Metric', 'AUROC', 'Protein']], x='Scoring Metric', y='AUROC', hue='Protein', hue_order=protein_order, legend=True, legend_out=True, sharex=True, sharey=False, estimator=np.mean)\n",
    "    auroc.set(ylim=(0, 1)).set_xticklabels(rotation=90).set_titles(\"Scoring Metric Comparison\")\n",
    "    auroc.savefig(os.path.join(characterization_out_dir, '{}_AUROC_Metric_Comparison_Hue_Mean.png'.format(separation)), dpi=600, bbox_inches='tight', transparent=True)\n",
    "    plt.close()\n",
    "    auroc = sns.catplot(data=sub_df[['Scoring Metric', 'AUROC', 'Protein']], x='Scoring Metric', y='AUROC', hue='Protein', hue_order=protein_order, legend=True, legend_out=True, sharex=True, sharey=False, estimator=np.median)\n",
    "    auroc.set(ylim=(0, 1)).set_xticklabels(rotation=90).set_titles(\"Scoring Metric Comparison\")\n",
    "    auroc.savefig(os.path.join(characterization_out_dir, '{}_AUROC_Metric_Comparison_Hue_Median.png'.format(separation)), dpi=600, bbox_inches='tight', transparent=True)\n",
    "    plt.close()\n",
    "    # Plot AUROC Data by Scoring Metric add col (distance model)\n",
    "    auroc = sns.catplot(data=sub_df, x='Scoring Metric', y='AUROC', col='Distance Model', col_order=distance_model_order,\n",
    "                        hue='Protein', hue_order=protein_order, legend=True, legend_out=True, sharex=True, sharey=False)\n",
    "    auroc.set(ylim=(0, 1)).set_xticklabels(rotation=90).set_titles(\"{col_var}:{col_name}\")\n",
    "    auroc.savefig(os.path.join(characterization_out_dir, '{}_AUROC_Metric_Comparison_Col.png'.format(separation)), dpi=600, bbox_inches='tight', transparent=True)\n",
    "    plt.close()\n",
    "    # Plot AUROC Data by Scoring Metric add col (distance model) summarizing by mean/median\n",
    "    auroc = sns.catplot(data=sub_df[['Scoring Metric', 'AUROC', 'Protein', 'Distance Model']], x='Scoring Metric', y='AUROC', col='Distance Model', col_order=distance_model_order,\n",
    "                        hue='Protein', hue_order=protein_order, legend=True, legend_out=True, sharex=True, sharey=False, estimator=np.mean)\n",
    "    auroc.set(ylim=(0, 1)).set_xticklabels(rotation=90).set_titles(\"{col_var}:{col_name}\")\n",
    "    auroc.savefig(os.path.join(characterization_out_dir, '{}_AUROC_Metric_Comparison_Col_Mean.png'.format(separation)), dpi=600, bbox_inches='tight', transparent=True)\n",
    "    plt.close()\n",
    "    auroc = sns.catplot(data=sub_df[['Scoring Metric', 'AUROC', 'Protein', 'Distance Model']], x='Scoring Metric', y='AUROC', col='Distance Model', col_order=distance_model_order,\n",
    "                        hue='Protein', hue_order=protein_order, legend=True, legend_out=True, sharex=True, sharey=False, estimator=np.median)\n",
    "    auroc.set(ylim=(0, 1)).set_xticklabels(rotation=90).set_titles(\"{col_var}:{col_name}\")\n",
    "    auroc.savefig(os.path.join(characterization_out_dir, '{}_AUROC_Metric_Comparison_Col_Median.png'.format(separation)), dpi=600, bbox_inches='tight', transparent=True)\n",
    "    plt.close()\n",
    "    # Plot AUROC Data by Scoring Metric add row (tree type)\n",
    "    auroc = sns.catplot(data=sub_df, x='Scoring Metric', y='AUROC', row='Tree Type', row_order=tree_type_order, col='Distance Model', col_order=distance_model_order,\n",
    "                        hue='Protein', hue_order=protein_order, legend=True, legend_out=True, sharex=True, sharey=False)\n",
    "    auroc.set(ylim=(0, 1)).set_xticklabels(rotation=90).set_titles(\"{col_name}:{row_name}\")\n",
    "    auroc.savefig(os.path.join(characterization_out_dir, '{}_AUROC_Metric_Comparison_Row.png'.format(separation)), dpi=600, bbox_inches='tight', transparent=True)\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/pandas/core/indexing.py:1404: FutureWarning: \n",
      "Passing list-likes to .loc or [] with any missing label will raise\n",
      "KeyError in the future, you can use .reindex() as an alternative.\n",
      "\n",
      "See the documentation here:\n",
      "https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#deprecate-loc-reindex-listlike\n",
      "  return self._getitem_tuple(key)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/scipy/stats/morestats.py:2391: RuntimeWarning: invalid value encountered in greater\n",
      "  r_plus = np.sum((d > 0) * r, axis=0)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/scipy/stats/morestats.py:2392: RuntimeWarning: invalid value encountered in less\n",
      "  r_minus = np.sum((d < 0) * r, axis=0)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/pandas/core/indexing.py:1404: FutureWarning: \n",
      "Passing list-likes to .loc or [] with any missing label will raise\n",
      "KeyError in the future, you can use .reindex() as an alternative.\n",
      "\n",
      "See the documentation here:\n",
      "https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#deprecate-loc-reindex-listlike\n",
      "  return self._getitem_tuple(key)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/pandas/core/indexing.py:1404: FutureWarning: \n",
      "Passing list-likes to .loc or [] with any missing label will raise\n",
      "KeyError in the future, you can use .reindex() as an alternative.\n",
      "\n",
      "See the documentation here:\n",
      "https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#deprecate-loc-reindex-listlike\n",
      "  return self._getitem_tuple(key)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/scipy/stats/morestats.py:2391: RuntimeWarning: invalid value encountered in greater\n",
      "  r_plus = np.sum((d > 0) * r, axis=0)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/scipy/stats/morestats.py:2392: RuntimeWarning: invalid value encountered in less\n",
      "  r_minus = np.sum((d < 0) * r, axis=0)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/pandas/core/indexing.py:1404: FutureWarning: \n",
      "Passing list-likes to .loc or [] with any missing label will raise\n",
      "KeyError in the future, you can use .reindex() as an alternative.\n",
      "\n",
      "See the documentation here:\n",
      "https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#deprecate-loc-reindex-listlike\n",
      "  return self._getitem_tuple(key)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/scipy/stats/morestats.py:2391: RuntimeWarning: invalid value encountered in greater\n",
      "  r_plus = np.sum((d > 0) * r, axis=0)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/scipy/stats/morestats.py:2392: RuntimeWarning: invalid value encountered in less\n",
      "  r_minus = np.sum((d < 0) * r, axis=0)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/pandas/core/indexing.py:1404: FutureWarning: \n",
      "Passing list-likes to .loc or [] with any missing label will raise\n",
      "KeyError in the future, you can use .reindex() as an alternative.\n",
      "\n",
      "See the documentation here:\n",
      "https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#deprecate-loc-reindex-listlike\n",
      "  return self._getitem_tuple(key)\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from scipy.stats import wilcoxon\n",
    "characterization_df['AUROC_Rank'] = characterization_df.groupby(['Protein', 'Sequence_Separation'])['AUROC'].rank(method='max')\n",
    "# method_order = ['identity_upgma_identity', 'identity_upgma_plain_entropy', 'identity_upgma_mutual_information', 'identity_upgma_normalized_mutual_information',\n",
    "#                 'identity_upgma_average_product_corrected_mutual_information', 'identity_upgma_filtered_average_product_corrected_mutual_information', 'identity_et_identity', 'identity_et_plain_entropy',\n",
    "#                 'identity_et_mutual_information', 'identity_et_normalized_mutual_information', 'identity_et_average_product_corrected_mutual_information',\n",
    "#                 'identity_et_filtered_average_product_corrected_mutual_information', 'identity_agglomerative_identity', 'identity_agglomerative_plain_entropy', 'identity_agglomerative_mutual_information',\n",
    "#                 'identity_agglomerative_normalized_mutual_information', 'identity_agglomerative_average_product_corrected_mutual_information',\n",
    "#                 'identity_agglomerative_filtered_average_product_corrected_mutual_information', 'blosum62_similarity_upgma_identity', 'blosum62_similarity_upgma_plain_entropy',\n",
    "#                 'blosum62_similarity_upgma_mutual_information', 'blosum62_similarity_upgma_normalized_mutual_information', 'blosum62_similarity_upgma_average_product_corrected_mutual_information',\n",
    "#                 'blosum62_similarity_upgma_filtered_average_product_corrected_mutual_information', 'blosum62_similarity_et_identity', 'blosum62_similarity_et_plain_entropy', 'blosum62_similarity_et_mutual_information',\n",
    "#                 'blosum62_similarity_et_normalized_mutual_information', 'blosum62_similarity_et_average_product_corrected_mutual_information',\n",
    "#                 'blosum62_similarity_et_filtered_average_product_corrected_mutual_information', 'blosum62_similarity_agglomerative_identity', 'blosum62_similarity_agglomerative_plain_entropy',\n",
    "#                 'blosum62_similarity_agglomerative_mutual_information', 'blosum62_similarity_agglomerative_normalized_mutual_information', 'blosum62_similarity_agglomerative_average_product_corrected_mutual_information',\n",
    "#                 'blosum62_similarity_agglomerative_filtered_average_product_corrected_mutual_information', 'blosum62_upgma_identity', 'blosum62_upgma_plain_entropy', 'blosum62_upgma_mutual_information',\n",
    "#                 'blosum62_upgma_normalized_mutual_information', 'blosum62_upgma_average_product_corrected_mutual_information', 'blosum62_upgma_filtered_average_product_corrected_mutual_information',\n",
    "#                 'blosum62_et_identity', 'blosum62_et_plain_entropy', 'blosum62_et_mutual_information', 'blosum62_et_normalized_mutual_information', 'blosum62_et_average_product_corrected_mutual_information',\n",
    "#                 'blosum62_et_filtered_average_product_corrected_mutual_information', 'blosum62_agglomerative_identity', 'blosum62_agglomerative_plain_entropy', 'blosum62_agglomerative_mutual_information',\n",
    "#                 'blosum62_agglomerative_normalized_mutual_information', 'blosum62_agglomerative_average_product_corrected_mutual_information', 'blosum62_agglomerative_filtered_average_product_corrected_mutual_information']\n",
    "method_order = ['identity_upgma_identity', 'identity_et_identity', 'identity_agglomerative_identity', 'blosum62_similarity_upgma_identity', 'blosum62_similarity_et_identity', 'blosum62_similarity_agglomerative_identity',\n",
    "                'blosum62_upgma_identity', 'blosum62_et_identity', 'blosum62_agglomerative_identity', 'identity_upgma_plain_entropy', 'identity_et_plain_entropy', 'identity_agglomerative_plain_entropy',\n",
    "                'blosum62_similarity_upgma_plain_entropy', 'blosum62_similarity_et_plain_entropy', 'blosum62_similarity_agglomerative_plain_entropy', 'blosum62_upgma_plain_entropy', 'blosum62_et_plain_entropy',\n",
    "                'blosum62_agglomerative_plain_entropy', 'identity_upgma_mutual_information', 'identity_et_mutual_information', 'identity_agglomerative_mutual_information', 'blosum62_similarity_upgma_mutual_information',\n",
    "                'blosum62_similarity_et_mutual_information', 'blosum62_similarity_agglomerative_mutual_information', 'blosum62_upgma_mutual_information', 'blosum62_et_mutual_information',\n",
    "                'blosum62_agglomerative_mutual_information', 'identity_upgma_normalized_mutual_information', 'identity_et_normalized_mutual_information', 'identity_agglomerative_normalized_mutual_information',\n",
    "                'blosum62_similarity_upgma_normalized_mutual_information', 'blosum62_similarity_et_normalized_mutual_information', 'blosum62_similarity_agglomerative_normalized_mutual_information',\n",
    "                'blosum62_upgma_normalized_mutual_information', 'blosum62_et_normalized_mutual_information', 'blosum62_agglomerative_normalized_mutual_information',\n",
    "                'identity_upgma_average_product_corrected_mutual_information', 'identity_et_average_product_corrected_mutual_information', 'identity_agglomerative_average_product_corrected_mutual_information',\n",
    "                'blosum62_similarity_upgma_average_product_corrected_mutual_information', 'blosum62_similarity_et_average_product_corrected_mutual_information',\n",
    "                'blosum62_similarity_agglomerative_average_product_corrected_mutual_information', 'blosum62_upgma_average_product_corrected_mutual_information', 'blosum62_et_average_product_corrected_mutual_information',\n",
    "                'blosum62_agglomerative_average_product_corrected_mutual_information', 'identity_upgma_filtered_average_product_corrected_mutual_information',\n",
    "                'identity_et_filtered_average_product_corrected_mutual_information', 'identity_agglomerative_filtered_average_product_corrected_mutual_information',\n",
    "                'blosum62_similarity_upgma_filtered_average_product_corrected_mutual_information', 'blosum62_similarity_et_filtered_average_product_corrected_mutual_information',\n",
    "                'blosum62_similarity_agglomerative_filtered_average_product_corrected_mutual_information', 'blosum62_upgma_filtered_average_product_corrected_mutual_information',\n",
    "                'blosum62_et_filtered_average_product_corrected_mutual_information', 'blosum62_agglomerative_filtered_average_product_corrected_mutual_information']\n",
    "dims = (11.25, 6.0)\n",
    "_, ax = plt.subplots(figsize=dims)\n",
    "g = sns.boxplot(data=characterization_df, x='AUROC', y='Method', hue='Sequence_Separation', order=method_order, hue_order=sequence_separation_order, width=1.6, orient='h', ax=ax)\n",
    "# g.set_xticklabels(g.get_xticklabels(),rotation=90)\n",
    "g.legend(loc='center right', bbox_to_anchor=(1.5, 0.5), ncol=1)\n",
    "plt.savefig(os.path.join(characterization_out_dir, 'AUROC_Method_Comparison.png'), dpi=600, bbox_inches='tight', transparent=True)\n",
    "plt.close()\n",
    "_, ax = plt.subplots(figsize=dims)\n",
    "g = sns.boxplot(data=characterization_df, x='AUROC_Rank', y='Method', hue='Sequence_Separation', order=method_order, hue_order=sequence_separation_order, width=1.6, orient='h', ax=ax)\n",
    "# g.set_xticklabels(g.get_xticklabels(),rotation=90)\n",
    "g.legend(loc='center right', bbox_to_anchor=(1.5, 0.5), ncol=1)\n",
    "plt.savefig(os.path.join(characterization_out_dir, 'AUROC_Rank_Method_Comparison.png'), dpi=600, bbox_inches='tight', transparent=True)\n",
    "plt.close()\n",
    "for separation in sequence_separation_order:\n",
    "    sub_df = characterization_df[characterization_df['Sequence_Separation'] == separation]\n",
    "    _, ax = plt.subplots(figsize=dims)\n",
    "    g = sns.boxplot(data=sub_df, x='AUROC', y='Method', order=method_order, orient='h', ax=ax, color='black')\n",
    "    # g.set_xticklabels(g.get_xticklabels(),rotation=90)\n",
    "    plt.savefig(os.path.join(characterization_out_dir, '{}_AUROC_Method_Comparison.png'.format(separation)), dpi=600, bbox_inches='tight', transparent=True)\n",
    "    plt.close()\n",
    "    auroc_stats = {'Method1': [], 'Method2': [], 'Statistic': [], 'P-Value': []}\n",
    "    _, ax = plt.subplots(figsize=dims)\n",
    "    g = sns.boxplot(data=sub_df, x='AUROC_Rank', y='Method', order=method_order, orient='h', ax=ax, color='black')\n",
    "    # g.set_xticklabels(g.get_xticklabels(),rotation=90)\n",
    "    plt.savefig(os.path.join(characterization_out_dir, '{}_AUROC_Rank_Method_Comparison.png'.format(separation)), dpi=600, bbox_inches='tight', transparent=True)\n",
    "    plt.close()\n",
    "    \n",
    "    heatmap_df1 = sub_df[['Method', 'Protein', 'AUROC']].pivot(index='Protein', columns='Method', values='AUROC')\n",
    "    sns.heatmap(data=heatmap_df1, vmin=0.0, vmax=1.0, cmap='seismic', center=0.5, cbar=True, square=True)\n",
    "    plt.savefig(os.path.join(characterization_out_dir, '{}_AUROC_Method_Heatmap.png'.format(separation)), dpi=600, bbox_inches='tight', transparent=True)\n",
    "    plt.close()\n",
    "    \n",
    "    heatmap_df2 = sub_df[['Method', 'Protein', 'AUROC_Rank']].pivot(index='Protein', columns='Method', values='AUROC_Rank')\n",
    "    sns.heatmap(data=heatmap_df2, vmin=0.0, vmax=np.max(sub_df['AUROC_Rank']), cmap='seismic', center=np.mean(sub_df['AUROC_Rank']), cbar=True, square=True)\n",
    "    plt.savefig(os.path.join(characterization_out_dir, '{}_AUROC_Rank_Method_Heatmap.png'.format(separation)), dpi=600, bbox_inches='tight', transparent=True)\n",
    "    plt.close()\n",
    "    \n",
    "    auroc_stats = {'Method1': [], 'Method2': [], 'AUROC_Statistic': [], 'AUROC_P-Value': [], 'Rank_Statistic': [], 'Rank_P-Value': []}\n",
    "    for i in range(len(method_order)):\n",
    "        for j in range(i + 1, len(method_order)):\n",
    "            auroc_stats['Method1'].append(method_order[i])\n",
    "            auroc_stats['Method2'].append(method_order[j])\n",
    "            aurocs1 = sub_df[sub_df['Method'] == method_order[i]]['AUROC'].values\n",
    "            aurocs2 = sub_df[sub_df['Method'] == method_order[j]]['AUROC'].values\n",
    "            stat, pval = wilcoxon(aurocs1, aurocs2)\n",
    "            auroc_stats['AUROC_Statistic'].append(stat)\n",
    "            auroc_stats['AUROC_P-Value'].append(pval)\n",
    "            ranks1 = sub_df[sub_df['Method'] == method_order[i]]['AUROC_Rank'].values\n",
    "            ranks2 = sub_df[sub_df['Method'] == method_order[j]]['AUROC_Rank'].values\n",
    "            stat2, pval2 = wilcoxon(aurocs1, aurocs2)\n",
    "            auroc_stats['Rank_Statistic'].append(stat2)\n",
    "            auroc_stats['Rank_P-Value'].append(pval2)\n",
    "    pd.DataFrame(auroc_stats).to_csv(os.path.join(characterization_out_dir, '{}_AUROC_Method_Statistics.csv'.format(separation)), sep='\\t', index=False, header=True, columns=['Method1', 'Method2', 'AUROC_Statistic', 'AUROC_P-Value', 'RANK_Statistic', 'Rank_P-Value'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "156\n",
      "         Distance Model Sequence_Separation  AUROC_Max\n",
      "0              blosum62                 Any       14.0\n",
      "5   blosum62_similarity                 Any        9.0\n",
      "10             identity                 Any        8.0\n",
      "         Distance Model Sequence_Separation  AUROC_Max\n",
      "3              blosum62           Neighbors       16.0\n",
      "8   blosum62_similarity           Neighbors        7.0\n",
      "13             identity           Neighbors        9.0\n",
      "         Distance Model Sequence_Separation  AUROC_Max\n",
      "4              blosum62               Short       10.0\n",
      "9   blosum62_similarity               Short       12.0\n",
      "14             identity               Short       11.0\n",
      "         Distance Model Sequence_Separation  AUROC_Max\n",
      "2              blosum62              Medium       10.0\n",
      "7   blosum62_similarity              Medium       11.0\n",
      "12             identity              Medium        9.0\n",
      "         Distance Model Sequence_Separation  AUROC_Max\n",
      "1              blosum62                Long       12.0\n",
      "6   blosum62_similarity                Long       10.0\n",
      "11             identity                Long        8.0\n",
      "         Distance Model Sequence_Separation  AUROC_Max\n",
      "1              blosum62                Long       12.0\n",
      "6   blosum62_similarity                Long       10.0\n",
      "11             identity                Long        8.0\n",
      "        Tree Type Sequence_Separation  AUROC_Max\n",
      "0   agglomerative                 Any        7.0\n",
      "5              et                 Any        7.0\n",
      "10          upgma                 Any       17.0\n",
      "        Tree Type Sequence_Separation  AUROC_Max\n",
      "3   agglomerative           Neighbors       18.0\n",
      "8              et           Neighbors        6.0\n",
      "13          upgma           Neighbors        8.0\n",
      "        Tree Type Sequence_Separation  AUROC_Max\n",
      "4   agglomerative               Short       11.0\n",
      "9              et               Short        7.0\n",
      "14          upgma               Short       15.0\n",
      "        Tree Type Sequence_Separation  AUROC_Max\n",
      "2   agglomerative              Medium       13.0\n",
      "7              et              Medium        6.0\n",
      "12          upgma              Medium       11.0\n",
      "        Tree Type Sequence_Separation  AUROC_Max\n",
      "1   agglomerative                Long        7.0\n",
      "6              et                Long        8.0\n",
      "11          upgma                Long       15.0\n",
      "                                       Scoring Metric Sequence_Separation  \\\n",
      "0        average_product_corrected_mutual_information                 Any   \n",
      "5   filtered_average_product_corrected_mutual_info...                 Any   \n",
      "10                                           identity                 Any   \n",
      "15                                 mutual_information                 Any   \n",
      "20                      normalized_mutual_information                 Any   \n",
      "25                                      plain_entropy                 Any   \n",
      "\n",
      "    AUROC_Max  \n",
      "0        12.0  \n",
      "5         4.0  \n",
      "10        0.0  \n",
      "15        1.0  \n",
      "20        5.0  \n",
      "25        9.0  \n",
      "                                       Scoring Metric Sequence_Separation  \\\n",
      "3        average_product_corrected_mutual_information           Neighbors   \n",
      "8   filtered_average_product_corrected_mutual_info...           Neighbors   \n",
      "13                                           identity           Neighbors   \n",
      "18                                 mutual_information           Neighbors   \n",
      "23                      normalized_mutual_information           Neighbors   \n",
      "28                                      plain_entropy           Neighbors   \n",
      "\n",
      "    AUROC_Max  \n",
      "3         5.0  \n",
      "8         6.0  \n",
      "13        4.0  \n",
      "18        9.0  \n",
      "23        6.0  \n",
      "28        2.0  \n",
      "                                       Scoring Metric Sequence_Separation  \\\n",
      "4        average_product_corrected_mutual_information               Short   \n",
      "9   filtered_average_product_corrected_mutual_info...               Short   \n",
      "14                                           identity               Short   \n",
      "19                                 mutual_information               Short   \n",
      "24                      normalized_mutual_information               Short   \n",
      "29                                      plain_entropy               Short   \n",
      "\n",
      "    AUROC_Max  \n",
      "4         1.0  \n",
      "9         8.0  \n",
      "14        2.0  \n",
      "19        2.0  \n",
      "24       18.0  \n",
      "29        2.0  \n",
      "                                       Scoring Metric Sequence_Separation  \\\n",
      "2        average_product_corrected_mutual_information              Medium   \n",
      "7   filtered_average_product_corrected_mutual_info...              Medium   \n",
      "12                                           identity              Medium   \n",
      "17                                 mutual_information              Medium   \n",
      "22                      normalized_mutual_information              Medium   \n",
      "27                                      plain_entropy              Medium   \n",
      "\n",
      "    AUROC_Max  \n",
      "2         5.0  \n",
      "7         1.0  \n",
      "12        1.0  \n",
      "17        1.0  \n",
      "22        6.0  \n",
      "27       16.0  \n",
      "                                       Scoring Metric Sequence_Separation  \\\n",
      "1        average_product_corrected_mutual_information                Long   \n",
      "6   filtered_average_product_corrected_mutual_info...                Long   \n",
      "11                                           identity                Long   \n",
      "16                                 mutual_information                Long   \n",
      "21                      normalized_mutual_information                Long   \n",
      "26                                      plain_entropy                Long   \n",
      "\n",
      "    AUROC_Max  \n",
      "1         5.0  \n",
      "6         2.0  \n",
      "11        3.0  \n",
      "16        0.0  \n",
      "21        8.0  \n",
      "26       12.0  \n",
      "                                                Method Sequence_Separation  \\\n",
      "0    blosum62_agglomerative_average_product_correct...                 Any   \n",
      "5    blosum62_agglomerative_filtered_average_produc...                 Any   \n",
      "10                     blosum62_agglomerative_identity                 Any   \n",
      "15           blosum62_agglomerative_mutual_information                 Any   \n",
      "20   blosum62_agglomerative_normalized_mutual_infor...                 Any   \n",
      "25                blosum62_agglomerative_plain_entropy                 Any   \n",
      "30   blosum62_et_average_product_corrected_mutual_i...                 Any   \n",
      "35   blosum62_et_filtered_average_product_corrected...                 Any   \n",
      "40                                blosum62_et_identity                 Any   \n",
      "45                      blosum62_et_mutual_information                 Any   \n",
      "50           blosum62_et_normalized_mutual_information                 Any   \n",
      "55                           blosum62_et_plain_entropy                 Any   \n",
      "60   blosum62_similarity_agglomerative_average_prod...                 Any   \n",
      "65   blosum62_similarity_agglomerative_filtered_ave...                 Any   \n",
      "70          blosum62_similarity_agglomerative_identity                 Any   \n",
      "75   blosum62_similarity_agglomerative_mutual_infor...                 Any   \n",
      "80   blosum62_similarity_agglomerative_normalized_m...                 Any   \n",
      "85     blosum62_similarity_agglomerative_plain_entropy                 Any   \n",
      "90   blosum62_similarity_et_average_product_correct...                 Any   \n",
      "95   blosum62_similarity_et_filtered_average_produc...                 Any   \n",
      "100                    blosum62_similarity_et_identity                 Any   \n",
      "105          blosum62_similarity_et_mutual_information                 Any   \n",
      "110  blosum62_similarity_et_normalized_mutual_infor...                 Any   \n",
      "115               blosum62_similarity_et_plain_entropy                 Any   \n",
      "120  blosum62_similarity_upgma_average_product_corr...                 Any   \n",
      "125  blosum62_similarity_upgma_filtered_average_pro...                 Any   \n",
      "130                 blosum62_similarity_upgma_identity                 Any   \n",
      "135       blosum62_similarity_upgma_mutual_information                 Any   \n",
      "140  blosum62_similarity_upgma_normalized_mutual_in...                 Any   \n",
      "145            blosum62_similarity_upgma_plain_entropy                 Any   \n",
      "150  blosum62_upgma_average_product_corrected_mutua...                 Any   \n",
      "155  blosum62_upgma_filtered_average_product_correc...                 Any   \n",
      "160                            blosum62_upgma_identity                 Any   \n",
      "165                  blosum62_upgma_mutual_information                 Any   \n",
      "170       blosum62_upgma_normalized_mutual_information                 Any   \n",
      "175                       blosum62_upgma_plain_entropy                 Any   \n",
      "180  identity_agglomerative_average_product_correct...                 Any   \n",
      "185  identity_agglomerative_filtered_average_produc...                 Any   \n",
      "190                    identity_agglomerative_identity                 Any   \n",
      "195          identity_agglomerative_mutual_information                 Any   \n",
      "200  identity_agglomerative_normalized_mutual_infor...                 Any   \n",
      "205               identity_agglomerative_plain_entropy                 Any   \n",
      "210  identity_et_average_product_corrected_mutual_i...                 Any   \n",
      "215  identity_et_filtered_average_product_corrected...                 Any   \n",
      "220                               identity_et_identity                 Any   \n",
      "225                     identity_et_mutual_information                 Any   \n",
      "230          identity_et_normalized_mutual_information                 Any   \n",
      "235                          identity_et_plain_entropy                 Any   \n",
      "240  identity_upgma_average_product_corrected_mutua...                 Any   \n",
      "245  identity_upgma_filtered_average_product_correc...                 Any   \n",
      "250                            identity_upgma_identity                 Any   \n",
      "255                  identity_upgma_mutual_information                 Any   \n",
      "260       identity_upgma_normalized_mutual_information                 Any   \n",
      "265                       identity_upgma_plain_entropy                 Any   \n",
      "\n",
      "     AUROC_Max  \n",
      "0          0.0  \n",
      "5          1.0  \n",
      "10         0.0  \n",
      "15         0.0  \n",
      "20         2.0  \n",
      "25         1.0  \n",
      "30         1.0  \n",
      "35         0.0  \n",
      "40         0.0  \n",
      "45         0.0  \n",
      "50         1.0  \n",
      "55         1.0  \n",
      "60         1.0  \n",
      "65         0.0  \n",
      "70         0.0  \n",
      "75         0.0  \n",
      "80         0.0  \n",
      "85         1.0  \n",
      "90         0.0  \n",
      "95         1.0  \n",
      "100        0.0  \n",
      "105        0.0  \n",
      "110        0.0  \n",
      "115        1.0  \n",
      "120        4.0  \n",
      "125        0.0  \n",
      "130        0.0  \n",
      "135        0.0  \n",
      "140        0.0  \n",
      "145        1.0  \n",
      "150        4.0  \n",
      "155        0.0  \n",
      "160        0.0  \n",
      "165        1.0  \n",
      "170        1.0  \n",
      "175        1.0  \n",
      "180        0.0  \n",
      "185        0.0  \n",
      "190        0.0  \n",
      "195        0.0  \n",
      "200        0.0  \n",
      "205        1.0  \n",
      "210        0.0  \n",
      "215        0.0  \n",
      "220        0.0  \n",
      "225        0.0  \n",
      "230        1.0  \n",
      "235        1.0  \n",
      "240        2.0  \n",
      "245        2.0  \n",
      "250        0.0  \n",
      "255        0.0  \n",
      "260        0.0  \n",
      "265        1.0  \n",
      "                                                Method Sequence_Separation  \\\n",
      "3    blosum62_agglomerative_average_product_correct...           Neighbors   \n",
      "8    blosum62_agglomerative_filtered_average_produc...           Neighbors   \n",
      "13                     blosum62_agglomerative_identity           Neighbors   \n",
      "18           blosum62_agglomerative_mutual_information           Neighbors   \n",
      "23   blosum62_agglomerative_normalized_mutual_infor...           Neighbors   \n",
      "28                blosum62_agglomerative_plain_entropy           Neighbors   \n",
      "33   blosum62_et_average_product_corrected_mutual_i...           Neighbors   \n",
      "38   blosum62_et_filtered_average_product_corrected...           Neighbors   \n",
      "43                                blosum62_et_identity           Neighbors   \n",
      "48                      blosum62_et_mutual_information           Neighbors   \n",
      "53           blosum62_et_normalized_mutual_information           Neighbors   \n",
      "58                           blosum62_et_plain_entropy           Neighbors   \n",
      "63   blosum62_similarity_agglomerative_average_prod...           Neighbors   \n",
      "68   blosum62_similarity_agglomerative_filtered_ave...           Neighbors   \n",
      "73          blosum62_similarity_agglomerative_identity           Neighbors   \n",
      "78   blosum62_similarity_agglomerative_mutual_infor...           Neighbors   \n",
      "83   blosum62_similarity_agglomerative_normalized_m...           Neighbors   \n",
      "88     blosum62_similarity_agglomerative_plain_entropy           Neighbors   \n",
      "93   blosum62_similarity_et_average_product_correct...           Neighbors   \n",
      "98   blosum62_similarity_et_filtered_average_produc...           Neighbors   \n",
      "103                    blosum62_similarity_et_identity           Neighbors   \n",
      "108          blosum62_similarity_et_mutual_information           Neighbors   \n",
      "113  blosum62_similarity_et_normalized_mutual_infor...           Neighbors   \n",
      "118               blosum62_similarity_et_plain_entropy           Neighbors   \n",
      "123  blosum62_similarity_upgma_average_product_corr...           Neighbors   \n",
      "128  blosum62_similarity_upgma_filtered_average_pro...           Neighbors   \n",
      "133                 blosum62_similarity_upgma_identity           Neighbors   \n",
      "138       blosum62_similarity_upgma_mutual_information           Neighbors   \n",
      "143  blosum62_similarity_upgma_normalized_mutual_in...           Neighbors   \n",
      "148            blosum62_similarity_upgma_plain_entropy           Neighbors   \n",
      "153  blosum62_upgma_average_product_corrected_mutua...           Neighbors   \n",
      "158  blosum62_upgma_filtered_average_product_correc...           Neighbors   \n",
      "163                            blosum62_upgma_identity           Neighbors   \n",
      "168                  blosum62_upgma_mutual_information           Neighbors   \n",
      "173       blosum62_upgma_normalized_mutual_information           Neighbors   \n",
      "178                       blosum62_upgma_plain_entropy           Neighbors   \n",
      "183  identity_agglomerative_average_product_correct...           Neighbors   \n",
      "188  identity_agglomerative_filtered_average_produc...           Neighbors   \n",
      "193                    identity_agglomerative_identity           Neighbors   \n",
      "198          identity_agglomerative_mutual_information           Neighbors   \n",
      "203  identity_agglomerative_normalized_mutual_infor...           Neighbors   \n",
      "208               identity_agglomerative_plain_entropy           Neighbors   \n",
      "213  identity_et_average_product_corrected_mutual_i...           Neighbors   \n",
      "218  identity_et_filtered_average_product_corrected...           Neighbors   \n",
      "223                               identity_et_identity           Neighbors   \n",
      "228                     identity_et_mutual_information           Neighbors   \n",
      "233          identity_et_normalized_mutual_information           Neighbors   \n",
      "238                          identity_et_plain_entropy           Neighbors   \n",
      "243  identity_upgma_average_product_corrected_mutua...           Neighbors   \n",
      "248  identity_upgma_filtered_average_product_correc...           Neighbors   \n",
      "253                            identity_upgma_identity           Neighbors   \n",
      "258                  identity_upgma_mutual_information           Neighbors   \n",
      "263       identity_upgma_normalized_mutual_information           Neighbors   \n",
      "268                       identity_upgma_plain_entropy           Neighbors   \n",
      "\n",
      "     AUROC_Max  \n",
      "3          3.0  \n",
      "8          2.0  \n",
      "13         0.0  \n",
      "18         1.0  \n",
      "23         4.0  \n",
      "28         1.0  \n",
      "33         0.0  \n",
      "38         0.0  \n",
      "43         1.0  \n",
      "48         1.0  \n",
      "53         0.0  \n",
      "58         1.0  \n",
      "63         1.0  \n",
      "68         0.0  \n",
      "73         0.0  \n",
      "78         1.0  \n",
      "83         0.0  \n",
      "88         0.0  \n",
      "93         0.0  \n",
      "98         0.0  \n",
      "103        1.0  \n",
      "108        1.0  \n",
      "113        0.0  \n",
      "118        0.0  \n",
      "123        0.0  \n",
      "128        1.0  \n",
      "133        1.0  \n",
      "138        1.0  \n",
      "143        0.0  \n",
      "148        0.0  \n",
      "153        0.0  \n",
      "158        0.0  \n",
      "163        1.0  \n",
      "168        1.0  \n",
      "173        0.0  \n",
      "178        0.0  \n",
      "183        1.0  \n",
      "188        1.0  \n",
      "193        0.0  \n",
      "198        1.0  \n",
      "203        2.0  \n",
      "208        0.0  \n",
      "213        0.0  \n",
      "218        0.0  \n",
      "223        0.0  \n",
      "228        1.0  \n",
      "233        0.0  \n",
      "238        0.0  \n",
      "243        0.0  \n",
      "248        2.0  \n",
      "253        0.0  \n",
      "258        1.0  \n",
      "263        0.0  \n",
      "268        0.0  \n",
      "                                                Method Sequence_Separation  \\\n",
      "4    blosum62_agglomerative_average_product_correct...               Short   \n",
      "9    blosum62_agglomerative_filtered_average_produc...               Short   \n",
      "14                     blosum62_agglomerative_identity               Short   \n",
      "19           blosum62_agglomerative_mutual_information               Short   \n",
      "24   blosum62_agglomerative_normalized_mutual_infor...               Short   \n",
      "29                blosum62_agglomerative_plain_entropy               Short   \n",
      "34   blosum62_et_average_product_corrected_mutual_i...               Short   \n",
      "39   blosum62_et_filtered_average_product_corrected...               Short   \n",
      "44                                blosum62_et_identity               Short   \n",
      "49                      blosum62_et_mutual_information               Short   \n",
      "54           blosum62_et_normalized_mutual_information               Short   \n",
      "59                           blosum62_et_plain_entropy               Short   \n",
      "64   blosum62_similarity_agglomerative_average_prod...               Short   \n",
      "69   blosum62_similarity_agglomerative_filtered_ave...               Short   \n",
      "74          blosum62_similarity_agglomerative_identity               Short   \n",
      "79   blosum62_similarity_agglomerative_mutual_infor...               Short   \n",
      "84   blosum62_similarity_agglomerative_normalized_m...               Short   \n",
      "89     blosum62_similarity_agglomerative_plain_entropy               Short   \n",
      "94   blosum62_similarity_et_average_product_correct...               Short   \n",
      "99   blosum62_similarity_et_filtered_average_produc...               Short   \n",
      "104                    blosum62_similarity_et_identity               Short   \n",
      "109          blosum62_similarity_et_mutual_information               Short   \n",
      "114  blosum62_similarity_et_normalized_mutual_infor...               Short   \n",
      "119               blosum62_similarity_et_plain_entropy               Short   \n",
      "124  blosum62_similarity_upgma_average_product_corr...               Short   \n",
      "129  blosum62_similarity_upgma_filtered_average_pro...               Short   \n",
      "134                 blosum62_similarity_upgma_identity               Short   \n",
      "139       blosum62_similarity_upgma_mutual_information               Short   \n",
      "144  blosum62_similarity_upgma_normalized_mutual_in...               Short   \n",
      "149            blosum62_similarity_upgma_plain_entropy               Short   \n",
      "154  blosum62_upgma_average_product_corrected_mutua...               Short   \n",
      "159  blosum62_upgma_filtered_average_product_correc...               Short   \n",
      "164                            blosum62_upgma_identity               Short   \n",
      "169                  blosum62_upgma_mutual_information               Short   \n",
      "174       blosum62_upgma_normalized_mutual_information               Short   \n",
      "179                       blosum62_upgma_plain_entropy               Short   \n",
      "184  identity_agglomerative_average_product_correct...               Short   \n",
      "189  identity_agglomerative_filtered_average_produc...               Short   \n",
      "194                    identity_agglomerative_identity               Short   \n",
      "199          identity_agglomerative_mutual_information               Short   \n",
      "204  identity_agglomerative_normalized_mutual_infor...               Short   \n",
      "209               identity_agglomerative_plain_entropy               Short   \n",
      "214  identity_et_average_product_corrected_mutual_i...               Short   \n",
      "219  identity_et_filtered_average_product_corrected...               Short   \n",
      "224                               identity_et_identity               Short   \n",
      "229                     identity_et_mutual_information               Short   \n",
      "234          identity_et_normalized_mutual_information               Short   \n",
      "239                          identity_et_plain_entropy               Short   \n",
      "244  identity_upgma_average_product_corrected_mutua...               Short   \n",
      "249  identity_upgma_filtered_average_product_correc...               Short   \n",
      "254                            identity_upgma_identity               Short   \n",
      "259                  identity_upgma_mutual_information               Short   \n",
      "264       identity_upgma_normalized_mutual_information               Short   \n",
      "269                       identity_upgma_plain_entropy               Short   \n",
      "\n",
      "     AUROC_Max  \n",
      "4          0.0  \n",
      "9          1.0  \n",
      "14         0.0  \n",
      "19         0.0  \n",
      "24         2.0  \n",
      "29         1.0  \n",
      "34         0.0  \n",
      "39         0.0  \n",
      "44         0.0  \n",
      "49         0.0  \n",
      "54         1.0  \n",
      "59         1.0  \n",
      "64         0.0  \n",
      "69         1.0  \n",
      "74         0.0  \n",
      "79         0.0  \n",
      "84         1.0  \n",
      "89         0.0  \n",
      "94         0.0  \n",
      "99         0.0  \n",
      "104        1.0  \n",
      "109        0.0  \n",
      "114        1.0  \n",
      "119        0.0  \n",
      "124        0.0  \n",
      "129        2.0  \n",
      "134        1.0  \n",
      "139        1.0  \n",
      "144        4.0  \n",
      "149        0.0  \n",
      "154        0.0  \n",
      "159        1.0  \n",
      "164        0.0  \n",
      "169        0.0  \n",
      "174        3.0  \n",
      "179        0.0  \n",
      "184        0.0  \n",
      "189        2.0  \n",
      "194        0.0  \n",
      "199        0.0  \n",
      "204        3.0  \n",
      "209        0.0  \n",
      "214        0.0  \n",
      "219        1.0  \n",
      "224        0.0  \n",
      "229        1.0  \n",
      "234        1.0  \n",
      "239        0.0  \n",
      "244        1.0  \n",
      "249        0.0  \n",
      "254        0.0  \n",
      "259        0.0  \n",
      "264        2.0  \n",
      "269        0.0  \n",
      "                                                Method Sequence_Separation  \\\n",
      "2    blosum62_agglomerative_average_product_correct...              Medium   \n",
      "7    blosum62_agglomerative_filtered_average_produc...              Medium   \n",
      "12                     blosum62_agglomerative_identity              Medium   \n",
      "17           blosum62_agglomerative_mutual_information              Medium   \n",
      "22   blosum62_agglomerative_normalized_mutual_infor...              Medium   \n",
      "27                blosum62_agglomerative_plain_entropy              Medium   \n",
      "32   blosum62_et_average_product_corrected_mutual_i...              Medium   \n",
      "37   blosum62_et_filtered_average_product_corrected...              Medium   \n",
      "42                                blosum62_et_identity              Medium   \n",
      "47                      blosum62_et_mutual_information              Medium   \n",
      "52           blosum62_et_normalized_mutual_information              Medium   \n",
      "57                           blosum62_et_plain_entropy              Medium   \n",
      "62   blosum62_similarity_agglomerative_average_prod...              Medium   \n",
      "67   blosum62_similarity_agglomerative_filtered_ave...              Medium   \n",
      "72          blosum62_similarity_agglomerative_identity              Medium   \n",
      "77   blosum62_similarity_agglomerative_mutual_infor...              Medium   \n",
      "82   blosum62_similarity_agglomerative_normalized_m...              Medium   \n",
      "87     blosum62_similarity_agglomerative_plain_entropy              Medium   \n",
      "92   blosum62_similarity_et_average_product_correct...              Medium   \n",
      "97   blosum62_similarity_et_filtered_average_produc...              Medium   \n",
      "102                    blosum62_similarity_et_identity              Medium   \n",
      "107          blosum62_similarity_et_mutual_information              Medium   \n",
      "112  blosum62_similarity_et_normalized_mutual_infor...              Medium   \n",
      "117               blosum62_similarity_et_plain_entropy              Medium   \n",
      "122  blosum62_similarity_upgma_average_product_corr...              Medium   \n",
      "127  blosum62_similarity_upgma_filtered_average_pro...              Medium   \n",
      "132                 blosum62_similarity_upgma_identity              Medium   \n",
      "137       blosum62_similarity_upgma_mutual_information              Medium   \n",
      "142  blosum62_similarity_upgma_normalized_mutual_in...              Medium   \n",
      "147            blosum62_similarity_upgma_plain_entropy              Medium   \n",
      "152  blosum62_upgma_average_product_corrected_mutua...              Medium   \n",
      "157  blosum62_upgma_filtered_average_product_correc...              Medium   \n",
      "162                            blosum62_upgma_identity              Medium   \n",
      "167                  blosum62_upgma_mutual_information              Medium   \n",
      "172       blosum62_upgma_normalized_mutual_information              Medium   \n",
      "177                       blosum62_upgma_plain_entropy              Medium   \n",
      "182  identity_agglomerative_average_product_correct...              Medium   \n",
      "187  identity_agglomerative_filtered_average_produc...              Medium   \n",
      "192                    identity_agglomerative_identity              Medium   \n",
      "197          identity_agglomerative_mutual_information              Medium   \n",
      "202  identity_agglomerative_normalized_mutual_infor...              Medium   \n",
      "207               identity_agglomerative_plain_entropy              Medium   \n",
      "212  identity_et_average_product_corrected_mutual_i...              Medium   \n",
      "217  identity_et_filtered_average_product_corrected...              Medium   \n",
      "222                               identity_et_identity              Medium   \n",
      "227                     identity_et_mutual_information              Medium   \n",
      "232          identity_et_normalized_mutual_information              Medium   \n",
      "237                          identity_et_plain_entropy              Medium   \n",
      "242  identity_upgma_average_product_corrected_mutua...              Medium   \n",
      "247  identity_upgma_filtered_average_product_correc...              Medium   \n",
      "252                            identity_upgma_identity              Medium   \n",
      "257                  identity_upgma_mutual_information              Medium   \n",
      "262       identity_upgma_normalized_mutual_information              Medium   \n",
      "267                       identity_upgma_plain_entropy              Medium   \n",
      "\n",
      "     AUROC_Max  \n",
      "2          0.0  \n",
      "7          0.0  \n",
      "12         1.0  \n",
      "17         0.0  \n",
      "22         0.0  \n",
      "27         3.0  \n",
      "32         0.0  \n",
      "37         0.0  \n",
      "42         0.0  \n",
      "47         0.0  \n",
      "52         1.0  \n",
      "57         1.0  \n",
      "62         0.0  \n",
      "67         0.0  \n",
      "72         0.0  \n",
      "77         0.0  \n",
      "82         1.0  \n",
      "87         3.0  \n",
      "92         0.0  \n",
      "97         0.0  \n",
      "102        0.0  \n",
      "107        0.0  \n",
      "112        1.0  \n",
      "117        2.0  \n",
      "122        3.0  \n",
      "127        0.0  \n",
      "132        0.0  \n",
      "137        0.0  \n",
      "142        0.0  \n",
      "147        1.0  \n",
      "152        0.0  \n",
      "157        0.0  \n",
      "162        0.0  \n",
      "167        1.0  \n",
      "172        1.0  \n",
      "177        2.0  \n",
      "182        1.0  \n",
      "187        1.0  \n",
      "192        0.0  \n",
      "197        0.0  \n",
      "202        1.0  \n",
      "207        2.0  \n",
      "212        0.0  \n",
      "217        0.0  \n",
      "222        0.0  \n",
      "227        0.0  \n",
      "232        0.0  \n",
      "237        1.0  \n",
      "242        1.0  \n",
      "247        0.0  \n",
      "252        0.0  \n",
      "257        0.0  \n",
      "262        1.0  \n",
      "267        1.0  \n"
     ]
    }
   ],
   "source": [
    "characterization_df['AUROC_Max'] = characterization_df['AUROC_Rank'] == max(characterization_df['AUROC_Rank'])\n",
    "print(sum(characterization_df['AUROC_Max']))\n",
    "# Compare which distance models achieved the highest scores\n",
    "best_aurocs = characterization_df.groupby(['Distance Model', 'Sequence_Separation'])['AUROC_Max'].sum().reset_index()\n",
    "for separation in sequence_separation_order:\n",
    "    sub_df = best_aurocs[best_aurocs['Sequence_Separation'] == separation]\n",
    "    print(sub_df)\n",
    "    _, ax = plt.subplots(figsize=dims)\n",
    "    g = sns.barplot(data=sub_df, x='AUROC_Max', y='Distance Model', order=distance_model_order, orient='h', ax=ax, color='black')\n",
    "    plt.savefig(os.path.join(characterization_out_dir, '{}_Best_AUROC_Distance_Model_Comparison.png'.format(separation)), dpi=600, bbox_inches='tight', transparent=True)\n",
    "    plt.close()\n",
    "# Compare which tree construction method achieved the highest scores\n",
    "best_aurocs = characterization_df.groupby(['Tree Type', 'Sequence_Separation'])['AUROC_Max'].sum().reset_index()\n",
    "for separation in sequence_separation_order:\n",
    "    print(sub_df)\n",
    "    sub_df = best_aurocs[best_aurocs['Sequence_Separation'] == separation]\n",
    "    _, ax = plt.subplots(figsize=dims)\n",
    "    g = sns.barplot(data=sub_df, x='AUROC_Max', y='Tree Type', order=tree_type_order, orient='h', ax=ax, color='black')\n",
    "    plt.savefig(os.path.join(characterization_out_dir, '{}_Best_AUROC_Tree_Type_Comparison.png'.format(separation)), dpi=600, bbox_inches='tight', transparent=True)\n",
    "    plt.close()\n",
    "# Compare which scoring metric achieved the highest scores\n",
    "best_aurocs = characterization_df.groupby(['Scoring Metric', 'Sequence_Separation'])['AUROC_Max'].sum().reset_index()\n",
    "for separation in sequence_separation_order:\n",
    "    print(sub_df)\n",
    "    sub_df = best_aurocs[best_aurocs['Sequence_Separation'] == separation]\n",
    "    _, ax = plt.subplots(figsize=dims)\n",
    "    g = sns.barplot(data=sub_df, x='AUROC_Max', y='Scoring Metric', order=scoring_metric_order, orient='h', ax=ax, color='black')\n",
    "    plt.savefig(os.path.join(characterization_out_dir, '{}_Best_AUROC_Scoring_Metric_Comparison.png'.format(separation)), dpi=600, bbox_inches='tight', transparent=True)\n",
    "    plt.close()\n",
    "# Compare which overall method achieved the highest scores\n",
    "best_aurocs = characterization_df.groupby(['Method', 'Sequence_Separation'])['AUROC_Max'].sum().reset_index()\n",
    "for separation in sequence_separation_order:\n",
    "    print(sub_df)\n",
    "    sub_df = best_aurocs[best_aurocs['Sequence_Separation'] == separation]\n",
    "    _, ax = plt.subplots(figsize=dims)\n",
    "    g = sns.barplot(data=sub_df, x='AUROC_Max', y='Method', order=method_order, orient='h', ax=ax, color='black')\n",
    "    plt.savefig(os.path.join(characterization_out_dir, '{}_Best_AUROC_Method_Comparison.png'.format(separation)), dpi=600, bbox_inches='tight', transparent=True)\n",
    "    plt.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python3 (PyET3)",
   "language": "python",
   "name": "pyet3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
