{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Small Test Set #\n",
    "## Primary Goal ##\n",
    "Check if I can use the data generated by Benu and I before my code overhaul to reproduce the figures and trends we had shown before.\n",
    "## Secondary Goal ##\n",
    "The goal of this test set is to perform proof of concept testing on a small number of proteins with a wide range of sizes and available homologs, orthologs, and paralogs. By doing so it should be possible to test the best parameterization for this tool as well as identifying the strengths and weaknesses of the tool using various measurments as end points.\n",
    "## Warning ##\n",
    "Before attempting to use this notebook make sure that your .env file has been properly setup to reflect the correct locations of command line tools and the location of files and directories needed for execution.\n",
    "### Initial Import###\n",
    "This first cell performs the necessary imports required to begin this notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from dotenv import find_dotenv, load_dotenv\n",
    "try:\n",
    "    dotenv_path = find_dotenv(raise_error_if_not_found=True)\n",
    "except IOError:\n",
    "    dotenv_path = find_dotenv(raise_error_if_not_found=True, usecwd=True)\n",
    "load_dotenv(dotenv_path)\n",
    "import os\n",
    "import sys\n",
    "sys.path.append(os.path.join(os.environ.get('PROJECT_PATH'), 'src'))\n",
    "sys.path.append(os.path.join(os.environ.get('PROJECT_PATH'), 'src', 'SupportingClasses'))\n",
    "input_dir = os.environ.get('INPUT_PATH')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Set Construction ##\n",
    "The first task required to test the data set is to download the required data and construct any necessary input files for all down stream analyses.\n",
    "In this case that means:\n",
    "* Downloading PDB files for the proteins in our small test set.\n",
    "* Extracting a query sequence from each PDB file.\n",
    "* Searching for paralogs, homologs, and orthologs in a custom BLAST database built by filtering the Uniref90 database.\n",
    "* Filtering the hits from the BLAST search to meet minimum and maximum length requirements, as well as minimum and maximum identity requirements.\n",
    "* Building alignments using CLUSTALW in both the fasta and msf formats since some of the tools which will be used for comparison need different formats.\n",
    "* Filtering the alignment for maximum identity similarity between seqeunces.\n",
    "* Re-aligning the filtered sequences using CLUSTALW.\n",
    "This is all handeled by the DataSetGenerator class found in the src/SupportingClasses folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Removing gaps took 0.003695507844289144 min\n",
      "Removing gaps took 1.5934308369954426e-05 min\n",
      "Removing gaps took 0.006346674760182699 min\n",
      "Removing gaps took 0.00010115305582682291 min\n",
      "Removing gaps took 0.008964864412943523 min\n",
      "Removing gaps took 0.00042479832967122396 min\n",
      "Removing gaps took 0.025551732381184897 min\n",
      "Removing gaps took 0.003337192535400391 min\n",
      "Removing gaps took 7.469256718953451e-05 min\n",
      "Removing gaps took 0.009037490685780842 min\n",
      "Removing gaps took 0.009725987911224365 min\n",
      "Removing gaps took 0.038866110642751056 min\n",
      "Removing gaps took 0.03223922650019328 min\n",
      "Removing gaps took 0.011290466785430909 min\n",
      "Removing gaps took 0.030785882472991945 min\n",
      "Removing gaps took 0.03383784691492717 min\n",
      "Removing gaps took 0.04176588853200277 min\n",
      "Removing gaps took 0.01701324780782064 min\n",
      "Removing gaps took 0.056421323617299395 min\n",
      "Removing gaps took 0.0675422708193461 min\n",
      "Removing gaps took 0.07285692691802978 min\n",
      "Removing gaps took 0.00010458628336588542 min\n",
      "Removing gaps took 0.060426489512125654 min\n",
      "   Protein_ID Accession BLAST_Hits Filtered_BLAST  Filtered_Alignment  Length  \\\n",
      "8       3q05A      None       None           None                  27     234   \n",
      "21      2b59A      None       None           None                  28     166   \n",
      "3       7hvpA      None       None           None                  31      97   \n",
      "1       1c17A      None       None           None                  49      79   \n",
      "5       206lA      None       None           None                  70     162   \n",
      "7       1bolA      None       None           None                 162     222   \n",
      "13      2z0eA      None       None           None                 329     315   \n",
      "9       1axbA      None       None           None                 344     263   \n",
      "4       135lA      None       None           None                 424     129   \n",
      "10      2rh1A      None       None           None                 427     282   \n",
      "17      4lliA      None       None           None                 460     377   \n",
      "15      1a26A      None       None           None                 520     351   \n",
      "16      1c0kA      None       None           None                 601     363   \n",
      "20      2zxeA      None       None           None                 623     992   \n",
      "14      1jwlA      None       None           None                 760     329   \n",
      "11      1hckA      None       None           None                 778     294   \n",
      "22      1h1vA      None       None           None                 785     368   \n",
      "0       2ysdA      None       None           None                 861      57   \n",
      "19      2iopA      None       None           None                 871     618   \n",
      "12      3b6vA      None       None           None                 876     306   \n",
      "18      4ycuA      None       None           None                 878     505   \n",
      "6       2werA      None       None           None                 881     214   \n",
      "2       3tnuA      None       None           None                 945      90   \n",
      "\n",
      "    Total_Size  \n",
      "8         6318  \n",
      "21        4648  \n",
      "3         3007  \n",
      "1         3871  \n",
      "5        11340  \n",
      "7        35964  \n",
      "13      103635  \n",
      "9        90472  \n",
      "4        54696  \n",
      "10      120414  \n",
      "17      173420  \n",
      "15      182520  \n",
      "16      218163  \n",
      "20      618016  \n",
      "14      250040  \n",
      "11      228732  \n",
      "22      288880  \n",
      "0        49077  \n",
      "19      538278  \n",
      "12      268056  \n",
      "18      443390  \n",
      "6       188534  \n",
      "2        85050  \n",
      "It took 0.5505787094434103 min to generate the data set.\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from time import time\n",
    "from Bio.Align.Applications import ClustalwCommandline\n",
    "from DataSetGenerator import DataSetGenerator\n",
    "from SeqAlignment import SeqAlignment\n",
    "clustalw_path = os.environ.get('CLUSTALW_PATH')\n",
    "protein_list_dir = os.path.join(input_dir, 'ProteinLists')\n",
    "if not os.path.isdir(protein_list_dir):\n",
    "    os.makedirs(protein_list_dir)\n",
    "small_list_fn = os.path.join(protein_list_dir, 'SmallDataSetOld.txt')\n",
    "proteins_of_interest = ['2ysdA', '1c17A', '3tnuA', '7hvpA', '135lA', '206lA', '2werA', '1bolA', '3q05A', '1axbA',\n",
    "                        '2rh1A', '1hckA', '3b6vA', '2z0eA', '1jwlA', '1a26A', '1c0kA', '4lliA', '4ycuA', '2iopA',\n",
    "                        '2zxeA', '2b59A', '1h1vA']\n",
    "if not os.path.isfile(small_list_fn):\n",
    "    with open(small_list_fn, 'w') as small_list_handle:\n",
    "        for p_id in proteins_of_interest:\n",
    "            small_list_handle.write('{}\\n'.format(p_id))\n",
    "generator = DataSetGenerator(input_dir)\n",
    "start = time()\n",
    "generator.pdb_path = os.path.join(input_dir, 'Old_Data')\n",
    "generator.sequence_path = None\n",
    "generator.blast_path = None\n",
    "generator.filtered_blast_path = None\n",
    "generator.alignment_path = None\n",
    "generator.filtered_alignment_path = None\n",
    "generator.final_alingment_path = generator.pdb_path\n",
    "generator.protein_data = {}\n",
    "summary = {'Protein_ID': [], 'Accession': [], 'BLAST_Hits': [], 'Filtered_BLAST': [],\n",
    "           'Length': [], 'Filtered_Alignment': [], 'Total_Size': []}\n",
    "for p_id in proteins_of_interest:\n",
    "    generator.protein_data[p_id] = {}\n",
    "    if p_id == '2rh1A':\n",
    "        generator.protein_data[p_id]['PDB'] = os.path.join(input_dir, 'Old_Data', '{}_NEW.pdb'.format(p_id))\n",
    "        generator.protein_data[p_id]['Final_FA_Aln'] = os.path.join(input_dir, 'Old_Data', '{}_NEW.fa'.format(p_id))\n",
    "    else:\n",
    "        generator.protein_data[p_id]['PDB'] = os.path.join(input_dir, 'Old_Data', 'query_{}.pdb'.format(p_id))\n",
    "        generator.protein_data[p_id]['Final_FA_Aln'] = os.path.join(input_dir, 'Old_Data', '{}.fa'.format(p_id))\n",
    "    generator.protein_data[p_id]['Chain'] = 'A'\n",
    "    seq_aln = SeqAlignment(file_name=generator.protein_data[p_id]['Final_FA_Aln'], query_id='query_' + p_id)\n",
    "    seq_aln.import_alignment()\n",
    "    seq_aln_non_gap = seq_aln.remove_gaps()\n",
    "    generator.protein_data[p_id]['Sequence'] = seq_aln_non_gap.query_sequence\n",
    "    generator.protein_data[p_id]['Length'] = seq_aln_non_gap.seq_length\n",
    "    generator.protein_data[p_id]['Seq_Fasta'] = None\n",
    "    generator.protein_data[p_id]['Accession'] = None\n",
    "    generator.protein_data[p_id]['Filter_Count'] = None\n",
    "    generator.protein_data[p_id]['Filtered_BLAST'] = None\n",
    "    generator.protein_data[p_id]['MSF_Aln'] = None\n",
    "    generator.protein_data[p_id]['FA_Aln'] = None\n",
    "    generator.protein_data[p_id]['Final_Count'] = seq_aln_non_gap.size\n",
    "    generator.protein_data[p_id]['Filtered_Alignment'] = None\n",
    "    generator.protein_data[p_id]['Final_MSF_Aln'] = os.path.splitext(generator.protein_data[p_id]['Final_FA_Aln'])[0] + '.msf'\n",
    "    if not os.path.isfile(generator.protein_data[p_id]['Final_MSF_Aln']):\n",
    "        msf_cline = ClustalwCommandline(clustalw_path, infile=generator.protein_data[p_id]['Final_FA_Aln'], convert=True,\n",
    "                                        outfile=generator.protein_data[p_id]['Final_MSF_Aln'], output='GCG')\n",
    "        print(msf_cline)\n",
    "        stdout, stderr = msf_cline()\n",
    "        print(stdout)\n",
    "        print(stderr)\n",
    "    summary['Protein_ID'].append(p_id)\n",
    "    summary['Accession'].append(None)\n",
    "    summary['BLAST_Hits'].append(None)\n",
    "    summary['Filtered_BLAST'].append(None)\n",
    "    summary['Length'].append(seq_aln_non_gap.seq_length)\n",
    "    summary['Filtered_Alignment'].append(seq_aln_non_gap.size)\n",
    "    summary['Total_Size'].append(seq_aln_non_gap.seq_length * seq_aln_non_gap.size)\n",
    "summary = pd.DataFrame(summary)\n",
    "summary.sort_values(by=['Filtered_Alignment', 'Length'], axis=0, inplace=True)\n",
    "summary_columns = ['Protein_ID', 'Accession', 'BLAST_Hits', 'Filtered_BLAST', 'Filtered_Alignment', 'Length',\n",
    "                   'Total_Size']\n",
    "print(summary[summary_columns])\n",
    "end = time()\n",
    "print('It took {} min to generate the data set.'.format((end - start) / 60.0))\n",
    "summary.to_csv(os.path.join(input_dir, 'small_data_set_old_summary.tsv'), sep='\\t', index=False, header=True,\n",
    "               columns=summary_columns)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a location to store the output of this method comparison."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_dir = os.environ.get('OUTPUT_PATH')\n",
    "small_set_out_dir = os.path.join(output_dir, 'ReproducibilityCheck')\n",
    "if not os.path.isdir(small_set_out_dir):\n",
    "    os.makedirs(small_set_out_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setting Up Scoring For Each Method\n",
    "To reduce memory load during prediction and evaluation, the scoring objects needed to compute the metrics used to compare methods will be created ahead of time so they are available to each method when it computes its predictions for a given protein. This will ensure that results do not need to be kept in memory while waiting for all other results to be computed, only the metrics measured for each method will be recorded."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Removing gaps took 0.00015128850936889647 min\n",
      "Importing the PDB file took 0.0015863418579101563 min\n",
      "Removing gaps took 0.00014876921971638998 min\n",
      "Importing the PDB file took 0.0007414420445760092 min\n",
      "Mapping query sequence and pdb took 0.0010100364685058593 min\n",
      "Computing the distance matrix based on the PDB file took 0.007689571380615235 min\n",
      "Removing gaps took 7.832050323486328e-05 min\n",
      "Importing the PDB file took 0.0004199981689453125 min\n",
      "Mapping query sequence and pdb took 0.0005836168924967448 min\n",
      "Computing the distance matrix based on the PDB file took 0.008163972695668539 min\n",
      "Removing gaps took 0.00011109113693237304 min\n",
      "Importing the PDB file took 0.0012097835540771484 min\n",
      "Removing gaps took 0.00010420878728230794 min\n",
      "Importing the PDB file took 0.0002633174260457357 min\n",
      "Mapping query sequence and pdb took 0.0004086931546529134 min\n",
      "Computing the distance matrix based on the PDB file took 0.0029820760091145834 min\n",
      "Removing gaps took 0.00010976394017537435 min\n",
      "Importing the PDB file took 0.0002760489781697591 min\n",
      "Mapping query sequence and pdb took 0.00043853521347045896 min\n",
      "Computing the distance matrix based on the PDB file took 0.0033655325571695966 min\n",
      "Removing gaps took 9.974241256713868e-05 min\n",
      "Importing the PDB file took 0.0003166834513346354 min\n",
      "Removing gaps took 9.769598642985026e-05 min\n",
      "Importing the PDB file took 0.00017432371775309244 min\n",
      "Mapping query sequence and pdb took 0.0002983450889587402 min\n",
      "Computing the distance matrix based on the PDB file took 0.0011258880297342936 min\n",
      "Removing gaps took 9.67264175415039e-05 min\n",
      "Importing the PDB file took 0.00016605456670125326 min\n",
      "Mapping query sequence and pdb took 0.0002905766169230143 min\n",
      "Computing the distance matrix based on the PDB file took 0.0011988600095113118 min\n",
      "Removing gaps took 1.575946807861328e-05 min\n",
      "Importing the PDB file took 0.0006127278010050456 min\n",
      "Removing gaps took 1.5191237131754558e-05 min\n",
      "Importing the PDB file took 0.0002516468365987142 min\n",
      "Mapping query sequence and pdb took 0.00029601653416951496 min\n",
      "Computing the distance matrix based on the PDB file took 0.0007677992184956868 min\n",
      "Removing gaps took 1.479784647623698e-05 min\n",
      "Importing the PDB file took 0.0002495090166727702 min\n",
      "Mapping query sequence and pdb took 0.000298468271891276 min\n",
      "Computing the distance matrix based on the PDB file took 0.0010553359985351562 min\n",
      "Removing gaps took 0.0004123131434122721 min\n",
      "Importing the PDB file took 0.0005922516187032064 min\n",
      "Removing gaps took 0.0004132429758707682 min\n",
      "Importing the PDB file took 0.0011314829190572104 min\n",
      "Mapping query sequence and pdb took 0.0016051411628723144 min\n",
      "Computing the distance matrix based on the PDB file took 0.003145007292429606 min\n",
      "Removing gaps took 0.00040454864501953124 min\n",
      "Importing the PDB file took 0.0002782464027404785 min\n",
      "Mapping query sequence and pdb took 0.0007462064425150553 min\n",
      "Computing the distance matrix based on the PDB file took 0.00332107941309611 min\n",
      "Removing gaps took 0.002289740244547526 min\n",
      "Importing the PDB file took 0.0007091760635375977 min\n",
      "Removing gaps took 0.003148821989695231 min\n",
      "Importing the PDB file took 0.00035395224889119464 min\n",
      "Mapping query sequence and pdb took 0.00360033909479777 min\n",
      "Computing the distance matrix based on the PDB file took 0.005823894341786703 min\n",
      "Removing gaps took 0.0023309906323750814 min\n",
      "Importing the PDB file took 0.000357965628306071 min\n",
      "Mapping query sequence and pdb took 0.0027911822001139325 min\n",
      "Computing the distance matrix based on the PDB file took 0.006057556470235189 min\n",
      "Removing gaps took 0.011691689491271973 min\n",
      "Importing the PDB file took 0.0011844833691914876 min\n",
      "Removing gaps took 0.011260000864664714 min\n",
      "Importing the PDB file took 0.0012832919756571451 min\n",
      "Mapping query sequence and pdb took 0.012799596786499024 min\n",
      "Computing the distance matrix based on the PDB file took 0.011490456263224284 min\n",
      "Removing gaps took 0.011189484596252441 min\n",
      "Importing the PDB file took 0.0004982550938924154 min\n",
      "Mapping query sequence and pdb took 0.01188809076944987 min\n",
      "Computing the distance matrix based on the PDB file took 0.012132561206817627 min\n",
      "Removing gaps took 0.009271021684010823 min\n",
      "Importing the PDB file took 0.0008386532465616862 min\n",
      "Removing gaps took 0.010121603806813559 min\n",
      "Importing the PDB file took 0.00044875542322794597 min\n",
      "Mapping query sequence and pdb took 0.010734625657399495 min\n",
      "Computing the distance matrix based on the PDB file took 0.008946510155995686 min\n",
      "Removing gaps took 0.010367349783579508 min\n",
      "Importing the PDB file took 0.0004936536153157552 min\n",
      "Mapping query sequence and pdb took 0.01102585792541504 min\n",
      "Computing the distance matrix based on the PDB file took 0.009395869572957356 min\n",
      "Removing gaps took 0.009054772059122721 min\n",
      "Importing the PDB file took 0.0008456269900004069 min\n",
      "Removing gaps took 0.009570443630218506 min\n",
      "Importing the PDB file took 0.00025235811869303386 min\n",
      "Mapping query sequence and pdb took 0.009958577156066895 min\n",
      "Computing the distance matrix based on the PDB file took 0.0018975337346394856 min\n",
      "Removing gaps took 0.008937613169352213 min\n",
      "Importing the PDB file took 0.0002433617909749349 min\n",
      "Mapping query sequence and pdb took 0.010238099098205566 min\n",
      "Computing the distance matrix based on the PDB file took 0.0024361411730448404 min\n",
      "Removing gaps took 0.011674777666727702 min\n",
      "Importing the PDB file took 0.0014211694399515788 min\n",
      "Removing gaps took 0.011203293005625408 min\n",
      "Importing the PDB file took 0.0005176663398742676 min\n",
      "Mapping query sequence and pdb took 0.011925371487935384 min\n",
      "Computing the distance matrix based on the PDB file took 0.010023272037506104 min\n",
      "Removing gaps took 0.01127164363861084 min\n",
      "Importing the PDB file took 0.0005290428797403971 min\n",
      "Mapping query sequence and pdb took 0.012004172801971436 min\n",
      "Computing the distance matrix based on the PDB file took 0.010578819115956624 min\n",
      "Removing gaps took 0.019879945119222007 min\n",
      "Importing the PDB file took 0.0008288900057474772 min\n",
      "Removing gaps took 0.019798568884531655 min\n",
      "Importing the PDB file took 0.0006444573402404786 min\n",
      "Mapping query sequence and pdb took 0.020721463362375896 min\n",
      "Computing the distance matrix based on the PDB file took 0.017471468448638915 min\n",
      "Removing gaps took 0.01979379653930664 min\n",
      "Importing the PDB file took 0.0006362120310465495 min\n",
      "Mapping query sequence and pdb took 0.02071596384048462 min\n",
      "Computing the distance matrix based on the PDB file took 0.018480118115743002 min\n",
      "Removing gaps took 0.036669663588205975 min\n",
      "Importing the PDB file took 0.0013719558715820312 min\n",
      "Removing gaps took 0.03781141837437948 min\n",
      "Importing the PDB file took 0.0016089042027791342 min\n",
      "Mapping query sequence and pdb took 0.03978670040766398 min\n",
      "Computing the distance matrix based on the PDB file took 0.014880486329396566 min\n",
      "Removing gaps took 0.037081519762674965 min\n",
      "Importing the PDB file took 0.000583799680074056 min\n",
      "Mapping query sequence and pdb took 0.03797954718271891 min\n",
      "Computing the distance matrix based on the PDB file took 0.01608220338821411 min\n",
      "Removing gaps took 0.05000158548355103 min\n",
      "Importing the PDB file took 0.0016498764355977376 min\n",
      "Removing gaps took 0.05172636906305949 min\n",
      "Importing the PDB file took 0.0006096760431925455 min\n",
      "Mapping query sequence and pdb took 0.05272779067357381 min\n",
      "Computing the distance matrix based on the PDB file took 0.017558443546295165 min\n",
      "Removing gaps took 0.05011043548583984 min\n",
      "Importing the PDB file took 0.0005768497784932454 min\n",
      "Mapping query sequence and pdb took 0.05103367567062378 min\n",
      "Computing the distance matrix based on the PDB file took 0.016798893610636394 min\n",
      "Removing gaps took 0.08226517041524252 min\n",
      "Importing the PDB file took 0.0033445040384928386 min\n",
      "Removing gaps took 0.08645814657211304 min\n",
      "Importing the PDB file took 0.0016240874926249187 min\n",
      "Mapping query sequence and pdb took 0.08961191574732462 min\n",
      "Computing the distance matrix based on the PDB file took 0.10786241292953491 min\n",
      "Removing gaps took 0.0911953051884969 min\n",
      "Importing the PDB file took 0.001643971602121989 min\n",
      "Mapping query sequence and pdb took 0.09435048500696817 min\n",
      "Computing the distance matrix based on the PDB file took 0.1280157446861267 min\n",
      "Removing gaps took 0.036799458662668864 min\n",
      "Importing the PDB file took 0.0012536287307739259 min\n",
      "Removing gaps took 0.03822381099065145 min\n",
      "Importing the PDB file took 0.0004938880602518718 min\n",
      "Mapping query sequence and pdb took 0.03906736373901367 min\n",
      "Computing the distance matrix based on the PDB file took 0.012557299931844075 min\n",
      "Removing gaps took 0.04072740872701009 min\n",
      "Importing the PDB file took 0.000498815377553304 min\n",
      "Mapping query sequence and pdb took 0.04157165288925171 min\n",
      "Computing the distance matrix based on the PDB file took 0.01380529006322225 min\n",
      "Removing gaps took 0.04752435286839803 min\n",
      "Importing the PDB file took 0.0010587135950724284 min\n",
      "Removing gaps took 0.048106682300567624 min\n",
      "Importing the PDB file took 0.0005520502726236979 min\n",
      "Mapping query sequence and pdb took 0.0490255872408549 min\n",
      "Computing the distance matrix based on the PDB file took 0.013204137484232584 min\n",
      "Removing gaps took 0.04806185166041056 min\n",
      "Importing the PDB file took 0.0004815181096394857 min\n",
      "Mapping query sequence and pdb took 0.048842302958170575 min\n",
      "Computing the distance matrix based on the PDB file took 0.011400055885314942 min\n",
      "Removing gaps took 0.0678117831548055 min\n",
      "Importing the PDB file took 0.0012845079104105632 min\n",
      "Removing gaps took 0.0717729926109314 min\n",
      "Importing the PDB file took 0.0006138165791829427 min\n",
      "Mapping query sequence and pdb took 0.0728104035059611 min\n",
      "Computing the distance matrix based on the PDB file took 0.016098121802012127 min\n",
      "Removing gaps took 0.0689048965771993 min\n",
      "Importing the PDB file took 0.0005916953086853027 min\n",
      "Mapping query sequence and pdb took 0.069890828927358 min\n",
      "Computing the distance matrix based on the PDB file took 0.017603866259256997 min\n",
      "Removing gaps took 0.004226477940877279 min\n",
      "Importing the PDB file took 0.0018123030662536622 min\n",
      "Removing gaps took 0.004342806339263916 min\n",
      "Importing the PDB file took 0.0001635313034057617 min\n",
      "Mapping query sequence and pdb took 0.0046605229377746586 min\n",
      "Computing the distance matrix based on the PDB file took 0.00040134588877360024 min\n",
      "Removing gaps took 0.004208179314931234 min\n",
      "Importing the PDB file took 0.00017137924830118816 min\n",
      "Mapping query sequence and pdb took 0.0045238931973775225 min\n",
      "Computing the distance matrix based on the PDB file took 0.000549014409383138 min\n",
      "Removing gaps took 0.07724250157674153 min\n",
      "Importing the PDB file took 0.0020646135012308757 min\n",
      "Removing gaps took 0.07956176996231079 min\n",
      "Importing the PDB file took 0.0009925524393717448 min\n",
      "Mapping query sequence and pdb took 0.08132067521413168 min\n",
      "Computing the distance matrix based on the PDB file took 0.042129298051198326 min\n",
      "Removing gaps took 0.07826225360234579 min\n",
      "Importing the PDB file took 0.0010079781214396158 min\n",
      "Mapping query sequence and pdb took 0.07995787064234415 min\n",
      "Computing the distance matrix based on the PDB file took 0.04776991208394368 min\n",
      "Removing gaps took 0.03956576188405355 min\n",
      "Importing the PDB file took 0.000639192263285319 min\n",
      "Removing gaps took 0.03747611443201701 min\n",
      "Importing the PDB file took 0.0004844029744466146 min\n",
      "Mapping query sequence and pdb took 0.03828970988591512 min\n",
      "Computing the distance matrix based on the PDB file took 0.01091768741607666 min\n",
      "Removing gaps took 0.0374687393506368 min\n",
      "Importing the PDB file took 0.0020137429237365723 min\n",
      "Mapping query sequence and pdb took 0.03989338080088298 min\n",
      "Computing the distance matrix based on the PDB file took 0.012774785359700521 min\n",
      "Removing gaps took 0.06473476489384969 min\n",
      "Importing the PDB file took 0.000818022092183431 min\n",
      "Removing gaps took 0.0653862754503886 min\n",
      "Importing the PDB file took 0.0008222778638203939 min\n",
      "Mapping query sequence and pdb took 0.06672046581904094 min\n",
      "Computing the distance matrix based on the PDB file took 0.030453733603159585 min\n",
      "Removing gaps took 0.06497817436854045 min\n",
      "Importing the PDB file took 0.0008139053980509441 min\n",
      "Mapping query sequence and pdb took 0.06630780696868896 min\n",
      "Computing the distance matrix based on the PDB file took 0.03278400500615438 min\n",
      "Removing gaps took 0.028588688373565672 min\n",
      "Importing the PDB file took 0.001029805342356364 min\n",
      "Removing gaps took 0.029303157329559328 min\n",
      "Importing the PDB file took 0.0003646930058797201 min\n",
      "Mapping query sequence and pdb took 0.029899962743123374 min\n",
      "Computing the distance matrix based on the PDB file took 0.005608868598937988 min\n",
      "Removing gaps took 0.02880646785100301 min\n",
      "Importing the PDB file took 0.0003719727198282878 min\n",
      "Mapping query sequence and pdb took 0.029413278897603354 min\n",
      "Computing the distance matrix based on the PDB file took 0.006022334098815918 min\n",
      "Removing gaps took 0.007897583643595378 min\n",
      "Importing the PDB file took 0.00031244754791259766 min\n",
      "Removing gaps took 0.0061527649561564125 min\n",
      "Importing the PDB file took 0.0001608411471048991 min\n",
      "Mapping query sequence and pdb took 0.0064805428187052405 min\n",
      "Computing the distance matrix based on the PDB file took 0.0010006229082743328 min\n",
      "Removing gaps took 0.006303807099660237 min\n",
      "Importing the PDB file took 0.00016285181045532225 min\n",
      "Mapping query sequence and pdb took 0.00663527250289917 min\n",
      "Computing the distance matrix based on the PDB file took 0.001206513245900472 min\n"
     ]
    }
   ],
   "source": [
    "from SeqAlignment import SeqAlignment\n",
    "from PDBReference import PDBReference\n",
    "from ContactScorer import ContactScorer, plot_z_scores\n",
    "protein_order = list(summary['Protein_ID'])\n",
    "method_order = ['MI', 'DI', 'DCA', 'ET-MIp', 'cET-MIp']\n",
    "sequence_separation_order = ['Any', 'Neighbors', 'Short', 'Medium', 'Long']\n",
    "protein_scorers = {}\n",
    "small_comparison_df = None\n",
    "small_comparison_fn = os.path.join(small_set_out_dir, 'Reproducibility_Check_Data.csv')\n",
    "if os.path.isfile(small_comparison_fn):\n",
    "    small_comparison_df = pd.read_csv(small_comparison_fn, sep='\\t', header=0, index_col=False)\n",
    "else:\n",
    "    for p_id in summary['Protein_ID']:\n",
    "        protein_scorers[p_id] = {}\n",
    "        # Import alignment and remove gaps\n",
    "        full_aln = SeqAlignment(file_name=generator.protein_data[p_id]['Final_FA_Aln'], query_id='query_' + p_id)\n",
    "        full_aln.import_alignment()\n",
    "        non_gap_aln = full_aln.remove_gaps()\n",
    "        # Import structure\n",
    "        pdb_structure = PDBReference(pdb_file=generator.protein_data[p_id]['PDB'])\n",
    "        pdb_structure.import_pdb(structure_id=p_id)\n",
    "        protein_scorers[p_id]['Structure'] = pdb_structure\n",
    "        # Initialize Beta Carbon distance scorer\n",
    "        contact_scorer_cb = ContactScorer(query=p_id, seq_alignment=non_gap_aln,\n",
    "                                          pdb_reference=pdb_structure, cutoff=8.0)\n",
    "        contact_scorer_cb.best_chain = generator.protein_data[p_id]['Chain']\n",
    "        contact_scorer_cb.fit()\n",
    "        contact_scorer_cb.measure_distance(method='CB')\n",
    "        protein_scorers[p_id]['Scorer_CB'] = contact_scorer_cb\n",
    "        # Initialize distance scorer minimizing distance between any atoms\n",
    "        contact_scorer_any = ContactScorer(query=p_id, seq_alignment=non_gap_aln,\n",
    "                                           pdb_reference=pdb_structure, cutoff=8.0)\n",
    "        contact_scorer_any.best_chain = generator.protein_data[p_id]['Chain']\n",
    "        contact_scorer_any.fit()\n",
    "        contact_scorer_any.measure_distance(method='Any')\n",
    "        protein_scorers[p_id]['Scorer_Any'] = contact_scorer_any\n",
    "        # Initialize z-scoring subproblems\n",
    "        protein_scorers[p_id]['biased_w2_ave'] = None\n",
    "        protein_scorers[p_id]['unbiased_w2_ave'] = None\n",
    "output_columns = ['Protein', 'Protein Length', 'Alignment Size', 'Method', 'Distance', 'Sequence_Separation',\n",
    "                  'AUROC', 'AUPRC',\n",
    "                  'Top K Predictions', 'Precision', 'Recall', 'F1 Score',\n",
    "                  'Biased Z-Score at 10%', 'Biased Z-Score at 30%', 'Max Biased Z-Score', 'AUC Biased Z-Score',\n",
    "                  'Unbiased Z-Score at 10%', 'Biased Z-Score at 30%', 'Max Unbiased Z-Score', 'AUC Unbiased Z-Score']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generating Values For Comparision#\n",
    "To determine the effectiveness of the new method and implementation the covariation of the same proteins will be computed using the previous Evolutionary Trace covariation method (ET-MIp) and other methods in the field.\n",
    "\n",
    "## ET-MIp##\n",
    "Scoring the the covariation of the proteins using the previous Evolutionary Trace covariation method (ET-MIp)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from ETMIPWrapper import ETMIPWrapper\n",
    "import numpy as np\n",
    "if not os.path.isfile(small_comparison_fn):\n",
    "    old_etmip_out_dir = os.path.join(small_set_out_dir, 'ET-MIp')\n",
    "    if not os.path.isdir(old_etmip_out_dir):\n",
    "        os.makedirs(old_etmip_out_dir)\n",
    "    old_etmip_method_fn = os.path.join(old_etmip_out_dir, 'wetc_ET-MIp_Method_Data.csv')\n",
    "    if os.path.isfile(old_etmip_method_fn):\n",
    "        old_etmip_method_df = pd.read_csv(old_etmip_method_fn, sep='\\t', header=0, index_col=False)\n",
    "    else:    \n",
    "        old_etmip_method_df = None\n",
    "        counts = {'success':0, 'value': 0, 'attribute':0}\n",
    "        for p_id in summary['Protein_ID']:\n",
    "            if p_id == '2rh1A':\n",
    "                continue\n",
    "            print('Attempting to calculate ET-MIp covariance for: {}'.format(p_id))\n",
    "            protein_dir = os.path.join(old_etmip_out_dir, p_id)\n",
    "            if not os.path.isdir(protein_dir):\n",
    "                os.makedirs(protein_dir)\n",
    "            protein_fn = os.path.join(protein_dir, '{}_Protein_Data.csv'.format(p_id))\n",
    "            if os.path.isfile(protein_fn):\n",
    "                protein_df = pd.read_csv(protein_fn, sep='\\t', header=0, index_col=False)\n",
    "                counts['success'] += 1\n",
    "            else:\n",
    "#                 curr_aln = SeqAlignment(file_name=generator.protein_data[p_id]['Final_FA_Aln'], query_id='query_' + p_id,\n",
    "#                                         polymer_type='Protein')\n",
    "#                 curr_aln.import_alignment()\n",
    "                curr_etmip = ETMIPWrapper(query='query_' + p_id, aln_file=generator.protein_data[p_id]['Final_FA_Aln'],\n",
    "                                          out_dir=protein_dir)\n",
    "                curr_etmip.import_scores(method='ET-MIp', prefix=p_id)\n",
    "                np.savez(os.path.join(protein_dir, 'ET-MIp.npz'), time=curr_etmip.time,\n",
    "                         scores=curr_etmip.scores, coverage=curr_etmip.coverage)\n",
    "                import pickle\n",
    "                with open(os.path.join(protein_dir, 'ET-MIp.pkl'), 'wb') as handle:\n",
    "                    pickle.dump((curr_etmip.distance_matrix, curr_etmip.tree, curr_etmip.rank_group_assignments),\n",
    "                                handle, pickle.HIGHEST_PROTOCOL)\n",
    "                print('Successfully computed ET-MIp covariance for: {} in {} sec'.format(p_id, end-start))\n",
    "                # Compute statistics for the final scores of the ET-MIp model\n",
    "                protein_df, _, _ = protein_scorers[p_id]['Scorer_CB'].evaluate_predictor(\n",
    "                    predictor=curr_etmip, verbosity=2, out_dir=protein_dir, dist='CB', biased_w2_ave=None,\n",
    "                    unbiased_w2_ave=None, processes=10, threshold=0.5, pos_size=2, rank_type='max',\n",
    "                    file_prefix='ET-MIp_Scores_', plots=True)\n",
    "                protein_df2, _, _ = protein_scorers[p_id]['Scorer_Any'].evaluate_predictor(\n",
    "                    predictor=curr_etmip, verbosity=2, out_dir=protein_dir, dist='Any', biased_w2_ave=None,\n",
    "                    unbiased_w2_ave=None, processes=10, threshold=0.5, pos_size=2, rank_type='max',\n",
    "                    file_prefix='ET-MIp_Scores_Dist_Any', plots=True)\n",
    "                protein_df = protein_df.append(protein_df2)\n",
    "                # Score Prediction Clustering\n",
    "                z_score_fn = os.path.join(protein_dir, 'ET-MIp_Scores_Dist-Any_{}_ZScores.tsv')\n",
    "                z_score_plot_fn = os.path.join(protein_dir, 'ET-MIp_Scores_Dist-Any_{}_ZScores.png')\n",
    "                z_score_biased, biased_w2_ave, biased_scw_z_auc = protein_scorers[p_id]['Scorer_Any'].score_clustering_of_contact_predictions(\n",
    "                    1.0 - curr_etmip.coverage, bias=True, file_path=z_score_fn.format('Biased'),\n",
    "                    w2_ave_sub=protein_scorers[p_id]['biased_w2_ave'], processes=10)\n",
    "                if protein_scorers[p_id]['biased_w2_ave'] is None:\n",
    "                    protein_scorers[p_id]['biased_w2_ave'] = biased_w2_ave\n",
    "                biased_z_score_array = np.array(pd.to_numeric(z_score_biased['Z-Score'], errors='coerce'))\n",
    "                protein_df['Max Biased Z-Score'] = np.nanmax(biased_z_score_array)\n",
    "                protein_df['Biased Z-Score at 10%'] = biased_z_score_array[z_score_biased['Num_Residues'] >= float(len(protein_scorers[p_id]['Scorer_Any'].query_structure.seq[protein_scorers[p_id]['Scorer_Any'].best_chain])) * 0.10][0]\n",
    "                protein_df['Biased Z-Score at 30%'] = biased_z_score_array[z_score_biased['Num_Residues'] >= float(len(protein_scorers[p_id]['Scorer_Any'].query_structure.seq[protein_scorers[p_id]['Scorer_Any'].best_chain])) * 0.30][0]\n",
    "                protein_df['AUC Biased Z-Score'] = biased_scw_z_auc\n",
    "                plot_z_scores(z_score_biased, z_score_plot_fn.format('Biased'))\n",
    "                z_score_unbiased, unbiased_w2_ave, unbiased_scw_z_auc = protein_scorers[p_id]['Scorer_Any'].score_clustering_of_contact_predictions(\n",
    "                    1.0 - curr_etmip.coverage, bias=False, file_path=z_score_fn.format('Unbiased'),\n",
    "                    w2_ave_sub=protein_scorers[p_id]['unbiased_w2_ave'], processes=10)\n",
    "                if protein_scorers[p_id]['unbiased_w2_ave'] is None:\n",
    "                    protein_scorers[p_id]['unbiased_w2_ave'] = unbiased_w2_ave\n",
    "                unbiased_z_score_array = np.array(pd.to_numeric(z_score_unbiased['Z-Score'], errors='coerce'))\n",
    "                protein_df['Max Unbiased Z-Score'] = np.nanmax(unbiased_z_score_array)\n",
    "                protein_df['Unbiased Z-Score at 10%'] = unbiased_z_score_array[z_score_unbiased['Num_Residues'] >= float(len(protein_scorers[p_id]['Scorer_Any'].query_structure.seq[protein_scorers[p_id]['Scorer_Any'].best_chain])) * 0.10][0]\n",
    "                protein_df['Unbiased Z-Score at 30%'] = unbiased_z_score_array[z_score_unbiased['Num_Residues'] >= float(len(protein_scorers[p_id]['Scorer_Any'].query_structure.seq[protein_scorers[p_id]['Scorer_Any'].best_chain])) * 0.30][0]\n",
    "                protein_df['AUC Unbiased Z-Score'] = unbiased_scw_z_auc\n",
    "                plot_z_scores(z_score_unbiased, z_score_plot_fn.format('Unbiased'))\n",
    "                # Record static data for this protein\n",
    "                protein_df['Protein'] = p_id\n",
    "                protein_df['Method'] = 'ET-MIp'\n",
    "                protein_df['Alignment Size'] = summary['Filtered_Alignment'].values[summary['Protein_ID'] == p_id][0]\n",
    "                protein_df.to_csv(protein_fn, sep='\\t', header=True, index=False, columns=output_columns)\n",
    "                print('Metrics meastured for ET-MIp covariance for: {}'.format(p_id))\n",
    "                counts['success'] += 1\n",
    "            if old_etmip_method_df is None:\n",
    "                old_etmip_method_df = protein_df\n",
    "            else:\n",
    "                old_etmip_method_df = old_etmip_method_df.append(protein_df)\n",
    "        print('{}\\tSuccesses\\n{}\\tValue Errors\\n{}\\tAttribute Errors'.format(counts['success'], counts['value'],\n",
    "                                                                             counts['attribute']))\n",
    "        old_etmip_method_df.to_csv(old_etmip_method_fn, sep='\\t', header=True, index=False, columns=output_columns)\n",
    "    if small_comparison_df is None:\n",
    "        small_comparison_df = old_etmip_method_df\n",
    "    else:\n",
    "        small_comparison_df = small_comparison_df.append(old_etmip_method_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DCA##\n",
    "Scoring the the covariation of the proteins using a DCA julia implementation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Attempting to calculate DCA covariance for: 3q05A\n",
      "Attempting to calculate DCA covariance for: 2b59A\n",
      "Attempting to calculate DCA covariance for: 7hvpA\n",
      "Attempting to calculate DCA covariance for: 1c17A\n",
      "Attempting to calculate DCA covariance for: 206lA\n",
      "Attempting to calculate DCA covariance for: 1bolA\n",
      "Attempting to calculate DCA covariance for: 2z0eA\n",
      "Attempting to calculate DCA covariance for: 1axbA\n",
      "Removing gaps took 0.012909332911173502 min\n",
      "Compute SCW Z-Score took 0.021410731474558513 min\n",
      "Compute SCW Z-Score took 0.011387709776560466 min\n",
      "Successfully computed DCA covariance for: 1axbA\n",
      "Attempting to calculate DCA covariance for: 135lA\n",
      "Attempting to calculate DCA covariance for: 2rh1A\n",
      "Attempting to calculate DCA covariance for: 4lliA\n",
      "Attempting to calculate DCA covariance for: 1a26A\n",
      "Attempting to calculate DCA covariance for: 1c0kA\n",
      "Attempting to calculate DCA covariance for: 2zxeA\n",
      "Attempting to calculate DCA covariance for: 1jwlA\n",
      "Attempting to calculate DCA covariance for: 1hckA\n",
      "Attempting to calculate DCA covariance for: 1h1vA\n",
      "Attempting to calculate DCA covariance for: 2ysdA\n",
      "Attempting to calculate DCA covariance for: 2iopA\n",
      "Attempting to calculate DCA covariance for: 3b6vA\n",
      "Attempting to calculate DCA covariance for: 4ycuA\n",
      "Attempting to calculate DCA covariance for: 2werA\n",
      "Attempting to calculate DCA covariance for: 3tnuA\n",
      "23\tSuccesses\n",
      "0\tValue Errors\n",
      "0\tAttribute Errors\n"
     ]
    }
   ],
   "source": [
    "from DCAWrapper import DCAWrapper\n",
    "if not os.path.isfile(small_comparison_fn):\n",
    "    dca_out_dir = os.path.join(small_set_out_dir, 'DCA')\n",
    "    if not os.path.isdir(dca_out_dir):\n",
    "        os.makedirs(dca_out_dir)\n",
    "    dca_method_fn = os.path.join(dca_out_dir, 'DCA_Method_Data.csv')\n",
    "    if os.path.isfile(dca_method_fn):\n",
    "        dca_method_df = pd.read_csv(dca_method_fn, sep='\\t', header=0, index_col=False)\n",
    "    else:\n",
    "        dca_method_df = None\n",
    "        counts = {'success':0, 'value': 0, 'attribute':0}\n",
    "        for p_id in summary['Protein_ID']:\n",
    "            print('Attempting to calculate DCA covariance for: {}'.format(p_id))\n",
    "            protein_dir = os.path.join(dca_out_dir, p_id)\n",
    "            if not os.path.isdir(protein_dir):\n",
    "                os.makedirs(protein_dir)\n",
    "            protein_fn = os.path.join(protein_dir, '{}_Protein_Data.csv'.format(p_id))\n",
    "            if os.path.isfile(protein_fn):\n",
    "                protein_df = pd.read_csv(protein_fn, sep='\\t', header=0, index_col=False)\n",
    "                counts['success'] += 1\n",
    "            else:\n",
    "#                 curr_aln = SeqAlignment(file_name=generator.protein_data[p_id]['Final_FA_Aln'], query_id='query_' + p_id,\n",
    "#                                         polymer_type='Protein')\n",
    "#                 curr_aln.import_alignment()\n",
    "                # Since the DCA implementation used here does not provide a way to specify the query sequence we remove the gaps\n",
    "                # from the query sequences so positions will be referenced correctly for that sequence (and unnecessary\n",
    "                # computations can be avoided).\n",
    "#                 curr_aln = curr_aln.remove_gaps()\n",
    "#                 new_aln_fn = os.path.join(protein_dir, '{}_no_gap.fasta'.format(p_id))\n",
    "#                 curr_aln.write_out_alignment(new_aln_fn)\n",
    "#                 curr_aln.file_name = new_aln_fn\n",
    "#                 curr_dca = DCAWrapper(alignment=curr_aln)\n",
    "                curr_dca = DCAWrapper(query=p_id, aln_file=generator.protein_data[p_id]['Final_FA_Aln'], out_dir=protein_dir)\n",
    "#                 curr_dca.import_covariance_scores(out_path=os.path.join(protein_dir, '{}.DCA.txt'.format(p_id)))\n",
    "                curr_dca.calculate_scores(delete_file=False)\n",
    "                # Compute statistics for the final scores of the ET-MIp model\n",
    "                protein_df, _, _ = protein_scorers[p_id]['Scorer_CB'].evaluate_predictor(\n",
    "                    predictor=curr_dca, verbosity=2, out_dir=protein_dir, dist='CB', biased_w2_ave=None,\n",
    "                    unbiased_w2_ave=None, processes=10, threshold=0.5, pos_size=2, rank_type='max', file_prefix='DCA_Scores_', plots=True)\n",
    "                protein_df2, _, _ = protein_scorers[p_id]['Scorer_Any'].evaluate_predictor(\n",
    "                    predictor=curr_dca, verbosity=2, out_dir=protein_dir, dist='Any', biased_w2_ave=None,\n",
    "                    unbiased_w2_ave=None, processes=10, threshold=0.5, pos_size=2, rank_type='max', file_prefix='DCA_Scores_Dist_Any', plots=True)\n",
    "                protein_df = protein_df.append(protein_df2)\n",
    "                # Score Prediction Clustering\n",
    "                _, dca_coverage  = compute_rank_and_coverage(seq_length=curr_dca.alignment.seq_length, scores=curr_dca.scores, pos_size=2,\n",
    "                    rank_type='max')\n",
    "                z_score_fn = os.path.join(protein_dir, 'DCA_Scores_Dist-Any_{}_ZScores.tsv')\n",
    "                z_score_plot_fn = os.path.join(protein_dir, 'DCA_Scores_Dist-Any_{}_ZScores.png')\n",
    "                z_score_biased, biased_w2_ave, biased_scw_z_auc = protein_scorers[p_id]['Scorer_Any'].score_clustering_of_contact_predictions(\n",
    "                    1.0 - dca_coverage, bias=True, file_path=z_score_fn.format('Biased'),\n",
    "                    w2_ave_sub=protein_scorers[p_id]['biased_w2_ave'], processes=10)\n",
    "                if protein_scorers[p_id]['biased_w2_ave'] is None:\n",
    "                    protein_scorers[p_id]['biased_w2_ave'] = biased_w2_ave\n",
    "                biased_z_score_array = np.array(pd.to_numeric(z_score_biased['Z-Score'], errors='coerce'))\n",
    "                protein_df['Max Biased Z-Score'] = np.nanmax(biased_z_score_array)\n",
    "                protein_df['Biased Z-Score at 10%'] = biased_z_score_array[z_score_biased['Num_Residues'] >= float(len(protein_scorers[p_id]['Scorer_Any'].query_structure.seq[protein_scorers[p_id]['Scorer_Any'].best_chain])) * 0.10][0]\n",
    "                protein_df['Biased Z-Score at 30%'] = biased_z_score_array[z_score_biased['Num_Residues'] >= float(len(protein_scorers[p_id]['Scorer_Any'].query_structure.seq[protein_scorers[p_id]['Scorer_Any'].best_chain])) * 0.30][0]\n",
    "                protein_df['AUC Biased Z-Score'] = biased_scw_z_auc\n",
    "                plot_z_scores(z_score_biased, z_score_plot_fn.format('Biased'))\n",
    "                z_score_unbiased, unbiased_w2_ave, unbiased_scw_z_auc = protein_scorers[p_id]['Scorer_Any'].score_clustering_of_contact_predictions(\n",
    "                    1.0 - dca_coverage, bias=False, file_path=z_score_fn.format('Unbiased'),\n",
    "                    w2_ave_sub=protein_scorers[p_id]['unbiased_w2_ave'], processes=10)\n",
    "                if protein_scorers[p_id]['unbiased_w2_ave'] is None:\n",
    "                    protein_scorers[p_id]['unbiased_w2_ave'] = unbiased_w2_ave\n",
    "                unbiased_z_score_array = np.array(pd.to_numeric(z_score_unbiased['Z-Score'], errors='coerce'))\n",
    "                protein_df['Max Unbiased Z-Score'] = np.nanmax(unbiased_z_score_array)\n",
    "                protein_df['Unbiased Z-Score at 10%'] = unbiased_z_score_array[z_score_unbiased['Num_Residues'] >= float(len(protein_scorers[p_id]['Scorer_Any'].query_structure.seq[protein_scorers[p_id]['Scorer_Any'].best_chain])) * 0.10][0]\n",
    "                protein_df['Unbiased Z-Score at 30%'] = unbiased_z_score_array[z_score_unbiased['Num_Residues'] >= float(len(protein_scorers[p_id]['Scorer_Any'].query_structure.seq[protein_scorers[p_id]['Scorer_Any'].best_chain])) * 0.30][0]\n",
    "                protein_df['AUC Unbiased Z-Score'] = unbiased_scw_z_auc\n",
    "                plot_z_scores(z_score_unbiased, z_score_plot_fn.format('Unbiased'))\n",
    "                # Record execution times\n",
    "                protein_df['Init Time'] = None\n",
    "                protein_df['Import Time'] = None\n",
    "                protein_df['Dist Tree Time'] = None\n",
    "                protein_df['Trace Time'] = None\n",
    "                protein_df['Total Time'] = None\n",
    "                # Record static data for this protein\n",
    "                protein_df['Protein'] = p_id\n",
    "                protein_df['Method'] = 'DCA'\n",
    "                protein_df['Alignment Size'] = summary['Filtered_Alignment'].values[summary['Protein_ID'] == p_id][0]\n",
    "                protein_df.to_csv(protein_fn, sep='\\t', header=True, index=False, columns=output_columns)\n",
    "                print('Successfully computed DCA covariance for: {}'.format(p_id))\n",
    "                counts['success'] += 1\n",
    "            if dca_method_df is None:\n",
    "                dca_method_df = protein_df\n",
    "            else:\n",
    "                dca_method_df = dca_method_df.append(protein_df)\n",
    "        print('{}\\tSuccesses\\n{}\\tValue Errors\\n{}\\tAttribute Errors'.format(counts['success'], counts['value'],\n",
    "                                                                             counts['attribute']))\n",
    "        dca_method_df.to_csv(dca_method_fn, sep='\\t', header=True, index=False, columns=output_columns)\n",
    "    if small_comparison_df is None:\n",
    "        small_comparison_df = dca_method_df\n",
    "    else:\n",
    "        small_comparison_df = small_comparison_df.append(dca_method_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## cET-MIp\n",
    "This segment the ET-MIp method, when constrained to an arbitrary set of nodes (1, 2, 3, 5, 7, 10, 25) at the top of the phylogenetic tree."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Attempting to calculate cET-MIp covariance for: 3q05A\n",
      "Successfully computed cET-MIp covariance for: 3q05A in 33.034722566604614 sec\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Compute SCW Z-Score took 0.01861115296681722 min\n",
      "Compute SCW Z-Score took 0.013350764910380045 min\n",
      "Metrics meastured for cET-MIp covariance for: 3q05A\n",
      "Attempting to calculate cET-MIp covariance for: 2b59A\n",
      "Successfully computed cET-MIp covariance for: 2b59A in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.009998353322347005 min\n",
      "Compute SCW Z-Score took 0.010991899172465007 min\n",
      "Metrics meastured for cET-MIp covariance for: 2b59A\n",
      "Attempting to calculate cET-MIp covariance for: 7hvpA\n",
      "Successfully computed cET-MIp covariance for: 7hvpA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.007678552468617757 min\n",
      "Compute SCW Z-Score took 0.008338566621144612 min\n",
      "Metrics meastured for cET-MIp covariance for: 7hvpA\n",
      "Attempting to calculate cET-MIp covariance for: 1c17A\n",
      "Successfully computed cET-MIp covariance for: 1c17A in 33.034722566604614 sec\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Compute SCW Z-Score took 0.00685197114944458 min\n",
      "Compute SCW Z-Score took 0.007689126332600911 min\n",
      "Metrics meastured for cET-MIp covariance for: 1c17A\n",
      "Attempting to calculate cET-MIp covariance for: 206lA\n",
      "Successfully computed cET-MIp covariance for: 206lA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.014568122227986653 min\n",
      "Compute SCW Z-Score took 0.010161197185516358 min\n",
      "Metrics meastured for cET-MIp covariance for: 206lA\n",
      "Attempting to calculate cET-MIp covariance for: 1bolA\n",
      "Successfully computed cET-MIp covariance for: 1bolA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.01884549856185913 min\n",
      "Compute SCW Z-Score took 0.016194522380828857 min\n",
      "Metrics meastured for cET-MIp covariance for: 1bolA\n",
      "Attempting to calculate cET-MIp covariance for: 2z0eA\n",
      "Successfully computed cET-MIp covariance for: 2z0eA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.09317988554636637 min\n",
      "Compute SCW Z-Score took 0.11031901439030965 min\n",
      "Metrics meastured for cET-MIp covariance for: 2z0eA\n",
      "Attempting to calculate cET-MIp covariance for: 1axbA\n",
      "Successfully computed cET-MIp covariance for: 1axbA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.02589957316716512 min\n",
      "Compute SCW Z-Score took 0.02780466874440511 min\n",
      "Metrics meastured for cET-MIp covariance for: 1axbA\n",
      "Attempting to calculate cET-MIp covariance for: 135lA\n",
      "Successfully computed cET-MIp covariance for: 135lA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.010491303602854411 min\n",
      "Compute SCW Z-Score took 0.011964937051137289 min\n",
      "Metrics meastured for cET-MIp covariance for: 135lA\n",
      "Attempting to calculate cET-MIp covariance for: 2rh1A\n",
      "Successfully computed cET-MIp covariance for: 2rh1A in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.04623865286509196 min\n",
      "Compute SCW Z-Score took 0.03604763746261597 min\n",
      "Metrics meastured for cET-MIp covariance for: 2rh1A\n",
      "Attempting to calculate cET-MIp covariance for: 4lliA\n",
      "Successfully computed cET-MIp covariance for: 4lliA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.06802795330683391 min\n",
      "Compute SCW Z-Score took 0.084891939163208 min\n",
      "Metrics meastured for cET-MIp covariance for: 4lliA\n",
      "Attempting to calculate cET-MIp covariance for: 1a26A\n",
      "Successfully computed cET-MIp covariance for: 1a26A in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.14380310773849486 min\n",
      "Compute SCW Z-Score took 0.13278024196624755 min\n",
      "Metrics meastured for cET-MIp covariance for: 1a26A\n",
      "Attempting to calculate cET-MIp covariance for: 1c0kA\n",
      "Successfully computed cET-MIp covariance for: 1c0kA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.14334330558776856 min\n",
      "Compute SCW Z-Score took 0.15344902674357097 min\n",
      "Metrics meastured for cET-MIp covariance for: 1c0kA\n",
      "Attempting to calculate cET-MIp covariance for: 2zxeA\n",
      "Successfully computed cET-MIp covariance for: 2zxeA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.673426361878713 min\n",
      "Compute SCW Z-Score took 0.6852782090504964 min\n",
      "Metrics meastured for cET-MIp covariance for: 2zxeA\n",
      "Attempting to calculate cET-MIp covariance for: 1jwlA\n",
      "Successfully computed cET-MIp covariance for: 1jwlA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.054970427354176836 min\n",
      "Compute SCW Z-Score took 0.046740583578745526 min\n",
      "Metrics meastured for cET-MIp covariance for: 1jwlA\n",
      "Attempting to calculate cET-MIp covariance for: 1hckA\n",
      "Successfully computed cET-MIp covariance for: 1hckA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.07582351764043173 min\n",
      "Compute SCW Z-Score took 0.07804312705993652 min\n",
      "Metrics meastured for cET-MIp covariance for: 1hckA\n",
      "Attempting to calculate cET-MIp covariance for: 1h1vA\n",
      "Successfully computed cET-MIp covariance for: 1h1vA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.1515523354212443 min\n",
      "Compute SCW Z-Score took 0.156049907207489 min\n",
      "Metrics meastured for cET-MIp covariance for: 1h1vA\n",
      "Attempting to calculate cET-MIp covariance for: 2ysdA\n",
      "Successfully computed cET-MIp covariance for: 2ysdA in 33.034722566604614 sec\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Compute SCW Z-Score took 0.008471016089121501 min\n",
      "Compute SCW Z-Score took 0.009182437260945638 min\n",
      "Metrics meastured for cET-MIp covariance for: 2ysdA\n",
      "Attempting to calculate cET-MIp covariance for: 2iopA\n",
      "Successfully computed cET-MIp covariance for: 2iopA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.22693382501602172 min\n",
      "Compute SCW Z-Score took 0.214168651898702 min\n",
      "Metrics meastured for cET-MIp covariance for: 2iopA\n",
      "Attempting to calculate cET-MIp covariance for: 3b6vA\n",
      "Successfully computed cET-MIp covariance for: 3b6vA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.08326433499654134 min\n",
      "Compute SCW Z-Score took 0.08648212353388468 min\n",
      "Metrics meastured for cET-MIp covariance for: 3b6vA\n",
      "Attempting to calculate cET-MIp covariance for: 4ycuA\n",
      "Successfully computed cET-MIp covariance for: 4ycuA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.13809781869252521 min\n",
      "Compute SCW Z-Score took 0.12764538923899332 min\n",
      "Metrics meastured for cET-MIp covariance for: 4ycuA\n",
      "Attempting to calculate cET-MIp covariance for: 2werA\n",
      "Successfully computed cET-MIp covariance for: 2werA in 33.034722566604614 sec\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Compute SCW Z-Score took 0.022269082069396973 min\n",
      "Compute SCW Z-Score took 0.025943811734517416 min\n",
      "Metrics meastured for cET-MIp covariance for: 2werA\n",
      "Attempting to calculate cET-MIp covariance for: 3tnuA\n",
      "Successfully computed cET-MIp covariance for: 3tnuA in 33.034722566604614 sec\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/ranking.py:571: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/home/daniel/Documents/git/ETMIP/src/SupportingClasses/ContactScorer.py:768: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/ranking.py:571: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/home/daniel/Documents/git/ETMIP/src/SupportingClasses/ContactScorer.py:768: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/ranking.py:571: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/home/daniel/Documents/git/ETMIP/src/SupportingClasses/ContactScorer.py:768: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/ranking.py:563: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/home/daniel/Documents/git/ETMIP/src/SupportingClasses/ContactScorer.py:754: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/ranking.py:571: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/home/daniel/Documents/git/ETMIP/src/SupportingClasses/ContactScorer.py:768: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/ranking.py:571: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/home/daniel/Documents/git/ETMIP/src/SupportingClasses/ContactScorer.py:768: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Compute SCW Z-Score took 0.013906260331471762 min\n",
      "Compute SCW Z-Score took 0.010667447249094646 min\n",
      "Metrics meastured for cET-MIp covariance for: 3tnuA\n",
      "23\tSuccesses\n",
      "0\tValue Errors\n",
      "0\tAttribute Errors\n"
     ]
    }
   ],
   "source": [
    "from utils import compute_rank_and_coverage\n",
    "def import_cETMIp_cov_scores(path, aln):\n",
    "    data = pd.read_csv(path, sep='\\t', header=0, index_col=None)\n",
    "    scores = np.zeros((aln.seq_length, aln.seq_length))\n",
    "    positions = sorted(set(data['Pos1']) | set(data['Pos2']))\n",
    "    pos_map = {x: i for i, x in enumerate(positions)}\n",
    "    data['Final_Pos1'] = data['Pos1'].apply(lambda x: pos_map[x])\n",
    "    data['Final_Pos2'] = data['Pos2'].apply(lambda x: pos_map[x])\n",
    "    scores[data['Final_Pos1'].values, data['Final_Pos2'].values] = data['ETMIp_Score'].values\n",
    "    _, coverages = compute_rank_and_coverage(seq_length=aln.seq_length, scores=scores, pos_size=2, rank_type='max')\n",
    "    return scores, coverages\n",
    "\n",
    "if not os.path.isfile(small_comparison_fn):\n",
    "    old_cetmip_out_dir = os.path.join(small_set_out_dir, 'cET-MIp')\n",
    "    if not os.path.isdir(old_cetmip_out_dir):\n",
    "        os.makedirs(old_cetmip_out_dir)\n",
    "    old_cetmip_method_fn = os.path.join(old_cetmip_out_dir, 'cET-MIp_Method_Data.csv')\n",
    "    if os.path.isfile(old_cetmip_method_fn):\n",
    "        old_cetmip_method_df = pd.read_csv(old_cetmip_method_fn, sep='\\t', header=0, index_col=False)\n",
    "    else:    \n",
    "        old_cetmip_method_df = None\n",
    "        counts = {'success':0, 'value': 0, 'attribute':0}\n",
    "        for p_id in summary['Protein_ID']:\n",
    "            print('Attempting to calculate cET-MIp covariance for: {}'.format(p_id))\n",
    "            protein_dir = os.path.join(old_cetmip_out_dir, p_id)\n",
    "            if not os.path.isdir(protein_dir):\n",
    "                os.makedirs(protein_dir)\n",
    "            protein_fn = os.path.join(protein_dir, '{}_Protein_Data.csv'.format(p_id))\n",
    "            if os.path.isfile(protein_fn):\n",
    "                protein_df = pd.read_csv(protein_fn, sep='\\t', header=0, index_col=False)\n",
    "                counts['success'] += 1\n",
    "            else:\n",
    "#                 curr_aln = SeqAlignment(file_name=generator.protein_data[p_id]['Final_FA_Aln'], query_id='query_' + p_id,\n",
    "#                                         polymer_type='Protein')\n",
    "#                 curr_aln.import_alignment()\n",
    "                curr_etmip = ETMIPWrapper(query='query_' + p_id, aln_file=generator.protein_data[p_id]['Final_FA_Aln'],\n",
    "                                          out_dir=protein_dir)\n",
    "                curr_etmip.scores, curr_etmip.coverage = import_cETMIp_cov_scores(\n",
    "                    os.path.join(protein_dir, '{}.cetmip.txt'.format(p_id)), curr_aln)\n",
    "                np.savez(os.path.join(protein_dir, 'cET-MIp.npz'), time=curr_etmip.time,\n",
    "                         scores=curr_etmip.scores, coverage=curr_etmip.coverage)\n",
    "                import pickle\n",
    "                with open(os.path.join(protein_dir, 'cET-MIp.pkl'), 'wb') as handle:\n",
    "                    pickle.dump((curr_etmip.distance_matrix, curr_etmip.tree, curr_etmip.rank_group_assignments),\n",
    "                                handle, pickle.HIGHEST_PROTOCOL)\n",
    "                print('Successfully computed cET-MIp covariance for: {} in {} sec'.format(p_id, end-start))\n",
    "                # Compute statistics for the final scores of the ET-MIp model\n",
    "                protein_df, _, _ = protein_scorers[p_id]['Scorer_CB'].evaluate_predictor(\n",
    "                    predictor=curr_etmip, verbosity=2, out_dir=protein_dir, dist='CB', biased_w2_ave=None,\n",
    "                    unbiased_w2_ave=None, processes=10, threshold=0.5, pos_size=2, rank_type='max',\n",
    "                    file_prefix='cET-MIp_Scores_', plots=True)\n",
    "                protein_df2, _, _ = protein_scorers[p_id]['Scorer_Any'].evaluate_predictor(\n",
    "                    predictor=curr_etmip, verbosity=2, out_dir=protein_dir, dist='Any', biased_w2_ave=None,\n",
    "                    unbiased_w2_ave=None, processes=10, threshold=0.5, pos_size=2, rank_type='max',\n",
    "                    file_prefix='cET-MIp_Scores_Dist_Any', plots=True)\n",
    "                protein_df = protein_df.append(protein_df2)\n",
    "                # Score Prediction Clustering\n",
    "                z_score_fn = os.path.join(protein_dir, 'cET-MIp_Scores_Dist-Any_{}_ZScores.tsv')\n",
    "                z_score_plot_fn = os.path.join(protein_dir, 'cET-MIp_Scores_Dist-Any_{}_ZScores.png')\n",
    "                z_score_biased, biased_w2_ave, biased_scw_z_auc = protein_scorers[p_id]['Scorer_Any'].score_clustering_of_contact_predictions(\n",
    "                    1.0 - curr_etmip.coverage, bias=True, file_path=z_score_fn.format('Biased'),\n",
    "                    w2_ave_sub=protein_scorers[p_id]['biased_w2_ave'], processes=10)\n",
    "                if protein_scorers[p_id]['biased_w2_ave'] is None:\n",
    "                    protein_scorers[p_id]['biased_w2_ave'] = biased_w2_ave\n",
    "                biased_z_score_array = np.array(pd.to_numeric(z_score_biased['Z-Score'], errors='coerce'))\n",
    "                protein_df['Max Biased Z-Score'] = np.nanmax(biased_z_score_array)\n",
    "                protein_df['Biased Z-Score at 10%'] = biased_z_score_array[z_score_biased['Num_Residues'] >= float(len(protein_scorers[p_id]['Scorer_Any'].query_structure.seq[protein_scorers[p_id]['Scorer_Any'].best_chain])) * 0.10][0]\n",
    "                protein_df['Biased Z-Score at 30%'] = biased_z_score_array[z_score_biased['Num_Residues'] >= float(len(protein_scorers[p_id]['Scorer_Any'].query_structure.seq[protein_scorers[p_id]['Scorer_Any'].best_chain])) * 0.30][0]\n",
    "                protein_df['AUC Biased Z-Score'] = biased_scw_z_auc\n",
    "                plot_z_scores(z_score_biased, z_score_plot_fn.format('Biased'))\n",
    "                z_score_unbiased, unbiased_w2_ave, unbiased_scw_z_auc = protein_scorers[p_id]['Scorer_Any'].score_clustering_of_contact_predictions(\n",
    "                    1.0 - curr_etmip.coverage, bias=False, file_path=z_score_fn.format('Unbiased'),\n",
    "                    w2_ave_sub=protein_scorers[p_id]['unbiased_w2_ave'], processes=10)\n",
    "                if protein_scorers[p_id]['unbiased_w2_ave'] is None:\n",
    "                    protein_scorers[p_id]['unbiased_w2_ave'] = unbiased_w2_ave\n",
    "                unbiased_z_score_array = np.array(pd.to_numeric(z_score_unbiased['Z-Score'], errors='coerce'))\n",
    "                protein_df['Max Unbiased Z-Score'] = np.nanmax(unbiased_z_score_array)\n",
    "                protein_df['Unbiased Z-Score at 10%'] = unbiased_z_score_array[z_score_unbiased['Num_Residues'] >= float(len(protein_scorers[p_id]['Scorer_Any'].query_structure.seq[protein_scorers[p_id]['Scorer_Any'].best_chain])) * 0.10][0]\n",
    "                protein_df['Unbiased Z-Score at 30%'] = unbiased_z_score_array[z_score_unbiased['Num_Residues'] >= float(len(protein_scorers[p_id]['Scorer_Any'].query_structure.seq[protein_scorers[p_id]['Scorer_Any'].best_chain])) * 0.30][0]\n",
    "                protein_df['AUC Unbiased Z-Score'] = unbiased_scw_z_auc\n",
    "                plot_z_scores(z_score_unbiased, z_score_plot_fn.format('Unbiased'))\n",
    "                # Record static data for this protein\n",
    "                protein_df['Protein'] = p_id\n",
    "                protein_df['Method'] = 'cET-MIp'\n",
    "                protein_df['Alignment Size'] = summary['Filtered_Alignment'].values[summary['Protein_ID'] == p_id][0]\n",
    "                protein_df.to_csv(protein_fn, sep='\\t', header=True, index=False, columns=output_columns)\n",
    "                print('Metrics meastured for cET-MIp covariance for: {}'.format(p_id))\n",
    "                counts['success'] += 1\n",
    "            if old_cetmip_method_df is None:\n",
    "                old_cetmip_method_df = protein_df\n",
    "            else:\n",
    "                old_cetmip_method_df = old_cetmip_method_df.append(protein_df)\n",
    "        print('{}\\tSuccesses\\n{}\\tValue Errors\\n{}\\tAttribute Errors'.format(counts['success'], counts['value'],\n",
    "                                                                             counts['attribute']))\n",
    "        old_cetmip_method_df.to_csv(old_cetmip_method_fn, sep='\\t', header=True, index=False, columns=output_columns)\n",
    "    if small_comparison_df is None:\n",
    "        small_comparison_df = old_cetmip_method_df\n",
    "    else:\n",
    "        small_comparison_df = small_comparison_df.append(old_cetmip_method_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MI\n",
    "This segment imports MI scores generated as intermediate values by EV Couplings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Attempting to calculate MI covariance for: 3q05A\n",
      "Successfully computed MI covariance for: 3q05A in 33.034722566604614 sec\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Compute SCW Z-Score took 0.019817384084065755 min\n",
      "Compute SCW Z-Score took 0.0148736039797465 min\n",
      "Metrics meastured for MI covariance for: 3q05A\n",
      "Attempting to calculate MI covariance for: 2b59A\n",
      "Successfully computed MI covariance for: 2b59A in 33.034722566604614 sec\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Compute SCW Z-Score took 0.01036144495010376 min\n",
      "Compute SCW Z-Score took 0.010043176015218098 min\n",
      "Metrics meastured for MI covariance for: 2b59A\n",
      "Attempting to calculate MI covariance for: 7hvpA\n",
      "Successfully computed MI covariance for: 7hvpA in 33.034722566604614 sec\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Compute SCW Z-Score took 0.007858351866404215 min\n",
      "Compute SCW Z-Score took 0.007786663373311361 min\n",
      "Metrics meastured for MI covariance for: 7hvpA\n",
      "Attempting to calculate MI covariance for: 1c17A\n",
      "Successfully computed MI covariance for: 1c17A in 33.034722566604614 sec\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Compute SCW Z-Score took 0.007564135392506917 min\n",
      "Compute SCW Z-Score took 0.0077718575795491535 min\n",
      "Metrics meastured for MI covariance for: 1c17A\n",
      "Attempting to calculate MI covariance for: 206lA\n",
      "Successfully computed MI covariance for: 206lA in 33.034722566604614 sec\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Compute SCW Z-Score took 0.014216025670369467 min\n",
      "Compute SCW Z-Score took 0.010912875334421793 min\n",
      "Metrics meastured for MI covariance for: 206lA\n",
      "Attempting to calculate MI covariance for: 1bolA\n",
      "Successfully computed MI covariance for: 1bolA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.014923528830210368 min\n",
      "Compute SCW Z-Score took 0.02121281623840332 min\n",
      "Metrics meastured for MI covariance for: 1bolA\n",
      "Attempting to calculate MI covariance for: 2z0eA\n",
      "Successfully computed MI covariance for: 2z0eA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.08788102865219116 min\n",
      "Compute SCW Z-Score took 0.0910598874092102 min\n",
      "Metrics meastured for MI covariance for: 2z0eA\n",
      "Attempting to calculate MI covariance for: 1axbA\n",
      "Successfully computed MI covariance for: 1axbA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.026435808340708414 min\n",
      "Compute SCW Z-Score took 0.023765432834625243 min\n",
      "Metrics meastured for MI covariance for: 1axbA\n",
      "Attempting to calculate MI covariance for: 135lA\n",
      "Successfully computed MI covariance for: 135lA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.01333990494410197 min\n",
      "Compute SCW Z-Score took 0.013058805465698242 min\n",
      "Metrics meastured for MI covariance for: 135lA\n",
      "Attempting to calculate MI covariance for: 2rh1A\n",
      "Successfully computed MI covariance for: 2rh1A in 33.034722566604614 sec\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Compute SCW Z-Score took 0.040933644771575926 min\n",
      "Compute SCW Z-Score took 0.04050557613372803 min\n",
      "Metrics meastured for MI covariance for: 2rh1A\n",
      "Attempting to calculate MI covariance for: 4lliA\n",
      "Successfully computed MI covariance for: 4lliA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.06761047840118409 min\n",
      "Compute SCW Z-Score took 0.07163734833399454 min\n",
      "Metrics meastured for MI covariance for: 4lliA\n",
      "Attempting to calculate MI covariance for: 1a26A\n",
      "Successfully computed MI covariance for: 1a26A in 33.034722566604614 sec\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Compute SCW Z-Score took 0.13854771455128986 min\n",
      "Compute SCW Z-Score took 0.1327943444252014 min\n",
      "Metrics meastured for MI covariance for: 1a26A\n",
      "Attempting to calculate MI covariance for: 1c0kA\n",
      "Successfully computed MI covariance for: 1c0kA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.1327655553817749 min\n",
      "Compute SCW Z-Score took 0.14542426665623984 min\n",
      "Metrics meastured for MI covariance for: 1c0kA\n",
      "Attempting to calculate MI covariance for: 2zxeA\n",
      "Successfully computed MI covariance for: 2zxeA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.617483651638031 min\n",
      "Compute SCW Z-Score took 0.5900325576464335 min\n",
      "Metrics meastured for MI covariance for: 2zxeA\n",
      "Attempting to calculate MI covariance for: 1jwlA\n",
      "Successfully computed MI covariance for: 1jwlA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.0475423534711202 min\n",
      "Compute SCW Z-Score took 0.04368751049041748 min\n",
      "Metrics meastured for MI covariance for: 1jwlA\n",
      "Attempting to calculate MI covariance for: 1hckA\n",
      "Successfully computed MI covariance for: 1hckA in 33.034722566604614 sec\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Compute SCW Z-Score took 0.06903199354807536 min\n",
      "Compute SCW Z-Score took 0.06816544930140177 min\n",
      "Metrics meastured for MI covariance for: 1hckA\n",
      "Attempting to calculate MI covariance for: 1h1vA\n",
      "Successfully computed MI covariance for: 1h1vA in 33.034722566604614 sec\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Compute SCW Z-Score took 0.13476604223251343 min\n",
      "Compute SCW Z-Score took 0.1351438045501709 min\n",
      "Metrics meastured for MI covariance for: 1h1vA\n",
      "Attempting to calculate MI covariance for: 2ysdA\n",
      "Successfully computed MI covariance for: 2ysdA in 33.034722566604614 sec\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Compute SCW Z-Score took 0.008566888173421223 min\n",
      "Compute SCW Z-Score took 0.008618477980295818 min\n",
      "Metrics meastured for MI covariance for: 2ysdA\n",
      "Attempting to calculate MI covariance for: 2iopA\n",
      "Successfully computed MI covariance for: 2iopA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.21591771841049195 min\n",
      "Compute SCW Z-Score took 0.20942237774531047 min\n",
      "Metrics meastured for MI covariance for: 2iopA\n",
      "Attempting to calculate MI covariance for: 3b6vA\n",
      "Successfully computed MI covariance for: 3b6vA in 33.034722566604614 sec\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Compute SCW Z-Score took 0.0802385409673055 min\n",
      "Compute SCW Z-Score took 0.07609207232793172 min\n",
      "Metrics meastured for MI covariance for: 3b6vA\n",
      "Attempting to calculate MI covariance for: 4ycuA\n",
      "Successfully computed MI covariance for: 4ycuA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.13149775664011637 min\n",
      "Compute SCW Z-Score took 0.14062191247940065 min\n",
      "Metrics meastured for MI covariance for: 4ycuA\n",
      "Attempting to calculate MI covariance for: 2werA\n",
      "Successfully computed MI covariance for: 2werA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.024577407042185466 min\n",
      "Compute SCW Z-Score took 0.022806549072265626 min\n",
      "Metrics meastured for MI covariance for: 2werA\n",
      "Attempting to calculate MI covariance for: 3tnuA\n",
      "Successfully computed MI covariance for: 3tnuA in 33.034722566604614 sec\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/ranking.py:571: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/home/daniel/Documents/git/ETMIP/src/SupportingClasses/ContactScorer.py:768: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/ranking.py:571: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/home/daniel/Documents/git/ETMIP/src/SupportingClasses/ContactScorer.py:768: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/ranking.py:571: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/home/daniel/Documents/git/ETMIP/src/SupportingClasses/ContactScorer.py:768: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/ranking.py:563: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/home/daniel/Documents/git/ETMIP/src/SupportingClasses/ContactScorer.py:754: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/ranking.py:571: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/home/daniel/Documents/git/ETMIP/src/SupportingClasses/ContactScorer.py:768: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/ranking.py:571: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/home/daniel/Documents/git/ETMIP/src/SupportingClasses/ContactScorer.py:768: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Compute SCW Z-Score took 0.009751419226328531 min\n",
      "Compute SCW Z-Score took 0.009075323740641275 min\n",
      "Metrics meastured for MI covariance for: 3tnuA\n",
      "23\tSuccesses\n",
      "0\tValue Errors\n",
      "0\tAttribute Errors\n"
     ]
    }
   ],
   "source": [
    "from utils import compute_rank_and_coverage\n",
    "def import_MI_cov_scores(path, aln):\n",
    "    data = pd.read_csv(path, sep='\\s+', header=None, index_col=None, names=['Pos1', 'AA1', 'Pos2', 'AA2', 'MI', 'DI'])\n",
    "    scores = np.zeros((aln.seq_length, aln.seq_length))\n",
    "    positions = sorted(set(data['Pos1']) | set(data['Pos2']))\n",
    "    pos_map = {x: i for i, x in enumerate(positions)}\n",
    "    data['Final_Pos1'] = data['Pos1'].apply(lambda x: pos_map[x])\n",
    "    data['Final_Pos2'] = data['Pos2'].apply(lambda x: pos_map[x])\n",
    "    scores[data['Final_Pos1'].values, data['Final_Pos2'].values] = data['MI'].values\n",
    "    _, coverages = compute_rank_and_coverage(seq_length=aln.seq_length, scores=scores, pos_size=2, rank_type='max')\n",
    "    return scores, coverages\n",
    "\n",
    "if not os.path.isfile(small_comparison_fn):\n",
    "    old_mi_out_dir = os.path.join(small_set_out_dir, 'EVCouplings_MI')\n",
    "    if not os.path.isdir(old_mi_out_dir):\n",
    "        os.makedirs(old_mi_out_dir)\n",
    "    old_mi_method_fn = os.path.join(old_mi_out_dir, 'MI_Method_Data.csv')\n",
    "    if os.path.isfile(old_mi_method_fn):\n",
    "        old_mi_method_df = pd.read_csv(old_mi_method_fn, sep='\\t', header=0, index_col=False)\n",
    "    else:    \n",
    "        old_mi_method_df = None\n",
    "        counts = {'success':0, 'value': 0, 'attribute':0}\n",
    "        for p_id in summary['Protein_ID']:\n",
    "            print('Attempting to calculate MI covariance for: {}'.format(p_id))\n",
    "            protein_dir = os.path.join(old_mi_out_dir, p_id)\n",
    "            if not os.path.isdir(protein_dir):\n",
    "                os.makedirs(protein_dir)\n",
    "            protein_fn = os.path.join(protein_dir, '{}_Protein_Data.csv'.format(p_id))\n",
    "            if os.path.isfile(protein_fn):\n",
    "                protein_df = pd.read_csv(protein_fn, sep='\\t', header=0, index_col=False)\n",
    "                counts['success'] += 1\n",
    "            else:\n",
    "#                 curr_aln = SeqAlignment(file_name=generator.protein_data[p_id]['Final_FA_Aln'], query_id='query_' + p_id,\n",
    "#                                         polymer_type='Protein')\n",
    "#                 curr_aln.import_alignment()\n",
    "                curr_etmip = ETMIPWrapper(query='query_' + p_id, aln_file=generator.protein_data[p_id]['Final_FA_Aln'],\n",
    "                                          out_dir=protein_dir)\n",
    "                curr_etmip.scores, curr_etmip.coverage = import_MI_cov_scores(\n",
    "                    os.path.join(protein_dir, '{}_MI_DI.txt'.format(p_id)), curr_aln)\n",
    "                np.savez(os.path.join(protein_dir, 'MI.npz'), time=curr_etmip.time,\n",
    "                         scores=curr_etmip.scores, coverage=curr_etmip.coverage)\n",
    "                import pickle\n",
    "                with open(os.path.join(protein_dir, 'MI.pkl'), 'wb') as handle:\n",
    "                    pickle.dump((curr_etmip.distance_matrix, curr_etmip.tree, curr_etmip.rank_group_assignments),\n",
    "                                handle, pickle.HIGHEST_PROTOCOL)\n",
    "                print('Successfully computed MI covariance for: {} in {} sec'.format(p_id, end-start))\n",
    "                # Compute statistics for the final scores of the ET-MIp model\n",
    "                protein_df, _, _ = protein_scorers[p_id]['Scorer_CB'].evaluate_predictor(\n",
    "                    predictor=curr_etmip, verbosity=2, out_dir=protein_dir, dist='CB', biased_w2_ave=None,\n",
    "                    unbiased_w2_ave=None, processes=10, threshold=0.5, pos_size=2, rank_type='max',\n",
    "                    file_prefix='MI_Scores_', plots=True)\n",
    "                protein_df2, _, _ = protein_scorers[p_id]['Scorer_Any'].evaluate_predictor(\n",
    "                    predictor=curr_etmip, verbosity=2, out_dir=protein_dir, dist='Any', biased_w2_ave=None,\n",
    "                    unbiased_w2_ave=None, processes=10, threshold=0.5, pos_size=2, rank_type='max',\n",
    "                    file_prefix='MI_Scores_Dist_Any', plots=True)\n",
    "                protein_df = protein_df.append(protein_df2)\n",
    "                # Score Prediction Clustering\n",
    "                z_score_fn = os.path.join(protein_dir, 'MI_Scores_Dist-Any_{}_ZScores.tsv')\n",
    "                z_score_plot_fn = os.path.join(protein_dir, 'MI_Scores_Dist-Any_{}_ZScores.png')\n",
    "                z_score_biased, biased_w2_ave, biased_scw_z_auc = protein_scorers[p_id]['Scorer_Any'].score_clustering_of_contact_predictions(\n",
    "                    1.0 - curr_etmip.coverage, bias=True, file_path=z_score_fn.format('Biased'),\n",
    "                    w2_ave_sub=protein_scorers[p_id]['biased_w2_ave'], processes=10)\n",
    "                if protein_scorers[p_id]['biased_w2_ave'] is None:\n",
    "                    protein_scorers[p_id]['biased_w2_ave'] = biased_w2_ave\n",
    "                biased_z_score_array = np.array(pd.to_numeric(z_score_biased['Z-Score'], errors='coerce'))\n",
    "                protein_df['Max Biased Z-Score'] = np.nanmax(biased_z_score_array)\n",
    "                protein_df['Biased Z-Score at 10%'] = biased_z_score_array[z_score_biased['Num_Residues'] >= float(len(protein_scorers[p_id]['Scorer_Any'].query_structure.seq[protein_scorers[p_id]['Scorer_Any'].best_chain])) * 0.10][0]\n",
    "                protein_df['Biased Z-Score at 30%'] = biased_z_score_array[z_score_biased['Num_Residues'] >= float(len(protein_scorers[p_id]['Scorer_Any'].query_structure.seq[protein_scorers[p_id]['Scorer_Any'].best_chain])) * 0.30][0]\n",
    "                protein_df['AUC Biased Z-Score'] = biased_scw_z_auc\n",
    "                plot_z_scores(z_score_biased, z_score_plot_fn.format('Biased'))\n",
    "                z_score_unbiased, unbiased_w2_ave, unbiased_scw_z_auc = protein_scorers[p_id]['Scorer_Any'].score_clustering_of_contact_predictions(\n",
    "                    1.0 - curr_etmip.coverage, bias=False, file_path=z_score_fn.format('Unbiased'),\n",
    "                    w2_ave_sub=protein_scorers[p_id]['unbiased_w2_ave'], processes=10)\n",
    "                if protein_scorers[p_id]['unbiased_w2_ave'] is None:\n",
    "                    protein_scorers[p_id]['unbiased_w2_ave'] = unbiased_w2_ave\n",
    "                unbiased_z_score_array = np.array(pd.to_numeric(z_score_unbiased['Z-Score'], errors='coerce'))\n",
    "                protein_df['Max Unbiased Z-Score'] = np.nanmax(unbiased_z_score_array)\n",
    "                protein_df['Unbiased Z-Score at 10%'] = unbiased_z_score_array[z_score_unbiased['Num_Residues'] >= float(len(protein_scorers[p_id]['Scorer_Any'].query_structure.seq[protein_scorers[p_id]['Scorer_Any'].best_chain])) * 0.10][0]\n",
    "                protein_df['Unbiased Z-Score at 30%'] = unbiased_z_score_array[z_score_unbiased['Num_Residues'] >= float(len(protein_scorers[p_id]['Scorer_Any'].query_structure.seq[protein_scorers[p_id]['Scorer_Any'].best_chain])) * 0.30][0]\n",
    "                protein_df['AUC Unbiased Z-Score'] = unbiased_scw_z_auc\n",
    "                plot_z_scores(z_score_unbiased, z_score_plot_fn.format('Unbiased'))\n",
    "                # Record static data for this protein\n",
    "                protein_df['Protein'] = p_id\n",
    "                protein_df['Method'] = 'MI'\n",
    "                protein_df['Alignment Size'] = summary['Filtered_Alignment'].values[summary['Protein_ID'] == p_id][0]\n",
    "                protein_df.to_csv(protein_fn, sep='\\t', header=True, index=False, columns=output_columns)\n",
    "                print('Metrics meastured for MI covariance for: {}'.format(p_id))\n",
    "                counts['success'] += 1\n",
    "            if old_mi_method_df is None:\n",
    "                old_mi_method_df = protein_df\n",
    "            else:\n",
    "                old_mi_method_df = old_mi_method_df.append(protein_df)\n",
    "        print('{}\\tSuccesses\\n{}\\tValue Errors\\n{}\\tAttribute Errors'.format(counts['success'], counts['value'],\n",
    "                                                                             counts['attribute']))\n",
    "        old_mi_method_df.to_csv(old_mi_method_fn, sep='\\t', header=True, index=False, columns=output_columns)\n",
    "    if small_comparison_df is None:\n",
    "        small_comparison_df = old_mi_method_df\n",
    "    else:\n",
    "        small_comparison_df = small_comparison_df.append(old_mi_method_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DI\n",
    "This segment imports DI scores generated as intermediate values by EV Couplings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Attempting to calculate DI covariance for: 3q05A\n",
      "Successfully computed dI covariance for: 3q05A in 33.034722566604614 sec\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Compute SCW Z-Score took 0.014219431082407634 min\n",
      "Compute SCW Z-Score took 0.01660402218500773 min\n",
      "Metrics meastured for DI covariance for: 3q05A\n",
      "Attempting to calculate DI covariance for: 2b59A\n",
      "Successfully computed dI covariance for: 2b59A in 33.034722566604614 sec\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Compute SCW Z-Score took 0.011115042368570964 min\n",
      "Compute SCW Z-Score took 0.01041925350824992 min\n",
      "Metrics meastured for DI covariance for: 2b59A\n",
      "Attempting to calculate DI covariance for: 7hvpA\n",
      "Successfully computed dI covariance for: 7hvpA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.007634969552357992 min\n",
      "Compute SCW Z-Score took 0.007882972558339437 min\n",
      "Metrics meastured for DI covariance for: 7hvpA\n",
      "Attempting to calculate DI covariance for: 1c17A\n",
      "Successfully computed dI covariance for: 1c17A in 33.034722566604614 sec\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Compute SCW Z-Score took 0.0075304587682088215 min\n",
      "Compute SCW Z-Score took 0.011876654624938966 min\n",
      "Metrics meastured for DI covariance for: 1c17A\n",
      "Attempting to calculate DI covariance for: 206lA\n",
      "Successfully computed dI covariance for: 206lA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.015273336569468181 min\n",
      "Compute SCW Z-Score took 0.010190558433532716 min\n",
      "Metrics meastured for DI covariance for: 206lA\n",
      "Attempting to calculate DI covariance for: 1bolA\n",
      "Successfully computed dI covariance for: 1bolA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.014323906103769938 min\n",
      "Compute SCW Z-Score took 0.01763444741566976 min\n",
      "Metrics meastured for DI covariance for: 1bolA\n",
      "Attempting to calculate DI covariance for: 2z0eA\n",
      "Successfully computed dI covariance for: 2z0eA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.08505027294158936 min\n",
      "Compute SCW Z-Score took 0.09089635610580445 min\n",
      "Metrics meastured for DI covariance for: 2z0eA\n",
      "Attempting to calculate DI covariance for: 1axbA\n",
      "Successfully computed dI covariance for: 1axbA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.026619390646616618 min\n",
      "Compute SCW Z-Score took 0.020876983801523846 min\n",
      "Metrics meastured for DI covariance for: 1axbA\n",
      "Attempting to calculate DI covariance for: 135lA\n",
      "Successfully computed dI covariance for: 135lA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.01137075424194336 min\n",
      "Compute SCW Z-Score took 0.011153435707092286 min\n",
      "Metrics meastured for DI covariance for: 135lA\n",
      "Attempting to calculate DI covariance for: 2rh1A\n",
      "Successfully computed dI covariance for: 2rh1A in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.034355664253234865 min\n",
      "Compute SCW Z-Score took 0.04020524024963379 min\n",
      "Metrics meastured for DI covariance for: 2rh1A\n",
      "Attempting to calculate DI covariance for: 4lliA\n",
      "Successfully computed dI covariance for: 4lliA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.06086438496907552 min\n",
      "Compute SCW Z-Score took 0.06843577226003011 min\n",
      "Metrics meastured for DI covariance for: 4lliA\n",
      "Attempting to calculate DI covariance for: 1a26A\n",
      "Successfully computed dI covariance for: 1a26A in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.1372578223546346 min\n",
      "Compute SCW Z-Score took 0.12263751824696859 min\n",
      "Metrics meastured for DI covariance for: 1a26A\n",
      "Attempting to calculate DI covariance for: 1c0kA\n",
      "Successfully computed dI covariance for: 1c0kA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.1326773802439372 min\n",
      "Compute SCW Z-Score took 0.1228037158648173 min\n",
      "Metrics meastured for DI covariance for: 1c0kA\n",
      "Attempting to calculate DI covariance for: 2zxeA\n",
      "Successfully computed dI covariance for: 2zxeA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.6343970417976379 min\n",
      "Compute SCW Z-Score took 0.5679650902748108 min\n",
      "Metrics meastured for DI covariance for: 2zxeA\n",
      "Attempting to calculate DI covariance for: 1jwlA\n",
      "Successfully computed dI covariance for: 1jwlA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.041952657699584964 min\n",
      "Compute SCW Z-Score took 0.0448902686436971 min\n",
      "Metrics meastured for DI covariance for: 1jwlA\n",
      "Attempting to calculate DI covariance for: 1hckA\n",
      "Successfully computed dI covariance for: 1hckA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.06991593837738037 min\n",
      "Compute SCW Z-Score took 0.0637492815653483 min\n",
      "Metrics meastured for DI covariance for: 1hckA\n",
      "Attempting to calculate DI covariance for: 1h1vA\n",
      "Successfully computed dI covariance for: 1h1vA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.12593545913696289 min\n",
      "Compute SCW Z-Score took 0.14182532628377278 min\n",
      "Metrics meastured for DI covariance for: 1h1vA\n",
      "Attempting to calculate DI covariance for: 2ysdA\n",
      "Successfully computed dI covariance for: 2ysdA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.008576003710428874 min\n",
      "Compute SCW Z-Score took 0.009652336438496908 min\n",
      "Metrics meastured for DI covariance for: 2ysdA\n",
      "Attempting to calculate DI covariance for: 2iopA\n",
      "Successfully computed dI covariance for: 2iopA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.20125606854756672 min\n",
      "Compute SCW Z-Score took 0.20031355222066244 min\n",
      "Metrics meastured for DI covariance for: 2iopA\n",
      "Attempting to calculate DI covariance for: 3b6vA\n",
      "Successfully computed dI covariance for: 3b6vA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.0737372080485026 min\n",
      "Compute SCW Z-Score took 0.07843906879425049 min\n",
      "Metrics meastured for DI covariance for: 3b6vA\n",
      "Attempting to calculate DI covariance for: 4ycuA\n",
      "Successfully computed dI covariance for: 4ycuA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.13827242453893027 min\n",
      "Compute SCW Z-Score took 0.12808012167612712 min\n",
      "Metrics meastured for DI covariance for: 4ycuA\n",
      "Attempting to calculate DI covariance for: 2werA\n",
      "Successfully computed dI covariance for: 2werA in 33.034722566604614 sec\n",
      "Compute SCW Z-Score took 0.025107852617899575 min\n",
      "Compute SCW Z-Score took 0.02021290858586629 min\n",
      "Metrics meastured for DI covariance for: 2werA\n",
      "Attempting to calculate DI covariance for: 3tnuA\n",
      "Successfully computed dI covariance for: 3tnuA in 33.034722566604614 sec\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/ranking.py:571: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/home/daniel/Documents/git/ETMIP/src/SupportingClasses/ContactScorer.py:768: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/ranking.py:571: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/home/daniel/Documents/git/ETMIP/src/SupportingClasses/ContactScorer.py:768: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/ranking.py:571: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/home/daniel/Documents/git/ETMIP/src/SupportingClasses/ContactScorer.py:768: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/ranking.py:563: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/home/daniel/Documents/git/ETMIP/src/SupportingClasses/ContactScorer.py:754: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/ranking.py:571: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/home/daniel/Documents/git/ETMIP/src/SupportingClasses/ContactScorer.py:768: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/ranking.py:571: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/home/daniel/Documents/git/ETMIP/src/SupportingClasses/ContactScorer.py:768: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless\n",
      "  UndefinedMetricWarning)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 due to no true samples.\n",
      "  'recall', 'true', average, warn_for)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Compute SCW Z-Score took 0.008866433302561443 min\n",
      "Compute SCW Z-Score took 0.010055756568908692 min\n",
      "Metrics meastured for DI covariance for: 3tnuA\n",
      "23\tSuccesses\n",
      "0\tValue Errors\n",
      "0\tAttribute Errors\n"
     ]
    }
   ],
   "source": [
    "from utils import compute_rank_and_coverage\n",
    "def import_DI_cov_scores(path, aln):\n",
    "    data = pd.read_csv(path, sep='\\s+', header=None, index_col=None, names=['Pos1', 'AA1', 'Pos2', 'AA2', 'MI', 'DI'])\n",
    "    scores = np.zeros((aln.seq_length, aln.seq_length))\n",
    "    positions = sorted(set(data['Pos1']) | set(data['Pos2']))\n",
    "    pos_map = {x: i for i, x in enumerate(positions)}\n",
    "    data['Final_Pos1'] = data['Pos1'].apply(lambda x: pos_map[x])\n",
    "    data['Final_Pos2'] = data['Pos2'].apply(lambda x: pos_map[x])\n",
    "    scores[data['Final_Pos1'].values, data['Final_Pos2'].values] = data['DI'].values\n",
    "    _, coverages = compute_rank_and_coverage(seq_length=aln.seq_length, scores=scores, pos_size=2, rank_type='max')\n",
    "    return scores, coverages\n",
    "\n",
    "if not os.path.isfile(small_comparison_fn):\n",
    "    old_di_out_dir = os.path.join(small_set_out_dir, 'EVCouplings_DI')\n",
    "    if not os.path.isdir(old_di_out_dir):\n",
    "        os.makedirs(old_di_out_dir)\n",
    "    old_di_method_fn = os.path.join(old_di_out_dir, 'DI_Method_Data.csv')\n",
    "    if os.path.isfile(old_di_method_fn):\n",
    "        old_di_method_df = pd.read_csv(old_di_method_fn, sep='\\t', header=0, index_col=False)\n",
    "    else:    \n",
    "        old_di_method_df = None\n",
    "        counts = {'success':0, 'value': 0, 'attribute':0}\n",
    "        for p_id in summary['Protein_ID']:\n",
    "            print('Attempting to calculate DI covariance for: {}'.format(p_id))\n",
    "            protein_dir = os.path.join(old_di_out_dir, p_id)\n",
    "            if not os.path.isdir(protein_dir):\n",
    "                os.makedirs(protein_dir)\n",
    "            protein_fn = os.path.join(protein_dir, '{}_Protein_Data.csv'.format(p_id))\n",
    "            if os.path.isfile(protein_fn):\n",
    "                protein_df = pd.read_csv(protein_fn, sep='\\t', header=0, index_col=False)\n",
    "                counts['success'] += 1\n",
    "            else:\n",
    "#                 curr_aln = SeqAlignment(file_name=generator.protein_data[p_id]['Final_FA_Aln'], query_id='query_' + p_id,\n",
    "#                                         polymer_type='Protein')\n",
    "#                 curr_aln.import_alignment()\n",
    "                curr_etmip = ETMIPWrapper(query='query_' + p_id, aln_file=generator.protein_data[p_id]['Final_FA_Aln']protein_dir,\n",
    "                                          out_dir=protein_dir)\n",
    "                curr_etmip.scores, curr_etmip.coverage = import_DI_cov_scores(\n",
    "                    os.path.join(protein_dir, '{}_MI_DI.txt'.format(p_id)), curr_aln)\n",
    "                np.savez(os.path.join(protein_dir, 'dI.npz'), time=curr_etmip.time,\n",
    "                         scores=curr_etmip.scores, coverage=curr_etmip.coverage)\n",
    "                import pickle\n",
    "                with open(os.path.join(protein_dir, 'dI.pkl'), 'wb') as handle:\n",
    "                    pickle.dump((curr_etmip.distance_matrix, curr_etmip.tree, curr_etmip.rank_group_assignments),\n",
    "                                handle, pickle.HIGHEST_PROTOCOL)\n",
    "                print('Successfully computed dI covariance for: {} in {} sec'.format(p_id, end-start))\n",
    "                # Compute statistics for the final scores of the ET-MIp model\n",
    "                protein_df, _, _ = protein_scorers[p_id]['Scorer_CB'].evaluate_predictor(\n",
    "                    predictor=curr_etmip, verbosity=2, out_dir=protein_dir, dist='CB', biased_w2_ave=None,\n",
    "                    unbiased_w2_ave=None, processes=10, threshold=0.5, pos_size=2, rank_type='max',\n",
    "                    file_prefix='DI_Scores_', plots=True)\n",
    "                protein_df2, _, _ = protein_scorers[p_id]['Scorer_Any'].evaluate_predictor(\n",
    "                    predictor=curr_etmip, verbosity=2, out_dir=protein_dir, dist='Any', biased_w2_ave=None,\n",
    "                    unbiased_w2_ave=None, processes=10, threshold=0.5, pos_size=2, rank_type='max',\n",
    "                    file_prefix='DI_Scores_Dist_Any', plots=True)\n",
    "                protein_df = protein_df.append(protein_df2)\n",
    "                # Score Prediction Clustering\n",
    "                z_score_fn = os.path.join(protein_dir, 'DI_Scores_Dist-Any_{}_ZScores.tsv')\n",
    "                z_score_plot_fn = os.path.join(protein_dir, 'DI_Scores_Dist-Any_{}_ZScores.png')\n",
    "                z_score_biased, biased_w2_ave, biased_scw_z_auc = protein_scorers[p_id]['Scorer_Any'].score_clustering_of_contact_predictions(\n",
    "                    1.0 - curr_etmip.coverage, bias=True, file_path=z_score_fn.format('Biased'),\n",
    "                    w2_ave_sub=protein_scorers[p_id]['biased_w2_ave'], processes=10)\n",
    "                if protein_scorers[p_id]['biased_w2_ave'] is None:\n",
    "                    protein_scorers[p_id]['biased_w2_ave'] = biased_w2_ave\n",
    "                biased_z_score_array = np.array(pd.to_numeric(z_score_biased['Z-Score'], errors='coerce'))\n",
    "                protein_df['Max Biased Z-Score'] = np.nanmax(biased_z_score_array)\n",
    "                protein_df['Biased Z-Score at 10%'] = biased_z_score_array[z_score_biased['Num_Residues'] >= float(len(protein_scorers[p_id]['Scorer_Any'].query_structure.seq[protein_scorers[p_id]['Scorer_Any'].best_chain])) * 0.10][0]\n",
    "                protein_df['Biased Z-Score at 30%'] = biased_z_score_array[z_score_biased['Num_Residues'] >= float(len(protein_scorers[p_id]['Scorer_Any'].query_structure.seq[protein_scorers[p_id]['Scorer_Any'].best_chain])) * 0.30][0]\n",
    "                protein_df['AUC Biased Z-Score'] = biased_scw_z_auc\n",
    "                plot_z_scores(z_score_biased, z_score_plot_fn.format('Biased'))\n",
    "                z_score_unbiased, unbiased_w2_ave, unbiased_scw_z_auc = protein_scorers[p_id]['Scorer_Any'].score_clustering_of_contact_predictions(\n",
    "                    1.0 - curr_etmip.coverage, bias=False, file_path=z_score_fn.format('Unbiased'),\n",
    "                    w2_ave_sub=protein_scorers[p_id]['unbiased_w2_ave'], processes=10)\n",
    "                if protein_scorers[p_id]['unbiased_w2_ave'] is None:\n",
    "                    protein_scorers[p_id]['unbiased_w2_ave'] = unbiased_w2_ave\n",
    "                unbiased_z_score_array = np.array(pd.to_numeric(z_score_unbiased['Z-Score'], errors='coerce'))\n",
    "                protein_df['Max Unbiased Z-Score'] = np.nanmax(unbiased_z_score_array)\n",
    "                protein_df['Unbiased Z-Score at 10%'] = unbiased_z_score_array[z_score_unbiased['Num_Residues'] >= float(len(protein_scorers[p_id]['Scorer_Any'].query_structure.seq[protein_scorers[p_id]['Scorer_Any'].best_chain])) * 0.10][0]\n",
    "                protein_df['Unbiased Z-Score at 30%'] = unbiased_z_score_array[z_score_unbiased['Num_Residues'] >= float(len(protein_scorers[p_id]['Scorer_Any'].query_structure.seq[protein_scorers[p_id]['Scorer_Any'].best_chain])) * 0.30][0]\n",
    "                protein_df['AUC Unbiased Z-Score'] = unbiased_scw_z_auc\n",
    "                plot_z_scores(z_score_unbiased, z_score_plot_fn.format('Unbiased'))\n",
    "                # Record static data for this protein\n",
    "                protein_df['Protein'] = p_id\n",
    "                protein_df['Method'] = 'DI'\n",
    "                protein_df['Alignment Size'] = summary['Filtered_Alignment'].values[summary['Protein_ID'] == p_id][0]\n",
    "                protein_df.to_csv(protein_fn, sep='\\t', header=True, index=False, columns=output_columns)\n",
    "                print('Metrics meastured for DI covariance for: {}'.format(p_id))\n",
    "                counts['success'] += 1\n",
    "            if old_di_method_df is None:\n",
    "                old_di_method_df = protein_df\n",
    "            else:\n",
    "                old_di_method_df = old_di_method_df.append(protein_df)\n",
    "        print('{}\\tSuccesses\\n{}\\tValue Errors\\n{}\\tAttribute Errors'.format(counts['success'], counts['value'],\n",
    "                                                                             counts['attribute']))\n",
    "        old_di_method_df.to_csv(old_di_method_fn, sep='\\t', header=True, index=False, columns=output_columns)\n",
    "    if small_comparison_df is None:\n",
    "        small_comparison_df = old_di_method_df\n",
    "    else:\n",
    "        small_comparison_df = small_comparison_df.append(old_di_method_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write out final comparison data so it can be loaded later for generating figures.\n",
    "if not os.path.isfile(small_comparison_fn):\n",
    "    small_comparison_df['Protein Length'] = small_comparison_df['Protein'].apply(lambda x: generator.protein_data[x]['Length'])\n",
    "    small_comparison_df.to_csv(small_comparison_fn, sep='\\t', header=True, index=False, columns=output_columns)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Method Comparison\n",
    "We now begin comparing methods based on their ability to predict the structural contacts in the proteins in this test set. There is an important consideration in the case of sequence separation and different measures by which to compare the methods.\n",
    "\n",
    "### Data Cleaning\n",
    "These data need an additional cleaning step beyond what was performed for the timing comparison. Since there are multiple categories of sequence separation and some proteins may not have any True Positive contacts for a category the scoring for that protein is incomplete. We will remove all such proteins from the comparison, performing the clean up separate for each metric of success. Another contributing factor which necessitates this kind of cleaning is assessment of the top K predictions for a protein or best L/K predictions which for poor predictions may not include any predictions of True Positives.\n",
    "\n",
    "### Sequence Separation\n",
    "One important consideration for the difficulty of prediction and interest in predictions is the distance between the residues for which coupling was predicted. As has been documented in the literature, especially in the CASP competitions, there are several categories of prediction:\n",
    "* Neighbors (1 - 5 residues apart) - This is the least interesting category of predictions. It is highly likely that residues this close together will show covariance signal. Predicting two residues are in contact that are this close together is trivial and uninformative.\n",
    "* Short (6 - 12 residues apart) - This is also not a very interesting type of prediction. Residues this close in proximity can be more easily modeled by alogrithms which focus on 2D protein structure modeling (identifying beta sheets, alpha helices, etc.).\n",
    "* Medium (13 - 24 residues apart) - This is a more interesting type of prediction. The resiudes in this range of separation are on the edge of the 2D protein structure prediction range.\n",
    "* Long (24 and more residues apart) - The most interesting category of predictions. Resiudes this far apart are not easily modeled by 2D protein structure modeling systems. They are also very useful for 3D and 4D protein structure prediction becausae they provide constraints on potential protein (similar to NMR data) folds which makes protein folding a more tractable problem for modelers.\n",
    "* Any/All - All categories can be considered at once, this provides a summary value, but is often skewed by one particularly good category of predictions.\n",
    "\n",
    "### Metrics of Success\n",
    "* AUROC - This measures the True Positive Rate vs the False Positive Rate of prediction, it can be considered a measure of the accuracy of the measure. This can be strongly influenced by the class imbalance which is present when predicting structural contacts since there are many fewer contacts than non-contacts. The True Positive case is if the C-beta of two amino acids is within 8.0 Angstroms of one another (as is done in the CASP competitions)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MI\n",
      "21\n",
      "   Protein Method Distance Sequence_Separation     AUROC\n",
      "10   3q05A     MI       CB           Neighbors  0.520454\n",
      "10   2b59A     MI       CB           Neighbors  0.513605\n",
      "10   7hvpA     MI       CB           Neighbors  0.513136\n",
      "10   1c17A     MI       CB           Neighbors  0.560120\n",
      "10   206lA     MI       CB           Neighbors  0.536861\n",
      "10   1bolA     MI       CB           Neighbors  0.505232\n",
      "10   2z0eA     MI       CB           Neighbors  0.548819\n",
      "10   1axbA     MI       CB           Neighbors  0.565785\n",
      "10   135lA     MI       CB           Neighbors  0.493339\n",
      "10   4lliA     MI       CB           Neighbors  0.478922\n",
      "10   1a26A     MI       CB           Neighbors  0.565014\n",
      "10   1c0kA     MI       CB           Neighbors  0.596392\n",
      "10   2zxeA     MI       CB           Neighbors  0.564293\n",
      "10   1jwlA     MI       CB           Neighbors  0.534891\n",
      "10   1hckA     MI       CB           Neighbors  0.506842\n",
      "10   1h1vA     MI       CB           Neighbors  0.537764\n",
      "10   2ysdA     MI       CB           Neighbors  0.516168\n",
      "10   2iopA     MI       CB           Neighbors  0.510709\n",
      "10   3b6vA     MI       CB           Neighbors  0.538781\n",
      "10   4ycuA     MI       CB           Neighbors  0.575158\n",
      "10   2werA     MI       CB           Neighbors  0.501789\n",
      "DCA\n",
      "22\n",
      "    Protein Method Distance Sequence_Separation     AUROC\n",
      "10    2ysdA    DCA       CB           Neighbors  0.608076\n",
      "10    1c17A    DCA       CB           Neighbors  0.534796\n",
      "10    7hvpA    DCA       CB           Neighbors  0.574923\n",
      "10    135lA    DCA       CB           Neighbors  0.638480\n",
      "10    206lA    DCA       CB           Neighbors  0.601593\n",
      "10    2werA    DCA       CB           Neighbors  0.643752\n",
      "10    1bolA    DCA       CB           Neighbors  0.598580\n",
      "10    3q05A    DCA       CB           Neighbors  0.563625\n",
      "10    1hckA    DCA       CB           Neighbors  0.622345\n",
      "10    3b6vA    DCA       CB           Neighbors  0.643803\n",
      "10    2z0eA    DCA       CB           Neighbors  0.612559\n",
      "10    1jwlA    DCA       CB           Neighbors  0.684952\n",
      "10    1a26A    DCA       CB           Neighbors  0.643024\n",
      "10    1c0kA    DCA       CB           Neighbors  0.650386\n",
      "10    4lliA    DCA       CB           Neighbors  0.578597\n",
      "10    4ycuA    DCA       CB           Neighbors  0.656378\n",
      "10    2iopA    DCA       CB           Neighbors  0.622600\n",
      "10    2zxeA    DCA       CB           Neighbors  0.624173\n",
      "10    2b59A    DCA       CB           Neighbors  0.548168\n",
      "10    1h1vA    DCA       CB           Neighbors  0.630677\n",
      "610   2werA    DCA       CB           Neighbors  0.643752\n",
      "10    1axbA    DCA       CB           Neighbors  0.656607\n",
      "DI\n",
      "21\n",
      "   Protein Method Distance Sequence_Separation     AUROC\n",
      "10   3q05A     DI       CB           Neighbors  0.562555\n",
      "10   2b59A     DI       CB           Neighbors  0.531583\n",
      "10   7hvpA     DI       CB           Neighbors  0.574649\n",
      "10   1c17A     DI       CB           Neighbors  0.566062\n",
      "10   206lA     DI       CB           Neighbors  0.556958\n",
      "10   1bolA     DI       CB           Neighbors  0.585795\n",
      "10   2z0eA     DI       CB           Neighbors  0.627612\n",
      "10   1axbA     DI       CB           Neighbors  0.633799\n",
      "10   135lA     DI       CB           Neighbors  0.605472\n",
      "10   4lliA     DI       CB           Neighbors  0.537827\n",
      "10   1a26A     DI       CB           Neighbors  0.647842\n",
      "10   1c0kA     DI       CB           Neighbors  0.664422\n",
      "10   2zxeA     DI       CB           Neighbors  0.603640\n",
      "10   1jwlA     DI       CB           Neighbors  0.646026\n",
      "10   1hckA     DI       CB           Neighbors  0.582186\n",
      "10   1h1vA     DI       CB           Neighbors  0.649313\n",
      "10   2ysdA     DI       CB           Neighbors  0.606911\n",
      "10   2iopA     DI       CB           Neighbors  0.583950\n",
      "10   3b6vA     DI       CB           Neighbors  0.608340\n",
      "10   4ycuA     DI       CB           Neighbors  0.637703\n",
      "10   2werA     DI       CB           Neighbors  0.601745\n",
      "DCA\n",
      "22\n",
      "    Protein Method Distance Sequence_Separation     AUROC\n",
      "10    2ysdA    DCA       CB           Neighbors  0.608076\n",
      "10    1c17A    DCA       CB           Neighbors  0.534796\n",
      "10    7hvpA    DCA       CB           Neighbors  0.574923\n",
      "10    135lA    DCA       CB           Neighbors  0.638480\n",
      "10    206lA    DCA       CB           Neighbors  0.601593\n",
      "10    2werA    DCA       CB           Neighbors  0.643752\n",
      "10    1bolA    DCA       CB           Neighbors  0.598580\n",
      "10    3q05A    DCA       CB           Neighbors  0.563625\n",
      "10    1hckA    DCA       CB           Neighbors  0.622345\n",
      "10    3b6vA    DCA       CB           Neighbors  0.643803\n",
      "10    2z0eA    DCA       CB           Neighbors  0.612559\n",
      "10    1jwlA    DCA       CB           Neighbors  0.684952\n",
      "10    1a26A    DCA       CB           Neighbors  0.643024\n",
      "10    1c0kA    DCA       CB           Neighbors  0.650386\n",
      "10    4lliA    DCA       CB           Neighbors  0.578597\n",
      "10    4ycuA    DCA       CB           Neighbors  0.656378\n",
      "10    2iopA    DCA       CB           Neighbors  0.622600\n",
      "10    2zxeA    DCA       CB           Neighbors  0.624173\n",
      "10    2b59A    DCA       CB           Neighbors  0.548168\n",
      "10    1h1vA    DCA       CB           Neighbors  0.630677\n",
      "610   2werA    DCA       CB           Neighbors  0.643752\n",
      "10    1axbA    DCA       CB           Neighbors  0.656607\n",
      "DCA\n",
      "22\n",
      "    Protein Method Distance Sequence_Separation     AUROC\n",
      "10    2ysdA    DCA       CB           Neighbors  0.608076\n",
      "10    1c17A    DCA       CB           Neighbors  0.534796\n",
      "10    7hvpA    DCA       CB           Neighbors  0.574923\n",
      "10    135lA    DCA       CB           Neighbors  0.638480\n",
      "10    206lA    DCA       CB           Neighbors  0.601593\n",
      "10    2werA    DCA       CB           Neighbors  0.643752\n",
      "10    1bolA    DCA       CB           Neighbors  0.598580\n",
      "10    3q05A    DCA       CB           Neighbors  0.563625\n",
      "10    1hckA    DCA       CB           Neighbors  0.622345\n",
      "10    3b6vA    DCA       CB           Neighbors  0.643803\n",
      "10    2z0eA    DCA       CB           Neighbors  0.612559\n",
      "10    1jwlA    DCA       CB           Neighbors  0.684952\n",
      "10    1a26A    DCA       CB           Neighbors  0.643024\n",
      "10    1c0kA    DCA       CB           Neighbors  0.650386\n",
      "10    4lliA    DCA       CB           Neighbors  0.578597\n",
      "10    4ycuA    DCA       CB           Neighbors  0.656378\n",
      "10    2iopA    DCA       CB           Neighbors  0.622600\n",
      "10    2zxeA    DCA       CB           Neighbors  0.624173\n",
      "10    2b59A    DCA       CB           Neighbors  0.548168\n",
      "10    1h1vA    DCA       CB           Neighbors  0.630677\n",
      "610   2werA    DCA       CB           Neighbors  0.643752\n",
      "10    1axbA    DCA       CB           Neighbors  0.656607\n",
      "ET-MIp\n",
      "21\n",
      "     Protein  Method Distance Sequence_Separation     AUROC\n",
      "10     3q05A  ET-MIp       CB           Neighbors  0.552664\n",
      "110    2b59A  ET-MIp       CB           Neighbors  0.557044\n",
      "210    7hvpA  ET-MIp       CB           Neighbors  0.501837\n",
      "310    1c17A  ET-MIp       CB           Neighbors  0.543410\n",
      "410    206lA  ET-MIp       CB           Neighbors  0.570281\n",
      "510    1bolA  ET-MIp       CB           Neighbors  0.572048\n",
      "610    2z0eA  ET-MIp       CB           Neighbors  0.584213\n",
      "710    1axbA  ET-MIp       CB           Neighbors  0.664097\n",
      "810    135lA  ET-MIp       CB           Neighbors  0.594869\n",
      "910    4lliA  ET-MIp       CB           Neighbors  0.530315\n",
      "1010   1a26A  ET-MIp       CB           Neighbors  0.659020\n",
      "1110   1c0kA  ET-MIp       CB           Neighbors  0.703755\n",
      "1210   2zxeA  ET-MIp       CB           Neighbors  0.622871\n",
      "1310   1jwlA  ET-MIp       CB           Neighbors  0.615950\n",
      "1410   1hckA  ET-MIp       CB           Neighbors  0.567997\n",
      "1510   1h1vA  ET-MIp       CB           Neighbors  0.614315\n",
      "1610   2ysdA  ET-MIp       CB           Neighbors  0.549476\n",
      "1710   2iopA  ET-MIp       CB           Neighbors  0.614713\n",
      "1810   3b6vA  ET-MIp       CB           Neighbors  0.580899\n",
      "1910   4ycuA  ET-MIp       CB           Neighbors  0.628906\n",
      "2010   2werA  ET-MIp       CB           Neighbors  0.564370\n",
      "DCA\n",
      "22\n",
      "    Protein Method Distance Sequence_Separation     AUROC\n",
      "10    2ysdA    DCA       CB           Neighbors  0.608076\n",
      "10    1c17A    DCA       CB           Neighbors  0.534796\n",
      "10    7hvpA    DCA       CB           Neighbors  0.574923\n",
      "10    135lA    DCA       CB           Neighbors  0.638480\n",
      "10    206lA    DCA       CB           Neighbors  0.601593\n",
      "10    2werA    DCA       CB           Neighbors  0.643752\n",
      "10    1bolA    DCA       CB           Neighbors  0.598580\n",
      "10    3q05A    DCA       CB           Neighbors  0.563625\n",
      "10    1hckA    DCA       CB           Neighbors  0.622345\n",
      "10    3b6vA    DCA       CB           Neighbors  0.643803\n",
      "10    2z0eA    DCA       CB           Neighbors  0.612559\n",
      "10    1jwlA    DCA       CB           Neighbors  0.684952\n",
      "10    1a26A    DCA       CB           Neighbors  0.643024\n",
      "10    1c0kA    DCA       CB           Neighbors  0.650386\n",
      "10    4lliA    DCA       CB           Neighbors  0.578597\n",
      "10    4ycuA    DCA       CB           Neighbors  0.656378\n",
      "10    2iopA    DCA       CB           Neighbors  0.622600\n",
      "10    2zxeA    DCA       CB           Neighbors  0.624173\n",
      "10    2b59A    DCA       CB           Neighbors  0.548168\n",
      "10    1h1vA    DCA       CB           Neighbors  0.630677\n",
      "610   2werA    DCA       CB           Neighbors  0.643752\n",
      "10    1axbA    DCA       CB           Neighbors  0.656607\n",
      "cET-MIp\n",
      "21\n",
      "   Protein   Method Distance Sequence_Separation     AUROC\n",
      "10   3q05A  cET-MIp       CB           Neighbors  0.565457\n",
      "10   2b59A  cET-MIp       CB           Neighbors  0.530026\n",
      "10   7hvpA  cET-MIp       CB           Neighbors  0.513822\n",
      "10   1c17A  cET-MIp       CB           Neighbors  0.543490\n",
      "10   206lA  cET-MIp       CB           Neighbors  0.600433\n",
      "10   1bolA  cET-MIp       CB           Neighbors  0.613281\n",
      "10   2z0eA  cET-MIp       CB           Neighbors  0.598695\n",
      "10   1axbA  cET-MIp       CB           Neighbors  0.694218\n",
      "10   135lA  cET-MIp       CB           Neighbors  0.608834\n",
      "10   4lliA  cET-MIp       CB           Neighbors  0.563249\n",
      "10   1a26A  cET-MIp       CB           Neighbors  0.682478\n",
      "10   1c0kA  cET-MIp       CB           Neighbors  0.714518\n",
      "10   2zxeA  cET-MIp       CB           Neighbors  0.638994\n",
      "10   1jwlA  cET-MIp       CB           Neighbors  0.691035\n",
      "10   1hckA  cET-MIp       CB           Neighbors  0.594412\n",
      "10   1h1vA  cET-MIp       CB           Neighbors  0.660527\n",
      "10   2ysdA  cET-MIp       CB           Neighbors  0.558461\n",
      "10   2iopA  cET-MIp       CB           Neighbors  0.610398\n",
      "10   3b6vA  cET-MIp       CB           Neighbors  0.619709\n",
      "10   4ycuA  cET-MIp       CB           Neighbors  0.656643\n",
      "10   2werA  cET-MIp       CB           Neighbors  0.572302\n",
      "MI\n",
      "21\n",
      "   Protein Method Distance Sequence_Separation     AUROC\n",
      "30   3q05A     MI       CB              Medium  0.457707\n",
      "30   2b59A     MI       CB              Medium  0.605095\n",
      "30   7hvpA     MI       CB              Medium  0.616456\n",
      "30   1c17A     MI       CB              Medium  0.550520\n",
      "30   206lA     MI       CB              Medium  0.475865\n",
      "30   1bolA     MI       CB              Medium  0.426646\n",
      "30   2z0eA     MI       CB              Medium  0.484164\n",
      "30   1axbA     MI       CB              Medium  0.512233\n",
      "30   135lA     MI       CB              Medium  0.305980\n",
      "30   4lliA     MI       CB              Medium  0.421664\n",
      "30   1a26A     MI       CB              Medium  0.562015\n",
      "30   1c0kA     MI       CB              Medium  0.501330\n",
      "30   2zxeA     MI       CB              Medium  0.481003\n",
      "30   1jwlA     MI       CB              Medium  0.456711\n",
      "30   1hckA     MI       CB              Medium  0.418942\n",
      "30   1h1vA     MI       CB              Medium  0.526087\n",
      "30   2ysdA     MI       CB              Medium  0.311874\n",
      "30   2iopA     MI       CB              Medium  0.564929\n",
      "30   3b6vA     MI       CB              Medium  0.453790\n",
      "30   4ycuA     MI       CB              Medium  0.519295\n",
      "30   2werA     MI       CB              Medium  0.577075\n",
      "DCA\n",
      "22\n",
      "    Protein Method Distance Sequence_Separation     AUROC\n",
      "30    2ysdA    DCA       CB              Medium  0.706033\n",
      "30    1c17A    DCA       CB              Medium  0.494461\n",
      "30    7hvpA    DCA       CB              Medium  0.620893\n",
      "30    135lA    DCA       CB              Medium  0.652836\n",
      "30    206lA    DCA       CB              Medium  0.723356\n",
      "30    2werA    DCA       CB              Medium  0.698971\n",
      "30    1bolA    DCA       CB              Medium  0.621448\n",
      "30    3q05A    DCA       CB              Medium  0.519122\n",
      "30    1hckA    DCA       CB              Medium  0.710699\n",
      "30    3b6vA    DCA       CB              Medium  0.622381\n",
      "30    2z0eA    DCA       CB              Medium  0.730846\n",
      "30    1jwlA    DCA       CB              Medium  0.730390\n",
      "30    1a26A    DCA       CB              Medium  0.712499\n",
      "30    1c0kA    DCA       CB              Medium  0.686232\n",
      "30    4lliA    DCA       CB              Medium  0.555267\n",
      "30    4ycuA    DCA       CB              Medium  0.707658\n",
      "30    2iopA    DCA       CB              Medium  0.710342\n",
      "30    2zxeA    DCA       CB              Medium  0.659997\n",
      "30    2b59A    DCA       CB              Medium  0.650424\n",
      "30    1h1vA    DCA       CB              Medium  0.675090\n",
      "130   1c17A    DCA       CB              Medium  0.494461\n",
      "30    1axbA    DCA       CB              Medium  0.707118\n",
      "DI\n",
      "21\n",
      "   Protein Method Distance Sequence_Separation     AUROC\n",
      "30   3q05A     DI       CB              Medium  0.470723\n",
      "30   2b59A     DI       CB              Medium  0.643178\n",
      "30   7hvpA     DI       CB              Medium  0.628690\n",
      "30   1c17A     DI       CB              Medium  0.479020\n",
      "30   206lA     DI       CB              Medium  0.686846\n",
      "30   1bolA     DI       CB              Medium  0.486684\n",
      "30   2z0eA     DI       CB              Medium  0.603320\n",
      "30   1axbA     DI       CB              Medium  0.620204\n",
      "30   135lA     DI       CB              Medium  0.473531\n",
      "30   4lliA     DI       CB              Medium  0.529569\n",
      "30   1a26A     DI       CB              Medium  0.657247\n",
      "30   1c0kA     DI       CB              Medium  0.625169\n",
      "30   2zxeA     DI       CB              Medium  0.531520\n",
      "30   1jwlA     DI       CB              Medium  0.603336\n",
      "30   1hckA     DI       CB              Medium  0.569306\n",
      "30   1h1vA     DI       CB              Medium  0.628293\n",
      "30   2ysdA     DI       CB              Medium  0.603209\n",
      "30   2iopA     DI       CB              Medium  0.646440\n",
      "30   3b6vA     DI       CB              Medium  0.530149\n",
      "30   4ycuA     DI       CB              Medium  0.620002\n",
      "30   2werA     DI       CB              Medium  0.649290\n",
      "DCA\n",
      "22\n",
      "    Protein Method Distance Sequence_Separation     AUROC\n",
      "30    2ysdA    DCA       CB              Medium  0.706033\n",
      "30    1c17A    DCA       CB              Medium  0.494461\n",
      "30    7hvpA    DCA       CB              Medium  0.620893\n",
      "30    135lA    DCA       CB              Medium  0.652836\n",
      "30    206lA    DCA       CB              Medium  0.723356\n",
      "30    2werA    DCA       CB              Medium  0.698971\n",
      "30    1bolA    DCA       CB              Medium  0.621448\n",
      "30    3q05A    DCA       CB              Medium  0.519122\n",
      "30    1hckA    DCA       CB              Medium  0.710699\n",
      "30    3b6vA    DCA       CB              Medium  0.622381\n",
      "30    2z0eA    DCA       CB              Medium  0.730846\n",
      "30    1jwlA    DCA       CB              Medium  0.730390\n",
      "30    1a26A    DCA       CB              Medium  0.712499\n",
      "30    1c0kA    DCA       CB              Medium  0.686232\n",
      "30    4lliA    DCA       CB              Medium  0.555267\n",
      "30    4ycuA    DCA       CB              Medium  0.707658\n",
      "30    2iopA    DCA       CB              Medium  0.710342\n",
      "30    2zxeA    DCA       CB              Medium  0.659997\n",
      "30    2b59A    DCA       CB              Medium  0.650424\n",
      "30    1h1vA    DCA       CB              Medium  0.675090\n",
      "130   1c17A    DCA       CB              Medium  0.494461\n",
      "30    1axbA    DCA       CB              Medium  0.707118\n",
      "DCA\n",
      "22\n",
      "    Protein Method Distance Sequence_Separation     AUROC\n",
      "30    2ysdA    DCA       CB              Medium  0.706033\n",
      "30    1c17A    DCA       CB              Medium  0.494461\n",
      "30    7hvpA    DCA       CB              Medium  0.620893\n",
      "30    135lA    DCA       CB              Medium  0.652836\n",
      "30    206lA    DCA       CB              Medium  0.723356\n",
      "30    2werA    DCA       CB              Medium  0.698971\n",
      "30    1bolA    DCA       CB              Medium  0.621448\n",
      "30    3q05A    DCA       CB              Medium  0.519122\n",
      "30    1hckA    DCA       CB              Medium  0.710699\n",
      "30    3b6vA    DCA       CB              Medium  0.622381\n",
      "30    2z0eA    DCA       CB              Medium  0.730846\n",
      "30    1jwlA    DCA       CB              Medium  0.730390\n",
      "30    1a26A    DCA       CB              Medium  0.712499\n",
      "30    1c0kA    DCA       CB              Medium  0.686232\n",
      "30    4lliA    DCA       CB              Medium  0.555267\n",
      "30    4ycuA    DCA       CB              Medium  0.707658\n",
      "30    2iopA    DCA       CB              Medium  0.710342\n",
      "30    2zxeA    DCA       CB              Medium  0.659997\n",
      "30    2b59A    DCA       CB              Medium  0.650424\n",
      "30    1h1vA    DCA       CB              Medium  0.675090\n",
      "130   1c17A    DCA       CB              Medium  0.494461\n",
      "30    1axbA    DCA       CB              Medium  0.707118\n",
      "ET-MIp\n",
      "21\n",
      "     Protein  Method Distance Sequence_Separation     AUROC\n",
      "30     3q05A  ET-MIp       CB              Medium  0.507015\n",
      "130    2b59A  ET-MIp       CB              Medium  0.603893\n",
      "230    7hvpA  ET-MIp       CB              Medium  0.589736\n",
      "330    1c17A  ET-MIp       CB              Medium  0.473565\n",
      "430    206lA  ET-MIp       CB              Medium  0.661969\n",
      "530    1bolA  ET-MIp       CB              Medium  0.603013\n",
      "630    2z0eA  ET-MIp       CB              Medium  0.599357\n",
      "730    1axbA  ET-MIp       CB              Medium  0.660507\n",
      "830    135lA  ET-MIp       CB              Medium  0.573286\n",
      "930    4lliA  ET-MIp       CB              Medium  0.510525\n",
      "1030   1a26A  ET-MIp       CB              Medium  0.675492\n",
      "1130   1c0kA  ET-MIp       CB              Medium  0.749174\n",
      "1230   2zxeA  ET-MIp       CB              Medium  0.610945\n",
      "1330   1jwlA  ET-MIp       CB              Medium  0.638604\n",
      "1430   1hckA  ET-MIp       CB              Medium  0.674007\n",
      "1530   1h1vA  ET-MIp       CB              Medium  0.654023\n",
      "1630   2ysdA  ET-MIp       CB              Medium  0.611040\n",
      "1730   2iopA  ET-MIp       CB              Medium  0.686441\n",
      "1830   3b6vA  ET-MIp       CB              Medium  0.586319\n",
      "1930   4ycuA  ET-MIp       CB              Medium  0.674607\n",
      "2030   2werA  ET-MIp       CB              Medium  0.643863\n",
      "DCA\n",
      "22\n",
      "    Protein Method Distance Sequence_Separation     AUROC\n",
      "30    2ysdA    DCA       CB              Medium  0.706033\n",
      "30    1c17A    DCA       CB              Medium  0.494461\n",
      "30    7hvpA    DCA       CB              Medium  0.620893\n",
      "30    135lA    DCA       CB              Medium  0.652836\n",
      "30    206lA    DCA       CB              Medium  0.723356\n",
      "30    2werA    DCA       CB              Medium  0.698971\n",
      "30    1bolA    DCA       CB              Medium  0.621448\n",
      "30    3q05A    DCA       CB              Medium  0.519122\n",
      "30    1hckA    DCA       CB              Medium  0.710699\n",
      "30    3b6vA    DCA       CB              Medium  0.622381\n",
      "30    2z0eA    DCA       CB              Medium  0.730846\n",
      "30    1jwlA    DCA       CB              Medium  0.730390\n",
      "30    1a26A    DCA       CB              Medium  0.712499\n",
      "30    1c0kA    DCA       CB              Medium  0.686232\n",
      "30    4lliA    DCA       CB              Medium  0.555267\n",
      "30    4ycuA    DCA       CB              Medium  0.707658\n",
      "30    2iopA    DCA       CB              Medium  0.710342\n",
      "30    2zxeA    DCA       CB              Medium  0.659997\n",
      "30    2b59A    DCA       CB              Medium  0.650424\n",
      "30    1h1vA    DCA       CB              Medium  0.675090\n",
      "130   1c17A    DCA       CB              Medium  0.494461\n",
      "30    1axbA    DCA       CB              Medium  0.707118\n",
      "cET-MIp\n",
      "21\n",
      "   Protein   Method Distance Sequence_Separation     AUROC\n",
      "30   3q05A  cET-MIp       CB              Medium  0.481320\n",
      "30   2b59A  cET-MIp       CB              Medium  0.598824\n",
      "30   7hvpA  cET-MIp       CB              Medium  0.627899\n",
      "30   1c17A  cET-MIp       CB              Medium  0.533233\n",
      "30   206lA  cET-MIp       CB              Medium  0.678387\n",
      "30   1bolA  cET-MIp       CB              Medium  0.645827\n",
      "30   2z0eA  cET-MIp       CB              Medium  0.693399\n",
      "30   1axbA  cET-MIp       CB              Medium  0.757404\n",
      "30   135lA  cET-MIp       CB              Medium  0.587816\n",
      "30   4lliA  cET-MIp       CB              Medium  0.595032\n",
      "30   1a26A  cET-MIp       CB              Medium  0.777076\n",
      "30   1c0kA  cET-MIp       CB              Medium  0.788115\n",
      "30   2zxeA  cET-MIp       CB              Medium  0.653950\n",
      "30   1jwlA  cET-MIp       CB              Medium  0.757387\n",
      "30   1hckA  cET-MIp       CB              Medium  0.678423\n",
      "30   1h1vA  cET-MIp       CB              Medium  0.750957\n",
      "30   2ysdA  cET-MIp       CB              Medium  0.663350\n",
      "30   2iopA  cET-MIp       CB              Medium  0.716670\n",
      "30   3b6vA  cET-MIp       CB              Medium  0.627042\n",
      "30   4ycuA  cET-MIp       CB              Medium  0.734387\n",
      "30   2werA  cET-MIp       CB              Medium  0.472983\n"
     ]
    }
   ],
   "source": [
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from scipy.stats import wilcoxon\n",
    "protein_method_groups = small_comparison_df[['Protein', 'Method']].drop_duplicates().groupby('Protein').count()\n",
    "method_max = protein_method_groups['Method'].max()\n",
    "proteins_to_keep = protein_method_groups.index[protein_method_groups['Method'] == method_max]\n",
    "comparable_method_proteins = small_comparison_df[small_comparison_df['Protein'].isin(proteins_to_keep)]\n",
    "protein_length_order = comparable_method_proteins.sort_values('Protein Length')['Protein'].unique()\n",
    "protein_alignment_order = comparable_method_proteins.sort_values('Alignment Size')['Protein'].unique()\n",
    "auroc_columns = ['Protein', 'Method', 'Distance', 'Sequence_Separation', 'AUROC']\n",
    "protein_auroc_groups = comparable_method_proteins[auroc_columns].drop_duplicates(\n",
    "    subset=['Protein', 'Method', 'Sequence_Separation', 'Distance']).groupby('Protein')['AUROC'].apply(\n",
    "    lambda x: not x.isnull().any())\n",
    "complete_proteins = protein_auroc_groups.index[protein_auroc_groups.values]\n",
    "comparable_auroc_proteins = small_comparison_df[small_comparison_df['Protein'].isin(complete_proteins)]\n",
    "auroc_protein_length_order = [x for x in protein_length_order if x in complete_proteins]\n",
    "auroc_protein_alignment_order = [x for x in protein_alignment_order if x in complete_proteins]\n",
    "auroc_subset_df = comparable_auroc_proteins.loc[:, auroc_columns].drop_duplicates()\n",
    "auroc_subset_df = auroc_subset_df.loc[auroc_subset_df['Distance'] == 'CB', :]\n",
    "# Plot the methods vs AUROC per protein ordered by protein length\n",
    "auroc_subset_df.to_csv(os.path.join(small_set_out_dir, 'Small_AUROC_Comaprison_Data.csv'), sep='\\t', header=True, index=False,\n",
    "                       columns=auroc_columns)\n",
    "protein_order_auroc_plot = sns.catplot(x=\"Protein\", y=\"AUROC\", hue=\"Method\", row=\"Sequence_Separation\", data=auroc_subset_df, kind=\"bar\",\n",
    "                                       ci=None, order=auroc_protein_length_order, hue_order=method_order, legend=True, legend_out=True)\n",
    "protein_order_auroc_plot.set_xticklabels(auroc_protein_length_order, rotation=90)\n",
    "protein_order_auroc_plot.savefig(os.path.join(small_set_out_dir, 'Protein_Method_AUROC_Comparison_Protein_Length_Order.png'),\n",
    "                                 bbox_inches='tight', transparent=True, dpi=300)\n",
    "plt.close()\n",
    "# Plot the methods vs AUROC per protein ordered by protein alignment size\n",
    "alignment_order_auroc_plot = sns.catplot(x=\"Protein\", y=\"AUROC\", hue=\"Method\", row=\"Sequence_Separation\", data=auroc_subset_df, kind=\"bar\",\n",
    "                                       ci=None, order=auroc_protein_alignment_order, hue_order=method_order, legend=True, legend_out=True)\n",
    "alignment_order_auroc_plot.set_xticklabels(auroc_protein_alignment_order, rotation=90)\n",
    "alignment_order_auroc_plot.savefig(os.path.join(small_set_out_dir, 'Protein_Method_AUROC_Comparison_Alignment_Size_Order.png'),\n",
    "                                   bbox_inches='tight', transparent=True, dpi=300)\n",
    "plt.close()\n",
    "# Plot the methods vs AUROC grouped together to see overall trends\n",
    "overall_auroc_plot = sns.boxplot(x=\"Sequence_Separation\", y=\"AUROC\", hue=\"Method\", data=auroc_subset_df,\n",
    "                                 order=sequence_separation_order, hue_order=method_order)\n",
    "overall_auroc_plot.set_xticklabels(sequence_separation_order, rotation=90)\n",
    "overall_auroc_plot.legend(loc='center left', bbox_to_anchor=(1.25, 0.5), ncol=1)\n",
    "plt.savefig(os.path.join(small_set_out_dir, 'Protein_Method_AUROC_Comparison.png'), bbox_inches='tight', transparent=True, dpi=300)\n",
    "plt.close()\n",
    "# Compute statistics comparing methods at each sequence separation\n",
    "auroc_statistics = {'Sequence Separation': [], 'Method 1': [], 'Method 2': [], 'Statistic': [], 'P-Value': []}\n",
    "for sep in sequence_separation_order:\n",
    "    sep_auroc_subset_df = auroc_subset_df.loc[auroc_subset_df['Sequence_Separation'] == sep, :]\n",
    "    for i in range(len(method_order)):\n",
    "        m1_sep_auroc_subset_df = sep_auroc_subset_df.loc[sep_auroc_subset_df['Method'] == method_order[i], :]\n",
    "        for j in range(i + 1, len(method_order)):\n",
    "            m2_sep_auroc_subset_df = sep_auroc_subset_df.loc[sep_auroc_subset_df['Method'] == method_order[j], :]\n",
    "            try:\n",
    "                stat, p_val = wilcoxon(x=m1_sep_auroc_subset_df['AUROC'], y=m2_sep_auroc_subset_df['AUROC'], zero_method='wilcox')\n",
    "            except ValueError:\n",
    "                print(method_order[i])\n",
    "                print(len(m1_sep_auroc_subset_df['AUROC']))\n",
    "                print(m1_sep_auroc_subset_df)\n",
    "                print(method_order[j])\n",
    "                print(len(m2_sep_auroc_subset_df['AUROC']))\n",
    "                print(m2_sep_auroc_subset_df)\n",
    "            auroc_statistics['Sequence Separation'].append(sep)\n",
    "            auroc_statistics['Method 1'].append(method_order[i])\n",
    "            auroc_statistics['Method 2'].append(method_order[j])\n",
    "            auroc_statistics['Statistic'].append(stat)\n",
    "            auroc_statistics['P-Value'].append(p_val)\n",
    "pd.DataFrame(auroc_statistics).to_csv(os.path.join(small_set_out_dir, 'Small_AUROC_Comaprison_Statistics.csv'), sep='\\t', header=True,\n",
    "                                      index=False, columns=['Sequence Separation', 'Method 1', 'Method 2', 'Statistic', 'P-Value'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/ipykernel_launcher.py:42: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/home/daniel/anaconda3/envs/PyET3/lib/python3.6/site-packages/ipykernel_launcher.py:45: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    }
   ],
   "source": [
    "auroc_columns = ['Protein', 'Method', 'Distance', 'Sequence_Separation', 'AUROC']\n",
    "protein_auroc_groups = comparable_method_proteins[auroc_columns].drop_duplicates().groupby('Protein')['AUROC'].apply(\n",
    "    lambda x: not x.isnull().any())\n",
    "complete_proteins = protein_auroc_groups.index[protein_auroc_groups.values]\n",
    "comparable_auroc_proteins = small_comparison_df[small_comparison_df['Protein'].isin(complete_proteins)]\n",
    "auroc_protein_length_order = [x for x in protein_length_order if x in complete_proteins]\n",
    "auroc_protein_alignment_order = [x for x in protein_alignment_order if x in complete_proteins]\n",
    "auroc_subset_df = comparable_auroc_proteins.loc[:, auroc_columns].drop_duplicates(subset=['Protein', 'Method', 'Sequence_Separation', 'Distance'])\n",
    "auroc_subset_df = auroc_subset_df.loc[auroc_subset_df['Distance'] == 'Any', :]\n",
    "# Plot the methods vs AUROC per protein ordered by protein length\n",
    "auroc_subset_df.to_csv(os.path.join(small_set_out_dir, 'Small_AUROC_Comaprison_Data_Dist_Any.csv'), sep='\\t', header=True, index=False,\n",
    "                       columns=auroc_columns)\n",
    "protein_order_auroc_plot = sns.catplot(x=\"Protein\", y=\"AUROC\", hue=\"Method\", row=\"Sequence_Separation\", data=auroc_subset_df, kind=\"bar\",\n",
    "                                       ci=None, order=auroc_protein_length_order, hue_order=method_order, legend=True, legend_out=True)\n",
    "protein_order_auroc_plot.set_xticklabels(auroc_protein_length_order, rotation=90)\n",
    "protein_order_auroc_plot.savefig(os.path.join(small_set_out_dir, 'Protein_Method_AUROC_Comparison_Protein_Length_Order_Dist_Any.png'),\n",
    "                                 bbox_inches='tight', transparent=True, dpi=300)\n",
    "plt.close()\n",
    "# Plot the methods vs AUROC per protein ordered by protein alignment size\n",
    "alignment_order_auroc_plot = sns.catplot(x=\"Protein\", y=\"AUROC\", hue=\"Method\", row=\"Sequence_Separation\", data=auroc_subset_df, kind=\"bar\",\n",
    "                                       ci=None, order=auroc_protein_alignment_order, hue_order=method_order, legend=True, legend_out=True)\n",
    "alignment_order_auroc_plot.set_xticklabels(auroc_protein_alignment_order, rotation=90)\n",
    "alignment_order_auroc_plot.savefig(os.path.join(small_set_out_dir, 'Protein_Method_AUROC_Comparison_Alignment_Size_Order_Dist_Any.png'),\n",
    "                                   bbox_inches='tight', transparent=True, dpi=300)\n",
    "plt.close()\n",
    "# Plot the methods vs AUROC grouped together to see overall trends\n",
    "overall_auroc_plot = sns.boxplot(x=\"Sequence_Separation\", y=\"AUROC\", hue=\"Method\", data=auroc_subset_df,\n",
    "                                 order=sequence_separation_order, hue_order=method_order)\n",
    "overall_auroc_plot.set_xticklabels(sequence_separation_order, rotation=90)\n",
    "overall_auroc_plot.legend(loc='center left', bbox_to_anchor=(1.25, 0.5), ncol=1)\n",
    "plt.savefig(os.path.join(small_set_out_dir, 'Protein_Method_AUROC_Comparison_Dist_Any.png'), bbox_inches='tight', transparent=True,\n",
    "            dpi=300)\n",
    "plt.close()\n",
    "# Compute statistics comparing methods at each sequence separation\n",
    "auroc_statistics = {'Sequence Separation': [], 'Method 1': [], 'Method 2': [], 'Statistic': [], 'P-Value': []}\n",
    "for sep in sequence_separation_order:\n",
    "    sep_auroc_subset_df = auroc_subset_df.loc[auroc_subset_df['Sequence_Separation'] == sep, :]\n",
    "    for i in range(len(method_order)):\n",
    "        m1_sep_auroc_subset_df = sep_auroc_subset_df.loc[sep_auroc_subset_df['Method'] == method_order[i], :]\n",
    "        m1_sep_auroc_subset_df.drop_duplicates(inplace=True)\n",
    "        for j in range(i + 1, len(method_order)):\n",
    "            m2_sep_auroc_subset_df = sep_auroc_subset_df.loc[sep_auroc_subset_df['Method'] == method_order[j], :]\n",
    "            m2_sep_auroc_subset_df.drop_duplicates(inplace=True)\n",
    "            try:\n",
    "                stat, p_val = wilcoxon(x=m1_sep_auroc_subset_df['AUROC'], y=m2_sep_auroc_subset_df['AUROC'], zero_method='wilcox')\n",
    "            except ValueError:\n",
    "                print(method_order[i])\n",
    "                print(len(m1_sep_auroc_subset_df['AUROC']))\n",
    "                print(m1_sep_auroc_subset_df)\n",
    "                print(method_order[j])\n",
    "                print(len(m2_sep_auroc_subset_df['AUROC']))\n",
    "                print(m2_sep_auroc_subset_df)\n",
    "            auroc_statistics['Sequence Separation'].append(sep)\n",
    "            auroc_statistics['Method 1'].append(method_order[i])\n",
    "            auroc_statistics['Method 2'].append(method_order[j])\n",
    "            auroc_statistics['Statistic'].append(stat)\n",
    "            auroc_statistics['P-Value'].append(p_val)\n",
    "pd.DataFrame(auroc_statistics).to_csv(os.path.join(small_set_out_dir, 'Small_AUROC_Comaprison_Statistics_Dist_any.csv'), sep='\\t',\n",
    "                                      header=True, index=False,\n",
    "                                      columns=['Sequence Separation', 'Method 1', 'Method 2', 'Statistic', 'P-Value'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Metrics of Success (Continued)\n",
    "* AUPRC - This measures the Precision vs the Recall of the predictions, it can be considered a measure of the accuracy of the measure. This is less strongly influenced by the class imbalance which is present when predicting structural contacts since there are many fewer contacts than non-contacts. The True Positive case is if the C-beta of two amino acids is within 8.0 Angstroms of one another (as is done in the CASP competitions)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "auprc_columns = ['Protein', 'Method', 'Distance', 'Sequence_Separation', 'AUPRC']\n",
    "protein_auprc_groups = comparable_method_proteins[auprc_columns].drop_duplicates().groupby('Protein')['AUPRC'].apply(\n",
    "    lambda x: not x.isnull().any())\n",
    "complete_proteins = protein_auprc_groups.index[protein_auprc_groups.values]\n",
    "comparable_auprc_proteins = small_comparison_df[small_comparison_df['Protein'].isin(complete_proteins)]\n",
    "auprc_protein_length_order = [x for x in protein_length_order if x in complete_proteins]\n",
    "auprc_protein_alignment_order = [x for x in protein_alignment_order if x in complete_proteins]\n",
    "auprc_subset_df = comparable_auprc_proteins.loc[:, auprc_columns].drop_duplicates(subset=['Protein', 'Method', 'Sequence_Separation', 'Distance'])\n",
    "auprc_subset_df = auprc_subset_df.loc[auprc_subset_df['Distance'] == 'CB', :]\n",
    "# Plot the methods vs AUPRC per protein ordered by protein length\n",
    "auprc_subset_df.to_csv(os.path.join(small_set_out_dir, 'Small_AUPRC_Comaprison_Data.csv'), sep='\\t', header=True, index=False,\n",
    "                       columns=auprc_columns)\n",
    "protein_order_auprc_plot = sns.catplot(x=\"Protein\", y=\"AUPRC\", hue=\"Method\", row=\"Sequence_Separation\", data=auprc_subset_df, kind=\"bar\",\n",
    "                                       ci=None, order=auprc_protein_length_order, hue_order=method_order, legend=True, legend_out=True)\n",
    "protein_order_auprc_plot.set_xticklabels(auprc_protein_length_order, rotation=90)\n",
    "protein_order_auprc_plot.savefig(os.path.join(small_set_out_dir, 'Protein_Method_AUPRC_Comparison_Protein_Length_Order.png'),\n",
    "                                 bbox_inches='tight', transparent=True, dpi=300)\n",
    "plt.close()\n",
    "# Plot the methods vs AUPRC per protein ordered by protein alignment size\n",
    "alignment_order_auprc_plot = sns.catplot(x=\"Protein\", y=\"AUPRC\", hue=\"Method\", row=\"Sequence_Separation\", data=auprc_subset_df, kind=\"bar\",\n",
    "                                       ci=None, order=auprc_protein_alignment_order, hue_order=method_order, legend=True, legend_out=True)\n",
    "alignment_order_auprc_plot.set_xticklabels(auprc_protein_alignment_order, rotation=90)\n",
    "alignment_order_auprc_plot.savefig(os.path.join(small_set_out_dir, 'Protein_Method_AUPRC_Comparison_Alignment_Size_Order.png'),\n",
    "                                   bbox_inches='tight', transparent=True, dpi=300)\n",
    "plt.close()\n",
    "# Plot the methods vs AUPRC grouped together to see overall trends\n",
    "overall_auprc_plot = sns.boxplot(x=\"Sequence_Separation\", y=\"AUPRC\", hue=\"Method\", data=auprc_subset_df,\n",
    "                                 order=sequence_separation_order, hue_order=method_order)\n",
    "overall_auprc_plot.set_xticklabels(sequence_separation_order, rotation=90)\n",
    "overall_auprc_plot.legend(loc='center left', bbox_to_anchor=(1.25, 0.5), ncol=1)\n",
    "plt.savefig(os.path.join(small_set_out_dir, 'Protein_Method_AUPRC_Comparison.png'), bbox_inches='tight', transparent=True, dpi=300)\n",
    "plt.close()\n",
    "# Compute statistics comparing methods at each sequence separation\n",
    "auprc_statistics = {'Sequence Separation': [], 'Method 1': [], 'Method 2': [], 'Statistic': [], 'P-Value': []}\n",
    "for sep in sequence_separation_order:\n",
    "    sep_auprc_subset_df = auprc_subset_df.loc[auprc_subset_df['Sequence_Separation'] == sep, :]\n",
    "    for i in range(len(method_order)):\n",
    "        m1_sep_auprc_subset_df = sep_auprc_subset_df.loc[sep_auprc_subset_df['Method'] == method_order[i], :]\n",
    "        for j in range(i + 1, len(method_order)):\n",
    "            m2_sep_auprc_subset_df = sep_auprc_subset_df.loc[sep_auprc_subset_df['Method'] == method_order[j], :]\n",
    "            try:\n",
    "                stat, p_val = wilcoxon(x=m1_sep_auprc_subset_df['AUPRC'], y=m2_sep_auprc_subset_df['AUPRC'], zero_method='wilcox')\n",
    "            except ValueError:\n",
    "                print(method_order[i])\n",
    "                print(len(m1_sep_auroc_subset_df['AUROC']))\n",
    "                print(m1_sep_auroc_subset_df)\n",
    "                print(method_order[j])\n",
    "                print(len(m2_sep_auroc_subset_df['AUROC']))\n",
    "                print(m2_sep_auroc_subset_df)\n",
    "            auprc_statistics['Sequence Separation'].append(sep)\n",
    "            auprc_statistics['Method 1'].append(method_order[i])\n",
    "            auprc_statistics['Method 2'].append(method_order[j])\n",
    "            auprc_statistics['Statistic'].append(stat)\n",
    "            auprc_statistics['P-Value'].append(p_val)\n",
    "pd.DataFrame(auprc_statistics).to_csv(os.path.join(small_set_out_dir, 'Small_AUPRC_Comaprison_Statistics.csv'), sep='\\t', header=True,\n",
    "                                      index=False, columns=['Sequence Separation', 'Method 1', 'Method 2', 'Statistic', 'P-Value'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "auprc_columns = ['Protein', 'Method', 'Distance', 'Sequence_Separation', 'AUPRC']\n",
    "protein_auprc_groups = comparable_method_proteins[auprc_columns].drop_duplicates().groupby('Protein')['AUPRC'].apply(\n",
    "    lambda x: not x.isnull().any())\n",
    "complete_proteins = protein_auprc_groups.index[protein_auprc_groups.values]\n",
    "comparable_auprc_proteins = small_comparison_df[small_comparison_df['Protein'].isin(complete_proteins)]\n",
    "auprc_protein_length_order = [x for x in protein_length_order if x in complete_proteins]\n",
    "auprc_protein_alignment_order = [x for x in protein_alignment_order if x in complete_proteins]\n",
    "auprc_subset_df = comparable_auprc_proteins.loc[:, auprc_columns].drop_duplicates(subset=['Protein', 'Method', 'Sequence_Separation', 'Distance'])\n",
    "auprc_subset_df = auprc_subset_df.loc[auprc_subset_df['Distance'] == 'Any', :]\n",
    "# Plot the methods vs AUPRC per protein ordered by protein length\n",
    "auprc_subset_df.to_csv(os.path.join(small_set_out_dir, 'Small_AUPRC_Comaprison_Data_Dist_Any.csv'), sep='\\t', header=True, index=False,\n",
    "                       columns=auprc_columns)\n",
    "protein_order_auprc_plot = sns.catplot(x=\"Protein\", y=\"AUPRC\", hue=\"Method\", row=\"Sequence_Separation\", data=auprc_subset_df, kind=\"bar\",\n",
    "                                       ci=None, order=auprc_protein_length_order, hue_order=method_order, legend=True, legend_out=True)\n",
    "protein_order_auprc_plot.set_xticklabels(auprc_protein_length_order, rotation=90)\n",
    "protein_order_auprc_plot.savefig(os.path.join(small_set_out_dir, 'Protein_Method_AUPRC_Comparison_Protein_Length_Order_Dist_Any.png'),\n",
    "                                 bbox_inches='tight', transparent=True, dpi=300)\n",
    "plt.close()\n",
    "# Plot the methods vs AUPRC per protein ordered by protein alignment size\n",
    "alignment_order_auprc_plot = sns.catplot(x=\"Protein\", y=\"AUPRC\", hue=\"Method\", row=\"Sequence_Separation\", data=auprc_subset_df, kind=\"bar\",\n",
    "                                       ci=None, order=auprc_protein_alignment_order, hue_order=method_order, legend=True, legend_out=True)\n",
    "alignment_order_auprc_plot.set_xticklabels(auprc_protein_alignment_order, rotation=90)\n",
    "alignment_order_auprc_plot.savefig(os.path.join(small_set_out_dir, 'Protein_Method_AUPRC_Comparison_Alignment_Size_Order_Dist_Any.png'),\n",
    "                                   bbox_inches='tight', transparent=True, dpi=300)\n",
    "plt.close()\n",
    "# Plot the methods vs AUPRC grouped together to see overall trends\n",
    "overall_auprc_plot = sns.boxplot(x=\"Sequence_Separation\", y=\"AUPRC\", hue=\"Method\", data=auprc_subset_df,\n",
    "                                 order=sequence_separation_order, hue_order=method_order)\n",
    "overall_auprc_plot.set_xticklabels(sequence_separation_order, rotation=90)\n",
    "overall_auprc_plot.legend(loc='center left', bbox_to_anchor=(1.25, 0.5), ncol=1)\n",
    "plt.savefig(os.path.join(small_set_out_dir, 'Protein_Method_AUPRC_Comparison_Dist_Any.png'), bbox_inches='tight', transparent=True, dpi=300)\n",
    "plt.close()\n",
    "# Compute statistics comparing methods at each sequence separation\n",
    "auprc_statistics = {'Sequence Separation': [], 'Method 1': [], 'Method 2': [], 'Statistic': [], 'P-Value': []}\n",
    "for sep in sequence_separation_order:\n",
    "    sep_auprc_subset_df = auprc_subset_df.loc[auprc_subset_df['Sequence_Separation'] == sep, :]\n",
    "    for i in range(len(method_order)):\n",
    "        m1_sep_auprc_subset_df = sep_auprc_subset_df.loc[sep_auprc_subset_df['Method'] == method_order[i], :]\n",
    "        for j in range(i + 1, len(method_order)):\n",
    "            m2_sep_auprc_subset_df = sep_auprc_subset_df.loc[sep_auprc_subset_df['Method'] == method_order[j], :]\n",
    "            try:\n",
    "                stat, p_val = wilcoxon(x=m1_sep_auprc_subset_df['AUPRC'], y=m2_sep_auprc_subset_df['AUPRC'], zero_method='wilcox')\n",
    "            except ValueError:\n",
    "                print(method_order[i])\n",
    "                print(len(m1_sep_auroc_subset_df['AUROC']))\n",
    "                print(m1_sep_auroc_subset_df)\n",
    "                print(method_order[j])\n",
    "                print(len(m2_sep_auroc_subset_df['AUROC']))\n",
    "                print(m2_sep_auroc_subset_df)\n",
    "            auprc_statistics['Sequence Separation'].append(sep)\n",
    "            auprc_statistics['Method 1'].append(method_order[i])\n",
    "            auprc_statistics['Method 2'].append(method_order[j])\n",
    "            auprc_statistics['Statistic'].append(stat)\n",
    "            auprc_statistics['P-Value'].append(p_val)\n",
    "pd.DataFrame(auprc_statistics).to_csv(os.path.join(small_set_out_dir, 'Small_AUPRC_Comaprison_Statistics_Dist_Any.csv'), sep='\\t',\n",
    "                                      header=True, index=False,\n",
    "                                      columns=['Sequence Separation', 'Method 1', 'Method 2', 'Statistic', 'P-Value'])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python3 (PyET3)",
   "language": "python",
   "name": "pyet3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
