{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Small Test Set #\n",
    "## Goal ##\n",
    "The goal of this test set is to perform proof of concept testing on a small number of proteins with a wide range of sizes and available homologs, orthologs, and paralogs. By doing so it should be possible to test the best parameterization for this tool as well as identifying the strengths and weaknesses of the tool using various measurments as end points.\n",
    "## Warning ##\n",
    "Before attempting to use this notebook make sure that your .env file has been properly setup to reflect the correct locations of command line tools and the location of files and directories needed for execution.\n",
    "### Initial Import###\n",
    "This first cell performs the necessary imports required to begin this notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from dotenv import find_dotenv, load_dotenv\n",
    "try:\n",
    "    dotenv_path = find_dotenv(raise_error_if_not_found=True)\n",
    "except IOError:\n",
    "    dotenv_path = find_dotenv(raise_error_if_not_found=True, usecwd=True)\n",
    "load_dotenv(dotenv_path)\n",
    "import os\n",
    "import sys\n",
    "sys.path.append(os.path.join(os.environ.get('PROJECT_PATH'), 'src'))\n",
    "sys.path.append(os.path.join(os.environ.get('PROJECT_PATH'), 'src', 'SupportingClasses'))\n",
    "input_dir = os.environ.get('INPUT_PATH')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Set Construction ##\n",
    "The first task required to test the data set is to download the required data and construct any necessary input files for all down stream analyses.\n",
    "In this case that means:\n",
    "* Downloading PDB files for the proteins in our small test set.\n",
    "* Extracting a query sequence from each PDB file.\n",
    "* Searching for paralogs, homologs, and orthologs in a custom BLAST database built by filtering the Uniref90 database.\n",
    "* Filtering the hits from the BLAST search to meet minimum and maximum length requirements, as well as minimum and maximum identity requirements.\n",
    "* Building alignments using CLUSTALW in both the fasta and msf formats since some of the tools which will be used for comparison need different formats.\n",
    "* Filtering the alignment for maximum identity similarity between seqeunces.\n",
    "* Re-aligning the filtered sequences using CLUSTALW.\n",
    "This is all handeled by the DataSetGenerator class found in the src/SupportingClasses folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Importing protein list\n",
      "Downloading structures and parsing in query sequences\n",
      "Unique Sequences Found: 23!\n",
      "BLASTing query sequences\n",
      "Filtering BLAST hits, aligning, filtering by identity, and re-aligning\n",
      "   Protein_ID     Accession  BLAST_Hits  Filtered_BLAST  Filtered_Alignment  \\\n",
      "18       2b59    CIPA_CLOTM        1606               4                   4   \n",
      "7        206l      LYS_BPT4        1039             131                  92   \n",
      "16       1c0k    OXDA_RHOTO        2500              51                  51   \n",
      "13       1bol    RNRH_RHINI        2500             131                 127   \n",
      "17       7hvp     POL_HV1A2        2500              33                  33   \n",
      "12       1c17    ATPL_ECOLI        1671             850                 813   \n",
      "2        1jwl    LACI_ECOLI        2500             228                 226   \n",
      "1        3q05     P53_HUMAN         932             306                 210   \n",
      "8        135l    LYSC_MELGA        1913             853                 818   \n",
      "4        2z0e   ATG4B_HUMAN        2111             360                 344   \n",
      "10       2rh1   ADRB2_HUMAN        2500             416                 399   \n",
      "9        1a26   PARP1_CHICK        2500             267                 261   \n",
      "22       3b6v   KIF3C_HUMAN        2500             481                 441   \n",
      "3        1h1v    GELS_HUMAN        2500             654                 599   \n",
      "20       2ysd   MAGI1_HUMAN        2500             412                 318   \n",
      "6        1hck    CDK2_HUMAN        2500            2468                1681   \n",
      "14       1axb    BLAT_ECOLI        2500            2102                2087   \n",
      "0        4lli   MYO5A_HUMAN        2500             441                 352   \n",
      "19       3tnu   K1C14_HUMAN        2500            1681                1518   \n",
      "5        2zxe  Q4H132_SQUAC        2500             880                 795   \n",
      "15       2wer   HSP82_YEAST        2500            1327                1207   \n",
      "21       4ycu     SYK_HUMAN        2500            2421                2347   \n",
      "11       2iop    HTPG_ECOLI        2500            2478                2472   \n",
      "\n",
      "    Length  Total_Size  \n",
      "18    1853      7412.0  \n",
      "7      164     15088.0  \n",
      "16     368     18768.0  \n",
      "13     238     30226.0  \n",
      "17    1437     47421.0  \n",
      "12      79     64227.0  \n",
      "2      360     81360.0  \n",
      "1      393     82530.0  \n",
      "8      147    120246.0  \n",
      "4      393    135192.0  \n",
      "10     413    164787.0  \n",
      "9     1011    263871.0  \n",
      "22     793    349713.0  \n",
      "3      782    468418.0  \n",
      "20    1491    474138.0  \n",
      "6      298    500938.0  \n",
      "14     286    596882.0  \n",
      "0     1855    652960.0  \n",
      "19     472    716496.0  \n",
      "5     1028    817260.0  \n",
      "15     709    855763.0  \n",
      "21     597   1401159.0  \n",
      "11     624   1542528.0  \n",
      "It took 0.14518646796544393 min to generate the data set.\n"
     ]
    }
   ],
   "source": [
    "from time import time\n",
    "from DataSetGenerator import DataSetGenerator\n",
    "protein_list_dir = os.path.join(input_dir, 'ProteinLists')\n",
    "if not os.path.isdir(protein_list_dir):\n",
    "    os.makedirs(protein_list_dir)\n",
    "small_list_fn = os.path.join(protein_list_dir, 'SmallDataSet.txt')\n",
    "if not os.path.isfile(small_list_fn):\n",
    "    proteins_of_interest = ['2ysdA', '1c17A', '3tnuA', '7hvpA', '135lA', '206lA', '2werA', '1bolA', '3q05A', '1axbA',\n",
    "                            '2rh1A', '1hckA', '3b6vA', '2z0eA', '1jwlA', '1a26A', '1c0kA', '4lliA', '4ycuA', '2iopA',\n",
    "                            '2zxeA', '2b59B', '1h1vG']\n",
    "    with open(small_list_fn, 'w') as small_list_handle:\n",
    "        for p_id in proteins_of_interest:\n",
    "            small_list_handle.write('{}\\n'.format(p_id))\n",
    "generator = DataSetGenerator(input_dir)\n",
    "start = time()\n",
    "summary = generator.build_pdb_alignment_dataset(protein_list_fn=os.path.basename(small_list_fn), num_threads=10,\n",
    "                                                database='customuniref90.fasta', max_target_seqs=2500, remote=False, verbose=False)\n",
    "summary['Accession'] = summary['Protein_ID'].apply(lambda x: generator.protein_data[x]['Accession'])\n",
    "summary['Length'] = summary['Protein_ID'].apply(lambda x: generator.protein_data[x]['Length'])\n",
    "summary['Total_Size'] = summary.apply(lambda x: float(x['Length']) * float(x['Filtered_Alignment']), axis=1)\n",
    "summary.sort_values(by=['Total_Size', 'Length', 'Filtered_Alignment'], axis=0, inplace=True)\n",
    "print(summary[['Protein_ID', 'Accession', 'BLAST_Hits', 'Filtered_BLAST', 'Filtered_Alignment', 'Length', 'Total_Size']])\n",
    "end = time()\n",
    "print('It took {} min to generate the data set.'.format((end - start) / 60.0))\n",
    "summary.to_csv(os.path.join(input_dir, 'small_data_set_summary.tsv'), sep='\\t', index=False, header=True,\n",
    "               columns=['Protein_ID', 'Accession', 'BLAST_Hits', 'Filtered_BLAST', 'Filtered_Alignment', 'Length', 'Total_Size'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a location to store the output of this method comparison."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_dir = os.environ.get('OUTPUT_PATH')\n",
    "small_set_out_dir = os.path.join(output_dir, 'SmallTestSet')\n",
    "if not os.path.isdir(small_set_out_dir):\n",
    "    os.makedirs(small_set_out_dir)\n",
    "from SeqAlignment import SeqAlignment"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generating Values For Comparision#\n",
    "To determine the effectiveness of the new method and implementation the covariation of the same proteins will be computed using the previous Evolutionary Trace covariation method (ET-MIp) and other methods in the field.\n",
    "\n",
    "## ET-MIp##\n",
    "Scoring the the covariation of the proteins using the previous Evolutionary Trace covariation method (ET-MIp)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Attempting to calculate ET-MIp covariance for: 2iop\n",
      "Output:\n",
      "\n",
      "Error:\n",
      "allocation failure 2 in dmatrix().\n",
      "\n",
      "Could not compute ET-MIp covariance for: 2iop with seq_length: 1690 and size: 2473\n",
      "Attempting to calculate ET-MIp covariance for: 7hvp\n",
      "15.208096981\n",
      "Successfully computed ET-MIp covariance for: 7hvp\n",
      "Attempting to calculate ET-MIp covariance for: 1c0k\n",
      "129.20468998\n",
      "Successfully computed ET-MIp covariance for: 1c0k\n",
      "Attempting to calculate ET-MIp covariance for: 1c17\n",
      "1031.44045806\n",
      "Successfully computed ET-MIp covariance for: 1c17\n",
      "Attempting to calculate ET-MIp covariance for: 135l\n",
      "3960.81864905\n",
      "Successfully computed ET-MIp covariance for: 135l\n",
      "Attempting to calculate ET-MIp covariance for: 2ysd\n",
      "319.227418184\n",
      "Successfully computed ET-MIp covariance for: 2ysd\n",
      "Attempting to calculate ET-MIp covariance for: 206l\n",
      "68.6146600246\n",
      "Successfully computed ET-MIp covariance for: 206l\n",
      "Attempting to calculate ET-MIp covariance for: 1hck\n",
      "Output:\n",
      "\n",
      "Error:\n",
      "allocation failure 2 in imatrix() for int.\n",
      "allocation failure 2 in dmatrix().\n",
      "allocation failure 2 in dmatrix().\n",
      "\n",
      "Could not compute ET-MIp covariance for: 1hck with seq_length: 4309 and size: 1657\n",
      "Attempting to calculate ET-MIp covariance for: 1axb\n",
      "110483.620779\n",
      "Successfully computed ET-MIp covariance for: 1axb\n",
      "Attempting to calculate ET-MIp covariance for: 1h1v\n",
      "Output:\n",
      "\n",
      "Error:\n",
      "allocation failure 2 in imatrix() for int.\n",
      "allocation failure 2 in dmatrix().\n",
      "allocation failure 2 in dmatrix().\n",
      "\n",
      "Could not compute ET-MIp covariance for: 1h1v with seq_length: 8760 and size: 1534\n",
      "Attempting to calculate ET-MIp covariance for: 1jwl\n",
      "1388.63932991\n",
      "Successfully computed ET-MIp covariance for: 1jwl\n",
      "Attempting to calculate ET-MIp covariance for: 2z0e\n",
      "3385.60032797\n",
      "Successfully computed ET-MIp covariance for: 2z0e\n",
      "Attempting to calculate ET-MIp covariance for: 2b59\n",
      "21.8229320049\n",
      "Successfully computed ET-MIp covariance for: 2b59\n",
      "Attempting to calculate ET-MIp covariance for: 3tnu\n",
      "8991.78355217\n",
      "Successfully computed ET-MIp covariance for: 3tnu\n",
      "Attempting to calculate ET-MIp covariance for: 4ycu\n",
      "Output:\n",
      "\n",
      "Error:\n",
      "allocation failure 2 in imatrix() for int.\n",
      "allocation failure 2 in dmatrix().\n",
      "allocation failure 2 in dmatrix().\n",
      "\n",
      "Could not compute ET-MIp covariance for: 4ycu with seq_length: 4374 and size: 2356\n",
      "Attempting to calculate ET-MIp covariance for: 4lli\n",
      "7098.01172996\n",
      "Successfully computed ET-MIp covariance for: 4lli\n",
      "Attempting to calculate ET-MIp covariance for: 1bol\n",
      "314.855408192\n",
      "Successfully computed ET-MIp covariance for: 1bol\n",
      "Attempting to calculate ET-MIp covariance for: 1a26\n",
      "Output:\n",
      "\n",
      "Error:\n",
      "allocation failure 2 in dmatrix().\n",
      "\n",
      "Could not compute ET-MIp covariance for: 1a26 with seq_length: 2664 and size: 902\n",
      "Attempting to calculate ET-MIp covariance for: 2wer\n",
      "Output:\n",
      "\n",
      "Error:\n",
      "allocation failure 2 in dmatrix().\n",
      "\n",
      "Could not compute ET-MIp covariance for: 2wer with seq_length: 1765 and size: 1957\n",
      "Attempting to calculate ET-MIp covariance for: 3b6v\n",
      "Output:\n",
      "\n",
      "Error:\n",
      "allocation failure 2 in imatrix() for int.\n",
      "allocation failure 2 in dmatrix().\n",
      "allocation failure 2 in dmatrix().\n",
      "\n",
      "Could not compute ET-MIp covariance for: 3b6v with seq_length: 3302 and size: 1358\n",
      "Attempting to calculate ET-MIp covariance for: 3q05\n",
      "1349.28829312\n",
      "Successfully computed ET-MIp covariance for: 3q05\n",
      "Attempting to calculate ET-MIp covariance for: 2rh1\n",
      "95.9581720829\n",
      "Successfully computed ET-MIp covariance for: 2rh1\n",
      "Attempting to calculate ET-MIp covariance for: 2zxe\n",
      "Output:\n",
      "\n",
      "Error:\n",
      "allocation failure 2 in imatrix() for int.\n",
      "allocation failure 2 in dmatrix().\n",
      "allocation failure 2 in dmatrix().\n",
      "\n",
      "Could not compute ET-MIp covariance for: 2zxe with seq_length: 7784 and size: 798\n"
     ]
    }
   ],
   "source": [
    "from ETMIPWrapper import ETMIPWrapper\n",
    "etmip_out_dir = os.path.join(small_set_out_dir, 'ET-MIp')\n",
    "if not os.path.isdir(etmip_out_dir):\n",
    "    os.makedirs(etmip_out_dir)\n",
    "etmip_scores = {}\n",
    "for p_id in generator.protein_data:\n",
    "    print('Attempting to calculate ET-MIp covariance for: {}'.format(p_id))\n",
    "    try:\n",
    "        protein_out_dir = os.path.join(etmip_out_dir, p_id)\n",
    "        if not os.path.isdir(protein_out_dir):\n",
    "            os.makedirs(protein_out_dir)\n",
    "        curr_aln = SeqAlignment(file_name=generator.protein_data[p_id]['Final_FA_Aln'], query_id=p_id, polymer_type='Protein')\n",
    "        curr_aln.import_alignment()\n",
    "        curr_etmip = ETMIPWrapper(alignment=curr_aln)\n",
    "        curr_etmip.calculate_scores(out_dir=protein_out_dir, delete_files=False)\n",
    "        etmip_scores[p_id] = curr_dca\n",
    "        print('Successfully computed ET-MIp covariance for: {}'.format(p_id))\n",
    "    except ValueError:\n",
    "        print('Could not compute ET-MIp covariance for: {} with seq_length: {} and size: {}'.format(\n",
    "            p_id, curr_aln.seq_length, curr_aln.size))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DCA##\n",
    "Scoring the the covariation of the proteins using a DCA julia implementation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Attempting to calculate DCA covariance for: 2ysd\n",
      "Removing gaps took 0.026671814918518066 min\n",
      "Output:\n",
      "theta = 0.2180278161939097 threshold = 325.0\n",
      "M = 318 N = 1491 Meff = 35.63992898084913\n",
      "\n",
      "Error:\n",
      "ERROR: LoadError: OutOfMemoryError()\n",
      "Stacktrace:\n",
      " [1] Type at ./boot.jl:396 [inlined]\n",
      " [2] Type at ./boot.jl:404 [inlined]\n",
      " [3] Type at ./boot.jl:411 [inlined]\n",
      " [4] similar at ./abstractarray.jl:618 [inlined]\n",
      " [5] similar at ./abstractarray.jl:617 [inlined]\n",
      " [6] similar at ./broadcast.jl:193 [inlined]\n",
      " [7] copy at ./broadcast.jl:776 [inlined]\n",
      " [8] materialize at ./broadcast.jl:756 [inlined]\n",
      " [9] broadcast(::typeof(*), ::Array{Float64,1}, ::LinearAlgebra.Adjoint{Float64,Array{Float64,1}}) at ./broadcast.jl:710\n",
      " [10] * at /home/conda/feedstock_root/build_artifacts/julia_1548684429855/work/usr/share/julia/stdlib/v1.0/LinearAlgebra/src/adjtrans.jl:205 [inlined]\n",
      " [11] compute_C at /home/daniel/.julia/dev/GaussDCA/src/GaussDCA.jl:176 [inlined]\n",
      " [12] #gDCA#1(::Float64, ::Symbol, ::Float64, ::Symbol, ::Int64, ::Bool, ::Function, ::String) at /home/daniel/.julia/dev/GaussDCA/src/GaussDCA.jl:46\n",
      " [13] gDCA(::String) at /home/daniel/.julia/dev/GaussDCA/src/GaussDCA.jl:30\n",
      " [14] top-level scope at none:0\n",
      " [15] include at ./boot.jl:317 [inlined]\n",
      " [16] include_relative(::Module, ::String) at ./loading.jl:1044\n",
      " [17] include(::Module, ::String) at ./sysimg.jl:29\n",
      " [18] exec_options(::Base.JLOptions) at ./client.jl:266\n",
      " [19] _start() at ./client.jl:425\n",
      "in expression starting at /home/daniel/Documents/git/ETMIP/src/SupportingClasses/cmd_line_GausDCA.jl:13\n",
      "\n",
      "Could not compute DCA covariance for: 2ysd with seq_length: 1491 and size: 318\n",
      "Attempting to calculate DCA covariance for: 1c17\n",
      "Removing gaps took 0.004229422410329183 min\n",
      "Output:\n",
      "theta = 0.25974123850629577 threshold = 20.0\n",
      "M = 813 N = 79 Meff = 164.31823105079212\n",
      "\n",
      "Error:\n",
      "\n",
      "6.029179573059082\n",
      "Successfully computed DCA covariance for: 1c17\n",
      "Attempting to calculate DCA covariance for: 3tnu\n",
      "Removing gaps took 0.057607607046763105 min\n",
      "Output:\n",
      "theta = 0.23577482649885514 threshold = 111.0\n",
      "M = 1518 N = 472 Meff = 295.4124406368615\n",
      "\n",
      "Error:\n",
      "\n",
      "16.284579753875732\n",
      "Successfully computed DCA covariance for: 3tnu\n",
      "Attempting to calculate DCA covariance for: 7hvp\n",
      "Removing gaps took 0.000983575979868571 min\n",
      "Output:\n",
      "theta = 0.17542928831787158 threshold = 252.0\n",
      "M = 33 N = 1437 Meff = 10.629003267973856\n",
      "\n",
      "Error:\n",
      "\n",
      "182.14166355133057\n",
      "Successfully computed DCA covariance for: 7hvp\n",
      "Attempting to calculate DCA covariance for: 135l\n",
      "Removing gaps took 0.011375614007314046 min\n",
      "Output:\n",
      "theta = 0.307386891588418 threshold = 45.0\n",
      "M = 818 N = 147 Meff = 155.49995145058068\n",
      "\n",
      "Error:\n",
      "\n",
      "6.307071208953857\n",
      "Successfully computed DCA covariance for: 135l\n",
      "Attempting to calculate DCA covariance for: 206l\n",
      "Removing gaps took 0.0018338402112325033 min\n",
      "Output:\n",
      "theta = 0.2513500653374363 threshold = 41.0\n",
      "M = 92 N = 164 Meff = 40.56154790771665\n",
      "\n",
      "Error:\n",
      "\n",
      "6.554086446762085\n",
      "Successfully computed DCA covariance for: 206l\n",
      "Attempting to calculate DCA covariance for: 2wer\n",
      "Removing gaps took 0.0865223487218221 min\n",
      "Output:\n",
      "theta = 0.2574036683998775 threshold = 182.0\n",
      "M = 1207 N = 709 Meff = 279.4606548688516\n",
      "\n",
      "Error:\n",
      "\n",
      "35.44221472740173\n",
      "Successfully computed DCA covariance for: 2wer\n",
      "Attempting to calculate DCA covariance for: 1bol\n",
      "Removing gaps took 0.0012925306955973307 min\n",
      "Output:\n",
      "theta = 0.27134702532943505 threshold = 64.0\n",
      "M = 127 N = 238 Meff = 70.26568194877018\n",
      "\n",
      "Error:\n",
      "\n",
      "7.674593210220337\n",
      "Successfully computed DCA covariance for: 1bol\n",
      "Attempting to calculate DCA covariance for: 3q05\n",
      "Removing gaps took 0.005605324109395345 min\n",
      "Output:\n",
      "theta = 0.2436204128879927 threshold = 95.0\n",
      "M = 210 N = 393 Meff = 69.92459982407618\n",
      "\n",
      "Error:\n",
      "\n",
      "12.025794982910156\n",
      "Successfully computed DCA covariance for: 3q05\n",
      "Attempting to calculate DCA covariance for: 1axb\n",
      "Removing gaps took 0.040909095605214434 min\n",
      "Output:\n",
      "theta = 0.23495474752516796 threshold = 67.0\n",
      "M = 2087 N = 286 Meff = 799.8372703831099\n",
      "\n",
      "Error:\n",
      "\n",
      "9.053714036941528\n",
      "Successfully computed DCA covariance for: 1axb\n",
      "Attempting to calculate DCA covariance for: 2rh1\n",
      "Removing gaps took 0.008382693926493327 min\n",
      "Output:\n",
      "theta = 0.2133874731946681 threshold = 88.0\n",
      "M = 399 N = 413 Meff = 72.16265302111556\n",
      "\n",
      "Error:\n",
      "\n",
      "12.945265531539917\n",
      "Successfully computed DCA covariance for: 2rh1\n",
      "Attempting to calculate DCA covariance for: 1hck\n",
      "Removing gaps took 0.060328900814056396 min\n",
      "Output:\n",
      "theta = 0.2342361240414314 threshold = 69.0\n",
      "M = 1681 N = 298 Meff = 477.80098296157166\n",
      "\n",
      "Error:\n",
      "\n",
      "9.635658740997314\n",
      "Successfully computed DCA covariance for: 1hck\n",
      "Attempting to calculate DCA covariance for: 3b6v\n",
      "Removing gaps took 0.026910297075907388 min\n",
      "Output:\n",
      "theta = 0.23323634540367308 threshold = 184.0\n",
      "M = 441 N = 793 Meff = 189.17153250402947\n",
      "\n",
      "Error:\n",
      "\n",
      "44.009138345718384\n",
      "Successfully computed DCA covariance for: 3b6v\n",
      "Attempting to calculate DCA covariance for: 2z0e\n",
      "Removing gaps took 0.00939255158106486 min\n",
      "Output:\n",
      "theta = 0.25495537563882453 threshold = 100.0\n",
      "M = 344 N = 393 Meff = 108.34600907839625\n",
      "\n",
      "Error:\n",
      "\n",
      "12.007091999053955\n",
      "Successfully computed DCA covariance for: 2z0e\n",
      "Attempting to calculate DCA covariance for: 1jwl\n",
      "Removing gaps took 0.003613301118214925 min\n",
      "Output:\n",
      "theta = 0.25033654713391756 threshold = 90.0\n",
      "M = 226 N = 360 Meff = 74.8712170526412\n",
      "\n",
      "Error:\n",
      "\n",
      "10.87816834449768\n",
      "Successfully computed DCA covariance for: 1jwl\n",
      "Attempting to calculate DCA covariance for: 1a26\n",
      "Removing gaps took 0.01639768679936727 min\n",
      "Output:\n",
      "theta = 0.25806261584141016 threshold = 260.0\n",
      "M = 261 N = 1011 Meff = 112.47182965306826\n",
      "\n",
      "Error:\n",
      "\n",
      "76.51840329170227\n",
      "Successfully computed DCA covariance for: 1a26\n",
      "Attempting to calculate DCA covariance for: 1c0k\n",
      "Removing gaps took 0.0011425336201985677 min\n",
      "Output:\n",
      "theta = 0.2580937474554649 threshold = 94.0\n",
      "M = 51 N = 368 Meff = 35.71666666666666\n",
      "\n",
      "Error:\n",
      "\n",
      "11.062349796295166\n",
      "Successfully computed DCA covariance for: 1c0k\n",
      "Attempting to calculate DCA covariance for: 4lli\n",
      "Removing gaps took 0.04841580788294474 min\n",
      "Output:\n",
      "theta = 0.24468297838671443 threshold = 453.0\n",
      "M = 352 N = 1855 Meff = 88.10520987057846\n",
      "\n",
      "Error:\n",
      "ERROR: LoadError: OutOfMemoryError()\n",
      "Stacktrace:\n",
      " [1] materialize at ./boot.jl:396 [inlined]\n",
      " [2] add_pseudocount(::Array{Float64,1}, ::Array{Float64,2}, ::Float64, ::Int64, ::Int64) at /home/daniel/.julia/dev/GaussDCA/src/GaussDCA.jl:157\n",
      " [3] #gDCA#1(::Float64, ::Symbol, ::Float64, ::Symbol, ::Int64, ::Bool, ::Function, ::String) at /home/daniel/.julia/dev/GaussDCA/src/GaussDCA.jl:44\n",
      " [4] gDCA(::String) at /home/daniel/.julia/dev/GaussDCA/src/GaussDCA.jl:30\n",
      " [5] top-level scope at none:0\n",
      " [6] include at ./boot.jl:317 [inlined]\n",
      " [7] include_relative(::Module, ::String) at ./loading.jl:1044\n",
      " [8] include(::Module, ::String) at ./sysimg.jl:29\n",
      " [9] exec_options(::Base.JLOptions) at ./client.jl:266\n",
      " [10] _start() at ./client.jl:425\n",
      "in expression starting at /home/daniel/Documents/git/ETMIP/src/SupportingClasses/cmd_line_GausDCA.jl:13\n",
      "\n",
      "Could not compute DCA covariance for: 4lli with seq_length: 1855 and size: 352\n",
      "Attempting to calculate DCA covariance for: 4ycu\n",
      "Removing gaps took 0.14794599215189616 min\n",
      "Output:\n",
      "theta = 0.23074377425863482 threshold = 137.0\n",
      "M = 2347 N = 597 Meff = 554.2573842112172\n",
      "\n",
      "Error:\n",
      "\n",
      "25.400590658187866\n",
      "Successfully computed DCA covariance for: 4ycu\n",
      "Attempting to calculate DCA covariance for: 2iop\n",
      "Removing gaps took 0.10019810199737549 min\n",
      "Output:\n",
      "theta = 0.2437993421853817 threshold = 152.0\n",
      "M = 2472 N = 624 Meff = 622.7788481908626\n",
      "\n",
      "Error:\n",
      "\n",
      "28.190845251083374\n",
      "Successfully computed DCA covariance for: 2iop\n",
      "Attempting to calculate DCA covariance for: 2zxe\n",
      "Removing gaps took 0.07456478675206503 min\n",
      "Output:\n",
      "theta = 0.23307614798911286 threshold = 239.0\n",
      "M = 795 N = 1028 Meff = 234.85546563765973\n",
      "\n",
      "Error:\n",
      "\n",
      "80.8471326828003\n",
      "Successfully computed DCA covariance for: 2zxe\n",
      "Attempting to calculate DCA covariance for: 2b59\n",
      "Removing gaps took 0.00035091638565063474 min\n",
      "Output:\n",
      "theta = 0.24293778975741243 threshold = 450.0\n",
      "M = 4 N = 1853 Meff = 4.0\n",
      "\n",
      "Error:\n",
      "ERROR: LoadError: OutOfMemoryError()\n",
      "Stacktrace:\n",
      " [1] materialize at ./boot.jl:396 [inlined]\n",
      " [2] add_pseudocount(::Array{Float64,1}, ::Array{Float64,2}, ::Float64, ::Int64, ::Int64) at /home/daniel/.julia/dev/GaussDCA/src/GaussDCA.jl:157\n",
      " [3] #gDCA#1(::Float64, ::Symbol, ::Float64, ::Symbol, ::Int64, ::Bool, ::Function, ::String) at /home/daniel/.julia/dev/GaussDCA/src/GaussDCA.jl:44\n",
      " [4] gDCA(::String) at /home/daniel/.julia/dev/GaussDCA/src/GaussDCA.jl:30\n",
      " [5] top-level scope at none:0\n",
      " [6] include at ./boot.jl:317 [inlined]\n",
      " [7] include_relative(::Module, ::String) at ./loading.jl:1044\n",
      " [8] include(::Module, ::String) at ./sysimg.jl:29\n",
      " [9] exec_options(::Base.JLOptions) at ./client.jl:266\n",
      " [10] _start() at ./client.jl:425\n",
      "in expression starting at /home/daniel/Documents/git/ETMIP/src/SupportingClasses/cmd_line_GausDCA.jl:13\n",
      "\n",
      "Could not compute DCA covariance for: 2b59 with seq_length: 1853 and size: 4\n",
      "Attempting to calculate DCA covariance for: 1h1v\n",
      "Removing gaps took 0.03923610051472982 min\n",
      "Output:\n",
      "theta = 0.23279931761124675 threshold = 182.0\n",
      "M = 599 N = 782 Meff = 122.3009409447558\n",
      "\n",
      "Error:\n",
      "\n",
      "43.00722551345825\n",
      "Successfully computed DCA covariance for: 1h1v\n"
     ]
    }
   ],
   "source": [
    "from DCAWrapper import DCAWrapper\n",
    "dca_out_dir = os.path.join(small_set_out_dir, 'DCA')\n",
    "if not os.path.isdir(dca_out_dir):\n",
    "    os.makedirs(dca_out_dir)\n",
    "dca_scores = {}\n",
    "for p_id in generator.protein_data:\n",
    "    print('Attempting to calculate DCA covariance for: {}'.format(p_id))\n",
    "    try:\n",
    "        protein_out_dir = os.path.join(dca_out_dir, p_id)\n",
    "        if not os.path.isdir(protein_out_dir):\n",
    "            os.makedirs(protein_out_dir)\n",
    "        curr_aln = SeqAlignment(file_name=generator.protein_data[p_id]['Final_FA_Aln'], query_id=p_id,\n",
    "                                polymer_type='Protein')\n",
    "        curr_aln.import_alignment()\n",
    "        # Since the DCA implementation used here does not provide a way to specify the query sequence we remove the gaps\n",
    "        # from the query sequences so positions will be referenced correctly for that sequence (and unnecessary\n",
    "        # computations can be avoided).\n",
    "        curr_aln = curr_aln.remove_gaps()\n",
    "        new_aln_fn = os.path.join(protein_out_dir, '{}_no_gap.fasta'.format(p_id))\n",
    "        curr_aln.write_out_alignment(new_aln_fn)\n",
    "        curr_aln.file_name = new_aln_fn\n",
    "        curr_dca = DCAWrapper(alignment=curr_aln)\n",
    "        curr_dca.calculate_scores(out_dir=protein_out_dir, delete_file=False)\n",
    "        dca_scores[p_id] = curr_dca\n",
    "        print('Successfully computed DCA covariance for: {}'.format(p_id))\n",
    "    except ValueError:\n",
    "        print('Could not compute DCA covariance for: {} with seq_length: {} and size: {}'.format(\n",
    "            p_id, curr_aln.seq_length, curr_aln.size))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## EVCouplings##\n",
    "Scoring the the covariation of the proteins using the EVCouplings method."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python3 (PyET3)",
   "language": "python",
   "name": "pyet3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
