{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Small Test Set #\n",
    "## Goal ##\n",
    "The goal of this test set is to perform proof of concept testing on a small number of proteins with a wide range of sizes and available homologs, orthologs, and paralogs. By doing so it should be possible to test the best parameterization for this tool as well as identifying the strengths and weaknesses of the tool using various measurments as end points.\n",
    "## Warning ##\n",
    "Before attempting to use this notebook make sure that your .env file has been properly setup to reflect the correct locations of command line tools and the location of files and directories needed for execution.\n",
    "### Initial Import###\n",
    "This first cell performs the necessary imports required to begin this notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "from dotenv import find_dotenv, load_dotenv\n",
    "try:\n",
    "    dotenv_path = find_dotenv(raise_error_if_not_found=True)\n",
    "except IOError:\n",
    "    dotenv_path = find_dotenv(raise_error_if_not_found=True, usecwd=True)\n",
    "load_dotenv(dotenv_path)\n",
    "import os\n",
    "import sys\n",
    "sys.path.append(os.path.join(os.environ.get('PROJECT_PATH'), 'src', 'SupportingClasses'))\n",
    "input_dir = os.environ.get('INPUT_PATH')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Set Construction ##\n",
    "The first task required to test the data set is to download the required data and construct any necessary input files for all down stream analyses.\n",
    "In this case that means:\n",
    "* Downloading PDB files for the proteins in our small test set.\n",
    "* Extracting a query sequence from each PDB file.\n",
    "* Searching for paralogs, homologs, and orthologs in a custom BLAST database built by filtering the Uniref90 database.\n",
    "* Filtering the hits from the BLAST search to meet minimum and maximum length requirements, as well as minimum and maximum identity requirements.\n",
    "* Building alignments using CLUSTALW in both the fasta and msf formats since some of the tools which will be used for comparison need different formats.\n",
    "* Filtering the alignment for maximum identity similarity between seqeunces.\n",
    "* Re-aligning the filtered sequences using CLUSTALW.\n",
    "This is all handeled by the DataSetGenerator class found in the src/SupportingClasses folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_list_dir = os.path.join(input_dir, 'ProteinLists')\n",
    "if not os.path.isdir(protein_list_dir):\n",
    "    os.makedirs(protein_list_dir)\n",
    "small_list_fn = os.path.join(protein_list_dir, 'SmallDataSet.txt')\n",
    "if not os.path.isfile(small_list_fn):\n",
    "    proteins_of_interest = ['2ysdA', '1c17A', '3tnuA', '7hvpA', '135lA', '206lA', '2b59A', '2werA', '1bolA', '3q05A',\n",
    "                            '1axbA', '2rh1A', '1hckA', '3b6vA', '2z0eA', '1jwlA', '1a26A', '1c0kA', '1h1vA', '4lliA',\n",
    "                            '4ycuA', '2iopA', '2zxeA']\n",
    "    with open(small_list_fn, 'wb') as small_list_handle:\n",
    "        for p_id in proteins_of_interest:\n",
    "            small_list_handle.write('{}\\n'.format(p_id))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Importing protein list\n",
      "Downloading structures and parsing in query sequences\n",
      "BLASTing query sequences\n",
      "Filtering BLAST hits, aligning, filtering by identity, and re-aligning\n",
      "It took 0.153495001793 min to generate the data set.\n"
     ]
    }
   ],
   "source": [
    "from time import time\n",
    "from DataSetGenerator import DataSetGenerator\n",
    "generator = DataSetGenerator(input_dir)\n",
    "start = time()\n",
    "generator.build_pdb_alignment_dataset(protein_list_fn=os.path.basename(small_list_fn), num_threads=10,\n",
    "                                      database='customuniref90.fasta', max_target_seqs=2500, remote=False, verbose=False)\n",
    "end = time()\n",
    "print('It took {} min to generate the data set.'.format((end - start) / 60.0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_dir = os.environ.get('OUTPUT_PATH')\n",
    "small_set_out_dir = os.path.join(output_dir, 'SmallTestSet')\n",
    "if not os.path.isdir(small_set_out_dir):\n",
    "    os.makedirs(small_set_out_dir)\n",
    "from SeqAlignment import SeqAlignment"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generating Values For Comparision#\n",
    "To determine the effectiveness of the new method and implementation the covariation of the same proteins will be computed using the previous Evolutionary Trace covariation method (ET-MIp) and other methods in the field.\n",
    "\n",
    "## ET-MIp##\n",
    "Scoring the the covariation of the proteins using the previous Evolutionary Trace covariation method (ET-MIp)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Attempting to calculate ET-MIp covariance for: 2iop\n",
      "Output:\n",
      "\n",
      "Error:\n",
      "allocation failure 2 in dmatrix().\n",
      "\n",
      "Could not compute ET-MIp covariance for: 2iop with seq_length: 1690 and size: 2473\n",
      "Attempting to calculate ET-MIp covariance for: 7hvp\n",
      "15.208096981\n",
      "Successfully computed ET-MIp covariance for: 7hvp\n",
      "Attempting to calculate ET-MIp covariance for: 1c0k\n",
      "129.20468998\n",
      "Successfully computed ET-MIp covariance for: 1c0k\n",
      "Attempting to calculate ET-MIp covariance for: 1c17\n",
      "1031.44045806\n",
      "Successfully computed ET-MIp covariance for: 1c17\n",
      "Attempting to calculate ET-MIp covariance for: 135l\n",
      "3960.81864905\n",
      "Successfully computed ET-MIp covariance for: 135l\n",
      "Attempting to calculate ET-MIp covariance for: 2ysd\n",
      "319.227418184\n",
      "Successfully computed ET-MIp covariance for: 2ysd\n",
      "Attempting to calculate ET-MIp covariance for: 206l\n",
      "68.6146600246\n",
      "Successfully computed ET-MIp covariance for: 206l\n",
      "Attempting to calculate ET-MIp covariance for: 1hck\n",
      "Output:\n",
      "\n",
      "Error:\n",
      "allocation failure 2 in imatrix() for int.\n",
      "allocation failure 2 in dmatrix().\n",
      "allocation failure 2 in dmatrix().\n",
      "\n",
      "Could not compute ET-MIp covariance for: 1hck with seq_length: 4309 and size: 1657\n",
      "Attempting to calculate ET-MIp covariance for: 1axb\n",
      "110483.620779\n",
      "Successfully computed ET-MIp covariance for: 1axb\n",
      "Attempting to calculate ET-MIp covariance for: 1h1v\n",
      "Output:\n",
      "\n",
      "Error:\n",
      "allocation failure 2 in imatrix() for int.\n",
      "allocation failure 2 in dmatrix().\n",
      "allocation failure 2 in dmatrix().\n",
      "\n",
      "Could not compute ET-MIp covariance for: 1h1v with seq_length: 8760 and size: 1534\n",
      "Attempting to calculate ET-MIp covariance for: 1jwl\n",
      "1388.63932991\n",
      "Successfully computed ET-MIp covariance for: 1jwl\n",
      "Attempting to calculate ET-MIp covariance for: 2z0e\n",
      "3385.60032797\n",
      "Successfully computed ET-MIp covariance for: 2z0e\n",
      "Attempting to calculate ET-MIp covariance for: 2b59\n",
      "21.8229320049\n",
      "Successfully computed ET-MIp covariance for: 2b59\n",
      "Attempting to calculate ET-MIp covariance for: 3tnu\n",
      "8991.78355217\n",
      "Successfully computed ET-MIp covariance for: 3tnu\n",
      "Attempting to calculate ET-MIp covariance for: 4ycu\n",
      "Output:\n",
      "\n",
      "Error:\n",
      "allocation failure 2 in imatrix() for int.\n",
      "allocation failure 2 in dmatrix().\n",
      "allocation failure 2 in dmatrix().\n",
      "\n",
      "Could not compute ET-MIp covariance for: 4ycu with seq_length: 4374 and size: 2356\n",
      "Attempting to calculate ET-MIp covariance for: 4lli\n",
      "7098.01172996\n",
      "Successfully computed ET-MIp covariance for: 4lli\n",
      "Attempting to calculate ET-MIp covariance for: 1bol\n",
      "314.855408192\n",
      "Successfully computed ET-MIp covariance for: 1bol\n",
      "Attempting to calculate ET-MIp covariance for: 1a26\n",
      "Output:\n",
      "\n",
      "Error:\n",
      "allocation failure 2 in dmatrix().\n",
      "\n",
      "Could not compute ET-MIp covariance for: 1a26 with seq_length: 2664 and size: 902\n",
      "Attempting to calculate ET-MIp covariance for: 2wer\n",
      "Output:\n",
      "\n",
      "Error:\n",
      "allocation failure 2 in dmatrix().\n",
      "\n",
      "Could not compute ET-MIp covariance for: 2wer with seq_length: 1765 and size: 1957\n",
      "Attempting to calculate ET-MIp covariance for: 3b6v\n",
      "Output:\n",
      "\n",
      "Error:\n",
      "allocation failure 2 in imatrix() for int.\n",
      "allocation failure 2 in dmatrix().\n",
      "allocation failure 2 in dmatrix().\n",
      "\n",
      "Could not compute ET-MIp covariance for: 3b6v with seq_length: 3302 and size: 1358\n",
      "Attempting to calculate ET-MIp covariance for: 3q05\n",
      "1349.28829312\n",
      "Successfully computed ET-MIp covariance for: 3q05\n",
      "Attempting to calculate ET-MIp covariance for: 2rh1\n",
      "95.9581720829\n",
      "Successfully computed ET-MIp covariance for: 2rh1\n",
      "Attempting to calculate ET-MIp covariance for: 2zxe\n",
      "Output:\n",
      "\n",
      "Error:\n",
      "allocation failure 2 in imatrix() for int.\n",
      "allocation failure 2 in dmatrix().\n",
      "allocation failure 2 in dmatrix().\n",
      "\n",
      "Could not compute ET-MIp covariance for: 2zxe with seq_length: 7784 and size: 798\n"
     ]
    }
   ],
   "source": [
    "from ETMIPWrapper import ETMIPWrapper\n",
    "etmip_out_dir = os.path.join(small_set_out_dir, 'ET-MIp')\n",
    "if not os.path.isdir(etmip_out_dir):\n",
    "    os.makedirs(etmip_out_dir)\n",
    "etmip_scores = {}\n",
    "for p_id in generator.protein_data:\n",
    "    print('Attempting to calculate ET-MIp covariance for: {}'.format(p_id))\n",
    "    try:\n",
    "        protein_out_dir = os.path.join(etmip_out_dir, p_id)\n",
    "        if not os.path.isdir(protein_out_dir):\n",
    "            os.makedirs(protein_out_dir)\n",
    "        curr_aln = SeqAlignment(file_name=generator.protein_data[p_id]['Final_FA_Aln'], query_id=p_id, polymer_type='Protein')\n",
    "        curr_aln.import_alignment()\n",
    "        curr_etmip = ETMIPWrapper(alignment=curr_aln)\n",
    "        curr_etmip.calculate_scores(out_dir=protein_out_dir, delete_files=False)\n",
    "        etmip_scores[p_id] = curr_dca\n",
    "        print('Successfully computed ET-MIp covariance for: {}'.format(p_id))\n",
    "    except ValueError:\n",
    "        print('Could not compute ET-MIp covariance for: {} with seq_length: {} and size: {}'.format(\n",
    "            p_id, curr_aln.seq_length, curr_aln.size))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DCA##\n",
    "Scoring the the covariation of the proteins using a DCA julia implementation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Attempting to calculate DCA covariance for: 2iop\n",
      "Removing gaps took 0.121132250627 min\n",
      "Output:\n",
      "theta = 0.24214823332465546 threshold = 149.0\n",
      "M = 2473 N = 618 Meff = 640.4080550722614\n",
      "\n",
      "Error:\n",
      "\n",
      "30.1861131191\n",
      "Successfully computed DCA covariance for: 2iop\n",
      "Attempting to calculate DCA covariance for: 7hvp\n",
      "Removing gaps took 0.000621664524078 min\n",
      "Output:\n",
      "theta = 0.19415079847616118 threshold = 18.0\n",
      "M = 59 N = 97 Meff = 24.145491838655456\n",
      "\n",
      "Error:\n",
      "\n",
      "5.05751204491\n",
      "Successfully computed DCA covariance for: 7hvp\n",
      "Attempting to calculate DCA covariance for: 1c0k\n",
      "Removing gaps took 0.000751682122548 min\n",
      "Output:\n",
      "theta = 0.26199960512671616 threshold = 95.0\n",
      "M = 54 N = 363 Meff = 36.266666666666666\n",
      "\n",
      "Error:\n",
      "\n",
      "10.8456110954\n",
      "Successfully computed DCA covariance for: 1c0k\n",
      "Attempting to calculate DCA covariance for: 1c17\n",
      "Removing gaps took 0.00439543326696 min\n",
      "Output:\n",
      "theta = 0.25974123850629577 threshold = 20.0\n",
      "M = 813 N = 79 Meff = 164.31823105079218\n",
      "\n",
      "Error:\n",
      "\n",
      "4.85462784767\n",
      "Successfully computed DCA covariance for: 1c17\n",
      "Attempting to calculate DCA covariance for: 135l\n",
      "Removing gaps took 0.0136780142784 min\n",
      "Output:\n",
      "theta = 0.29348014503263226 threshold = 37.0\n",
      "M = 946 N = 129 Meff = 174.94424227092037\n",
      "\n",
      "Error:\n",
      "\n",
      "5.37845802307\n",
      "Successfully computed DCA covariance for: 135l\n",
      "Attempting to calculate DCA covariance for: 2ysd\n",
      "Removing gaps took 0.00314469734828 min\n",
      "Output:\n",
      "theta = 0.23410477454152434 threshold = 13.0\n",
      "M = 624 N = 57 Meff = 161.18108441195093\n",
      "\n",
      "Error:\n",
      "\n",
      "4.76661801338\n",
      "Successfully computed DCA covariance for: 2ysd\n",
      "Attempting to calculate DCA covariance for: 206l\n",
      "Removing gaps took 0.000554434458415 min\n",
      "Output:\n",
      "theta = 0.2654356912028723 threshold = 43.0\n",
      "M = 93 N = 162 Meff = 45.40141108891109\n",
      "\n",
      "Error:\n",
      "\n",
      "5.60829591751\n",
      "Successfully computed DCA covariance for: 206l\n",
      "Attempting to calculate DCA covariance for: 1hck\n",
      "Removing gaps took 0.082338865598 min\n",
      "Output:\n",
      "theta = 0.22516467597825307 threshold = 66.0\n",
      "M = 1657 N = 294 Meff = 420.24338871737143\n",
      "\n",
      "Error:\n",
      "\n",
      "8.71556591988\n",
      "Successfully computed DCA covariance for: 1hck\n",
      "Attempting to calculate DCA covariance for: 1axb\n",
      "Removing gaps took 0.0645329157511 min\n",
      "Output:\n",
      "theta = 0.2593716458953484 threshold = 68.0\n",
      "M = 2140 N = 263 Meff = 728.2388549192701\n",
      "\n",
      "Error:\n",
      "\n",
      "7.98908090591\n",
      "Successfully computed DCA covariance for: 1axb\n",
      "Attempting to calculate DCA covariance for: 1h1v\n",
      "Removing gaps took 0.133141167959 min\n",
      "Output:\n",
      "theta = 0.2643931794789898 threshold = 97.0\n",
      "M = 1534 N = 368 Meff = 459.25591405916606\n",
      "\n",
      "Error:\n",
      "\n",
      "11.3196079731\n",
      "Successfully computed DCA covariance for: 1h1v\n",
      "Attempting to calculate DCA covariance for: 1jwl\n",
      "Removing gaps took 0.00339941978455 min\n",
      "Output:\n",
      "theta = 0.24558720082080387 threshold = 80.0\n",
      "M = 231 N = 329 Meff = 74.23954435507807\n",
      "\n",
      "Error:\n",
      "\n",
      "9.55069804192\n",
      "Successfully computed DCA covariance for: 1jwl\n",
      "Attempting to calculate DCA covariance for: 2z0e\n",
      "Removing gaps took 0.0126938819885 min\n",
      "Output:\n",
      "theta = 0.2426972050073856 threshold = 76.0\n",
      "M = 351 N = 315 Meff = 104.14350431382843\n",
      "\n",
      "Error:\n",
      "\n",
      "9.44197201729\n",
      "Successfully computed DCA covariance for: 2z0e\n",
      "Attempting to calculate DCA covariance for: 2b59\n",
      "Removing gaps took 0.000176417827606 min\n",
      "Output:\n",
      "theta = 0.2740624834950341 threshold = 45.0\n",
      "M = 44 N = 166 Meff = 35.333333333333336\n",
      "\n",
      "Error:\n",
      "\n",
      "5.85636520386\n",
      "Successfully computed DCA covariance for: 2b59\n",
      "Attempting to calculate DCA covariance for: 3tnu\n",
      "Removing gaps took 0.0171169002851 min\n",
      "Output:\n",
      "theta = 0.22193425812739595 threshold = 19.0\n",
      "M = 1835 N = 90 Meff = 291.59484913531685\n",
      "\n",
      "Error:\n",
      "\n",
      "5.24426698685\n",
      "Successfully computed DCA covariance for: 3tnu\n",
      "Attempting to calculate DCA covariance for: 4ycu\n",
      "Removing gaps took 0.189930816491 min\n",
      "Output:\n",
      "theta = 0.25155359638606367 threshold = 127.0\n",
      "M = 2356 N = 506 Meff = 544.3522187846356\n",
      "\n",
      "Error:\n",
      "\n",
      "19.5850419998\n",
      "Successfully computed DCA covariance for: 4ycu\n",
      "Attempting to calculate DCA covariance for: 4lli\n",
      "Removing gaps took 0.0201053818067 min\n",
      "Output:\n",
      "theta = 0.24643295576173735 threshold = 92.0\n",
      "M = 435 N = 377 Meff = 85.4011599712531\n",
      "\n",
      "Error:\n",
      "\n",
      "11.5103650093\n",
      "Successfully computed DCA covariance for: 4lli\n",
      "Attempting to calculate DCA covariance for: 1bol\n",
      "Removing gaps took 0.00128388007482 min\n",
      "Output:\n",
      "theta = 0.28095009567289936 threshold = 62.0\n",
      "M = 154 N = 222 Meff = 72.66038218919023\n",
      "\n",
      "Error:\n",
      "\n",
      "6.77346777916\n",
      "Successfully computed DCA covariance for: 1bol\n",
      "Attempting to calculate DCA covariance for: 1a26\n",
      "Removing gaps took 0.0502624511719 min\n",
      "Output:\n",
      "theta = 0.283441055340027 threshold = 99.0\n",
      "M = 902 N = 351 Meff = 294.92396662647207\n",
      "\n",
      "Error:\n",
      "\n",
      "10.5819900036\n",
      "Successfully computed DCA covariance for: 1a26\n",
      "Attempting to calculate DCA covariance for: 2wer\n",
      "Removing gaps took 0.0872077504794 min\n",
      "Output:\n",
      "theta = 0.24540013047645298 threshold = 52.0\n",
      "M = 1957 N = 214 Meff = 437.4907716035432\n",
      "\n",
      "Error:\n",
      "\n",
      "6.83188414574\n",
      "Successfully computed DCA covariance for: 2wer\n",
      "Attempting to calculate DCA covariance for: 3b6v\n",
      "Removing gaps took 0.0729323148727 min\n",
      "Output:\n",
      "theta = 0.23968718307364154 threshold = 73.0\n",
      "M = 1358 N = 306 Meff = 316.42494135406844\n",
      "\n",
      "Error:\n",
      "\n",
      "9.22730183601\n",
      "Successfully computed DCA covariance for: 3b6v\n",
      "Attempting to calculate DCA covariance for: 3q05\n",
      "Removing gaps took 0.00595390001933 min\n",
      "Output:\n",
      "theta = 0.21047783516485308 threshold = 49.0\n",
      "M = 316 N = 234 Meff = 61.962070678905775\n",
      "\n",
      "Error:\n",
      "\n",
      "7.0948600769\n",
      "Successfully computed DCA covariance for: 3q05\n",
      "Attempting to calculate DCA covariance for: 2rh1\n",
      "Removing gaps took 0.00158646504084 min\n",
      "Output:\n",
      "theta = 0.2991989820740242 threshold = 132.0\n",
      "M = 35 N = 442 Meff = 21.333333333333336\n",
      "\n",
      "Error:\n",
      "\n",
      "14.728703022\n",
      "Successfully computed DCA covariance for: 2rh1\n",
      "Attempting to calculate DCA covariance for: 2zxe\n",
      "Removing gaps took 0.118946218491 min\n",
      "Output:\n",
      "theta = 0.2376809941399701 threshold = 235.0\n",
      "M = 798 N = 992 Meff = 239.66057471238324\n",
      "\n",
      "Error:\n",
      "\n",
      "79.9719362259\n",
      "Successfully computed DCA covariance for: 2zxe\n"
     ]
    }
   ],
   "source": [
    "from DCAWrapper import DCAWrapper\n",
    "dca_out_dir = os.path.join(small_set_out_dir, 'DCA')\n",
    "if not os.path.isdir(dca_out_dir):\n",
    "    os.makedirs(dca_out_dir)\n",
    "dca_scores = {}\n",
    "for p_id in generator.protein_data:\n",
    "    print('Attempting to calculate DCA covariance for: {}'.format(p_id))\n",
    "    try:\n",
    "        protein_out_dir = os.path.join(dca_out_dir, p_id)\n",
    "        if not os.path.isdir(protein_out_dir):\n",
    "            os.makedirs(protein_out_dir)\n",
    "        curr_aln = SeqAlignment(file_name=generator.protein_data[p_id]['Final_FA_Aln'], query_id=p_id,\n",
    "                                polymer_type='Protein')\n",
    "        curr_aln.import_alignment()\n",
    "        # Since the DCA implementation used here does not provide a way to specify the query sequence we remove the gaps\n",
    "        # from the query sequences so positions will be referenced correctly for that sequence (and unnecessary\n",
    "        # computations can be avoided).\n",
    "        curr_aln = curr_aln.remove_gaps()\n",
    "        new_aln_fn = os.path.join(protein_out_dir, '{}_no_gap.fasta'.format(p_id))\n",
    "        curr_aln.write_out_alignment(new_aln_fn)\n",
    "        curr_aln.file_name = new_aln_fn\n",
    "        curr_dca = DCAWrapper(alignment=curr_aln)\n",
    "        curr_dca.calculate_scores(out_dir=protein_out_dir, delete_file=False)\n",
    "        dca_scores[p_id] = curr_dca\n",
    "        print('Successfully computed DCA covariance for: {}'.format(p_id))\n",
    "    except ValueError:\n",
    "        print('Could not compute DCA covariance for: {} with seq_length: {} and size: {}'.format(\n",
    "            p_id, curr_aln.seq_length, curr_aln.size))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## EVCouplings##\n",
    "Scoring the the covariation of the proteins using the EVCouplings method."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (cET-MIp)",
   "language": "python",
   "name": "cet-mip"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
