"""
Created on June 3, 2019

@author: Daniel Konecki
"""
import os
import heapq
import numpy as np
from Bio.Phylo import read, write
from Bio.Phylo.TreeConstruction import DistanceMatrix
from sklearn.cluster import AgglomerativeClustering
from Bio.Phylo.TreeConstruction import DistanceTreeConstructor


class PhylogeneticTree(object):
    """
    This class represents a phylogenetic tree as represented by the Bio.Phylo package. It allows for the construction of
    trees by several different methods, including loading a pre-existing tree.

    Attributes:
        distance_matrix (Bio.Phylo.TreeConstruction.DistanceMatrix) The distance matrix a tree is based on.
        tree_method (str): The method by which to construct a phylogenetic tree. The currently supported options are:
            'upgma'
            'agglomerative'
            'custom'
        tree_args (dict): The arguments needed to use each of these tree building methods.
            'upgma'
                No additional arguments.
            'agglomerative'
                'cache_dir' (str): The path to the directory where the agglomerative clustering data can be saved and
                loaded from.
                'affinity' (str): The affinity/distance calculation method to use when operating on the distance values
                for clustering. Further details can be found at: https://scikit-learn.org/stable/modules/generated/sklea
                rn.cluster.AgglomerativeClustering.html#sklearn.cluster.AgglomerativeClustering. The options are:
                    euclidean (default)
                    l1
                    l2
                    manhattan
                    cosin
                    precomputed
                'linkage' (str): The linkage algorithm to use when building the agglomerative clustering tree structure.
                Further details can be found at: https://scikit-learn.org/stable/modules/generated/sklearn.cluster.Agglo
                merativeClustering.html#sklearn.cluster.AgglomerativeClustering. The options are:
                    ward (default)
                    complete
                    average
                    single
            'custom'
                tree_path (str/path): The path to a file where the desired tree has been written in 'newick' format.
        tree (Bio.Phylo.Newick.Tree): The tree constructed using the specified method, until construct_tree has been
        called it is None.
        size (int): The number of terminal nodes in the tree (size of the alignment used to calculate the distance
        matrix used for tree construction).
    """

    def __init__(self, tree_building_method='upgma', tree_building_args={}):
        self.distance_matrix = None
        self.tree_method = tree_building_method
        self.tree_args = tree_building_args
        self.tree = None
        self.size = None

    def _custom_tree(self, tree_path):
        """
        Custom Tree

        This method loads a tree in the Newick format which should correspond to an alignment of interest.

        Args:
            tree_path (str): The full path to a Newick formatted tree which represents a target alignment.
        Returns:
            Bio.Phylo.Newick.Tree: Tree structure loaded from a file for a target alignment.
        """
        custom_tree = read(file=tree_path, format='newick')
        nodes_to_process = [custom_tree.root]
        while len(nodes_to_process) > 0:
            current_node = nodes_to_process.pop()
            if not current_node.is_terminal():
                nodes_to_process += current_node.clades
                if current_node.confidence:
                    current_node.name = 'Inner{}'.format(current_node.confidence)
                    current_node.confidence = None
        return custom_tree

    def _upgma_tree(self):
        """
        UPGMA Tree

        This method generates a UPGMA tree.

        Returns:
            Bio.Phylo.Newick.Tree: Tree structure generated by the UPGMA method from the distance matrix computed for
            the target alignment.
        """
        constructor = DistanceTreeConstructor()
        upgma_tree = constructor.upgma(distance_matrix=self.distance_matrix)
        return upgma_tree

    def _et_tree(self):
        """
        ET Tree

        This method reproduces the tree used by the Evolutionary Trace method in the past.This is a variant on the UPGMA
        tree construction method which joins nodes, not based on their minimum distance to each other but on the minimum
        average distance between their terminal descendants.
        """

        def height_of(clade):
            """
            Height Of

            Calculate clade height -- the longest path to any terminal (PRIVATE).

            Copied verbatim from Bio.Phylo.DistanceTreeConstructor

            Args:
                clade (Bio.Phylo.BaseTree.Clade): Clade object for which to determine the height.
            Returns:
                int: The height of the provided clade.
            """
            height = 0
            if clade.is_terminal():
                height = clade.branch_length
            else:
                height = height + max(height_of(c) for c in clade.clades)
            return height

        from Bio.Phylo import BaseTree
        from copy import deepcopy
        from itertools import product
        original_dist_mat = np.array(self.distance_matrix)
        index_map = {}
        clades = []
        for i in range(self.size):
            name = self.distance_matrix.names[i]
            index_map[name] = i
            clade = BaseTree.Clade(None, name)
            clades.append(clade)
        min_i = 0
        min_j = 0
        inner_count = self.size
        dm = deepcopy(self.distance_matrix)
        inner_clade = None
        while len(dm) > 1:
            # print('#' * 100)
            # print(len(dm))
            # print(len(clades))
            # Find minimum index
            min_i = 0
            min_j = 1
            min_dist = dm[min_i, min_j]
            for i in range(len(dm)):
                for j in range(i + 1, len(dm)):
                    # print(dm[i, j])
                    if min_dist > dm[i, j]:
                        min_dist = dm[i, j]
                        min_i = i
                        min_j = j
            # print(min_i)
            # print(min_j)
            # Create clade
            clade1 = clades[min_i]
            clade2 = clades[min_j]
            inner_count -= 1
            inner_clade = BaseTree.Clade(None, "Inner{}".format(inner_count))
            inner_clade.clades.append(clade1)
            inner_clade.clades.append(clade2)
            # Assign branch length
            if clade1.is_terminal():
                clade1.branch_length = min_dist * 1.0 / 2
            else:
                clade1.branch_length = min_dist * 1.0 / 2 - height_of(clade1)
            if clade2.is_terminal():
                clade2.branch_length = min_dist * 1.0 / 2
            else:
                clade2.branch_length = min_dist * 1.0 / 2 - height_of(clade2)
            # Rebuild distance mat set the distances of new node at the index of min_j
            for k in range(0, len(dm)):
                print('@' * 100)
                if k != min_i and k != min_j:
                    print(len(inner_clade.get_terminals()))
                    indices_inner = [index_map[node.name] for node in inner_clade.get_terminals()]
                    print(len(indices_inner))
                    print(len(clades[k].get_terminals()))
                    indices_k = [index_map[node.name] for node in clades[k].get_terminals()]
                    print(len(indices_k))
                    pos_inner, pos_k = zip(*product(indices_inner, indices_k))
                    print(len(indices_inner) * len(indices_k))
                    print(len(pos_inner))
                    dm[min_i, k] = np.mean(original_dist_mat[list(pos_inner), list(pos_k)])
            dm.names[min_i] = "Inner" + str(inner_count)
            del dm[min_j]
            # Update node list
            clades[min_i] = inner_clade
            del clades[min_j]
        inner_clade.branch_length = 0
        return BaseTree.Tree(inner_clade)

    def _agglomerative_clustering(self, cache_dir=None, affinity='euclidean', linkage='ward'):
        """
        Agglomerative Clustering

        Constructs a tree using agglomerative/hierarchical clustering (this requires significant conversion from the
        formatted provided by sklearn).

        Args:
            cache_dir (str): The path to the directory where the agglomerative clustering data can be saved and loaded
            from.
            affinity (str): The affinity/distance calculation method to use when operating on the distance values for
            clustering. Further details can be found at: https://scikit-learn.org/stable/modules/generated/sklearn.clust
            er.AgglomerativeClustering.html#sklearn.cluster.AgglomerativeClustering. The options are:
                euclidean (default)
                l1
                l2
                manhattan
                cosin
                precomputed
            linkage (str): The linkage algorithm to use when building the agglomerative clustering tree structure.
            Further details can be found at: https://scikit-learn.org/stable/modules/generated/sklearn.cluster.Agglomera
            tiveClustering.html#sklearn.cluster.AgglomerativeClustering. The options are:
                ward (default)
                complete
                average
                single
        Returns:
            Bio.Phylo.Newick.Tree: Tree structure generated by the agglomerative/hierarchical clustering method from the
            distance matrix computed for the target alignment.
        References:
            The solution for converting an agglomerative clustering tree from sklearn into a Newick formatted tree was
            taken from the following StackOverflow discussion, the solution used was provided by user: lucianopaz
            https://stackoverflow.com/questions/29127013/plot-dendrogram-using-sklearn-agglomerativeclustering
        """
        ml_model = AgglomerativeClustering(affinity=affinity, linkage=linkage, n_clusters=len(self.distance_matrix),
                                           memory=cache_dir, compute_full_tree=True)
        ml_model.fit(np.array(self.distance_matrix))
        newick_tree_string = convert_agglomerative_clustering_to_newick_tree(
            clusterer=ml_model, labels=self.distance_matrix.names, distance_matrix=np.array(self.distance_matrix))
        newick_fn = os.path.join(cache_dir, 'joblib', 'agg_clustering_{}_{}.newick'.format(affinity, linkage))
        with open(newick_fn, 'wb') as newick_handle:
            newick_handle.write(newick_tree_string)
        agg_clustering_tree = read(file=newick_fn, format='newick')
        return agg_clustering_tree

    def construct_tree(self, dm):
        """
        Construct Tree

        Construct the specified type of phylogenetic tree using the provided distance matrix.

        Args:
            dm (Bio.Phylo.TreeConstruction.DistanceMatrix): The distance matrix for an alignment of interest. If the
            tree building method is custom dm maybe be set to None. This method updates the __distance_matrix and tree
            attributes of the PhylogeneticTree instance.
        """
        method_dict = {'agglomerative': self._agglomerative_clustering, 'upgma': self._upgma_tree,
                       'custom': self._custom_tree, 'et': self._et_tree}
        if isinstance(dm, DistanceMatrix):
            self.distance_matrix = dm
        else:
            raise ValueError('Distance matrix (dm) must have type Bio.Phylo.TreeConstruction.DistanceMatrix!')
        self.size = len(dm)
        self.tree = method_dict[self.tree_method](**self.tree_args)

    def write_out_tree(self, filename):
        """
        Write Out Tree

        Writes the tree structure to file using the newick format.

        Args:
            filename (str): The full path to the file where the newick formatted tree should be written.
        """
        if self.tree is None:
            raise ValueError('Attempting to write tree before construction!')
        with open(filename, 'wb') as tree_handle:
            write(self.tree, file=tree_handle, format='newick')

    def traverse_top_down(self):
        """
        Traverse Base Tree

        This method acts an iterator for the tree traversing from the root to the leaves always yielding the next node
        according to path length.

        Returns:
            generator: A generator which will yield a new node ordered from least to greatest path length from the root.
        """
        nodes_to_process = []
        heapq.heappush(nodes_to_process, (0, self.tree.root))
        while len(nodes_to_process) > 0:
            curr_branch_len, curr_node = heapq.heappop(nodes_to_process)
            yield curr_node
            for node in curr_node.clades:
                heapq.heappush(nodes_to_process, (curr_branch_len + node.branch_length, node))

    def traverse_bottom_up(self):
        """
        Traverse Base Tree

        This method acts an iterator for the tree traversing from the leaves to the root always yielding the next node
        according to path length.

        Returns:
            generator: A generator which will yield a new node ordered from greatest to least path length from the root.
        """
        nodes_to_process = []
        nodes_visited = set()
        for leaf in self.tree.root.get_terminals():
            path = self.tree.get_path(leaf)
            dist = get_path_length(path)
            nodes_visited.add(leaf.name)
            heapq.heappush(nodes_to_process, (-1 * dist, path, leaf))
        while len(nodes_to_process) > 0:
            curr_dist, curr_path, curr_node = heapq.heappop(nodes_to_process)
            yield curr_node
            if len(curr_path) > 1:
                parent_node = curr_path[-2]
                parent_path = curr_path[:-1]
                parent_dist = curr_dist + curr_node.branch_length
                if parent_node.name not in nodes_visited:
                    nodes_visited.add(parent_node.name)
                    heapq.heappush(nodes_to_process, (parent_dist, parent_path, parent_node))
            else:
                if self.tree.root.name not in nodes_visited:
                    nodes_visited.add(self.tree.root.name)
                    heapq.heappush(nodes_to_process, (0.0, [], self.tree.root))

    def traverse_by_rank(self):
        """
        Traverse By Rank

        This method acts as an iterator for the tree, traversing from the root to the leaves always yielding a full list
        of nodes which constitute a rank (level in the tree). This means each new list generated should be larger than
        the previous list by 1 which is achieved by finding the shallowest inner node and replacing it with its two
        children. The node to replace is chosen by looking at the inner node names, the expected naming format is
        "Inner<int>" where <int> is the position of that node in the tree construction process. Nodes may be named in
        increasing order (self.tree.root.name="Inner1") or decreasing order (self.tree.root.name="Inner<self.size - 1>")

        Returns:
             generator: A generator which will yield a list of nodes corresponding to a rank/level in the tree.
        """
        root_pos = int(self.tree.root.name.strip('Inner'))
        reverse = -1 if (root_pos == (self.size - 1)) else 1
        next_node_to_split = [(reverse * root_pos, self.tree.root)]
        current_nodes = [self.tree.root]
        while len(current_nodes) < self.size:
            yield current_nodes
            pos, node_to_split = heapq.heappop(next_node_to_split)
            new_nodes = []
            for node in current_nodes:
                if node == node_to_split:
                    for child in node_to_split.clades:
                        new_nodes.append(child)
                        if not child.is_terminal():
                            heapq.heappush(next_node_to_split, (reverse * int(child.name.strip('Inner')), child))
                else:
                    new_nodes.append(node)
            current_nodes = new_nodes
        yield current_nodes

    def rename_internal_nodes(self):
        """
        Rename Internal Nodes

        This function renames all internal nodes so that trees constructed by different methods will follow the same
        internal node naming convention. The convention in this case is that every internal node has the name Inner{}
        where {} is replaced by an integer. The larger the integer the further down in the tree the node finds itself
        (i.e. the root node will have the integer 1 while the lowest will have the integer
        len(self.tree.get_terminals) - 1
        """
        starting_node_pos = int(self.tree.root.name.strip('Inner'))
        print('Starting Node Pos: {}'.format(starting_node_pos))
        if starting_node_pos == 1:
            return
        node_pattern = 'Inner{}'
        for node in self.traverse_bottom_up():
            if not node.is_terminal():
                node_pos = int(node.name.strip('Inner'))
                node.name = node_pattern.format(self.size - node_pos)

    def assign_group_rank(self, ranks=None):
        """
        Assign Group Rank

        This function traverses the tree from top to bottom assigning each node to a rank and group and tracking which
        leaf/terminal nodes are related to that node.

        Args:
            ranks (set/list): If this is set to None all ranks will be evaluated (default behavior for EvolutionaryTrace
            variants). If a set or list is passed only the ranks specified there will be have ranks and groups assigned.
            The lowest possible rank is 1 and the highest possible rank is the size of the alignment used to generate
            this tree (i.e. the number of leaf nodes/sequences). A rank and group will always be assigned for rank 1,
            even if it is not specified.
        Return:
            dict: First level of the dictionary maps a rank to another dictionary. The second level of the dictionary
            maps a group value to another dictionary. This third level of the dictionary maps the key 'node' to the node
            which is the root of the group at the given rank, 'terminals' to a list of node names for the leaf/
            terminal nodes which are ancestors of the root node, and 'descendants' to a list of nodes which are
            descendants of 'node' from the closest assigned rank (at the lowest rank this will be None).
        """
        # Create a set of the ranks for which to create assignments.
        if ranks is None:
            ranks = set(range(1, 1 + self.size))
        else:
            ranks = set(ranks)
            if 1 not in ranks:
                ranks.add(1)
        # Assign nodes and terminals to rank and group labels
        rank_group_mapping = {}
        iterator = self.traverse_by_rank()
        for rank_nodes in iterator:
            rank = len(rank_nodes)
            if rank not in ranks:
                continue
            rank_group_mapping[rank] = {}
            for group in range(1, rank + 1):
                current_node = rank_nodes[group - 1]
                rank_group_mapping[rank][group] = {'node': current_node, 'terminals': []}
                if current_node.is_bifurcating():
                    rank_group_mapping[rank][group]['terminals'] += [t.name for t in current_node.get_terminals()]
                else:
                    rank_group_mapping[rank][group]['terminals'].append(current_node.name)
        # For each node assigned in the rank_group_mapping determine the descendants in the next closest rank.
        prev_rank = None
        unique_descendants = {}
        for rank in sorted(ranks, reverse=True):
            curr_rank = set([])
            for group in rank_group_mapping[rank]:
                node = rank_group_mapping[rank][group]['node']
                curr_rank.add(node)
                descendants = None
                if node.name not in unique_descendants:
                    if prev_rank is not None:
                        descendants = []
                        for prev_node in prev_rank:
                            if (node == prev_node) or node.is_parent_of(target=prev_node):
                                descendants.append(prev_node)
                        prev_rank -= set(descendants)
                    unique_descendants[node.name] = descendants
                rank_group_mapping[rank][group]['descendants'] = unique_descendants[node.name]
            prev_rank = curr_rank
        return rank_group_mapping


def get_path_length(path):
    """
    Get Parent

    Get the parent of the passed in node.

    Args:
        path (list): A list of Bio.Phylo.BaseTree.Clade objects which constitute a path between the root and a target
        node.
    Returns:
         float: The length of the path passed in.
    """
    dist = 0
    for node in path:
        dist += node.branch_length
    return dist


def get_cluster_spanner(agg_clusterer):
    """
    Get Cluster Spanner

    Get a callable that computes a given cluster's span. To compute a cluster's span, call spanner(cluster). The cluster
    must be a 2D numpy array, where the axis=0 holds separate cluster members and the axis=1 holds the different
    variables.

    Args:
        agg_clusterer (sklearn.cluster.AgglomerativeClustering): Instance of the clustering model which has already been
        fit to data and which can be used to determine the proper way to extract distances from a given distance
        matrix.
    Returns:
        function: Callable to compute a given cluster's span.
    References:
        Taken from user lucianopaz's response on stackoverflow question:
        https://stackoverflow.com/questions/29127013/plot-dendrogram-using-sklearn-agglomerativeclustering
    """
    spanner = None
    if agg_clusterer.linkage == 'ward':
        if agg_clusterer.affinity == 'euclidean':
            spanner = lambda x: np.sum((x - agg_clusterer.pooling_func(x, axis=0)) ** 2)
    elif agg_clusterer.linkage == 'complete':
        if agg_clusterer.affinity == 'euclidean':
            spanner = lambda x: np.max(np.sum((x[:, None, :] - x[None, :, :]) ** 2, axis=2))
        elif agg_clusterer.affinity == 'l1' or agg_clusterer.affinity == 'manhattan':
            spanner = lambda x: np.max(np.sum(np.abs(x[:, None, :] - x[None, :, :]), axis=2))
        elif agg_clusterer.affinity == 'l2':
            spanner = lambda x: np.max(np.sqrt(np.sum((x[:, None, :] - x[None, :, :]) ** 2, axis=2)))
        elif agg_clusterer.affinity == 'cosine':
            spanner = lambda x: (np.max(np.sum((x[:, None, :] * x[None, :, :])) /
                                        (np.sqrt(np.sum(x[:, None, :] * x[:, None, :], axis=2, keepdims=True)) *
                                         np.sqrt(np.sum(x[None, :, :] * x[None, :, :], axis=2, keepdims=True)))))
        else:
            raise AttributeError('Unknown affinity attribute value {0}.'.format(agg_clusterer.affinity))
    elif agg_clusterer.linkage == 'average':
        if agg_clusterer.affinity == 'euclidean':
            spanner = lambda x: np.mean(np.sum((x[:, None, :] - x[None, :, :]) ** 2, axis=2))
        elif agg_clusterer.affinity == 'l1' or agg_clusterer.affinity == 'manhattan':
            spanner = lambda x: np.mean(np.sum(np.abs(x[:, None, :] - x[None, :, :]), axis=2))
        elif agg_clusterer.affinity == 'l2':
            spanner = lambda x: np.mean(np.sqrt(np.sum((x[:, None, :] - x[None, :, :]) ** 2, axis=2)))
        elif agg_clusterer.affinity == 'cosine':
            spanner = lambda x: (np.mean(np.sum((x[:, None, :] * x[None, :, :])) /
                                         (np.sqrt(np.sum(x[:, None, :] * x[:, None, :], axis=2, keepdims=True)) *
                                          np.sqrt(np.sum(x[None, :, :] * x[None, :, :], axis=2, keepdims=True)))))
        else:
            raise AttributeError('Unknown affinity attribute value {0}.'.format(agg_clusterer.affinity))
    else:
        raise AttributeError('Unknown linkage attribute value {0}.'.format(agg_clusterer.linkage))
    return spanner


def go_down_tree(children, n_leaves, x, leaf_labels, nodename, spanner, inner):
    """
    go_down_tree(children,n_leaves,X,leaf_labels,nodename,spanner)

    Iterative function that traverses the subtree that descends from
    nodename and returns the Newick representation of the subtree.

    Args:
        children (list): sklearn.cluster.AgglomerativeClustering.children_
        n_leaves (int): sklearn.cluster.AgglomerativeClustering.n_leaves_
        x (2D np.array): Parameters supplied to sklearn.cluster.AgglomerativeClustering.fit
        leaf_labels (list): The label for each parameter in array axis=0 of x.
        nodename (int): The intermediate node name whose children are located in children[nodename-n_leaves].
        spanner (function): Callable that computes the dendrite's span
        inner (list): A list of labels for inner nodes which is used every time an internal node needs to be named.
    Returns:
        str: The str representation of the agglomerative clutering tree in Newick format.
    References:
        Taken from user lucianopaz's response on stackoverflow question:
        https://stackoverflow.com/questions/29127013/plot-dendrogram-using-sklearn-agglomerativeclustering
    """
    nodeindex = nodename-n_leaves
    if nodename < n_leaves:
        return leaf_labels[nodeindex], np.array([x[nodeindex]])
    else:
        node_children = children[nodeindex]
        branch1, branch1samples = go_down_tree(children, n_leaves, x, leaf_labels, node_children[1], spanner, inner)
        if branch1.endswith(')'):
            branch1 += 'Inner{}'.format(inner.pop())
        branch0, branch0samples = go_down_tree(children, n_leaves, x, leaf_labels, node_children[0], spanner, inner)
        if branch0.endswith(')'):
            branch0 += 'Inner{}'.format(inner.pop())
        node = np.vstack((branch0samples, branch1samples))
        branch0span = spanner(branch0samples)
        branch1span = spanner(branch1samples)
        nodespan = spanner(node)
        branch0distance = nodespan-branch0span
        branch1distance = nodespan-branch1span
        nodename = '({branch0}:{branch0distance},{branch1}:{branch1distance})'.format(
            branch0=branch0, branch0distance=branch0distance, branch1=branch1, branch1distance=branch1distance)
        return nodename, node


def build_newick_tree(children, n_leaves, x, leaf_labels, spanner):
    """
    Build Newick Tree

    Generate a string representation (Newick tree) from the sklearn.cluster.AgglomerativeClustering.fit output.

    Args:
        children (list): sklearn.cluster.AgglomerativeClustering.children_
        n_leaves (int): sklearn.cluster.AgglomerativeClustering.n_leaves_
        x (2D np.array): Parameters supplied to sklearn.cluster.AgglomerativeClustering.fit
        leaf_labels (list): The label for each parameter in array axis=0 of x.
        spanner (function): Callable that computes the dendrite's span
    Returns:
        str: The final string with the Newick representation of the agglomerative clustering tree being converted.
    References:
        Taken from user lucianopaz's response on stackoverflow question:
        https://stackoverflow.com/questions/29127013/plot-dendrogram-using-sklearn-agglomerativeclustering
    """
    inner_labels = list(range(1, n_leaves))
    inner_string, inner_node = go_down_tree(children, n_leaves, x, leaf_labels, len(children)+n_leaves-1, spanner,
                                            inner_labels)
    prepend_to_root = inner_string + 'Inner{};'.format(inner_labels.pop())
    return prepend_to_root


def convert_agglomerative_clustering_to_newick_tree(clusterer, labels, distance_matrix):
    """
    Convert Agglomerative Clustering To Newick Tree

    This function converts the tree generated by sklearn.cluster.AgglomerativeClustering.fit() to a Newick formatted
    string.

    Args:
        clusterer (sklearn.cluster.AgglomerativeClustering): Instance of the clustering model which has already been
        fit to data and which can be used to determine the proper way to extract distances from a given distance
        matrix.
        labels (list): The label for each parameter in array axis=0 of distance_matrix.
        distance_matrix (2D np.array): Parameters supplied to sklearn.cluster.AgglomerativeClustering.fit
    Returns:
        str: The final string with the Newick representation of the agglomerative clustering tree being converted.
    References:
        Adapted from user lucianopaz's response on stackoverflow question:
        https://stackoverflow.com/questions/29127013/plot-dendrogram-using-sklearn-agglomerativeclustering
    """
    spanner = get_cluster_spanner(clusterer)
    newick_tree = build_newick_tree(clusterer.children_, clusterer.n_leaves_, distance_matrix, labels, spanner)
    return newick_tree
